// uglify-js/lib/ast.js
/***********************************************************************

  A JavaScript tokenizer / parser / beautifier / compressor.
  https://github.com/mishoo/UglifyJS2

  -------------------------------- (C***REMOVED*** ---------------------------------

                           Author: Mihai Bazon
                         <mihai.bazon@gmail.com>
                       http://mihai.bazon.net/blog

  Distributed under the BSD license:

    Copyright 2012 (c***REMOVED*** Mihai Bazon <mihai.bazon@gmail.com>

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

        * Redistributions of source code must retain the above
          copyright notice, this list of conditions and the following
          disclaimer.

        * Redistributions in binary form must reproduce the above
          copyright notice, this list of conditions and the following
          disclaimer in the documentation and/or other materials
          provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION***REMOVED*** HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
    TORT (INCLUDING NEGLIGENCE OR OTHERWISE***REMOVED*** ARISING IN ANY WAY OUT OF
    THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

 ***********************************************************************/

"use strict";

function DEFNODE(type, props, methods, base***REMOVED*** {
    if (arguments.length < 4***REMOVED*** base = AST_Node;
    if (!props***REMOVED*** props = [];
    else props = props.split(/\s+/***REMOVED***;
    var self_props = props;
    if (base && base.PROPS***REMOVED***
        props = props.concat(base.PROPS***REMOVED***;
    var code = "return function AST_" + type + "(props***REMOVED***{ if (props***REMOVED*** { ";
    for (var i = props.length; --i >= 0;***REMOVED*** {
        code += "this." + props[i] + " = props." + props[i] + ";";
  ***REMOVED***
    var proto = base && new base;
    if (proto && proto.initialize || (methods && methods.initialize***REMOVED******REMOVED***
        code += "this.initialize(***REMOVED***;";
    code += "}}";
    var ctor = new Function(code***REMOVED***(***REMOVED***;
    if (proto***REMOVED*** {
        ctor.prototype = proto;
        ctor.BASE = base;
  ***REMOVED***
    if (base***REMOVED*** base.SUBCLASSES.push(ctor***REMOVED***;
    ctor.prototype.CTOR = ctor;
    ctor.PROPS = props || null;
    ctor.SELF_PROPS = self_props;
    ctor.SUBCLASSES = [];
    if (type***REMOVED*** {
        ctor.prototype.TYPE = ctor.TYPE = type;
  ***REMOVED***
    if (methods***REMOVED*** for (i in methods***REMOVED*** if (methods.hasOwnProperty(i***REMOVED******REMOVED*** {
        if (/^\$/.test(i***REMOVED******REMOVED*** {
            ctor[i.substr(1***REMOVED***] = methods[i];
      ***REMOVED*** else {
            ctor.prototype[i] = methods[i];
      ***REMOVED***
  ***REMOVED***
    ctor.DEFMETHOD = function(name, method***REMOVED*** {
        this.prototype[name] = method;
  ***REMOVED***;
    return ctor;
};

var AST_Token = DEFNODE("Token", "type value line col pos endpos nlb comments_before file", {
}, null***REMOVED***;

var AST_Node = DEFNODE("Node", "start end", {
    clone: function(***REMOVED*** {
        return new this.CTOR(this***REMOVED***;
  ***REMOVED***,
    $documentation: "Base class of all AST nodes",
    $propdoc: {
        start: "[AST_Token] The first token of this node",
        end: "[AST_Token] The last token of this node"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this***REMOVED***;
  ***REMOVED***,
    walk: function(visitor***REMOVED*** {
        return this._walk(visitor***REMOVED***; // not sure the indirection will be any help
  ***REMOVED***
}, null***REMOVED***;

AST_Node.warn_function = null;
AST_Node.warn = function(txt, props***REMOVED*** {
    if (AST_Node.warn_function***REMOVED***
        AST_Node.warn_function(string_template(txt, props***REMOVED******REMOVED***;
};

/* -----[ statements ]----- */

var AST_Statement = DEFNODE("Statement", null, {
    $documentation: "Base class of all statements",
}***REMOVED***;

var AST_Debugger = DEFNODE("Debugger", null, {
    $documentation: "Represents a debugger statement",
}, AST_Statement***REMOVED***;

var AST_Directive = DEFNODE("Directive", "value scope", {
    $documentation: "Represents a directive, like \"use strict\";",
    $propdoc: {
        value: "[string] The value of this directive as a plain string (it's not an AST_String!***REMOVED***",
        scope: "[AST_Scope/S] The scope that this directive affects"
  ***REMOVED***,
}, AST_Statement***REMOVED***;

var AST_SimpleStatement = DEFNODE("SimpleStatement", "body", {
    $documentation: "A statement consisting of an expression, i.e. a = 1 + 2",
    $propdoc: {
        body: "[AST_Node] an expression node (should not be instanceof AST_Statement***REMOVED***"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Statement***REMOVED***;

function walk_body(node, visitor***REMOVED*** {
    if (node.body instanceof AST_Statement***REMOVED*** {
        node.body._walk(visitor***REMOVED***;
  ***REMOVED***
    else node.body.forEach(function(stat***REMOVED***{
        stat._walk(visitor***REMOVED***;
  ***REMOVED******REMOVED***;
};

var AST_Block = DEFNODE("Block", "body", {
    $documentation: "A body of statements (usually bracketed***REMOVED***",
    $propdoc: {
        body: "[AST_Statement*] an array of statements"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            walk_body(this, visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Statement***REMOVED***;

var AST_BlockStatement = DEFNODE("BlockStatement", null, {
    $documentation: "A block statement",
}, AST_Block***REMOVED***;

var AST_EmptyStatement = DEFNODE("EmptyStatement", null, {
    $documentation: "The empty statement (empty block or simply a semicolon***REMOVED***",
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this***REMOVED***;
  ***REMOVED***
}, AST_Statement***REMOVED***;

var AST_StatementWithBody = DEFNODE("StatementWithBody", "body", {
    $documentation: "Base class for all statements that contain one nested body: `For`, `ForIn`, `Do`, `While`, `With`",
    $propdoc: {
        body: "[AST_Statement] the body; this should always be present, even if it's an AST_EmptyStatement"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Statement***REMOVED***;

var AST_LabeledStatement = DEFNODE("LabeledStatement", "label", {
    $documentation: "Statement with a label",
    $propdoc: {
        label: "[AST_Label] a label definition"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.label._walk(visitor***REMOVED***;
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_StatementWithBody***REMOVED***;

var AST_DWLoop = DEFNODE("DWLoop", "condition", {
    $documentation: "Base class for do/while statements",
    $propdoc: {
        condition: "[AST_Node] the loop condition.  Should not be instanceof AST_Statement"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.condition._walk(visitor***REMOVED***;
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_StatementWithBody***REMOVED***;

var AST_Do = DEFNODE("Do", null, {
    $documentation: "A `do` statement",
}, AST_DWLoop***REMOVED***;

var AST_While = DEFNODE("While", null, {
    $documentation: "A `while` statement",
}, AST_DWLoop***REMOVED***;

var AST_For = DEFNODE("For", "init condition step", {
    $documentation: "A `for` statement",
    $propdoc: {
        init: "[AST_Node?] the `for` initialization code, or null if empty",
        condition: "[AST_Node?] the `for` termination clause, or null if empty",
        step: "[AST_Node?] the `for` update clause, or null if empty"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            if (this.init***REMOVED*** this.init._walk(visitor***REMOVED***;
            if (this.condition***REMOVED*** this.condition._walk(visitor***REMOVED***;
            if (this.step***REMOVED*** this.step._walk(visitor***REMOVED***;
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_StatementWithBody***REMOVED***;

var AST_ForIn = DEFNODE("ForIn", "init name object", {
    $documentation: "A `for ... in` statement",
    $propdoc: {
        init: "[AST_Node] the `for/in` initialization code",
        name: "[AST_SymbolRef?] the loop variable, only if `init` is AST_Var",
        object: "[AST_Node] the object that we're looping through"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.init._walk(visitor***REMOVED***;
            this.object._walk(visitor***REMOVED***;
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_StatementWithBody***REMOVED***;

var AST_With = DEFNODE("With", "expression", {
    $documentation: "A `with` statement",
    $propdoc: {
        expression: "[AST_Node] the `with` expression"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
            this.body._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_StatementWithBody***REMOVED***;

/* -----[ scope and functions ]----- */

var AST_Scope = DEFNODE("Scope", "directives variables functions uses_with uses_eval parent_scope enclosed cname", {
    $documentation: "Base class for all statements introducing a lexical scope",
    $propdoc: {
        directives: "[string*/S] an array of directives declared in this scope",
        variables: "[Object/S] a map of name -> SymbolDef for all variables/functions defined in this scope",
        functions: "[Object/S] like `variables`, but only lists function declarations",
        uses_with: "[boolean/S] tells whether this scope uses the `with` statement",
        uses_eval: "[boolean/S] tells whether this scope contains a direct call to the global `eval`",
        parent_scope: "[AST_Scope?/S] link to the parent scope",
        enclosed: "[SymbolDef*/S] a list of all symbol definitions that are accessed from this scope or any subscopes",
        cname: "[integer/S] current index for mangling variables (used internally by the mangler***REMOVED***",
  ***REMOVED***,
}, AST_Block***REMOVED***;

var AST_Toplevel = DEFNODE("Toplevel", "globals", {
    $documentation: "The toplevel scope",
    $propdoc: {
        globals: "[Object/S] a map of name -> SymbolDef for all undeclared names",
  ***REMOVED***,
    wrap_commonjs: function(name, export_all***REMOVED*** {
        var self = this;
        var to_export = [];
        if (export_all***REMOVED*** {
            self.figure_out_scope(***REMOVED***;
            self.walk(new TreeWalker(function(node***REMOVED***{
                if (node instanceof AST_SymbolDeclaration && node.definition(***REMOVED***.global***REMOVED*** {
                    if (!find_if(function(n***REMOVED***{ return n.name == node.name }, to_export***REMOVED******REMOVED***
                        to_export.push(node***REMOVED***;
              ***REMOVED***
          ***REMOVED******REMOVED******REMOVED***;
      ***REMOVED***
        var wrapped_tl = "(function(exports, global***REMOVED***{ global['" + name + "'] = exports; '$ORIG'; '$EXPORTS'; }({}, (function(***REMOVED***{return this}(***REMOVED******REMOVED******REMOVED******REMOVED***";
        wrapped_tl = parse(wrapped_tl***REMOVED***;
        wrapped_tl = wrapped_tl.transform(new TreeTransformer(function before(node***REMOVED***{
            if (node instanceof AST_SimpleStatement***REMOVED*** {
                node = node.body;
                if (node instanceof AST_String***REMOVED*** switch (node.getValue(***REMOVED******REMOVED*** {
                  case "$ORIG":
                    return MAP.splice(self.body***REMOVED***;
                  case "$EXPORTS":
                    var body = [];
                    to_export.forEach(function(sym***REMOVED***{
                        body.push(new AST_SimpleStatement({
                            body: new AST_Assign({
                                left: new AST_Sub({
                                    expression: new AST_SymbolRef({ name: "exports" }***REMOVED***,
                                    property: new AST_String({ value: sym.name }***REMOVED***,
                              ***REMOVED******REMOVED***,
                                operator: "=",
                                right: new AST_SymbolRef(sym***REMOVED***,
                          ***REMOVED******REMOVED***,
                      ***REMOVED******REMOVED******REMOVED***;
                  ***REMOVED******REMOVED***;
                    return MAP.splice(body***REMOVED***;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED******REMOVED******REMOVED***;
        return wrapped_tl;
  ***REMOVED***
}, AST_Scope***REMOVED***;

var AST_Lambda = DEFNODE("Lambda", "name argnames uses_arguments", {
    $documentation: "Base class for functions",
    $propdoc: {
        name: "[AST_SymbolDeclaration?] the name of this function",
        argnames: "[AST_SymbolFunarg*] array of function arguments",
        uses_arguments: "[boolean/S] tells whether this function accesses the arguments array"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            if (this.name***REMOVED*** this.name._walk(visitor***REMOVED***;
            this.argnames.forEach(function(arg***REMOVED***{
                arg._walk(visitor***REMOVED***;
          ***REMOVED******REMOVED***;
            walk_body(this, visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Scope***REMOVED***;

var AST_Accessor = DEFNODE("Accessor", null, {
    $documentation: "A setter/getter function"
}, AST_Lambda***REMOVED***;

var AST_Function = DEFNODE("Function", null, {
    $documentation: "A function expression"
}, AST_Lambda***REMOVED***;

var AST_Defun = DEFNODE("Defun", null, {
    $documentation: "A function definition"
}, AST_Lambda***REMOVED***;

/* -----[ JUMPS ]----- */

var AST_Jump = DEFNODE("Jump", null, {
    $documentation: "Base class for “jumps” (for now that's `return`, `throw`, `break` and `continue`***REMOVED***"
}, AST_Statement***REMOVED***;

var AST_Exit = DEFNODE("Exit", "value", {
    $documentation: "Base class for “exits” (`return` and `throw`***REMOVED***",
    $propdoc: {
        value: "[AST_Node?] the value returned or thrown by this statement; could be null for AST_Return"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, this.value && function(***REMOVED***{
            this.value._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Jump***REMOVED***;

var AST_Return = DEFNODE("Return", null, {
    $documentation: "A `return` statement"
}, AST_Exit***REMOVED***;

var AST_Throw = DEFNODE("Throw", null, {
    $documentation: "A `throw` statement"
}, AST_Exit***REMOVED***;

var AST_LoopControl = DEFNODE("LoopControl", "label", {
    $documentation: "Base class for loop control statements (`break` and `continue`***REMOVED***",
    $propdoc: {
        label: "[AST_LabelRef?] the label, or null if none",
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, this.label && function(***REMOVED***{
            this.label._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Jump***REMOVED***;

var AST_Break = DEFNODE("Break", null, {
    $documentation: "A `break` statement"
}, AST_LoopControl***REMOVED***;

var AST_Continue = DEFNODE("Continue", null, {
    $documentation: "A `continue` statement"
}, AST_LoopControl***REMOVED***;

/* -----[ IF ]----- */

var AST_If = DEFNODE("If", "condition alternative", {
    $documentation: "A `if` statement",
    $propdoc: {
        condition: "[AST_Node] the `if` condition",
        alternative: "[AST_Statement?] the `else` part, or null if not present"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.condition._walk(visitor***REMOVED***;
            this.body._walk(visitor***REMOVED***;
            if (this.alternative***REMOVED*** this.alternative._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_StatementWithBody***REMOVED***;

/* -----[ SWITCH ]----- */

var AST_Switch = DEFNODE("Switch", "expression", {
    $documentation: "A `switch` statement",
    $propdoc: {
        expression: "[AST_Node] the `switch` “discriminant”"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
            walk_body(this, visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Block***REMOVED***;

var AST_SwitchBranch = DEFNODE("SwitchBranch", null, {
    $documentation: "Base class for `switch` branches",
}, AST_Block***REMOVED***;

var AST_Default = DEFNODE("Default", null, {
    $documentation: "A `default` switch branch",
}, AST_SwitchBranch***REMOVED***;

var AST_Case = DEFNODE("Case", "expression", {
    $documentation: "A `case` switch branch",
    $propdoc: {
        expression: "[AST_Node] the `case` expression"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
            walk_body(this, visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_SwitchBranch***REMOVED***;

/* -----[ EXCEPTIONS ]----- */

var AST_Try = DEFNODE("Try", "bcatch bfinally", {
    $documentation: "A `try` statement",
    $propdoc: {
        bcatch: "[AST_Catch?] the catch block, or null if not present",
        bfinally: "[AST_Finally?] the finally block, or null if not present"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            walk_body(this, visitor***REMOVED***;
            if (this.bcatch***REMOVED*** this.bcatch._walk(visitor***REMOVED***;
            if (this.bfinally***REMOVED*** this.bfinally._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Block***REMOVED***;

// XXX: this is wrong according to ECMA-262 (12.4***REMOVED***.  the catch block
// should introduce another scope, as the argname should be visible
// only inside the catch block.  However, doing it this way because of
// IE which simply introduces the name in the surrounding scope.  If
// we ever want to fix this then AST_Catch should inherit from
// AST_Scope.
var AST_Catch = DEFNODE("Catch", "argname", {
    $documentation: "A `catch` node; only makes sense as part of a `try` statement",
    $propdoc: {
        argname: "[AST_SymbolCatch] symbol for the exception"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.argname._walk(visitor***REMOVED***;
            walk_body(this, visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Block***REMOVED***;

var AST_Finally = DEFNODE("Finally", null, {
    $documentation: "A `finally` node; only makes sense as part of a `try` statement"
}, AST_Block***REMOVED***;

/* -----[ VAR/CONST ]----- */

var AST_Definitions = DEFNODE("Definitions", "definitions", {
    $documentation: "Base class for `var` or `const` nodes (variable declarations/initializations***REMOVED***",
    $propdoc: {
        definitions: "[AST_VarDef*] array of variable definitions"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.definitions.forEach(function(def***REMOVED***{
                def._walk(visitor***REMOVED***;
          ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_Statement***REMOVED***;

var AST_Var = DEFNODE("Var", null, {
    $documentation: "A `var` statement"
}, AST_Definitions***REMOVED***;

var AST_Const = DEFNODE("Const", null, {
    $documentation: "A `const` statement"
}, AST_Definitions***REMOVED***;

var AST_VarDef = DEFNODE("VarDef", "name value", {
    $documentation: "A variable declaration; only appears in a AST_Definitions node",
    $propdoc: {
        name: "[AST_SymbolVar|AST_SymbolConst] name of the variable",
        value: "[AST_Node?] initializer, or null of there's no initializer"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.name._walk(visitor***REMOVED***;
            if (this.value***REMOVED*** this.value._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

/* -----[ OTHER ]----- */

var AST_Call = DEFNODE("Call", "expression args", {
    $documentation: "A function call expression",
    $propdoc: {
        expression: "[AST_Node] expression to invoke as function",
        args: "[AST_Node*] array of arguments"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
            this.args.forEach(function(arg***REMOVED***{
                arg._walk(visitor***REMOVED***;
          ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_New = DEFNODE("New", null, {
    $documentation: "An object instantiation.  Derives from a function call since it has exactly the same properties"
}, AST_Call***REMOVED***;

var AST_Seq = DEFNODE("Seq", "car cdr", {
    $documentation: "A sequence expression (two comma-separated expressions***REMOVED***",
    $propdoc: {
        car: "[AST_Node] first element in sequence",
        cdr: "[AST_Node] second element in sequence"
  ***REMOVED***,
    $cons: function(x, y***REMOVED*** {
        var seq = new AST_Seq(x***REMOVED***;
        seq.car = x;
        seq.cdr = y;
        return seq;
  ***REMOVED***,
    $from_array: function(array***REMOVED*** {
        if (array.length == 0***REMOVED*** return null;
        if (array.length == 1***REMOVED*** return array[0].clone(***REMOVED***;
        var list = null;
        for (var i = array.length; --i >= 0;***REMOVED*** {
            list = AST_Seq.cons(array[i], list***REMOVED***;
      ***REMOVED***
        var p = list;
        while (p***REMOVED*** {
            if (p.cdr && !p.cdr.cdr***REMOVED*** {
                p.cdr = p.cdr.car;
                break;
          ***REMOVED***
            p = p.cdr;
      ***REMOVED***
        return list;
  ***REMOVED***,
    to_array: function(***REMOVED*** {
        var p = this, a = [];
        while (p***REMOVED*** {
            a.push(p.car***REMOVED***;
            if (p.cdr && !(p.cdr instanceof AST_Seq***REMOVED******REMOVED*** {
                a.push(p.cdr***REMOVED***;
                break;
          ***REMOVED***
            p = p.cdr;
      ***REMOVED***
        return a;
  ***REMOVED***,
    add: function(node***REMOVED*** {
        var p = this;
        while (p***REMOVED*** {
            if (!(p.cdr instanceof AST_Seq***REMOVED******REMOVED*** {
                var cell = AST_Seq.cons(p.cdr, node***REMOVED***;
                return p.cdr = cell;
          ***REMOVED***
            p = p.cdr;
      ***REMOVED***
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.car._walk(visitor***REMOVED***;
            if (this.cdr***REMOVED*** this.cdr._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_PropAccess = DEFNODE("PropAccess", "expression property", {
    $documentation: "Base class for property access expressions, i.e. `a.foo` or `a[\"foo\"]`",
    $propdoc: {
        expression: "[AST_Node] the “container” expression",
        property: "[AST_Node|string] the property to access.  For AST_Dot this is always a plain string, while for AST_Sub it's an arbitrary AST_Node"
  ***REMOVED***
}***REMOVED***;

var AST_Dot = DEFNODE("Dot", null, {
    $documentation: "A dotted property access expression",
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_PropAccess***REMOVED***;

var AST_Sub = DEFNODE("Sub", null, {
    $documentation: "Index-style property access, i.e. `a[\"foo\"]`",
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
            this.property._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}, AST_PropAccess***REMOVED***;

var AST_Unary = DEFNODE("Unary", "operator expression", {
    $documentation: "Base class for unary expressions",
    $propdoc: {
        operator: "[string] the operator",
        expression: "[AST_Node] expression that this unary operator applies to"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.expression._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_UnaryPrefix = DEFNODE("UnaryPrefix", null, {
    $documentation: "Unary prefix expression, i.e. `typeof i` or `++i`"
}, AST_Unary***REMOVED***;

var AST_UnaryPostfix = DEFNODE("UnaryPostfix", null, {
    $documentation: "Unary postfix expression, i.e. `i++`"
}, AST_Unary***REMOVED***;

var AST_Binary = DEFNODE("Binary", "left operator right", {
    $documentation: "Binary expression, i.e. `a + b`",
    $propdoc: {
        left: "[AST_Node] left-hand side expression",
        operator: "[string] the operator",
        right: "[AST_Node] right-hand side expression"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.left._walk(visitor***REMOVED***;
            this.right._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_Conditional = DEFNODE("Conditional", "condition consequent alternative", {
    $documentation: "Conditional expression using the ternary operator, i.e. `a ? b : c`",
    $propdoc: {
        condition: "[AST_Node]",
        consequent: "[AST_Node]",
        alternative: "[AST_Node]"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.condition._walk(visitor***REMOVED***;
            this.consequent._walk(visitor***REMOVED***;
            this.alternative._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_Assign = DEFNODE("Assign", null, {
    $documentation: "An assignment expression — `a = b + 5`",
}, AST_Binary***REMOVED***;

/* -----[ LITERALS ]----- */

var AST_Array = DEFNODE("Array", "elements", {
    $documentation: "An array literal",
    $propdoc: {
        elements: "[AST_Node*] array of elements"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.elements.forEach(function(el***REMOVED***{
                el._walk(visitor***REMOVED***;
          ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_Object = DEFNODE("Object", "properties", {
    $documentation: "An object literal",
    $propdoc: {
        properties: "[AST_ObjectProperty*] array of properties"
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.properties.forEach(function(prop***REMOVED***{
                prop._walk(visitor***REMOVED***;
          ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_ObjectProperty = DEFNODE("ObjectProperty", "key value", {
    $documentation: "Base class for literal object properties",
    $propdoc: {
        key: "[string] the property name; it's always a plain string in our AST, no matter if it was a string, number or identifier in original code",
        value: "[AST_Node] property value.  For setters and getters this is an AST_Function."
  ***REMOVED***,
    _walk: function(visitor***REMOVED*** {
        return visitor._visit(this, function(***REMOVED***{
            this.value._walk(visitor***REMOVED***;
      ***REMOVED******REMOVED***;
  ***REMOVED***
}***REMOVED***;

var AST_ObjectKeyVal = DEFNODE("ObjectKeyVal", null, {
    $documentation: "A key: value object property",
}, AST_ObjectProperty***REMOVED***;

var AST_ObjectSetter = DEFNODE("ObjectSetter", null, {
    $documentation: "An object setter property",
}, AST_ObjectProperty***REMOVED***;

var AST_ObjectGetter = DEFNODE("ObjectGetter", null, {
    $documentation: "An object getter property",
}, AST_ObjectProperty***REMOVED***;

var AST_Symbol = DEFNODE("Symbol", "scope name thedef", {
    $propdoc: {
        name: "[string] name of this symbol",
        scope: "[AST_Scope/S] the current scope (not necessarily the definition scope***REMOVED***",
        thedef: "[SymbolDef/S] the definition of this symbol"
  ***REMOVED***,
    $documentation: "Base class for all symbols",
}***REMOVED***;

var AST_SymbolAccessor = DEFNODE("SymbolAccessor", null, {
    $documentation: "The name of a property accessor (setter/getter function***REMOVED***"
}, AST_Symbol***REMOVED***;

var AST_SymbolDeclaration = DEFNODE("SymbolDeclaration", "init", {
    $documentation: "A declaration symbol (symbol in var/const, function name or argument, symbol in catch***REMOVED***",
    $propdoc: {
        init: "[AST_Node*/S] array of initializers for this declaration."
  ***REMOVED***
}, AST_Symbol***REMOVED***;

var AST_SymbolVar = DEFNODE("SymbolVar", null, {
    $documentation: "Symbol defining a variable",
}, AST_SymbolDeclaration***REMOVED***;

var AST_SymbolConst = DEFNODE("SymbolConst", null, {
    $documentation: "A constant declaration"
}, AST_SymbolDeclaration***REMOVED***;

var AST_SymbolFunarg = DEFNODE("SymbolFunarg", null, {
    $documentation: "Symbol naming a function argument",
}, AST_SymbolVar***REMOVED***;

var AST_SymbolDefun = DEFNODE("SymbolDefun", null, {
    $documentation: "Symbol defining a function",
}, AST_SymbolDeclaration***REMOVED***;

var AST_SymbolLambda = DEFNODE("SymbolLambda", null, {
    $documentation: "Symbol naming a function expression",
}, AST_SymbolDeclaration***REMOVED***;

var AST_SymbolCatch = DEFNODE("SymbolCatch", null, {
    $documentation: "Symbol naming the exception in catch",
}, AST_SymbolDeclaration***REMOVED***;

var AST_Label = DEFNODE("Label", "references", {
    $documentation: "Symbol naming a label (declaration***REMOVED***",
    $propdoc: {
        references: "[AST_LabelRef*] a list of nodes referring to this label"
  ***REMOVED***
}, AST_Symbol***REMOVED***;

var AST_SymbolRef = DEFNODE("SymbolRef", null, {
    $documentation: "Reference to some symbol (not definition/declaration***REMOVED***",
}, AST_Symbol***REMOVED***;

var AST_LabelRef = DEFNODE("LabelRef", null, {
    $documentation: "Reference to a label symbol",
}, AST_Symbol***REMOVED***;

var AST_This = DEFNODE("This", null, {
    $documentation: "The `this` symbol",
}, AST_Symbol***REMOVED***;

var AST_Constant = DEFNODE("Constant", null, {
    $documentation: "Base class for all constants",
    getValue: function(***REMOVED*** {
        return this.value;
  ***REMOVED***
}***REMOVED***;

var AST_String = DEFNODE("String", "value", {
    $documentation: "A string literal",
    $propdoc: {
        value: "[string] the contents of this string"
  ***REMOVED***
}, AST_Constant***REMOVED***;

var AST_Number = DEFNODE("Number", "value", {
    $documentation: "A number literal",
    $propdoc: {
        value: "[number] the numeric value"
  ***REMOVED***
}, AST_Constant***REMOVED***;

var AST_RegExp = DEFNODE("RegExp", "value", {
    $documentation: "A regexp literal",
    $propdoc: {
        value: "[RegExp] the actual regexp"
  ***REMOVED***
}, AST_Constant***REMOVED***;

var AST_Atom = DEFNODE("Atom", null, {
    $documentation: "Base class for atoms",
}, AST_Constant***REMOVED***;

var AST_Null = DEFNODE("Null", null, {
    $documentation: "The `null` atom",
    value: null
}, AST_Atom***REMOVED***;

var AST_NaN = DEFNODE("NaN", null, {
    $documentation: "The impossible value",
    value: 0/0
}, AST_Atom***REMOVED***;

var AST_Undefined = DEFNODE("Undefined", null, {
    $documentation: "The `undefined` value",
    value: (function(***REMOVED***{}(***REMOVED******REMOVED***
}, AST_Atom***REMOVED***;

var AST_Hole = DEFNODE("Hole", null, {
    $documentation: "A hole in an array",
    value: (function(***REMOVED***{}(***REMOVED******REMOVED***
}, AST_Atom***REMOVED***;

var AST_Infinity = DEFNODE("Infinity", null, {
    $documentation: "The `Infinity` value",
    value: 1/0
}, AST_Atom***REMOVED***;

var AST_Boolean = DEFNODE("Boolean", null, {
    $documentation: "Base class for booleans",
}, AST_Atom***REMOVED***;

var AST_False = DEFNODE("False", null, {
    $documentation: "The `false` atom",
    value: false
}, AST_Boolean***REMOVED***;

var AST_True = DEFNODE("True", null, {
    $documentation: "The `true` atom",
    value: true
}, AST_Boolean***REMOVED***;

/* -----[ TreeWalker ]----- */

function TreeWalker(callback***REMOVED*** {
    this.visit = callback;
    this.stack = [];
};
TreeWalker.prototype = {
    _visit: function(node, descend***REMOVED*** {
        this.stack.push(node***REMOVED***;
        var ret = this.visit(node, descend ? function(***REMOVED***{
            descend.call(node***REMOVED***;
      ***REMOVED*** : noop***REMOVED***;
        if (!ret && descend***REMOVED*** {
            descend.call(node***REMOVED***;
      ***REMOVED***
        this.stack.pop(***REMOVED***;
        return ret;
  ***REMOVED***,
    parent: function(n***REMOVED*** {
        return this.stack[this.stack.length - 2 - (n || 0***REMOVED***];
  ***REMOVED***,
    push: function (node***REMOVED*** {
        this.stack.push(node***REMOVED***;
  ***REMOVED***,
    pop: function(***REMOVED*** {
        return this.stack.pop(***REMOVED***;
  ***REMOVED***,
    self: function(***REMOVED*** {
        return this.stack[this.stack.length - 1];
  ***REMOVED***,
    find_parent: function(type***REMOVED*** {
        var stack = this.stack;
        for (var i = stack.length; --i >= 0;***REMOVED*** {
            var x = stack[i];
            if (x instanceof type***REMOVED*** return x;
      ***REMOVED***
  ***REMOVED***,
    in_boolean_context: function(***REMOVED*** {
        var stack = this.stack;
        var i = stack.length, self = stack[--i];
        while (i > 0***REMOVED*** {
            var p = stack[--i];
            if ((p instanceof AST_If           && p.condition === self***REMOVED*** ||
                (p instanceof AST_Conditional  && p.condition === self***REMOVED*** ||
                (p instanceof AST_DWLoop       && p.condition === self***REMOVED*** ||
                (p instanceof AST_For          && p.condition === self***REMOVED*** ||
                (p instanceof AST_UnaryPrefix  && p.operator == "!" && p.expression === self***REMOVED******REMOVED***
            {
                return true;
          ***REMOVED***
            if (!(p instanceof AST_Binary && (p.operator == "&&" || p.operator == "||"***REMOVED******REMOVED******REMOVED***
                return false;
            self = p;
      ***REMOVED***
  ***REMOVED***,
    loopcontrol_target: function(label***REMOVED*** {
        var stack = this.stack;
        if (label***REMOVED*** {
            for (var i = stack.length; --i >= 0;***REMOVED*** {
                var x = stack[i];
                if (x instanceof AST_LabeledStatement && x.label.name == label.name***REMOVED*** {
                    return x.body;
              ***REMOVED***
          ***REMOVED***
      ***REMOVED*** else {
            for (var i = stack.length; --i >= 0;***REMOVED*** {
                var x = stack[i];
                if (x instanceof AST_Switch
                    || x instanceof AST_For
                    || x instanceof AST_ForIn
                    || x instanceof AST_DWLoop***REMOVED*** return x;
          ***REMOVED***
      ***REMOVED***
  ***REMOVED***
};
