// express/lib/router/route.js

/**
 * Module dependencies.
 */

var utils = require('../utils'***REMOVED***;

/**
 * Expose `Route`.
 */

module.exports = Route;

/**
 * Initialize `Route` with the given HTTP `method`, `path`,
 * and an array of `callbacks` and `options`.
 *
 * Options:
 *
 *   - `sensitive`    enable case-sensitive routes
 *   - `strict`       enable strict matching for trailing slashes
 *
 * @param {String} method
 * @param {String} path
 * @param {Array} callbacks
 * @param {Object} options.
 * @api private
 */

function Route(method, path, callbacks, options***REMOVED*** {
  options = options || {};
  this.path = path;
  this.method = method;
  this.callbacks = callbacks;
  this.regexp = utils.pathRegexp(path
    , this.keys = []
    , options.sensitive
    , options.strict***REMOVED***;
}

/**
 * Check if this route matches `path`, if so
 * populate `.params`.
 *
 * @param {String} path
 * @return {Boolean}
 * @api private
 */

Route.prototype.match = function(path***REMOVED***{
  var keys = this.keys
    , params = this.params = []
    , m = this.regexp.exec(path***REMOVED***;

  if (!m***REMOVED*** return false;

  for (var i = 1, len = m.length; i < len; ++i***REMOVED*** {
    var key = keys[i - 1];

    try {
      var val = 'string' == typeof m[i]
        ? decodeURIComponent(m[i]***REMOVED***
        : m[i];
  ***REMOVED*** catch(e***REMOVED*** {
      var err = new Error("Failed to decode param '" + m[i] + "'"***REMOVED***;
      err.status = 400;
      throw err;
  ***REMOVED***

    if (key***REMOVED*** {
      params[key.name] = val;
  ***REMOVED*** else {
      params.push(val***REMOVED***;
  ***REMOVED***
***REMOVED***

  return true;
};
