// core-util-is/util.js
// Copyright Joyent, Inc. and other Node contributors.
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"***REMOVED***, to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.

// NOTE: These type checking functions intentionally don't use `instanceof`
// because it is fragile and can be easily faked with `Object.create(***REMOVED***`.
function isArray(ar***REMOVED*** {
  return Array.isArray(ar***REMOVED***;
}
exports.isArray = isArray;

function isBoolean(arg***REMOVED*** {
  return typeof arg === 'boolean';
}
exports.isBoolean = isBoolean;

function isNull(arg***REMOVED*** {
  return arg === null;
}
exports.isNull = isNull;

function isNullOrUndefined(arg***REMOVED*** {
  return arg == null;
}
exports.isNullOrUndefined = isNullOrUndefined;

function isNumber(arg***REMOVED*** {
  return typeof arg === 'number';
}
exports.isNumber = isNumber;

function isString(arg***REMOVED*** {
  return typeof arg === 'string';
}
exports.isString = isString;

function isSymbol(arg***REMOVED*** {
  return typeof arg === 'symbol';
}
exports.isSymbol = isSymbol;

function isUndefined(arg***REMOVED*** {
  return arg === void 0;
}
exports.isUndefined = isUndefined;

function isRegExp(re***REMOVED*** {
  return isObject(re***REMOVED*** && objectToString(re***REMOVED*** === '[object RegExp]';
}
exports.isRegExp = isRegExp;

function isObject(arg***REMOVED*** {
  return typeof arg === 'object' && arg !== null;
}
exports.isObject = isObject;

function isDate(d***REMOVED*** {
  return isObject(d***REMOVED*** && objectToString(d***REMOVED*** === '[object Date]';
}
exports.isDate = isDate;

function isError(e***REMOVED*** {
  return isObject(e***REMOVED*** && objectToString(e***REMOVED*** === '[object Error]';
}
exports.isError = isError;

function isFunction(arg***REMOVED*** {
  return typeof arg === 'function';
}
exports.isFunction = isFunction;

function isPrimitive(arg***REMOVED*** {
  return arg === null ||
         typeof arg === 'boolean' ||
         typeof arg === 'number' ||
         typeof arg === 'string' ||
         typeof arg === 'symbol' ||  // ES6 symbol
         typeof arg === 'undefined';
}
exports.isPrimitive = isPrimitive;

function isBuffer(arg***REMOVED*** {
  return arg instanceof Buffer;
}
exports.isBuffer = isBuffer;

function objectToString(o***REMOVED*** {
  return Object.prototype.toString.call(o***REMOVED***;
}
