// mongodb/lib/mongodb/db.js
/**
 * Module dependencies.
 * @ignore
 */
var QueryCommand = require('./commands/query_command'***REMOVED***.QueryCommand
  , DbCommand = require('./commands/db_command'***REMOVED***.DbCommand
  , MongoReply = require('./responses/mongo_reply'***REMOVED***.MongoReply
  , Admin = require('./admin'***REMOVED***.Admin
  , Collection = require('./collection'***REMOVED***.Collection
  , Server = require('./connection/server'***REMOVED***.Server
  , ReplSet = require('./connection/repl_set/repl_set'***REMOVED***.ReplSet
  , ReadPreference = require('./connection/read_preference'***REMOVED***.ReadPreference
  , Mongos = require('./connection/mongos'***REMOVED***.Mongos
  , Cursor = require('./cursor'***REMOVED***.Cursor
  , EventEmitter = require('events'***REMOVED***.EventEmitter
  , InsertCommand = require('./commands/insert_command'***REMOVED***.InsertCommand
  , inherits = require('util'***REMOVED***.inherits
  , crypto = require('crypto'***REMOVED***
  , timers = require('timers'***REMOVED***
  , utils = require('./utils'***REMOVED***
  
  // Authentication methods
  , mongodb_cr_authenticate = require('./auth/mongodb_cr.js'***REMOVED***.authenticate
  , mongodb_gssapi_authenticate = require('./auth/mongodb_gssapi.js'***REMOVED***.authenticate
  , mongodb_sspi_authenticate = require('./auth/mongodb_sspi.js'***REMOVED***.authenticate
  , mongodb_plain_authenticate = require('./auth/mongodb_plain.js'***REMOVED***.authenticate
  , mongodb_x509_authenticate = require('./auth/mongodb_x509.js'***REMOVED***.authenticate;

var hasKerberos = false;
// Check if we have a the kerberos library
try {
  require('kerberos'***REMOVED***;
  hasKerberos = true;
} catch(err***REMOVED*** {}

// Set processor, setImmediate if 0.10 otherwise nextTick
var processor = require('./utils'***REMOVED***.processor(***REMOVED***;
var indexOptions = ["unique", "sparse", "background", "dropDups", "min", "max", "v", "expireAfterSeconds", "bucketSize"];

/**
 * Create a new Db instance.
 *
 * Options
 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *  - **readPreference** {String}, the preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *  - **native_parser** {Boolean, default:false}, use c++ bson parser.
 *  - **forceServerObjectId** {Boolean, default:false}, force server to create _id fields instead of client.
 *  - **pkFactory** {Object}, object overriding the basic ObjectID primary key generation.
 *  - **serializeFunctions** {Boolean, default:false}, serialize functions.
 *  - **raw** {Boolean, default:false}, perform operations using raw bson buffers.
 *  - **recordQueryStats** {Boolean, default:false}, record query statistics during execution.
 *  - **retryMiliSeconds** {Number, default:5000}, number of milliseconds between retries.
 *  - **numberOfRetries** {Number, default:5}, number of retries off connection.
 *  - **logger** {Object, default:null}, an object representing a logger that you want to use, needs to support functions debug, log, error **({error:function(message, object***REMOVED*** {}, log:function(message, object***REMOVED*** {}, debug:function(message, object***REMOVED*** {}}***REMOVED*****.
 *  - **slaveOk** {Number, default:null}, force setting of SlaveOk flag on queries (only use when explicitly connecting to a secondary server***REMOVED***.
 *  - **promoteLongs** {Boolean, default:true}, when deserializing a Long will fit it into a Number if it's smaller than 53 bits
 *  - **bufferMaxEntries** {Boolean, default: -1}, sets a cap on how many operations the driver will buffer up before giving up on getting a working connection, default is -1 which is unlimited
 *
 * @class Represents a Db
 * @param {String} databaseName name of the database.
 * @param {Object} serverConfig server config object.
 * @param {Object} [options] additional options for the collection.
 */
function Db(databaseName, serverConfig, options***REMOVED*** {
  if(!(this instanceof Db***REMOVED******REMOVED*** return new Db(databaseName, serverConfig, options***REMOVED***;
  EventEmitter.call(this***REMOVED***;
  var self = this;
  this.databaseName = databaseName;
  this.serverConfig = serverConfig;
  this.options = options == null ? {} : options;
  // State to check against if the user force closed db
  this._applicationClosed = false;
  // Fetch the override flag if any
  var overrideUsedFlag = this.options['override_used_flag'] == null ? false : this.options['override_used_flag'];

  // Verify that nobody is using this config
  if(!overrideUsedFlag && this.serverConfig != null && typeof this.serverConfig == 'object' && this.serverConfig._isUsed && this.serverConfig._isUsed(***REMOVED******REMOVED*** {    
    throw new Error('A Server or ReplSet instance cannot be shared across multiple Db instances'***REMOVED***;
***REMOVED*** else if(!overrideUsedFlag && typeof this.serverConfig == 'object'***REMOVED***{
    // Set being used
    this.serverConfig._used = true;
***REMOVED***

  // Allow slaveOk override
  this.slaveOk = this.options['slave_ok'] == null ? false : this.options['slave_ok'];
  this.slaveOk = this.options['slaveOk'] == null ? this.slaveOk : this.options['slaveOk'];
  
  // Number of operations to buffer before failure
  this.bufferMaxEntries = typeof this.options['bufferMaxEntries'] == 'number' ? this.options['bufferMaxEntries'] : -1;

  // Ensure we have a valid db name
  validateDatabaseName(databaseName***REMOVED***;

  // Contains all the connections for the db
  try {
    this.native_parser = this.options.native_parser;
    // The bson lib
    var bsonLib = this.bsonLib = this.options.native_parser ? require('bson'***REMOVED***.BSONNative : require('bson'***REMOVED***.BSONPure;
    bsonLib = require('bson'***REMOVED***.BSONPure;
    // Fetch the serializer object
    var BSON = bsonLib.BSON;
    
    // Create a new instance
    this.bson = new BSON([bsonLib.Long, bsonLib.ObjectID, bsonLib.Binary, bsonLib.Code, bsonLib.DBRef, bsonLib.Symbol, bsonLib.Double, bsonLib.Timestamp, bsonLib.MaxKey, bsonLib.MinKey]***REMOVED***;
    this.bson.promoteLongs = this.options.promoteLongs == null ? true : this.options.promoteLongs;
    
    // Backward compatibility to access types
    this.bson_deserializer = bsonLib;
    this.bson_serializer = bsonLib;
    
    // Add any overrides to the serializer and deserializer
    this.bson_deserializer.promoteLongs = this.options.promoteLongs == null ? true : this.options.promoteLongs;
***REMOVED*** catch (err***REMOVED*** {
    // If we tried to instantiate the native driver
    var msg = 'Native bson parser not compiled, please compile '
            + 'or avoid using native_parser=true';
    throw Error(msg***REMOVED***;
***REMOVED***

  // Internal state of the server
  this._state = 'disconnected';

  this.pkFactory = this.options.pkFactory == null ? bsonLib.ObjectID : this.options.pkFactory;
  this.forceServerObjectId = this.options.forceServerObjectId != null ? this.options.forceServerObjectId : false;

  // Added safe
  this.safe = this.options.safe == null ? false : this.options.safe;  

  // If we have not specified a "safe mode" we just print a warning to the console
  if(this.options.safe == null 
    && this.options.w == null 
    && this.options.j == null
    && this.options.journal == null 
    && this.options.fsync == null***REMOVED*** {
    console.log("========================================================================================"***REMOVED***;
    console.log("=  Please ensure that you set the default write concern for the database by setting    ="***REMOVED***;
    console.log("=   one of the options                                                                 ="***REMOVED***;
    console.log("=                                                                                      ="***REMOVED***;
    console.log("=     w: (value of > -1 or the string 'majority'***REMOVED***, where < 1 means                     ="***REMOVED***;
    console.log("=        no write acknowledgement                                                       ="***REMOVED***;
    console.log("=     journal: true/false, wait for flush to journal before acknowledgement             ="***REMOVED***;
    console.log("=     fsync: true/false, wait for flush to file system before acknowledgement           ="***REMOVED***;
    console.log("=                                                                                      ="***REMOVED***;
    console.log("=  For backward compatibility safe is still supported and                              ="***REMOVED***;
    console.log("=   allows values of [true | false | {j:true} | {w:n, wtimeout:n} | {fsync:true}]      ="***REMOVED***;
    console.log("=   the default value is false which means the driver receives does not                ="***REMOVED***;
    console.log("=   return the information of the success/error of the insert/update/remove            ="***REMOVED***;
    console.log("=                                                                                      ="***REMOVED***;
    console.log("=   ex: new Db(new Server('localhost', 27017***REMOVED***, {safe:false}***REMOVED***                           ="***REMOVED***;
    console.log("=                                                                                      ="***REMOVED***;
    console.log("=   http://www.mongodb.org/display/DOCS/getLastError+Command                           ="***REMOVED***;
    console.log("=                                                                                      ="***REMOVED***;
    console.log("=  The default of no acknowledgement will change in the very near future                ="***REMOVED***;
    console.log("=                                                                                      ="***REMOVED***;
    console.log("=  This message will disappear when the default safe is set on the driver Db           ="***REMOVED***;
    console.log("========================================================================================"***REMOVED***;
***REMOVED***

  // Internal states variables
  this.notReplied ={};
  this.isInitializing = true;
  this.openCalled = false;

  // Command queue, keeps a list of incoming commands that need to be executed once the connection is up
  this.commands = [];

  // Set up logger
  this.logger = this.options.logger != null
    && (typeof this.options.logger.debug == 'function'***REMOVED***
    && (typeof this.options.logger.error == 'function'***REMOVED***
    && (typeof this.options.logger.log == 'function'***REMOVED***
      ? this.options.logger : {error:function(message, object***REMOVED*** {}, log:function(message, object***REMOVED*** {}, debug:function(message, object***REMOVED*** {}};

  // Associate the logger with the server config
  this.serverConfig.logger = this.logger;
  if(this.serverConfig.strategyInstance***REMOVED*** this.serverConfig.strategyInstance.logger = this.logger;
  this.tag = new Date(***REMOVED***.getTime(***REMOVED***;
  // Just keeps list of events we allow
  this.eventHandlers = {error:[], parseError:[], poolReady:[], message:[], close:[]};

  // Controls serialization options
  this.serializeFunctions = this.options.serializeFunctions != null ? this.options.serializeFunctions : false;

  // Raw mode
  this.raw = this.options.raw != null ? this.options.raw : false;

  // Record query stats
  this.recordQueryStats = this.options.recordQueryStats != null ? this.options.recordQueryStats : false;

  // If we have server stats let's make sure the driver objects have it enabled
  if(this.recordQueryStats == true***REMOVED*** {
    this.serverConfig.enableRecordQueryStats(true***REMOVED***;
***REMOVED***

  // Retry information
  this.retryMiliSeconds = this.options.retryMiliSeconds != null ? this.options.retryMiliSeconds : 1000;
  this.numberOfRetries = this.options.numberOfRetries != null ? this.options.numberOfRetries : 60;

  // Set default read preference if any
  this.readPreference = this.options.readPreference;

  // Set slaveOk if we have specified a secondary or secondary preferred readPreference
  if(this.readPreference == ReadPreference.SECONDARY ||
    this.readPreference == ReadPreference.SECONDARY_PREFERRED***REMOVED*** {
    this.slaveOk = true;
***REMOVED***

  // Set read preference on serverConfig if none is set
  // but the db one was
  if(this.serverConfig.options.readPreference != null***REMOVED*** {
    this.serverConfig.setReadPreference(this.serverConfig.options.readPreference***REMOVED***;
***REMOVED*** else if(this.readPreference != null***REMOVED*** {
    this.serverConfig.setReadPreference(this.readPreference***REMOVED***;
***REMOVED***

  // Ensure we keep a reference to this db
  this.serverConfig._dbStore.add(this***REMOVED***;
};

/**
 * @ignore
 */
function validateDatabaseName(databaseName***REMOVED*** {
  if(typeof databaseName !== 'string'***REMOVED*** throw new Error("database name must be a string"***REMOVED***;
  if(databaseName.length === 0***REMOVED*** throw new Error("database name cannot be the empty string"***REMOVED***;
  if(databaseName == '$external'***REMOVED*** return;

  var invalidChars = [" ", ".", "$", "/", "\\"];
  for(var i = 0; i < invalidChars.length; i++***REMOVED*** {
    if(databaseName.indexOf(invalidChars[i]***REMOVED*** != -1***REMOVED*** throw new Error("database names cannot contain the character '" + invalidChars[i] + "'"***REMOVED***;
***REMOVED***
}

/**
 * @ignore
 */
inherits(Db, EventEmitter***REMOVED***;

/**
 * Initialize the database connection.
 *
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the index information or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.open = function(callback***REMOVED*** {
  var self = this;

  // Check that the user has not called this twice
  if(this.openCalled***REMOVED*** {
    // Close db
    this.close(***REMOVED***;
    // Throw error
    throw new Error("db object already connecting, open cannot be called multiple times"***REMOVED***;
***REMOVED***

  // If we have a specified read preference
  if(this.readPreference != null***REMOVED*** this.serverConfig.setReadPreference(this.readPreference***REMOVED***;

  // Set that db has been opened
  this.openCalled = true;

  // Set the status of the server
  self._state = 'connecting';
  
  // Set up connections
  if(self.serverConfig instanceof Server || self.serverConfig instanceof ReplSet || self.serverConfig instanceof Mongos***REMOVED*** {
    // Ensure we have the original options passed in for the server config
    var connect_options = {};
    for(var name in self.serverConfig.options***REMOVED*** {
      connect_options[name] = self.serverConfig.options[name]
  ***REMOVED***
    connect_options.firstCall = true;

    // Attempt to connect
    self.serverConfig.connect(self, connect_options, function(err, result***REMOVED*** {
      if(err != null***REMOVED*** {
        // Close db to reset connection
        return self.close(function (***REMOVED*** {
          // Return error from connection
          return callback(err, null***REMOVED***;
      ***REMOVED******REMOVED***;
    ***REMOVED***
      // Set the status of the server
      self._state = 'connected';
      // If we have queued up commands execute a command to trigger replays
      if(self.commands.length > 0***REMOVED*** _execute_queued_command(self***REMOVED***;
      // Callback
      process.nextTick(function(***REMOVED*** {
        try {
          callback(null, self***REMOVED***;
      ***REMOVED*** catch(err***REMOVED*** {
          self.close(***REMOVED***;
          throw err;
      ***REMOVED***
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED*** else {
    try {
      callback(Error("Server parameter must be of type Server, ReplSet or Mongos"***REMOVED***, null***REMOVED***;
  ***REMOVED*** catch(err***REMOVED*** {
      self.close(***REMOVED***;
      throw err;
  ***REMOVED***
***REMOVED***
};

/**
 * Create a new Db instance sharing the current socket connections.
 *
 * @param {String} dbName the name of the database we want to use.
 * @return {Db} a db instance using the new database.
 * @api public
 */
Db.prototype.db = function(dbName***REMOVED*** {
  // Copy the options and add out internal override of the not shared flag
  var options = {};
  for(var key in this.options***REMOVED*** {
    options[key] = this.options[key];
***REMOVED***

  // Add override flag
  options['override_used_flag'] = true;
  // Check if the db already exists and reuse if it's the case
  var db = this.serverConfig._dbStore.fetch(dbName***REMOVED***;

  // Create a new instance
  if(!db***REMOVED*** {
    db = new Db(dbName, this.serverConfig, options***REMOVED***;
***REMOVED***

  // Return the db object
  return db;  
};

/**
 * Close the current db connection, including all the child db instances. Emits close event if no callback is provided.
 *
 * @param {Boolean} [forceClose] connection can never be reused.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.close = function(forceClose, callback***REMOVED*** {
  var self = this;
  // Ensure we force close all connections
  this._applicationClosed = false;

  if(typeof forceClose == 'function'***REMOVED*** {
    callback = forceClose;
***REMOVED*** else if(typeof forceClose == 'boolean'***REMOVED*** {
    this._applicationClosed = forceClose;
***REMOVED***

  this.serverConfig.close(function(err, result***REMOVED*** {
    // You can reuse the db as everything is shut down
    self.openCalled = false;
    // If we have a callback call it
    if(callback***REMOVED*** callback(err, result***REMOVED***;    
***REMOVED******REMOVED***;
};

/**
 * Access the Admin database
 *
 * @param {Function} [callback] returns the results.
 * @return {Admin} the admin db object.
 * @api public
 */
Db.prototype.admin = function(callback***REMOVED*** {
  if(callback == null***REMOVED*** return new Admin(this***REMOVED***;
  callback(null, new Admin(this***REMOVED******REMOVED***;
};

/**
 * Returns a cursor to all the collection information.
 *
 * @param {String} [collectionName] the collection name we wish to retrieve the information from.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the options or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.collectionsInfo = function(collectionName, callback***REMOVED*** {
  if(callback == null && typeof collectionName == 'function'***REMOVED*** { callback = collectionName; collectionName = null; }
  // Create selector
  var selector = {};
  // If we are limiting the access to a specific collection name
  if(collectionName != null***REMOVED*** selector.name = this.databaseName + "." + collectionName;

  // Return Cursor
  // callback for backward compatibility
  if(callback***REMOVED*** {
    callback(null, new Cursor(this, new Collection(this, DbCommand.SYSTEM_NAMESPACE_COLLECTION***REMOVED***, selector***REMOVED******REMOVED***;
***REMOVED*** else {
    return new Cursor(this, new Collection(this, DbCommand.SYSTEM_NAMESPACE_COLLECTION***REMOVED***, selector***REMOVED***;
***REMOVED***
};

/**
 * Get the list of all collection names for the specified db
 *
 * Options
 *  - **namesOnly** {String, default:false}, Return only the full collection namespace.
 *
 * @param {String} [collectionName] the collection name we wish to filter by.
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the collection names or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.collectionNames = function(collectionName, options, callback***REMOVED*** {
  var self = this;
  var args = Array.prototype.slice.call(arguments, 0***REMOVED***;
  callback = args.pop(***REMOVED***;
  collectionName = args.length ? args.shift(***REMOVED*** : null;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // Ensure no breaking behavior
  if(collectionName != null && typeof collectionName == 'object'***REMOVED*** {
    options = collectionName;
    collectionName = null;
***REMOVED***

  // Let's make our own callback to reuse the existing collections info method
  self.collectionsInfo(collectionName, function(err, cursor***REMOVED*** {
    if(err != null***REMOVED*** return callback(err, null***REMOVED***;

    cursor.toArray(function(err, documents***REMOVED*** {
      if(err != null***REMOVED*** return callback(err, null***REMOVED***;

      // List of result documents that have been filtered
      var filtered_documents = documents.filter(function(document***REMOVED*** {
        return !(document.name.indexOf(self.databaseName***REMOVED*** == -1 || document.name.indexOf('$'***REMOVED*** != -1***REMOVED***;
    ***REMOVED******REMOVED***;

      // If we are returning only the names
      if(options.namesOnly***REMOVED*** {
        filtered_documents = filtered_documents.map(function(document***REMOVED*** { return document.name }***REMOVED***;
    ***REMOVED***

      // Return filtered items
      callback(null, filtered_documents***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Fetch a specific collection (containing the actual collection information***REMOVED***. If the application does not use strict mode you can
 * can use it without a callback in the following way. var collection = db.collection('mycollection'***REMOVED***;
 *
 * Options
*  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *  - **serializeFunctions** {Boolean, default:false}, serialize functions on the document.
 *  - **raw** {Boolean, default:false}, perform all operations using raw bson objects.
 *  - **pkFactory** {Object}, object overriding the basic ObjectID primary key generation.
 *  - **readPreference** {String}, the preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *  - **strict**, (Boolean, default:false***REMOVED*** returns an error if the collection does not exist
 *
 * @param {String} collectionName the collection name we wish to access.
 * @param {Object} [options] returns option results.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the collection or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.collection = function(collectionName, options, callback***REMOVED*** {
  var self = this;
  if(typeof options === "function"***REMOVED*** { callback = options; options = {}; }
  // Execute safe

  if(options && (options.strict***REMOVED******REMOVED*** {
    self.collectionNames(collectionName, function(err, collections***REMOVED*** {
      if(err != null***REMOVED*** return callback(err, null***REMOVED***;

      if(collections.length == 0***REMOVED*** {
        return callback(new Error("Collection " + collectionName + " does not exist. Currently in safe mode."***REMOVED***, null***REMOVED***;
    ***REMOVED*** else {
        try {
          var collection = new Collection(self, collectionName, self.pkFactory, options***REMOVED***;
      ***REMOVED*** catch(err***REMOVED*** {
          return callback(err, null***REMOVED***;
      ***REMOVED***
        return callback(null, collection***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED*** else {
    try {
      var collection = new Collection(self, collectionName, self.pkFactory, options***REMOVED***;
  ***REMOVED*** catch(err***REMOVED*** {
      if(callback == null***REMOVED*** {
        throw err;
    ***REMOVED*** else {
        return callback(err, null***REMOVED***;
    ***REMOVED***
  ***REMOVED***

    // If we have no callback return collection object
    return callback == null ? collection : callback(null, collection***REMOVED***;
***REMOVED***
};

/**
 * Fetch all collections for the current db.
 *
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the collections or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.collections = function(callback***REMOVED*** {
  var self = this;
  // Let's get the collection names
  self.collectionNames(function(err, documents***REMOVED*** {
    if(err != null***REMOVED*** return callback(err, null***REMOVED***;
    var collections = [];
    documents.forEach(function(document***REMOVED*** {
      collections.push(new Collection(self, document.name.replace(self.databaseName + ".", ''***REMOVED***, self.pkFactory***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
    // Return the collection objects
    callback(null, collections***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Evaluate javascript on the server
 *
 * Options
 *  - **nolock** {Boolean, default:false}, Tell MongoDB not to block on the evaulation of the javascript.
 *
 * @param {Code} code javascript to execute on server.
 * @param {Object|Array} [parameters] the parameters for the call.
 * @param {Object} [options] the options
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from eval or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.eval = function(code, parameters, options, callback***REMOVED*** {
  // Unpack calls
  var args = Array.prototype.slice.call(arguments, 1***REMOVED***;
  callback = args.pop(***REMOVED***;
  parameters = args.length ? args.shift(***REMOVED*** : parameters;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  var finalCode = code;
  var finalParameters = [];
  // If not a code object translate to one
  if(!(finalCode instanceof this.bsonLib.Code***REMOVED******REMOVED*** {
    finalCode = new this.bsonLib.Code(finalCode***REMOVED***;
***REMOVED***

  // Ensure the parameters are correct
  if(parameters != null && parameters.constructor != Array && typeof parameters !== 'function'***REMOVED*** {
    finalParameters = [parameters];
***REMOVED*** else if(parameters != null && parameters.constructor == Array && typeof parameters !== 'function'***REMOVED*** {
    finalParameters = parameters;
***REMOVED***

  // Create execution selector
  var cmd = {'$eval':finalCode, 'args':finalParameters};
  // Check if the nolock parameter is passed in
  if(options['nolock']***REMOVED*** {
    cmd['nolock'] = options['nolock'];
***REMOVED***

  // Set primary read preference
  options.readPreference = ReadPreference.PRIMARY;

  // Execute the command
  this.command(cmd, options, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    if(result && result.ok == 1***REMOVED*** return callback(null, result.retval***REMOVED***;
    if(result***REMOVED*** return callback(new Error("eval failed: " + result.errmsg***REMOVED***, null***REMOVED***;
    callback(err, result***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Dereference a dbref, against a db
 *
 * @param {DBRef} dbRef db reference object we wish to resolve.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from dereference or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.dereference = function(dbRef, callback***REMOVED*** {
  var db = this;
  // If we have a db reference then let's get the db first
  if(dbRef.db != null***REMOVED*** db = this.db(dbRef.db***REMOVED***;
  // Fetch the collection and find the reference
  var collection = db.collection(dbRef.namespace***REMOVED***;
  collection.findOne({'_id':dbRef.oid}, function(err, result***REMOVED*** {
    callback(err, result***REMOVED***;
***REMOVED******REMOVED***;
}

/**
 * Logout user from server, fire off on all connections and remove all auth info
 *
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from logout or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.logout = function(options, callback***REMOVED*** {
  var self = this;
  // Unpack calls
  var args = Array.prototype.slice.call(arguments, 0***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // Number of connections we need to logout from
  var numberOfConnections = this.serverConfig.allRawConnections(***REMOVED***.length;
  // logout command
  var cmd = {'logout':1};
  // Add onAll to login to ensure all connection are logged out
  options.onAll = true;

  // Execute the command
  this.command(cmd, options, function(err, result***REMOVED*** {
    // Count down
    numberOfConnections = numberOfConnections - 1;
    // Work around the case where the number of connections are 0
    if(numberOfConnections <= 0 && typeof callback == 'function'***REMOVED*** {
      var internalCallback = callback;
      callback = null;

      // Remove the db from auths
      self.serverConfig.auth.remove(self.databaseName***REMOVED***;
      // Callback with result
      internalCallback(null, result.ok == 1 ? true : false***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;
}

/**
 * Authenticate a user against the server.
 * authMechanism
 * Options
 *  - **authMechanism** {String, default:MONGODB-CR}, The authentication mechanism to use, GSSAPI or MONGODB-CR
 *
 * @param {String} username username.
 * @param {String} password password.
 * @param {Object} [options] the options
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from authentication or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.authenticate = function(username, password, options, callback***REMOVED*** {
  var self = this;

  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Set default mechanism
  if(!options.authMechanism***REMOVED*** {
    options.authMechanism = 'MONGODB-CR';
***REMOVED*** else if(options.authMechanism != 'GSSAPI' 
    && options.authMechanism != 'MONGODB-CR'
    && options.authMechanism != 'MONGODB-X509'
    && options.authMechanism != 'PLAIN'***REMOVED*** {
      return callback(new Error("only GSSAPI, PLAIN, MONGODB-X509 or MONGODB-CR is supported by authMechanism"***REMOVED******REMOVED***;
***REMOVED***

  // the default db to authenticate against is 'this'
  // if authententicate is called from a retry context, it may be another one, like admin
  var authdb = options.authdb ? options.authdb : self.databaseName;
  authdb = options.authSource ? options.authSource : authdb;

  // Callback
  var _callback = function(err, result***REMOVED*** {
    if(self.listeners("authenticated"***REMOVED***.length > 9***REMOVED*** {
      self.emit("authenticated", err, result***REMOVED***;
  ***REMOVED***

    // Return to caller
    callback(err, result***REMOVED***;
***REMOVED***

  // If classic auth delegate to auth command
  if(options.authMechanism == 'MONGODB-CR'***REMOVED*** {
    mongodb_cr_authenticate(self, username, password, authdb, options, _callback***REMOVED***;
***REMOVED*** else if(options.authMechanism == 'PLAIN'***REMOVED*** {
    mongodb_plain_authenticate(self, username, password, options, _callback***REMOVED***;
***REMOVED*** else if(options.authMechanism == 'MONGODB-X509'***REMOVED*** {
    mongodb_x509_authenticate(self, username, password, options, _callback***REMOVED***;
***REMOVED*** else if(options.authMechanism == 'GSSAPI'***REMOVED*** {
    //
    // Kerberos library is not installed, throw and error
    if(hasKerberos == false***REMOVED*** {
      console.log("========================================================================================"***REMOVED***;
      console.log("=  Please make sure that you install the Kerberos library to use GSSAPI                ="***REMOVED***;
      console.log("=                                                                                      ="***REMOVED***;
      console.log("=  npm install -g kerberos                                                             ="***REMOVED***;
      console.log("=                                                                                      ="***REMOVED***;
      console.log("=  The Kerberos package is not installed by default for simplicities sake              ="***REMOVED***;
      console.log("=  and needs to be global install                                                      ="***REMOVED***;
      console.log("========================================================================================"***REMOVED***;
      throw new Error("Kerberos library not installed"***REMOVED***;
  ***REMOVED***

    if(process.platform == 'win32'***REMOVED*** {
      mongodb_sspi_authenticate(self, username, password, authdb, options, _callback***REMOVED***;
  ***REMOVED*** else {
      // We have the kerberos library, execute auth process
      mongodb_gssapi_authenticate(self, username, password, authdb, options, _callback***REMOVED***;      
  ***REMOVED***
***REMOVED***
};

/**
 * Add a user to the database.
 *
 * Options
 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *  - **customData**, (Object, default:{}***REMOVED*** custom data associated with the user (only Mongodb 2.6 or higher***REMOVED***
 *  - **roles**, (Array, default:[]***REMOVED*** roles associated with the created user (only Mongodb 2.6 or higher***REMOVED***
 *
 * @param {String} username username.
 * @param {String} password password.
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from addUser or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.addUser = function(username, password, options, callback***REMOVED*** {
  // Checkout a write connection to get the server capabilities
  var connection = this.serverConfig.checkoutWriter(***REMOVED***;
  if(connection != null 
    && connection.serverCapabilities != null 
    && connection.serverCapabilities.hasAuthCommands***REMOVED*** {
      return _executeAuthCreateUserCommand(this, username, password, options, callback***REMOVED***;
***REMOVED***

  // Unpack the parameters
  var self = this;
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // Get the error options
  var errorOptions = _getWriteConcern(this, options***REMOVED***;
  errorOptions.w = errorOptions.w == null ? 1 : errorOptions.w;
  // Use node md5 generator
  var md5 = crypto.createHash('md5'***REMOVED***;
  // Generate keys used for authentication
  md5.update(username + ":mongo:" + password***REMOVED***;
  var userPassword = md5.digest('hex'***REMOVED***;
  // Fetch a user collection
  var collection = this.collection(DbCommand.SYSTEM_USER_COLLECTION***REMOVED***;
  // Check if we are inserting the first user
  collection.count({}, function(err, count***REMOVED*** {
    // We got an error (f.ex not authorized***REMOVED***
    if(err != null***REMOVED*** return callback(err, null***REMOVED***;
    // Check if the user exists and update i
    collection.find({user: username}, {dbName: options['dbName']}***REMOVED***.toArray(function(err, documents***REMOVED*** {
      // We got an error (f.ex not authorized***REMOVED***
      if(err != null***REMOVED*** return callback(err, null***REMOVED***;
      // Add command keys
      var commandOptions = errorOptions;
      commandOptions.dbName = options['dbName'];
      commandOptions.upsert = true;

      // We have a user, let's update the password or upsert if not
      collection.update({user: username},{$set: {user: username, pwd: userPassword}}, commandOptions, function(err, results, full***REMOVED*** {
        if(count == 0 && err***REMOVED*** {
          callback(null, [{user:username, pwd:userPassword}]***REMOVED***;
      ***REMOVED*** else if(err***REMOVED*** {
          callback(err, null***REMOVED***
      ***REMOVED*** else {
          callback(null, [{user:username, pwd:userPassword}]***REMOVED***;
      ***REMOVED***
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * @ignore
 */
var _executeAuthCreateUserCommand = function(self, username, password, options, callback***REMOVED*** {
  // Special case where there is no password ($external users***REMOVED***
  if(typeof username == 'string' 
    && password != null && typeof password == 'object'***REMOVED*** {
    callback = options;
    options = password;
    password = null;
***REMOVED***

  // Unpack all options
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***  

  // Error out if we digestPassword set
  if(options.digestPassword != null***REMOVED*** {
    throw utils.toError("The digestPassword option is not supported via add_user. Please use db.command('createUser', ...***REMOVED*** instead for this option."***REMOVED***;
***REMOVED***

  // Get additional values
  var customData = options.customData != null ? options.customData : {};
  var roles = Array.isArray(options.roles***REMOVED*** ? options.roles : [];
  var maxTimeMS = typeof options.maxTimeMS == 'number' ? options.maxTimeMS : null;

  // If not roles defined print deprecated message
  if(roles.length == 0***REMOVED*** {
    console.log("Creating a user without roles is deprecated in MongoDB >= 2.6"***REMOVED***;
***REMOVED***

  // Get the error options
  var writeConcern = _getWriteConcern(self, options***REMOVED***;
  var commandOptions = {writeCommand:true};
  if(options['dbName']***REMOVED*** commandOptions.dbName = options['dbName'];

  // Add maxTimeMS to options if set
  if(maxTimeMS != null***REMOVED*** commandOptions.maxTimeMS = maxTimeMS;

  // Check the db name and add roles if needed
  if((self.databaseName.toLowerCase(***REMOVED*** == 'admin' || options.dbName == 'admin'***REMOVED*** && !Array.isArray(options.roles***REMOVED******REMOVED*** {
    roles = ['root']
***REMOVED*** else if(!Array.isArray(options.roles***REMOVED******REMOVED*** {
    roles = ['dbOwner']
***REMOVED***

  // Build the command to execute
  var command = {
      createUser: username
    , customData: customData
    , roles: roles
    , digestPassword:false
    , writeConcern: writeConcern
***REMOVED***

  // Use node md5 generator
  var md5 = crypto.createHash('md5'***REMOVED***;
  // Generate keys used for authentication
  md5.update(username + ":mongo:" + password***REMOVED***;
  var userPassword = md5.digest('hex'***REMOVED***;

  // No password
  if(typeof password == 'string'***REMOVED*** {
    command.pwd = userPassword;
***REMOVED***

  // Execute the command
  self.command(command, commandOptions, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    callback(!result.ok ? utils.toError("Failed to add user " + username***REMOVED*** : null
      , result.ok ? [{user: username, pwd: ''}] : null***REMOVED***;
***REMOVED******REMOVED***
}

/**
 * Remove a user from a database
 *
 * Options
 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *
 * @param {String} username username.
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from removeUser or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.removeUser = function(username, options, callback***REMOVED*** {
  // Checkout a write connection to get the server capabilities
  var connection = this.serverConfig.checkoutWriter(***REMOVED***;
  if(connection != null && connection.serverCapabilities != null && connection.serverCapabilities.hasAuthCommands***REMOVED*** {
    return _executeAuthRemoveUserCommand(this, username, options, callback***REMOVED***;
***REMOVED***

  // Unpack the parameters
  var self = this;
  var args = Array.prototype.slice.call(arguments, 1***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // Figure out the safe mode settings
  var safe = self.safe != null && self.safe == false ? {w: 1} : self.safe;
  // Override with options passed in if applicable
  safe = options != null && options['safe'] != null ? options['safe'] : safe;
  // Ensure it's at least set to safe
  safe = safe == null ? {w: 1} : safe;

  // Fetch a user collection
  var collection = this.collection(DbCommand.SYSTEM_USER_COLLECTION***REMOVED***;
  collection.findOne({user: username}, {dbName: options['dbName']}, function(err, user***REMOVED*** {
    if(user != null***REMOVED*** {
      // Add command keys
      var commandOptions = safe;
      commandOptions.dbName = options['dbName'];

      collection.remove({user: username}, commandOptions, function(err, result***REMOVED*** {
        callback(err, true***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED*** else {
      callback(err, false***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;
};

var _executeAuthRemoveUserCommand = function(self, username, options, callback***REMOVED*** {
  // Unpack all options
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Get the error options
  var writeConcern = _getWriteConcern(self, options***REMOVED***;
  var commandOptions = {writeCommand:true};
  if(options['dbName']***REMOVED*** commandOptions.dbName = options['dbName'];

  // Get additional values
  var maxTimeMS = typeof options.maxTimeMS == 'number' ? options.maxTimeMS : null;

  // Add maxTimeMS to options if set
  if(maxTimeMS != null***REMOVED*** commandOptions.maxTimeMS = maxTimeMS;

  // Build the command to execute
  var command = {
      dropUser: username
    , writeConcern: writeConcern
***REMOVED***

  // Execute the command
  self.command(command, commandOptions, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    callback(null, result.ok ? true : false***REMOVED***;
***REMOVED******REMOVED***
}

/**
 * Creates a collection on a server pre-allocating space, need to create f.ex capped collections.
 *
 * Options
*  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *  - **serializeFunctions** {Boolean, default:false}, serialize functions on the document.
 *  - **raw** {Boolean, default:false}, perform all operations using raw bson objects.
 *  - **pkFactory** {Object}, object overriding the basic ObjectID primary key generation.
 *  - **capped** {Boolean, default:false}, create a capped collection.
 *  - **size** {Number}, the size of the capped collection in bytes.
 *  - **max** {Number}, the maximum number of documents in the capped collection.
 *  - **autoIndexId** {Boolean, default:true}, create an index on the _id field of the document, True by default on MongoDB 2.2 or higher off for version < 2.2.
 *  - **readPreference** {String}, the preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *  - **strict**, (Boolean, default:false***REMOVED*** throws an error if collection already exists
 *
 * @param {String} collectionName the collection name we wish to access.
 * @param {Object} [options] returns option results.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from createCollection or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.createCollection = function(collectionName, options, callback***REMOVED*** {
  var self = this;
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Figure out the safe mode settings
  var safe = self.safe != null && self.safe == false ? {w: 1} : self.safe;
  // Override with options passed in if applicable
  safe = options != null && options['safe'] != null ? options['safe'] : safe;
  // Ensure it's at least set to safe
  safe = safe == null ? {w: 1} : safe;

  // Check if we have the name
  this.collectionNames(collectionName, function(err, collections***REMOVED*** {
    if(err != null***REMOVED*** return callback(err, null***REMOVED***;

    var found = false;
    collections.forEach(function(collection***REMOVED*** {
      if(collection.name == self.databaseName + "." + collectionName***REMOVED*** found = true;
  ***REMOVED******REMOVED***;

    // If the collection exists either throw an exception (if db in safe mode***REMOVED*** or return the existing collection
    if(found && options && options.strict***REMOVED*** {
      return callback(new Error("Collection " + collectionName + " already exists. Currently in strict mode."***REMOVED***, null***REMOVED***;
  ***REMOVED*** else if(found***REMOVED***{
      try {
        var collection = new Collection(self, collectionName, self.pkFactory, options***REMOVED***;
    ***REMOVED*** catch(err***REMOVED*** {
        return callback(err, null***REMOVED***;
    ***REMOVED***
      return callback(null, collection***REMOVED***;
  ***REMOVED***

    // logout command
    var cmd = {'create':collectionName};

    for(var name in options***REMOVED*** {
      if(options[name] != null && options[name].constructor != Function***REMOVED*** cmd[name] = options[name];
  ***REMOVED***

    // Execute the command
    self.command(cmd, options, function(err, result***REMOVED*** {
      if(err***REMOVED*** return callback(err, null***REMOVED***;
      try {
        callback(null, new Collection(self, collectionName, self.pkFactory, options***REMOVED******REMOVED***;
    ***REMOVED*** catch(err***REMOVED*** {
        callback(utils.toError(err***REMOVED***, null***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

var _getReadConcern = function(self, options***REMOVED*** {
  if(options.readPreference***REMOVED*** return options.readPreference;
  if(self.readPreference***REMOVED*** return self.readPreference;
  return 'primary';
}

/**
 * Execute a command hash against MongoDB. This lets you acess any commands not available through the api on the server.
 *
 * Options
 *  - **readPreference** {String}, the preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *  - **maxTimeMS** {Number}, number of milliseconds to wait before aborting the query.
 *  - **ignoreCommandFilter** {Boolean}, overrides the default redirection of certain commands to primary.
 *  - **writeCommand** {Boolean, default: false}, signals this is a write command and to ignore read preferences
 *  - **checkKeys** {Boolean, default: false}, overrides the default not to check the key names for the command
 *
 * @param {Object} selector the command hash to send to the server, ex: {ping:1}.
 * @param {Object} [options] additional options for the command.
 * @param {Function} callback this will be called after executing this method. The command always return the whole result of the command as the second parameter.
 * @return {null}
 * @api public
 */
Db.prototype.command = function(selector, options, callback***REMOVED*** {
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Ignore command preference (I know what I'm doing***REMOVED***
  var ignoreCommandFilter = options.ignoreCommandFilter ? options.ignoreCommandFilter : false;

  // Get read preference if we set one
  var readPreference = _getReadConcern(this, options***REMOVED***;

  // Ensure only commands who support read Prefrences are exeuted otherwise override and use Primary
  if(readPreference != false && ignoreCommandFilter == false***REMOVED*** {
    if(selector['group'] || selector['aggregate'] || selector['collStats'] || selector['dbStats']
      || selector['count'] || selector['distinct'] || selector['geoNear'] || selector['geoSearch'] 
      || selector['geoWalk'] || selector['text'] || selector['cursorInfo']
      || selector['parallelCollectionScan']
      || (selector['mapreduce'] && (selector.out == 'inline' || selector.out.inline***REMOVED******REMOVED******REMOVED*** {
      // Set the read preference
      options.readPreference = readPreference;
  ***REMOVED*** else {
      options.readPreference = ReadPreference.PRIMARY;
  ***REMOVED***
***REMOVED*** else if(readPreference != false***REMOVED*** {
    options.readPreference = readPreference;
***REMOVED***

  // Add the maxTimeMS option to the command if specified
  if(typeof options.maxTimeMS == 'number'***REMOVED*** {
    selector.maxTimeMS = options.maxTimeMS    
***REMOVED***

  // Command options
  var command_options = {};

  // Do we have an override for checkKeys
  if(typeof options['checkKeys'] == 'boolean'***REMOVED*** command_options['checkKeys'] = options['checkKeys'];
  command_options['checkKeys'] = typeof options['checkKeys'] == 'boolean' ? options['checkKeys'] : false;
  if(typeof options['serializeFunctions'] == 'boolean'***REMOVED*** command_options['serializeFunctions'] = options['serializeFunctions'];
  if(options['dbName']***REMOVED*** command_options['dbName'] = options['dbName'];

  // If we have a write command, remove readPreference as an option
  if((options.writeCommand 
    || selector['findAndModify'] 
    || selector['insert'] || selector['update'] || selector['delete']
    || selector['createUser'] || selector['updateUser'] || selector['removeUser']***REMOVED***
    && options.readPreference***REMOVED*** {
    delete options['readPreference'];
***REMOVED***

  // Add a write concern if we have passed in any
  if(options.w || options.wtimeout || options.j || options.fsync || options.safe***REMOVED*** {
    selector.writeConcern = {};
    if(options.safe***REMOVED*** selector.writeConcern.w = 1;
    if(options.w***REMOVED*** selector.writeConcern.w = options.w;
    if(options.wtimeout***REMOVED*** selector.writeConcern.wtimeout = options.wtimeout;
    if(options.j***REMOVED*** selector.writeConcern.j = options.j;
    if(options.fsync***REMOVED*** selector.writeConcern.fsync = options.fsync;
***REMOVED***

  // If we have an actual writeConcern object override
  if(options.writeConcern***REMOVED*** {
    selector.writeConcern = writeConcern;
***REMOVED***

  // Check if we need to set slaveOk
  if(command_options.readPreference != 'primary'***REMOVED*** 
    command_options.slaveOk = true;

  // Execution db
  var execDb = typeof options.auth == 'string' ? this.db(options.auth***REMOVED*** : this;
  execDb = typeof options.authdb == 'string' ? this.db(options.authdb***REMOVED*** : execDb;

  // Execute a query command
  this._executeQueryCommand(DbCommand.createDbSlaveOkCommand(execDb, selector, command_options***REMOVED***, options, function(err, results, connection***REMOVED*** {
    if(options.returnConnection***REMOVED*** {
      if(err***REMOVED*** return callback(err, null, connection***REMOVED***;
      if(results.documents[0].errmsg***REMOVED*** 
        return callback(utils.toError(results.documents[0]***REMOVED***, null, connection***REMOVED***;
      callback(null, results.documents[0], connection***REMOVED***;
  ***REMOVED*** else {
      if(err***REMOVED*** return callback(err, null***REMOVED***;
      if(results.documents[0].errmsg***REMOVED*** 
        return callback(utils.toError(results.documents[0]***REMOVED***, null***REMOVED***;
      callback(null, results.documents[0]***REMOVED***;      
  ***REMOVED***
***REMOVED******REMOVED***;
};

/**
 * Drop a collection from the database, removing it permanently. New accesses will create a new collection.
 *
 * @param {String} collectionName the name of the collection we wish to drop.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from dropCollection or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.dropCollection = function(collectionName, callback***REMOVED*** {
  var self = this;
  callback || (callback = function(***REMOVED***{}***REMOVED***;

  // Command to execute
  var cmd = {'drop':collectionName}

  // Execute the command
  this.command(cmd, {}, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    if(result.ok***REMOVED*** return callback(null, true***REMOVED***;
    callback(null, false***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Rename a collection.
 * 
 * Options
 *  - **dropTarget** {Boolean, default:false}, drop the target name collection if it previously exists.
 *
 * @param {String} fromCollection the name of the current collection we wish to rename.
 * @param {String} toCollection the new name of the collection.
 * @param {Object} [options] returns option results.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from renameCollection or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.renameCollection = function(fromCollection, toCollection, options, callback***REMOVED*** {
  var self = this;

  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {}
***REMOVED***

  // Add return new collection
  options.new_collection = true;

  // Execute using the collection method
  this.collection(fromCollection***REMOVED***.rename(toCollection, options, callback***REMOVED***;
};

/**
 * Runs a command on the database.
 * @ignore
 * @api private
 */
Db.prototype.executeDbCommand = function(command_hash, options, callback***REMOVED*** {
  if(callback == null***REMOVED*** { callback = options; options = {}; }
  this._executeQueryCommand(DbCommand.createDbSlaveOkCommand(this, command_hash, options***REMOVED***, options, function(err, result***REMOVED*** {
    if(callback***REMOVED*** callback(err, result***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Runs a command on the database as admin.
 * @ignore
 * @api private
 */
Db.prototype.executeDbAdminCommand = function(command_hash, options, callback***REMOVED*** {
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {}
***REMOVED***

  if(options.readPreference***REMOVED*** {
    options.readPreference = options.readPreference;
***REMOVED***

  this._executeQueryCommand(DbCommand.createAdminDbCommand(this, command_hash***REMOVED***, options, function(err, result***REMOVED*** {
    if(callback***REMOVED*** callback(err, result***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Creates an index on the collection.
 *
 * Options
*  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *  - **unique** {Boolean, default:false}, creates an unique index.
 *  - **sparse** {Boolean, default:false}, creates a sparse index.
 *  - **background** {Boolean, default:false}, creates the index in the background, yielding whenever possible.
 *  - **dropDups** {Boolean, default:false}, a unique index cannot be created on a key that has pre-existing duplicate values. If you would like to create the index anyway, keeping the first document the database indexes and deleting all subsequent documents that have duplicate value
 *  - **min** {Number}, for geospatial indexes set the lower bound for the co-ordinates.
 *  - **max** {Number}, for geospatial indexes set the high bound for the co-ordinates.
 *  - **v** {Number}, specify the format version of the indexes.
 *  - **expireAfterSeconds** {Number}, allows you to expire data on indexes applied to a data (MongoDB 2.2 or higher***REMOVED***
 *  - **name** {String}, override the autogenerated index name (useful if the resulting name is larger than 128 bytes***REMOVED***
 *
 * @param {String} collectionName name of the collection to create the index on.
 * @param {Object} fieldOrSpec fieldOrSpec that defines the index.
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from createIndex or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.createIndex = function(collectionName, fieldOrSpec, options, callback***REMOVED*** {
  var self = this;
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** || {} : {};
  options = typeof callback === 'function' ? options : callback;
  options = options == null ? {} : options;

  // Get the error options
  var writeConcern = _getWriteConcern(self, options***REMOVED***;
  // Ensure we have a callback
  if(_hasWriteConcern(writeConcern***REMOVED*** && typeof callback != 'function'***REMOVED*** {
    throw new Error("Cannot use a writeConcern without a provided callback"***REMOVED***;
***REMOVED***

  // Attempt to run using createIndexes command
  createIndexUsingCreateIndexes(self, collectionName, fieldOrSpec, options, function(err, result***REMOVED*** {
    if(err == null***REMOVED*** {
      return callback(err, result***REMOVED***;
  ***REMOVED***
    
    // Create command
    var command = createCreateIndexCommand(self, collectionName, fieldOrSpec, options***REMOVED***;
    // Default command options
    var commandOptions = {};

    // If we have error conditions set handle them
    if(_hasWriteConcern(writeConcern***REMOVED*** && typeof callback == 'function'***REMOVED*** {
      // Set safe option
      commandOptions['safe'] = writeConcern;
      // If we have an error option
      if(typeof writeConcern == 'object'***REMOVED*** {
        var keys = Object.keys(writeConcern***REMOVED***;
        for(var i = 0; i < keys.length; i++***REMOVED*** {
          commandOptions[keys[i]] = writeConcern[keys[i]];
      ***REMOVED***
    ***REMOVED***

      // Execute insert command
      self._executeInsertCommand(command, commandOptions, function(err, result***REMOVED*** {
        if(err != null***REMOVED*** return callback(err, null***REMOVED***;

        result = result && result.documents;
        if (result[0].err***REMOVED*** {
          callback(utils.toError(result[0]***REMOVED******REMOVED***;
      ***REMOVED*** else {
          callback(null, command.documents[0].name***REMOVED***;
      ***REMOVED***
    ***REMOVED******REMOVED***;
  ***REMOVED*** else {
      // Execute insert command
      var result = self._executeInsertCommand(command, commandOptions, function(***REMOVED*** {}***REMOVED***;
      // If no callback just return
      if(!callback***REMOVED*** return;
      // If error return error
      if(result instanceof Error***REMOVED*** {
        return callback(result***REMOVED***;
    ***REMOVED***
      // Otherwise just return
      return callback(null, null***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;
};

var createCreateIndexCommand = function(db, collectionName, fieldOrSpec, options***REMOVED*** {
  var indexParameters = utils.parseIndexOptions(fieldOrSpec***REMOVED***;
  var fieldHash = indexParameters.fieldHash;
  var keys = indexParameters.keys;

  // Generate the index name
  var indexName = typeof options.name == 'string'
    ? options.name
    : indexParameters.name;

  var selector = {
    'ns': db.databaseName + "." + collectionName,
    'key': fieldHash,
    'name': indexName
***REMOVED***

  // Ensure we have a correct finalUnique
  var finalUnique = options == null || 'object' === typeof options
    ? false
    : options;

  // Set up options
  options = options == null || typeof options == 'boolean'
    ? {}
    : options;

  // Add all the options
  var keys = Object.keys(options***REMOVED***;
  for(var i = 0; i < keys.length; i++***REMOVED*** {
    if(indexOptions.indexOf(keys[i]***REMOVED*** != -1***REMOVED*** {
      selector[keys[i]] = options[keys[i]];
  ***REMOVED***
***REMOVED***

  if(selector['unique'] == null***REMOVED***
    selector['unique'] = finalUnique;

  var name = db.databaseName + "." + DbCommand.SYSTEM_INDEX_COLLECTION;
  var cmd = new InsertCommand(db, name, false***REMOVED***;
  return cmd.add(selector***REMOVED***;
}

var createIndexUsingCreateIndexes = function(self, collectionName, fieldOrSpec, options, callback***REMOVED*** {
  // Build the index
  var indexParameters = utils.parseIndexOptions(fieldOrSpec***REMOVED***;
  // Generate the index name
  var indexName = typeof options.name == 'string'
    ? options.name
    : indexParameters.name;

  // Set up the index
  var indexes = [{
      name: indexName
    , key: indexParameters.fieldHash
***REMOVED***];

  // merge all the options
  for(var name in options***REMOVED*** {
    if(indexOptions.indexOf(name***REMOVED*** != -1***REMOVED*** {
      indexes[0][name] = options[name];      
  ***REMOVED***
***REMOVED***

  // Create command
  var command = {createIndexes: collectionName, indexes: indexes};
  // Build the command
  self.command(command, options, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    if(result.ok == 0***REMOVED*** {
      return callback(utils.toError(result***REMOVED***, null***REMOVED***;
  ***REMOVED***

    // Return the indexName for backward compatibility
    callback(null, indexName***REMOVED***;
***REMOVED******REMOVED***;
}

/**
 * Ensures that an index exists, if it does not it creates it
 *
 * Options
 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowledgement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *  - **unique** {Boolean, default:false}, creates an unique index.
 *  - **sparse** {Boolean, default:false}, creates a sparse index.
 *  - **background** {Boolean, default:false}, creates the index in the background, yielding whenever possible.
 *  - **dropDups** {Boolean, default:false}, a unique index cannot be created on a key that has pre-existing duplicate values. If you would like to create the index anyway, keeping the first document the database indexes and deleting all subsequent documents that have duplicate value
 *  - **min** {Number}, for geospatial indexes set the lower bound for the co-ordinates.
 *  - **max** {Number}, for geospatial indexes set the high bound for the co-ordinates.
 *  - **v** {Number}, specify the format version of the indexes.
 *  - **expireAfterSeconds** {Number}, allows you to expire data on indexes applied to a data (MongoDB 2.2 or higher***REMOVED***
 *  - **name** {String}, override the autogenerated index name (useful if the resulting name is larger than 128 bytes***REMOVED***
 *
 * @param {String} collectionName name of the collection to create the index on.
 * @param {Object} fieldOrSpec fieldOrSpec that defines the index.
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from ensureIndex or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.ensureIndex = function(collectionName, fieldOrSpec, options, callback***REMOVED*** {
  var self = this;

  if(typeof callback === 'undefined' && typeof options === 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Ensure non empty options
  options = options || {};

  // Get the error options
  var writeConcern = _getWriteConcern(this, options***REMOVED***;
  // Make sure we don't try to do a write concern without a callback
  if(_hasWriteConcern(writeConcern***REMOVED*** && callback == null***REMOVED***
    throw new Error("Cannot use a writeConcern without a provided callback"***REMOVED***;
  
  // Create command
  var command = createCreateIndexCommand(this, collectionName, fieldOrSpec, options***REMOVED***;
  var index_name = command.documents[0].name;

  // Default command options
  var commandOptions = {};
  // Check if the index allready exists
  this.indexInformation(collectionName, writeConcern, function(err, collectionInfo***REMOVED*** {
    if(err != null***REMOVED*** return callback(err, null***REMOVED***;
    // If the index does not exist, create it
    if(!collectionInfo[index_name]***REMOVED***  {
      self.createIndex(collectionName, fieldOrSpec, options, callback***REMOVED***;
  ***REMOVED*** else {
      if(typeof callback === 'function'***REMOVED*** return callback(null, index_name***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;
};

/**
 * Returns the information available on allocated cursors.
 *
 * Options
 *  - **readPreference** {String}, the preferred read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from cursorInfo or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.cursorInfo = function(options, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 0***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // cursorInfo command
  var cmd = {'cursorInfo':1};

  // Execute the command
  this.command(cmd, options, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    callback(null, result***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Drop an index on a collection.
 *
 * @param {String} collectionName the name of the collection where the command will drop an index.
 * @param {String} indexName name of the index to drop.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from dropIndex or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.dropIndex = function(collectionName, indexName, options, callback***REMOVED*** {  
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // Delete index command
  var cmd = {'deleteIndexes':collectionName, 'index':indexName};  

  // Execute command
  this.command(cmd, options, function(err, result***REMOVED*** {
    if(callback == null***REMOVED*** return;
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    callback(null, result***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Reindex all indexes on the collection
 * Warning: reIndex is a blocking operation (indexes are rebuilt in the foreground***REMOVED*** and will be slow for large collections.
 *
 * @param {String} collectionName the name of the collection.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from reIndex or null if an error occurred.
 * @api public
**/
Db.prototype.reIndex = function(collectionName, options, callback***REMOVED*** {
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***
  
  // Reindex
  var cmd = {'reIndex':collectionName};

  // Execute the command
  this.command(cmd, options, function(err, result***REMOVED*** {
    if(callback == null***REMOVED*** return;
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    callback(null, result.ok ? true : false***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Retrieves this collections index info.
 *
 * Options
 *  - **full** {Boolean, default:false}, returns the full raw index information.
 *  - **readPreference** {String}, the preferred read preference ((Server.PRIMARY, Server.PRIMARY_PREFERRED, Server.SECONDARY, Server.SECONDARY_PREFERRED, Server.NEAREST***REMOVED***.
 *
 * @param {String} collectionName the name of the collection.
 * @param {Object} [options] additional options during update.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from indexInformation or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.indexInformation = function(collectionName, options, callback***REMOVED*** {
  if(typeof callback === 'undefined'***REMOVED*** {
    if(typeof options === 'undefined'***REMOVED*** {
      callback = collectionName;
      collectionName = null;
  ***REMOVED*** else {
      callback = options;
  ***REMOVED***
    options = {};
***REMOVED***

  // If we specified full information
  var full = options['full'] == null ? false : options['full'];
  // Build selector for the indexes
  var selector = collectionName != null ? {ns: (this.databaseName + "." + collectionName***REMOVED***} : {};

  // Get read preference if we set one
  var readPreference = _getReadConcern(this, options***REMOVED***;

  // Iterate through all the fields of the index
  var collection = this.collection(DbCommand.SYSTEM_INDEX_COLLECTION***REMOVED***;
  // Perform the find for the collection
  collection.find(selector***REMOVED***.setReadPreference(readPreference***REMOVED***.toArray(function(err, indexes***REMOVED*** {
    if(err != null***REMOVED*** return callback(err, null***REMOVED***;
    // Contains all the information
    var info = {};

    // if full defined just return all the indexes directly
    if(full***REMOVED*** return callback(null, indexes***REMOVED***;

    // Process all the indexes
    for(var i = 0; i < indexes.length; i++***REMOVED*** {
      var index = indexes[i];
      // Let's unpack the object
      info[index.name] = [];
      for(var name in index.key***REMOVED*** {
        info[index.name].push([name, index.key[name]]***REMOVED***;
    ***REMOVED***
  ***REMOVED***

    // Return all the indexes
    callback(null, info***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Drop a database.
 *
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from dropDatabase or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.dropDatabase = function(options, callback***REMOVED*** {
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Reindex
  var cmd = {'dropDatabase':1};

  // Execute the command
  this.command(cmd, options, function(err, result***REMOVED*** {
    if(callback == null***REMOVED*** return;
    if(err***REMOVED*** return callback(err, null***REMOVED***;
    callback(null, result.ok ? true : false***REMOVED***;
***REMOVED******REMOVED***;
}

/**
 * Get all the db statistics.
 *
 * Options
 *  - **scale** {Number}, divide the returned sizes by scale value.
 *  - **readPreference** {String}, the preferred read preference ((Server.PRIMARY, Server.PRIMARY_PREFERRED, Server.SECONDARY, Server.SECONDARY_PREFERRED, Server.NEAREST***REMOVED***.
 *
 * @param {Objects} [options] options for the stats command
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the results from stats or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.prototype.stats = function stats(options, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 0***REMOVED***;
  callback = args.pop(***REMOVED***;
  // Fetch all commands
  options = args.length ? args.shift(***REMOVED*** || {} : {};

  // Build command object
  var commandObject = {
    dbStats:true
***REMOVED***;

  // Check if we have the scale value
  if(options['scale'] != null***REMOVED*** commandObject['scale'] = options['scale'];

  // Execute the command
  this.command(commandObject, options, callback***REMOVED***;
}

/**
 * @ignore
 */
var bindToCurrentDomain = function(callback***REMOVED*** {
  var domain = process.domain;
  if(domain == null || callback == null***REMOVED*** {
    return callback;
***REMOVED*** else {
    return domain.bind(callback***REMOVED***;
***REMOVED***
}

/**
 * @ignore
 */
var __executeQueryCommand = function(self, db_command, options, callback***REMOVED*** {
  // Options unpacking
  var readPreference = options.readPreference != null ? options.readPreference : 'primary';
  var onAll = options['onAll'] != null ? options['onAll'] : false;
  var specifiedConnection = options['connection'] != null ? options['connection'] : null;
  var raw = typeof options.raw == 'boolean' ? options.raw : false;

  // Correct readPreference preference to default primary if set to false, null or primary
  if(!(typeof readPreference == 'object'***REMOVED*** && readPreference._type == 'ReadPreference'***REMOVED*** {
    readPreference = (readPreference == null || readPreference == 'primary' || readPreference == false***REMOVED*** ? ReadPreference.PRIMARY : readPreference;
    if(!ReadPreference.isValid(readPreference***REMOVED******REMOVED*** return callback(new Error("Illegal readPreference mode specified, " + JSON.stringify(readPreference***REMOVED******REMOVED******REMOVED***;
***REMOVED*** else if(typeof readPreference == 'object' && readPreference._type == 'ReadPreference'***REMOVED*** {
    if(!readPreference.isValid(***REMOVED******REMOVED*** return callback(new Error("Illegal readPreference mode specified, " + JSON.stringify(readPreference***REMOVED******REMOVED******REMOVED***;
***REMOVED***

  // If we have a read preference set and we are a mongos pass the read preference on to the mongos instance,
  if(self.serverConfig.isMongos(***REMOVED*** && readPreference != null && readPreference != 'primary'***REMOVED*** {
    db_command.setMongosReadPreference(readPreference***REMOVED***;
***REMOVED***

  // If we got a callback object
  if(typeof callback === 'function' && !onAll***REMOVED*** {
    callback = bindToCurrentDomain(callback***REMOVED***;    
    // Override connection if we passed in a specific connection
    var connection = specifiedConnection != null ? specifiedConnection : null;

    if(connection instanceof Error***REMOVED*** return callback(connection, null***REMOVED***;

    // Fetch either a reader or writer dependent on the specified readPreference option if no connection
    // was passed in
    if(connection == null***REMOVED*** {
      connection = self.serverConfig.checkoutReader(readPreference***REMOVED***;
  ***REMOVED***

    if(connection == null***REMOVED*** {
      return callback(new Error("no open connections"***REMOVED******REMOVED***;
  ***REMOVED*** else if(connection instanceof Error || connection['message'] != null***REMOVED*** {
      return callback(connection***REMOVED***;
  ***REMOVED***

    // Exhaust Option
    var exhaust = options.exhaust || false;
    
    // Register the handler in the data structure
    self.serverConfig._registerHandler(db_command, raw, connection, exhaust, callback***REMOVED***;
    
    // Write the message out and handle any errors if there are any
    connection.write(db_command, function(err***REMOVED*** {
      if(err != null***REMOVED*** {
        // Call the handler with an error
        if(Array.isArray(db_command***REMOVED******REMOVED***
          self.serverConfig._callHandler(db_command[0].getRequestId(***REMOVED***, null, err***REMOVED***;
        else
          self.serverConfig._callHandler(db_command.getRequestId(***REMOVED***, null, err***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED*** else if(typeof callback === 'function' && onAll***REMOVED*** {
    callback = bindToCurrentDomain(callback***REMOVED***;
    var connections = self.serverConfig.allRawConnections(***REMOVED***;
    var numberOfEntries = connections.length;
    // Go through all the connections
    for(var i = 0; i < connections.length; i++***REMOVED*** {
      // Fetch a connection
      var connection = connections[i];

      // Ensure we have a valid connection
      if(connection == null***REMOVED*** {
        return callback(new Error("no open connections"***REMOVED******REMOVED***;
    ***REMOVED*** else if(connection instanceof Error***REMOVED*** {
        return callback(connection***REMOVED***;
    ***REMOVED***

      // Register the handler in the data structure
      self.serverConfig._registerHandler(db_command, raw, connection, callback***REMOVED***;

      // Write the message out
      connection.write(db_command, function(err***REMOVED*** {
        // Adjust the number of entries we need to process
        numberOfEntries = numberOfEntries - 1;
        // Remove listener
        if(err != null***REMOVED*** {
          // Clean up listener and return error
          self.serverConfig._removeHandler(db_command.getRequestId(***REMOVED******REMOVED***;
      ***REMOVED***

        // No more entries to process callback with the error
        if(numberOfEntries <= 0***REMOVED*** {
          callback(err***REMOVED***;
      ***REMOVED***
    ***REMOVED******REMOVED***;

      // Update the db_command request id
      db_command.updateRequestId(***REMOVED***;
  ***REMOVED***
***REMOVED*** else {
    // Fetch either a reader or writer dependent on the specified read option
    var connection = self.serverConfig.checkoutReader(readPreference***REMOVED***;
    // Override connection if needed
    connection = specifiedConnection != null ? specifiedConnection : connection;
    // Ensure we have a valid connection
    if(connection == null || connection instanceof Error || connection['message'] != null***REMOVED*** return null;
    // Write the message out
    connection.write(db_command, function(err***REMOVED*** {
      if(err != null***REMOVED*** {
        // Emit the error
        self.emit("error", err***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED***
};

/**
 * Execute db query command (not safe***REMOVED***
 * @ignore
 * @api private
 */
Db.prototype._executeQueryCommand = function(db_command, options, callback***REMOVED*** {
  var self = this;

  // Unpack the parameters
  if(typeof options === 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // fast fail option used for HA, no retry
  var failFast = options['failFast'] != null
    ? options['failFast']
    : false;

  // Check if the user force closed the command
  if(this._applicationClosed***REMOVED*** {
    var err = new Error("db closed by application"***REMOVED***;
    if('function' == typeof callback***REMOVED*** {
      return callback(err, null***REMOVED***;
  ***REMOVED*** else {
      throw err;
  ***REMOVED***
***REMOVED***

  if(this.serverConfig.isDestroyed(***REMOVED******REMOVED*** 
    return callback(new Error("Connection was destroyed by application"***REMOVED******REMOVED***;

  // Specific connection
  var connection = options.connection;
  // Check if the connection is actually live
  if(connection 
    && (!connection.isConnected || !connection.isConnected(***REMOVED******REMOVED******REMOVED*** connection = null;

  // Get the configuration
  var config = this.serverConfig;
  var readPreference = options.readPreference;
  // Allow for the usage of the readPreference model
  if(readPreference == null***REMOVED*** {
    readPreference = options.readPreference;
***REMOVED***

  if(!connection && !config.canRead(readPreference***REMOVED*** && !config.canWrite(***REMOVED*** && config.isAutoReconnect(***REMOVED******REMOVED*** {

    if(readPreference == ReadPreference.PRIMARY 
      || readPreference == ReadPreference.PRIMARY_PREFERRED
      || (readPreference != null && typeof readPreference == 'object' && readPreference.mode***REMOVED***
      || readPreference == null***REMOVED*** {
      
      // Save the command
      self.serverConfig._commandsStore.read_from_writer(
        {   type: 'query'
          , db_command: db_command
          , options: options
          , callback: callback
          , db: self
          , executeQueryCommand: __executeQueryCommand
          , executeInsertCommand: __executeInsertCommand
      ***REMOVED***
      ***REMOVED***;
  ***REMOVED*** else {
      self.serverConfig._commandsStore.read(
        {   type: 'query'
          , db_command: db_command
          , options: options
          , callback: callback 
          , db: self
          , executeQueryCommand: __executeQueryCommand
          , executeInsertCommand: __executeInsertCommand
      ***REMOVED***
      ***REMOVED***;
  ***REMOVED***

    // If we have blown through the number of items let's 
    if(!self.serverConfig._commandsStore.validateBufferLimit(self.bufferMaxEntries***REMOVED******REMOVED*** {
      self.close(***REMOVED***;
  ***REMOVED***    
***REMOVED*** else if(!connection && !config.canRead(readPreference***REMOVED*** && !config.canWrite(***REMOVED*** && !config.isAutoReconnect(***REMOVED******REMOVED*** {
    return callback(new Error("no open connections"***REMOVED***, null***REMOVED***;
***REMOVED*** else {
    if(typeof callback == 'function'***REMOVED*** {
      __executeQueryCommand(self, db_command, options, function (err, result, conn***REMOVED*** {
        callback(err, result, conn***REMOVED***;
    ***REMOVED******REMOVED***;          
  ***REMOVED*** else {
      __executeQueryCommand(self, db_command, options***REMOVED***;
  ***REMOVED***
***REMOVED***
};

/**
 * @ignore
 */
var __executeInsertCommand = function(self, db_command, options, callback***REMOVED*** {
  // Always checkout a writer for this kind of operations
  var connection = self.serverConfig.checkoutWriter(***REMOVED***;
  // Get safe mode
  var safe = options['safe'] != null ? options['safe'] : false;
  var specifiedConnection = options['connection'] != null ? options['connection'] : null;
  // Override connection if needed
  connection = specifiedConnection != null ? specifiedConnection : connection;

  // Validate if we can use this server 2.6 wire protocol
  if(!connection.isCompatible(***REMOVED******REMOVED*** {
    return callback(utils.toError("driver is incompatible with this server version"***REMOVED***, null***REMOVED***;
***REMOVED***

  // Ensure we have a valid connection
  if(typeof callback === 'function'***REMOVED*** {
    callback = bindToCurrentDomain(callback***REMOVED***;
    // Ensure we have a valid connection
    if(connection == null***REMOVED*** {
      return callback(new Error("no open connections"***REMOVED******REMOVED***;
  ***REMOVED*** else if(connection instanceof Error***REMOVED*** {
      return callback(connection***REMOVED***;
  ***REMOVED***

    var errorOptions = _getWriteConcern(self, options***REMOVED***;
    if(errorOptions.w > 0 || errorOptions.w == 'majority' || errorOptions.j || errorOptions.journal || errorOptions.fsync***REMOVED*** {      
      // db command is now an array of commands (original command + lastError***REMOVED***
      db_command = [db_command, DbCommand.createGetLastErrorCommand(errorOptions, self***REMOVED***];
      // Register the handler in the data structure
      self.serverConfig._registerHandler(db_command[1], false, connection, callback***REMOVED***;      
  ***REMOVED***
***REMOVED***

  // If we have no callback and there is no connection
  if(connection == null***REMOVED*** return null;
  if(connection instanceof Error && typeof callback == 'function'***REMOVED*** return callback(connection, null***REMOVED***;
  if(connection instanceof Error***REMOVED*** return null;
  if(connection == null && typeof callback == 'function'***REMOVED*** return callback(new Error("no primary server found"***REMOVED***, null***REMOVED***;

  // Write the message out
  connection.write(db_command, function(err***REMOVED*** {
    // Return the callback if it's not a safe operation and the callback is defined
    if(typeof callback === 'function' && (safe == null || safe == false***REMOVED******REMOVED*** {
      // Perform the callback
      callback(err, null***REMOVED***;
  ***REMOVED*** else if(typeof callback === 'function'***REMOVED*** {
      // Call the handler with an error
      self.serverConfig._callHandler(db_command[1].getRequestId(***REMOVED***, null, err***REMOVED***;
  ***REMOVED*** else if(typeof callback == 'function' && safe && safe.w == -1***REMOVED*** {
      // Call the handler with no error
      self.serverConfig._callHandler(db_command[1].getRequestId(***REMOVED***, null, null***REMOVED***;
  ***REMOVED*** else if(!safe || safe.w == -1***REMOVED*** {
      self.emit("error", err***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;
};

/**
 * Execute an insert Command
 * @ignore
 * @api private
 */
Db.prototype._executeInsertCommand = function(db_command, options, callback***REMOVED*** {
  var self = this;

  // Unpack the parameters
  if(callback == null && typeof options === 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Ensure options are not null
  options = options == null ? {} : options;

  // Check if the user force closed the command
  if(this._applicationClosed***REMOVED*** {
    if(typeof callback == 'function'***REMOVED*** {
      return callback(new Error("db closed by application"***REMOVED***, null***REMOVED***;
  ***REMOVED*** else {
      throw new Error("db closed by application"***REMOVED***;
  ***REMOVED***
***REMOVED***

  if(this.serverConfig.isDestroyed(***REMOVED******REMOVED*** return callback(new Error("Connection was destroyed by application"***REMOVED******REMOVED***;

  // Specific connection
  var connection = options.connection;
  // Check if the connection is actually live
  if(connection 
    && (!connection.isConnected || !connection.isConnected(***REMOVED******REMOVED******REMOVED*** connection = null;

  // Get config
  var config = self.serverConfig;
  // Check if we are connected
  if(!connection && !config.canWrite(***REMOVED*** && config.isAutoReconnect(***REMOVED******REMOVED*** {
    self.serverConfig._commandsStore.write(
      {   type:'insert'
        , 'db_command':db_command
        , 'options':options
        , 'callback':callback
        , db: self
        , executeQueryCommand: __executeQueryCommand
        , executeInsertCommand: __executeInsertCommand
    ***REMOVED***
    ***REMOVED***;

    // If we have blown through the number of items let's 
    if(!self.serverConfig._commandsStore.validateBufferLimit(self.bufferMaxEntries***REMOVED******REMOVED*** {
      self.close(***REMOVED***;
  ***REMOVED***        
***REMOVED*** else if(!connection && !config.canWrite(***REMOVED*** && !config.isAutoReconnect(***REMOVED******REMOVED*** {
    return callback(new Error("no open connections"***REMOVED***, null***REMOVED***;
***REMOVED*** else {
    __executeInsertCommand(self, db_command, options, callback***REMOVED***;
***REMOVED***
};

/**
 * Update command is the same
 * @ignore
 * @api private
 */
Db.prototype._executeUpdateCommand = Db.prototype._executeInsertCommand;
/**
 * Remove command is the same
 * @ignore
 * @api private
 */
Db.prototype._executeRemoveCommand = Db.prototype._executeInsertCommand;

/**
 * Wrap a Mongo error document into an Error instance.
 * Deprecated. Use utils.toError instead.
 *
 * @ignore
 * @api private
 * @deprecated
 */
Db.prototype.wrap = utils.toError;

/**
 * Default URL
 *
 * @classconstant DEFAULT_URL
 **/
Db.DEFAULT_URL = 'mongodb://localhost:27017/default';

/**
 * Connect to MongoDB using a url as documented at
 *
 *  docs.mongodb.org/manual/reference/connection-string/
 *
 * Options
 *  - **uri_decode_auth** {Boolean, default:false} uri decode the user name and password for authentication
 *  - **db** {Object, default: null} a hash off options to set on the db object, see **Db constructor**
 *  - **server** {Object, default: null} a hash off options to set on the server objects, see **Server** constructor**
 *  - **replSet** {Object, default: null} a hash off options to set on the replSet object, see **ReplSet** constructor**
 *  - **mongos** {Object, default: null} a hash off options to set on the mongos object, see **Mongos** constructor**
 *
 * @param {String} url connection url for MongoDB.
 * @param {Object} [options] optional options for insert command
 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occurred, or null otherwise. While the second parameter will contain the db instance or null if an error occurred.
 * @return {null}
 * @api public
 */
Db.connect = function(url, options, callback***REMOVED*** {
  // Ensure correct mapping of the callback
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  // Ensure same behavior as previous version w:0
  if(url.indexOf("safe"***REMOVED*** == -1 
    && url.indexOf("w"***REMOVED*** == -1 
    && url.indexOf("journal"***REMOVED*** == -1 && url.indexOf("j"***REMOVED*** == -1
    && url.indexOf("fsync"***REMOVED*** == -1***REMOVED*** options.w = 1;

  // Avoid circular require problem
  var MongoClient = require('./mongo_client.js'***REMOVED***.MongoClient;
  // Attempt to connect
  MongoClient.connect.call(MongoClient, url, options, callback***REMOVED***;
};

/**
 * State of the db connection
 * @ignore
 */
Object.defineProperty(Db.prototype, "state", { enumerable: true
  , get: function (***REMOVED*** {
      return this.serverConfig._serverState;
  ***REMOVED***
}***REMOVED***;

/**
 * @ignore
 */
var _hasWriteConcern = function(errorOptions***REMOVED*** {
  return errorOptions == true
    || errorOptions.w > 0
    || errorOptions.w == 'majority'
    || errorOptions.j == true
    || errorOptions.journal == true
    || errorOptions.fsync == true
};

/**
 * @ignore
 */
var _setWriteConcernHash = function(options***REMOVED*** {
  var finalOptions = {};
  if(options.w != null***REMOVED*** finalOptions.w = options.w;  
  if(options.journal == true***REMOVED*** finalOptions.j = options.journal;
  if(options.j == true***REMOVED*** finalOptions.j = options.j;
  if(options.fsync == true***REMOVED*** finalOptions.fsync = options.fsync;
  if(options.wtimeout != null***REMOVED*** finalOptions.wtimeout = options.wtimeout;  
  return finalOptions;
};

/**
 * @ignore
 */
var _getWriteConcern = function(self, options***REMOVED*** {
  // Final options
  var finalOptions = {w:1};
  // Local options verification
  if(options.w != null || typeof options.j == 'boolean' || typeof options.journal == 'boolean' || typeof options.fsync == 'boolean'***REMOVED*** {
    finalOptions = _setWriteConcernHash(options***REMOVED***;
***REMOVED*** else if(options.safe != null && typeof options.safe == 'object'***REMOVED*** {
    finalOptions = _setWriteConcernHash(options.safe***REMOVED***;
***REMOVED*** else if(typeof options.safe == "boolean"***REMOVED*** {
    finalOptions = {w: (options.safe ? 1 : 0***REMOVED***};
***REMOVED*** else if(self.options.w != null || typeof self.options.j == 'boolean' || typeof self.options.journal == 'boolean' || typeof self.options.fsync == 'boolean'***REMOVED*** {
    finalOptions = _setWriteConcernHash(self.options***REMOVED***;
***REMOVED*** else if(self.safe.w != null || typeof self.safe.j == 'boolean' || typeof self.safe.journal == 'boolean' || typeof self.safe.fsync == 'boolean'***REMOVED*** {
    finalOptions = _setWriteConcernHash(self.safe***REMOVED***;
***REMOVED*** else if(typeof self.safe == "boolean"***REMOVED*** {
    finalOptions = {w: (self.safe ? 1 : 0***REMOVED***};
***REMOVED***

  // Ensure we don't have an invalid combination of write concerns
  if(finalOptions.w < 1 
    && (finalOptions.journal == true || finalOptions.j == true || finalOptions.fsync == true***REMOVED******REMOVED*** throw new Error("No acknowledgement using w < 1 cannot be combined with journal:true or fsync:true"***REMOVED***;

  // Return the options
  return finalOptions;
}

/**
 * Legacy support
 *
 * @ignore
 * @api private
 */
exports.connect = Db.connect;
exports.Db = Db;

/**
 * Remove all listeners to the db instance.
 * @ignore
 * @api private
 */
Db.prototype.removeAllEventListeners = function(***REMOVED*** {
  this.removeAllListeners("close"***REMOVED***;
  this.removeAllListeners("error"***REMOVED***;
  this.removeAllListeners("timeout"***REMOVED***;
  this.removeAllListeners("parseError"***REMOVED***;
  this.removeAllListeners("poolReady"***REMOVED***;
  this.removeAllListeners("message"***REMOVED***;
};
