// mongodb/lib/mongodb/index.js
try {
  exports.BSONPure = require('bson'***REMOVED***.BSONPure;
  exports.BSONNative = require('bson'***REMOVED***.BSONNative;
} catch(err***REMOVED*** {
  // do nothing
}

// export the driver version
exports.version = require('../../package'***REMOVED***.version;

[ 'commands/base_command'
  , 'admin'
  , 'collection'
  , 'connection/read_preference'
  , 'connection/connection'
  , 'connection/server'
  , 'connection/mongos'
  , 'connection/repl_set/repl_set'
  , 'mongo_client'
  , 'cursor'
  , 'db'
  , 'mongo_client'
  , 'gridfs/grid'
  ,	'gridfs/chunk'
  , 'gridfs/gridstore'].forEach(function (path***REMOVED*** {
  	var module = require('./' + path***REMOVED***;
  	for (var i in module***REMOVED*** {
  		exports[i] = module[i];
  ***REMOVED***
}***REMOVED***;

// backwards compat
exports.ReplSetServers = exports.ReplSet;
// Add BSON Classes
exports.Binary = require('bson'***REMOVED***.Binary;
exports.Code = require('bson'***REMOVED***.Code;
exports.DBRef = require('bson'***REMOVED***.DBRef;
exports.Double = require('bson'***REMOVED***.Double;
exports.Long = require('bson'***REMOVED***.Long;
exports.MinKey = require('bson'***REMOVED***.MinKey;
exports.MaxKey = require('bson'***REMOVED***.MaxKey;
exports.ObjectID = require('bson'***REMOVED***.ObjectID;
exports.Symbol = require('bson'***REMOVED***.Symbol;
exports.Timestamp = require('bson'***REMOVED***.Timestamp;  
// Add BSON Parser
exports.BSON = require('bson'***REMOVED***.BSONPure.BSON;

// Set up the connect function
var connect = exports.Db.connect;

// Add the pure and native backward compatible functions
exports.pure = exports.native = function(***REMOVED*** {
  return connect;
}

// Map all values to the exports value
for(var name in exports***REMOVED*** {
  connect[name] = exports[name];
}

// Set our exports to be the connect function
module.exports = connect;
