// bson/ext/index.js
var bson = null;

try {
	// Load the precompiled win32 binary
	if(process.platform == "win32" && process.arch == "x64"***REMOVED*** {
	  bson = require('./win32/x64/bson'***REMOVED***;  
	} else if(process.platform == "win32" && process.arch == "ia32"***REMOVED*** {
	  bson = require('./win32/ia32/bson'***REMOVED***;  
	} else {
	  bson = require('../build/Release/bson'***REMOVED***;  
	}	
} catch(err***REMOVED*** {
	// Attempt to load the release bson version
	try {
		bson = require('../build/Release/bson'***REMOVED***;
	} catch (err***REMOVED*** {
		console.error("Failed to load c++ bson extension, using pure JS version"***REMOVED***;
		bson = require('../lib/bson/bson'***REMOVED***;
	}
}

exports.BSON = bson.BSON;
exports.Long = require('../lib/bson/long'***REMOVED***.Long;
exports.ObjectID = require('../lib/bson/objectid'***REMOVED***.ObjectID;
exports.DBRef = require('../lib/bson/db_ref'***REMOVED***.DBRef;
exports.Code = require('../lib/bson/code'***REMOVED***.Code;
exports.Timestamp = require('../lib/bson/timestamp'***REMOVED***.Timestamp;
exports.Binary = require('../lib/bson/binary'***REMOVED***.Binary;
exports.Double = require('../lib/bson/double'***REMOVED***.Double;
exports.MaxKey = require('../lib/bson/max_key'***REMOVED***.MaxKey;
exports.MinKey = require('../lib/bson/min_key'***REMOVED***.MinKey;
exports.Symbol = require('../lib/bson/symbol'***REMOVED***.Symbol;

// Just add constants tot he Native BSON parser
exports.BSON.BSON_BINARY_SUBTYPE_DEFAULT = 0;
exports.BSON.BSON_BINARY_SUBTYPE_FUNCTION = 1;
exports.BSON.BSON_BINARY_SUBTYPE_BYTE_ARRAY = 2;
exports.BSON.BSON_BINARY_SUBTYPE_UUID = 3;
exports.BSON.BSON_BINARY_SUBTYPE_MD5 = 4;
exports.BSON.BSON_BINARY_SUBTYPE_USER_DEFINED = 128;
