// negotiator/lib/mediaType.js
module.exports = preferredMediaTypes;
preferredMediaTypes.preferredMediaTypes = preferredMediaTypes;

function parseAccept(accept***REMOVED*** {
  return accept.split(','***REMOVED***.map(function(e***REMOVED*** {
    return parseMediaType(e.trim(***REMOVED******REMOVED***;
***REMOVED******REMOVED***.filter(function(e***REMOVED*** {
    return e && e.q > 0;
***REMOVED******REMOVED***;
};

function parseMediaType(s***REMOVED*** {
  var match = s.match(/\s*(\S+***REMOVED***\/([^;\s]+***REMOVED***\s*(?:;(.****REMOVED******REMOVED***?/***REMOVED***;
  if (!match***REMOVED*** return null;

  var type = match[1],
      subtype = match[2],
      full = "" + type + "/" + subtype,
      params = {},
      q = 1;

  if (match[3]***REMOVED*** {
    params = match[3].split(';'***REMOVED***.map(function(s***REMOVED*** {
      return s.trim(***REMOVED***.split('='***REMOVED***;
  ***REMOVED******REMOVED***.reduce(function (set, p***REMOVED*** {
      set[p[0]] = p[1];
      return set
  ***REMOVED***, params***REMOVED***;

    if (params.q != null***REMOVED*** {
      q = parseFloat(params.q***REMOVED***;
      delete params.q;
  ***REMOVED***
***REMOVED***

  return {
    type: type,
    subtype: subtype,
    params: params,
    q: q,
    full: full
***REMOVED***;
}

function getMediaTypePriority(type, accepted***REMOVED*** {
  return (accepted.filter(function(a***REMOVED*** {
    return specify(type, a***REMOVED***;
***REMOVED******REMOVED***.sort(function (a, b***REMOVED*** {
    // revsort
    return a.q > b.q ? -1 : 1;
***REMOVED******REMOVED***[0] || {q:0}***REMOVED***.q;
}

function specifies(spec, type***REMOVED*** {
  return spec === '*' || spec === type;
}

function specify(type, spec***REMOVED*** {
  var p = parseMediaType(type***REMOVED***;

  if (spec.params***REMOVED*** {
    var keys = Object.keys(spec.params***REMOVED***;
    if (keys.some(function (k***REMOVED*** {
      return !specifies(spec.params[k], p.params[k]***REMOVED***;
  ***REMOVED******REMOVED******REMOVED*** {
      // some didn't specify.
      return null;
  ***REMOVED***
***REMOVED***

  if (specifies(spec.type, p.type***REMOVED*** &&
      specifies(spec.subtype, p.subtype***REMOVED******REMOVED*** {
    return spec;
***REMOVED***
}

function preferredMediaTypes(accept, provided***REMOVED*** {
  accept = parseAccept(accept || ''***REMOVED***;
  if (provided***REMOVED*** {
    return provided.map(function(type***REMOVED*** {
      return [type, getMediaTypePriority(type, accept***REMOVED***];
  ***REMOVED******REMOVED***.filter(function(pair***REMOVED*** {
      return pair[1] > 0;
  ***REMOVED******REMOVED***.sort(function(a, b***REMOVED*** {
      // revsort
      return a[1] === b[1] ? 0 : a[1] > b[1] ? -1 : 1;
  ***REMOVED******REMOVED***.map(function(pair***REMOVED*** {
      return pair[0];
  ***REMOVED******REMOVED***;

***REMOVED*** else {
    return accept.sort(function (a, b***REMOVED*** {
      // revsort
      return a.q < b.q ? 1 : -1;
  ***REMOVED******REMOVED***.map(function(type***REMOVED*** {
      return type.full;
  ***REMOVED******REMOVED***;
***REMOVED***
}
