// mongodb/lib/mongodb/collection/batch/unordered.js
var shared = require('../shared'***REMOVED***
	, common = require('./common'***REMOVED***
	, utils = require('../../utils'***REMOVED***
  , hasWriteCommands = utils.hasWriteCommands
  , WriteError = common.WriteError
  , BatchWriteResult = common.BatchWriteResult
  , LegacyOp = common.LegacyOp
  , ObjectID = require('bson'***REMOVED***.ObjectID
  , Batch = common.Batch
  , mergeBatchResults = common.mergeBatchResults;

/**
 * Create a new UnorderedBulkOperation instance (INTERNAL TYPE, do not instantiate directly***REMOVED***
 *
 * Options
 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowlegement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *
 * @class Represents a UnorderedBulkOperation
 * @param {Object} collection collection instance.
 * @param {Object} [options] additional options for the collection.
 * @return {Object} a ordered bulk operation instance.
 */
var UnorderedBulkOperation = function(collection, options***REMOVED*** {
	options = options == null ? {} : options;

	// Contains reference to self
	var self = this;
	// Get the namesspace for the write operations
  var namespace = collection.collectionName;
  // Used to mark operation as executed
  var executed = false;

	// Current item
  // var currentBatch = null;
	var currentOp = null;
	var currentIndex = 0;
  var currentBatchSize = 0;
  var currentBatchSizeBytes = 0;
  var batches = [];

  // The current Batches for the different operations
  var currentInsertBatch = null;
  var currentUpdateBatch = null;
  var currentRemoveBatch = null;

	// Handle to the bson serializer, used to calculate running sizes
  var db = collection.db;
	var bson = db.bson;

  // Set max byte size
	var maxBatchSizeBytes = db.serverConfig.checkoutWriter(***REMOVED***.maxBsonSize;
	var maxWriteBatchSize = db.serverConfig.checkoutWriter(***REMOVED***.maxWriteBatchSize || 1000;

  // Get the write concern
  var writeConcern = shared._getWriteConcern(collection, options***REMOVED***;

  // Final results
  var bulkResult = {
  	  ok: 1
    , writeErrors: []
    , writeConcernErrors: []
    , nInserted: 0
    , nUpserted: 0
    , nMatched: 0
    , nModified: 0
    , nRemoved: 0
    , upserted: []
***REMOVED***;

  // Specify a full class so we can generate documentation correctly
	var FindOperators = function(***REMOVED*** {
		/**
		 * Add a single update document to the bulk operation
		 *
		 * @param {Object} doc update operations
		 * @return {UnorderedBulkOperation}
		 * @api public
		 */
		this.update = function(updateDocument***REMOVED*** {
			// Perform upsert
			var upsert = typeof currentOp.upsert == 'boolean' ? currentOp.upsert : false;
			
			// Establish the update command
			var document = {
					q: currentOp.selector
				, u: updateDocument
				, multi: true
				, upsert: upsert
			}

			// Clear out current Op
			currentOp = null;
			// Add the update document to the list
			return addToOperationsList(self, common.UPDATE, document***REMOVED***;
		}	

		/**
		 * Add a single update one document to the bulk operation
		 *
		 * @param {Object} doc update operations
		 * @return {UnorderedBulkOperation}
		 * @api public
		 */
		this.updateOne = function(updateDocument***REMOVED*** {
			// Perform upsert
			var upsert = typeof currentOp.upsert == 'boolean' ? currentOp.upsert : false;
			
			// Establish the update command
			var document = {
					q: currentOp.selector
				, u: updateDocument
				, multi: false
				, upsert: upsert
			}

			// Clear out current Op
			currentOp = null;
			// Add the update document to the list
			return addToOperationsList(self, common.UPDATE, document***REMOVED***;
		}

		/**
		 * Add a replace one operation to the bulk operation
		 *
		 * @param {Object} doc the new document to replace the existing one with
		 * @return {UnorderedBulkOperation}
		 * @api public
		 */
		this.replaceOne = function(updateDocument***REMOVED*** {
			this.updateOne(updateDocument***REMOVED***;
		}

		/**
		 * Upsert modifier for update bulk operation
		 *
		 * @return {UnorderedBulkOperation}
		 * @api public
		 */
		this.upsert = function(***REMOVED*** {
			currentOp.upsert = true;
			return this;
		}

		/**
		 * Add a remove one operation to the bulk operation
		 *
		 * @param {Object} doc selector for the removal of documents
		 * @return {UnorderedBulkOperation}
		 * @api public
		 */
		this.removeOne = function(***REMOVED*** {		
			// Establish the update command
			var document = {
					q: currentOp.selector
				, limit: 1
			}

			// Clear out current Op
			currentOp = null;
			// Add the remove document to the list
			return addToOperationsList(self, common.REMOVE, document***REMOVED***;
		}

		/**
		 * Add a remove operation to the bulk operation
		 *
		 * @param {Object} doc selector for the single document to remove
		 * @return {UnorderedBulkOperation}
		 * @api public
		 */
		this.remove = function(***REMOVED*** {
			// Establish the update command
			var document = {
					q: currentOp.selector
				, limit: 0
			}

			// Clear out current Op
			currentOp = null;
			// Add the remove document to the list
			return addToOperationsList(self, common.REMOVE, document***REMOVED***;				
		}
	}

	// 
	// Add to the operations list
	//
	var addToOperationsList = function(_self, docType, document***REMOVED*** {
    // Get the bsonSize
    var bsonSize = bson.calculateObjectSize(document, false***REMOVED***;
    // Throw error if the doc is bigger than the max BSON size
    if(bsonSize >= maxBatchSizeBytes***REMOVED*** throw utils.toError("document is larger than the maximum size " + maxBatchSizeBytes***REMOVED***;
    // Holds the current batch
    var currentBatch = null;
    // Get the right type of batch
    if(docType == common.INSERT***REMOVED*** {
    	currentBatch = currentInsertBatch;
  ***REMOVED*** else if(docType == common.UPDATE***REMOVED*** {
    	currentBatch = currentUpdateBatch;
  ***REMOVED*** else if(docType == common.REMOVE***REMOVED*** {
    	currentBatch = currentRemoveBatch;
  ***REMOVED***

    // Create a new batch object if we don't have a current one
    if(currentBatch == null***REMOVED*** currentBatch = new Batch(docType, currentIndex***REMOVED***;
    
    // Check if we need to switch batch type
    if(currentBatch.batchType != docType***REMOVED*** {
      // Save current batch
      batches.push(currentBatch***REMOVED***;
      // Create a new batch
      currentBatch = new Batch(docType, currentIndex***REMOVED***;  

      // Reset the current size trackers
      currentBatchSize = 0;
      currentBatchSizeBytes = 0;
  ***REMOVED***

    // Check if we need to create a new batch
    if(((currentBatchSize + 1***REMOVED*** >= maxWriteBatchSize***REMOVED***
      || ((currentBatchSizeBytes +  currentBatchSizeBytes***REMOVED*** >= maxBatchSizeBytes***REMOVED***
      || (currentBatch.batchType != docType***REMOVED******REMOVED*** {
      // Save the batch to the execution stack
      batches.push(currentBatch***REMOVED***;
      
      // Create a new batch
      currentBatch = new Batch(docType, currentIndex***REMOVED***;
      
      // Reset the current size trackers
      currentBatchSize = 0;
      currentBatchSizeBytes = 0;
  ***REMOVED*** else {
	    // Update current batch size
	    currentBatchSize = currentBatchSize + 1;
	    currentBatchSizeBytes = currentBatchSizeBytes + bsonSize;
  ***REMOVED***

    // We have an array of documents
    if(Array.isArray(document***REMOVED******REMOVED*** {
    	throw util.toError("operation passed in cannot be an Array"***REMOVED***;
  ***REMOVED*** else {
      currentBatch.operations.push(document***REMOVED***;
      currentBatch.originalIndexes.push(currentIndex***REMOVED***;
      currentIndex = currentIndex + 1;
  ***REMOVED***

    // Save back the current Batch to the right type
    if(docType == common.INSERT***REMOVED*** {
    	currentInsertBatch = currentBatch;
  ***REMOVED*** else if(docType == common.UPDATE***REMOVED*** {
    	currentUpdateBatch = currentBatch;
  ***REMOVED*** else if(docType == common.REMOVE***REMOVED*** {
    	currentRemoveBatch = currentBatch;
  ***REMOVED***

    // Update current batch size
    currentBatchSize = currentBatchSize + 1;
    currentBatchSizeBytes = currentBatchSizeBytes + bsonSize;

    // Return self
		return _self;
	}

	/**
	 * Add a single insert document to the bulk operation
	 *
	 * @param {Object} doc the document to insert
	 * @return {UnorderedBulkOperation}
	 * @api public
	 */
	this.insert = function(document***REMOVED*** {
    if(document._id == null***REMOVED*** document._id = new ObjectID(***REMOVED***;
		return addToOperationsList(self, common.INSERT, document***REMOVED***;
	}

	/**
	 * Initiate a find operation for an update/updateOne/remove/removeOne/replaceOne
	 *
	 * @param {Object} selector the selector used to locate documents for the operation
	 * @return {UnorderedBulkOperation}
	 * @api public
	 */
	this.find = function(selector***REMOVED*** {
		if (!selector***REMOVED*** {
			throw utils.toError("Bulk find operation must specify a selector"***REMOVED***;
		}

		// Save a current selector
		currentOp = {
			selector: selector
		}

		return new FindOperators(***REMOVED***;
	}	

	//
	// Execute the command
	var executeBatch = function(batch, callback***REMOVED*** {
		// Contains the command we are going to execute
		var cmd = null;

    // Generate the right update
    if(batch.batchType == common.UPDATE***REMOVED*** {
      cmd = { update: namespace, updates: batch.operations, ordered: false }
  ***REMOVED*** else if(batch.batchType == common.INSERT***REMOVED*** {
      cmd = { insert: namespace, documents: batch.operations, ordered: false }
  ***REMOVED*** else if(batch.batchType == common.REMOVE***REMOVED*** {
      cmd = { delete: namespace, deletes: batch.operations, ordered: false }
  ***REMOVED***

    // If we have a write concern
    if(writeConcern != null***REMOVED*** {
      cmd.writeConcern = writeConcern;
  ***REMOVED***    

		// Execute the write command
		db.command(cmd, function(err, result***REMOVED*** {
			callback(null, mergeBatchResults(false, batch, bulkResult, err, result***REMOVED******REMOVED***;
		}***REMOVED***;
	}

	//
	// Execute all the commands
	var executeBatches = function(callback***REMOVED*** {
		var numberOfCommandsToExecute = batches.length;
		// Execute over all the batches
		for(var i = 0; i < batches.length; i++***REMOVED*** {
			executeBatch(batches[i], function(err, result***REMOVED*** {
				numberOfCommandsToExecute = numberOfCommandsToExecute - 1;

				// Execute
				if(numberOfCommandsToExecute == 0***REMOVED*** {
					// If we have an error stop
					if(bulkResult.ok == 0 && callback***REMOVED*** {
						return callback(bulkResult.error, null***REMOVED***;
					}

					callback(null, new BatchWriteResult(bulkResult***REMOVED******REMOVED***;
				}
			}***REMOVED***;
		}
	}

	/**
	 * Execute the unordered bulk operation
	 *
	 * Options
	 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowlegement of write and w >= 1, w = 'majority' or tag acknowledges the write
	 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
	 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
	 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
	 *
	 * @param {Object} [options] additional options during update.
	 * @param {Function} callback this will be called after executing this method. The first parameter will contain the Error object if an error occured, or null otherwise. While the second parameter will contain the results from from the unordered bulk operation.
	 * @return {null}
	 * @api public
	 */
	this.execute = function(_writeConcern, callback***REMOVED*** {
		if(executed***REMOVED*** throw util.toError("batch cannot be re-executed"***REMOVED***;
		if(typeof _writeConcern == 'function'***REMOVED*** {
			callback = _writeConcern;
		} else {
			writeConcern = _writeConcern;
		}

    // If we have current batch
    if(currentInsertBatch***REMOVED*** batches.push(currentInsertBatch***REMOVED***;
    if(currentUpdateBatch***REMOVED*** batches.push(currentUpdateBatch***REMOVED***;
    if(currentRemoveBatch***REMOVED*** batches.push(currentRemoveBatch***REMOVED***;

		// If we have no operations in the bulk raise an error
		if(batches.length == 0***REMOVED*** {
			throw utils.toError("Invalid Operation, No operations in bulk"***REMOVED***;
		}

		// Check if we support bulk commands
		if(hasWriteCommands(db.serverConfig.checkoutWriter(***REMOVED******REMOVED******REMOVED***
			return executeBatches(function(err, result***REMOVED*** {
				callback(err, result***REMOVED***;
			}***REMOVED***;

		// Set nModified to null as we don't support this field
		bulkResult.nModified = null;

		// Run in backward compatibility mode
		backwardsCompatibilityExecuteCommands(function(err, result***REMOVED*** {
			callback(err, result***REMOVED***;
		}***REMOVED***;
	}	

	// 
	// Execute the inserts
	var executeInserts = function(_collection, _batch, _result, _callback***REMOVED*** {
		var totalNumberOfInserts = _batch.operations.length;

		// Options for the update operation
		var batchOptions = writeConcern || {};

		// Execute the op
		var executeLegacyInsert = function(_i, _op, _options, __callback***REMOVED*** {
			// Execute the update
			_collection.insert(_op.operation, _options, function(err, r***REMOVED*** {
				// If we have don't have w:0 merge the result
				if(_options.w == null || _options.w != 0***REMOVED*** {
					// Merge the results in 
					var result = common.mergeLegacyResults(false, _op.operation, _batch, bulkResult, err || r, _op.index***REMOVED***;
					if(result == false***REMOVED*** {
						return _callback(null, new BatchWriteResult(bulkResult***REMOVED******REMOVED***;
					}
				}

				__callback(null, _result***REMOVED***;
			}***REMOVED***;
		}

		// Execute all the insert operations
		for(var i = 0; i < _batch.operations.length; i++***REMOVED*** {
			var legacyOp = new LegacyOp(_batch.batchType, _batch.operations[i], _batch.originalIndexes[i]***REMOVED***;
			executeLegacyInsert(i, legacyOp, batchOptions, function(err, result***REMOVED*** {
				totalNumberOfInserts = totalNumberOfInserts - 1;
				
				// No more inserts
				if(totalNumberOfInserts == 0***REMOVED*** {
					_callback(null, _result***REMOVED***;
				}
			}***REMOVED***;
		}
	}

	//
	// Execute updates
	var executeUpdates = function(_collection, _batch, _result, _callback***REMOVED*** {
		var totalNumberOfUpdates = _batch.operations.length;
		// Options for the update operation
		var batchOptions = writeConcern || {};

		// Execute the op
		var executeLegacyUpdate = function(_i, _op, _options, __callback***REMOVED*** {
			var options = common.cloneOptions(batchOptions***REMOVED***;

			// Add any additional options
			if(_op.operation.multi != null***REMOVED*** options.multi = _op.operation.multi ? true : false;
			if(_op.operation.upsert != null***REMOVED*** options.upsert = _op.operation.upsert;

			// Execute the update
			_collection.update(_op.operation.q, _op.operation.u, options, function(err, r, full***REMOVED*** {
				// If we have don't have w:0 merge the result
				if(options.w == null || options.w != 0***REMOVED*** {
					// Merge the results in 
					var result = common.mergeLegacyResults(false, _op.operation, _batch, bulkResult, err || full, _op.index***REMOVED***;
					if(result == false***REMOVED*** {
						return _callback(null, new BatchWriteResult(bulkResult***REMOVED******REMOVED***;
					}
				}

				return __callback(null, _result***REMOVED***;
			}***REMOVED***;
		}

		// Execute all the insert operations
		for(var i = 0; i < _batch.operations.length; i++***REMOVED*** {
			var legacyOp = new LegacyOp(_batch.batchType, _batch.operations[i], _batch.originalIndexes[i]***REMOVED***;
			executeLegacyUpdate(i, legacyOp, options, function(err, result***REMOVED*** {
				totalNumberOfUpdates = totalNumberOfUpdates - 1;
				
				// No more inserts
				if(totalNumberOfUpdates == 0***REMOVED*** {
					_callback(null, _result***REMOVED***;
				}
			}***REMOVED***;
		}
	}

	//
	// Execute updates
	var executeRemoves = function(_collection, _batch, _result, _callback***REMOVED*** {
		var totalNumberOfRemoves = _batch.operations.length;
		// Options for the update operation
		var batchOptions = writeConcern || {};

		// Execute the op
		var executeLegacyRemove = function(_i, _op, _options, __callback***REMOVED*** {
			var options = common.cloneOptions(batchOptions***REMOVED***;

			// Add any additional options
			if(_op.operation.limit != null***REMOVED*** options.single = _op.operation.limit == 1 ? true : false;

			// Execute the update
			_collection.remove(_op.operation.q, options, function(err, r***REMOVED*** {
				// If we have don't have w:0 merge the result
				if(options.w == null || options.w != 0***REMOVED*** {
					// Merge the results in 
					var result = common.mergeLegacyResults(false, _op.operation, _batch, bulkResult, err || r, _op.index***REMOVED***;
					if(result == false***REMOVED*** {
						return _callback(null, new BatchWriteResult(bulkResult***REMOVED******REMOVED***;
					}
				}

				return __callback(null, _result***REMOVED***;
			}***REMOVED***;
		}

		// Execute all the insert operations
		for(var i = 0; i < _batch.operations.length; i++***REMOVED*** {
			var legacyOp = new LegacyOp(_batch.batchType, _batch.operations[i], _batch.originalIndexes[i]***REMOVED***;
			executeLegacyRemove(i, legacyOp, options, function(err, result***REMOVED*** {
				totalNumberOfRemoves = totalNumberOfRemoves - 1;
				
				// No more inserts
				if(totalNumberOfRemoves == 0***REMOVED*** {
					_callback(null, _result***REMOVED***;
				}
			}***REMOVED***;
		}
	}

	//
	// Execute all operation in backwards compatible fashion
	var backwardsCompatibilityExecuteCommands = function(callback***REMOVED*** {
		if(batches.length == 0***REMOVED*** {
			return callback(null, new BatchWriteResult(bulkResult***REMOVED******REMOVED***;
		}

		// Ordered execution of the command
		var batch = batches.shift(***REMOVED***;

		// Process the legacy operations
		var processLegacyOperations = function(err, results***REMOVED*** {
			// Merge the results together
			var mergeResult = mergeBatchResults(false, batch, bulkResult, err, results***REMOVED***;
			if(mergeResult != null***REMOVED*** {
				return callback(null, mergeResult***REMOVED***
			}

			// If we have an error stop
			if(bulkResult.ok == 0 && callback***REMOVED*** {
				var internalCallback = callback;
				callback = null;
				return internalCallback(bulkResult.error, null***REMOVED***;
			} else if(bulkResult.ok == 0 && callback == null***REMOVED*** {
				return;
			}

			// Execute the next step
			backwardsCompatibilityExecuteCommands(callback***REMOVED***;			
		}

		// Execute an insert batch
		if(batch.batchType == common.INSERT***REMOVED*** {
			return executeInserts(collection, batch, {n: 0}, processLegacyOperations***REMOVED***;
		}

		// Execute an update batch
		if(batch.batchType == common.UPDATE***REMOVED*** {
			return executeUpdates(collection, batch, {n: 0}, processLegacyOperations***REMOVED***;
		}

		// Execute an update batch
		if(batch.batchType == common.REMOVE***REMOVED*** {
			return executeRemoves(collection, batch, {n: 0}, processLegacyOperations***REMOVED***;
		}
	}	
}

/**
 * Returns an unordered batch object
 *
 */
var initializeUnorderedBulkOp = function(options***REMOVED*** {
	return new UnorderedBulkOperation(this, options***REMOVED***;
}

exports.initializeUnorderedBulkOp = initializeUnorderedBulkOp;
