// stylus/lib/stylus.js
/*!
 * Stylus
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Renderer = require('./renderer'***REMOVED***
  , Parser = require('./parser'***REMOVED***
  , nodes = require('./nodes'***REMOVED***
  , utils = require('./utils'***REMOVED***;

/**
 * Export render as the module.
 */

exports = module.exports = render;

/**
 * Library version.
 */

exports.version = require('../package'***REMOVED***.version;

/**
 * Expose nodes.
 */

exports.nodes = nodes;

/**
 * Expose BIFs.
 */

exports.functions = require('./functions'***REMOVED***;

/**
 * Expose utils.
 */

exports.utils = require('./utils'***REMOVED***;

/**
 * Expose middleware.
 */

exports.middleware = require('./middleware'***REMOVED***;

/**
 * Expose constructors.
 */

exports.Visitor = require('./visitor'***REMOVED***;
exports.Parser = require('./parser'***REMOVED***;
exports.Evaluator = require('./visitor/evaluator'***REMOVED***;
exports.Compiler = require('./visitor/compiler'***REMOVED***;

/**
 * Convert the given `css` to `stylus` source.
 *
 * @param {String} css
 * @return {String}
 * @api public
 */

exports.convertCSS = require('./convert/css'***REMOVED***;

/**
 * Render the given `str` with `options` and callback `fn(err, css***REMOVED***`.
 *
 * @param {String} str
 * @param {Object|Function} options
 * @param {Function} fn
 * @api public
 */

exports.render = function(str, options, fn***REMOVED***{
  if ('function' == typeof options***REMOVED*** fn = options, options = {};
  return new Renderer(str, options***REMOVED***.render(fn***REMOVED***;
};

/**
 * Return a new `Renderer` for the given `str` and `options`.
 *
 * @param {String} str
 * @param {Object} options
 * @return {Renderer}
 * @api public
 */

function render(str, options***REMOVED*** {
  return new Renderer(str, options***REMOVED***;
}

/**
 * Expose optional functions.
 */

exports.url = require('./functions/url'***REMOVED***;
exports.resolver = require('./functions/resolver'***REMOVED***;
