// should/lib/should.js
/*!
 * Should
 * Copyright(c***REMOVED*** 2010-2012 TJ Holowaychuk <tj@vision-media.ca>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var util = require('./util'***REMOVED***
  , assert = require('assert'***REMOVED***
  , AssertionError = assert.AssertionError
  , statusCodes = require('./http'***REMOVED***.STATUS_CODES
  , eql = require('./eql'***REMOVED***
  , inspect = require('util'***REMOVED***.inspect;

/**
 * Our function should
 * @param obj
 * @returns {Assertion}
 */
var should = function(obj***REMOVED*** {
  return new Assertion(util.isWrapperType(obj***REMOVED*** ? obj.valueOf(***REMOVED***: obj***REMOVED***;
};

should.inspect = inspect;

/**
 * Expose assert to should
 *
 * This allows you to do things like below
 * without require(***REMOVED***ing the assert module.
 *
 *    should.equal(foo.bar, undefined***REMOVED***;
 *
 */
util.merge(should, assert***REMOVED***;


/**
 * Assert _obj_ exists, with optional message.
 *
 * @param {*} obj
 * @param {String} [msg]
 * @api public
 */
should.exist = should.exists = function(obj, msg***REMOVED*** {
  if(null == obj***REMOVED*** {
    throw new AssertionError({
      message: msg || ('expected ' + should.inspect(obj***REMOVED*** + ' to exist'***REMOVED***
      , stackStartFunction: should.exist
  ***REMOVED******REMOVED***;
***REMOVED***
};

/**
 * Asserts _obj_ does not exist, with optional message.
 *
 * @param {*} obj
 * @param {String} [msg]
 * @api public
 */

should.not = {};
should.not.exist = should.not.exists = function(obj, msg***REMOVED***{
  if (null != obj***REMOVED*** {
    throw new AssertionError({
      message: msg || ('expected ' + should.inspect(obj***REMOVED*** + ' to not exist'***REMOVED***
      , stackStartFunction: should.not.exist
  ***REMOVED******REMOVED***;
***REMOVED***
};

/**
 * Expose should to external world.
 */
exports = module.exports = should;


/**
 * Expose api via `Object#should`.
 *
 * @api public
 */

Object.defineProperty(Object.prototype, 'should', {
  set: function(***REMOVED***{},
  get: function(***REMOVED***{
    return should(this***REMOVED***;
***REMOVED***,
  configurable: true
}***REMOVED***;

/**
 * Initialize a new `Assertion` with the given _obj_.
 *
 * @param {*} obj
 * @api private
 */

var Assertion = should.Assertion = function Assertion(obj***REMOVED*** {
  this.obj = obj;
};

var hasOwnProperty = Object.prototype.hasOwnProperty;

/**
 * Prototype.
 */

Assertion.prototype = {

  /**
   * Assert _expr_ with the given _msg_ and _negatedMsg_.
   *
   * @param {Boolean} expr
   * @param {function} msg
   * @param {function} negatedMsg
   * @param {Object} [expected]
   * @param {Boolean} [showDiff]
   * @param {String} [description]
   * @api private
   */

  assert: function(expr, msg, negatedMsg, expected, showDiff, description***REMOVED***{
    msg = this.negate ? negatedMsg : msg

    var ok = this.negate ? !expr : expr
      , obj = this.obj;

    if (ok***REMOVED*** return;

    var err = new AssertionError({
        message: msg.call(this***REMOVED***
      , actual: obj
      , expected: expected
      , stackStartFunction: this.assert
      , negated: this.negate
  ***REMOVED******REMOVED***;

    err.showDiff = showDiff;
    err.description = description

    throw err;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get an(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get of(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get a(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get and(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get be(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get have(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Dummy getter.
   *
   * @api public
   */

  get with(***REMOVED*** {
    return this;
***REMOVED***,

  /**
   * Negation modifier.
   *
   * @api public
   */

  get not(***REMOVED*** {
    this.negate = true;
    return this;
***REMOVED***,

  /**
   * Get object inspection string.
   *
   * @return {String}
   * @api private
   */

  get inspect(***REMOVED*** {
    return should.inspect(this.obj***REMOVED***;
***REMOVED***,

  /**
   * Assert instanceof `Arguments`.
   *
   * @api public
   */

  get arguments(***REMOVED*** {
    this.assert(
        util.isArguments(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be arguments' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not be arguments' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert that object is empty.
   *
   * @api public
   */

  get empty(***REMOVED*** {
    var length = this.obj.length;

    if(util.isString(this.obj***REMOVED*** || Array.isArray(this.obj***REMOVED*** || util.isArguments(this.obj***REMOVED******REMOVED*** {
      this.assert(
        0 === length
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be empty' }
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be empty' }***REMOVED***;
  ***REMOVED*** else {
      var ok = true;
      for (var prop in this.obj***REMOVED*** {
        if(hasOwnProperty.call(this.obj, prop***REMOVED******REMOVED*** {
          ok = false;
          break;
      ***REMOVED***
    ***REMOVED***

      this.assert(
        ok
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be empty' }
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be empty' }***REMOVED***;

  ***REMOVED***
    return this;
***REMOVED***,

  /**
   * Assert ok.
   *
   * @api public
   */

  get ok(***REMOVED*** {
    this.assert(
        this.obj
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be truthy' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be falsey' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert true.
   *
   * @api public
   */

  get true(***REMOVED*** {
    this.assert(
        true === this.obj
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be true' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be true' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert false.
   *
   * @api public
   */

  get false(***REMOVED*** {
    this.assert(
        false === this.obj
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be false' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be false' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert NaN.
   *
   * @api public
   */

  get NaN(***REMOVED*** {
    this.assert(
        util.isNumber(this.obj***REMOVED*** && isNaN(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be NaN' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be NaN' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert Infinity.
   *
   * @api public
   */

  get Infinity(***REMOVED*** {
    this.assert(
      util.isNumber(this.obj***REMOVED*** && !isNaN(this.obj***REMOVED*** && !isFinite(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be Infinity' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be Infinity' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert equal.
   *
   * @param {*} val
   * @param {String} description
   * @api public
   */

  eql: function(val, description***REMOVED***{
    this.assert(
        eql(val, this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to equal ' + should.inspect(val***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not equal ' + should.inspect(val***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , val
      , true
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert strict equal.
   *
   * @param {*} val
   * @param {String} description
   * @api public
   */

  equal: function(val, description***REMOVED***{
    this.assert(
        val === this.obj
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to equal ' + should.inspect(val***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not equal ' + should.inspect(val***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , val
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert within start to finish (inclusive***REMOVED***.
   *
   * @param {Number} start
   * @param {Number} finish
   * @param {String} description
   * @api public
   */

  within: function(start, finish, description***REMOVED***{
    var range = start + '..' + finish;
    this.assert(
        this.obj >= start && this.obj <= finish
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be within ' + range + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not be within ' + range + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert within value +- delta (inclusive***REMOVED***.
   *
   * @param {Number} value
   * @param {Number} delta
   * @param {String} description
   * @api public
   */

  approximately: function(value, delta, description***REMOVED*** {
    this.assert(
      Math.abs(this.obj - value***REMOVED*** <= delta
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be approximately ' + value + " +- " + delta + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not be approximately ' + value + " +- " + delta + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert typeof.
   *
   * @param {*} type
   * @param {String} description
   * @api public
   */
  type: function(type, description***REMOVED***{
    this.assert(
        type == typeof this.obj
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to have type ' + type + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to have type ' + type  + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert instanceof.
   *
   * @param {Function} constructor
   * @param {String} description
   * @api public
   */

  instanceof: function(constructor, description***REMOVED***{
    var name = constructor.name;
    this.assert(
        this.obj instanceof constructor
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be an instance of ' + name + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be an instance of ' + name + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert if given object is a function.
   */
  get Function(***REMOVED***{
    this.assert(
      util.isFunction(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be a function' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be a function' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert given object is an object.
   */
  get Object(***REMOVED***{
    this.assert(
      util.isObject(this.obj***REMOVED*** && !Array.isArray(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be an object' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be an object' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert given object is a string
   */
  get String(***REMOVED***{
    this.assert(
      util.isString(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be a string' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be a string' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert given object is an array
   */
  get Array(***REMOVED***{
    this.assert(
      Array.isArray(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be an array' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be an array' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert given object is a number. NaN and Infinity are not numbers.
   */
  get Number(***REMOVED***{
    this.assert(
      util.isNumber(this.obj***REMOVED*** && isFinite(this.obj***REMOVED*** && !isNaN(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be a number' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be a number' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert given object is a boolean
   */
  get Boolean(***REMOVED***{
    this.assert(
      util.isBoolean(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be a boolean' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be a boolean' }***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert given object is an error
   */
  get Error(***REMOVED*** {
    this.assert(
      util.isError(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be an error' }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to be an error' }***REMOVED***;
    return this;
***REMOVED***,
  /**
   * Assert numeric value above _n_.
   *
   * @param {Number} n
   * @param {String} description
   * @api public
   */

  above: function(n, description***REMOVED***{
    this.assert(
        this.obj > n
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be above ' + n + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be below ' + n + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert numeric value below _n_.
   *
   * @param {Number} n
   * @param {String} description
   * @api public
   */

  below: function(n, description***REMOVED***{
    this.assert(
        this.obj < n
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be below ' + n + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to be above ' + n + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert string value matches _regexp_.
   *
   * @param {RegExp} regexp
   * @param {String} description
   * @api public
   */

  match: function(regexp, description***REMOVED***{
    this.assert(
        regexp.exec(this.obj***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to match ' + regexp + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' not to match ' + regexp + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert property "length" exists and has value of _n_.
   *
   * @param {Number} n
   * @param {String} description
   * @api public
   */

  length: function(n, description***REMOVED***{
    this.obj.should.have.property('length'***REMOVED***;
    var len = this.obj.length;
    this.assert(
        n == len
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to have a length of ' + n + ' but got ' + len + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not have a length of ' + len + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert property _name_ exists, with optional _val_.
   *
   * @param {String} name
   * @param {*} [val]
   * @param {String} description
   * @api public
   */

  property: function(name, val, description***REMOVED***{
    if (this.negate && undefined !== val***REMOVED*** {
      if (undefined === this.obj[name]***REMOVED*** {
        throw new Error(this.inspect + ' has no property ' + should.inspect(name***REMOVED*** + (description ? " | " + description : ""***REMOVED******REMOVED***;
    ***REMOVED***
  ***REMOVED*** else {
      this.assert(
          undefined !== this.obj[name]
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to have a property ' + should.inspect(name***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not have a property ' + should.inspect(name***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , void 0
        , void 0
        , description***REMOVED***;
  ***REMOVED***

    if (undefined !== val***REMOVED*** {
      this.assert(
          val === this.obj[name]
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to have a property ' + should.inspect(name***REMOVED***
          + ' of ' + should.inspect(val***REMOVED*** + ', but got ' + should.inspect(this.obj[name]***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not have a property ' + should.inspect(name***REMOVED*** + ' of ' + should.inspect(val***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , void 0
        , void 0
        , description***REMOVED***;
  ***REMOVED***

    this.obj = this.obj[name];
    return this;
***REMOVED***,
  /**
   * Asset have given properties
   * @param {Array|String ...} names
   * @api public
   */
  properties: function(names***REMOVED*** {
    var str
      , ok = true;

    names = names instanceof Array
      ? names
      : Array.prototype.slice.call(arguments***REMOVED***;

    var len = names.length;

    if (!len***REMOVED*** throw new Error('names required'***REMOVED***;

    // make sure they're all present
    ok = names.every(function(name***REMOVED***{
      return this.obj[name] !== undefined;
  ***REMOVED***, this***REMOVED***;

    // key string
    if (len > 1***REMOVED*** {
      names = names.map(function(name***REMOVED***{
        return should.inspect(name***REMOVED***;
    ***REMOVED******REMOVED***;
      var last = names.pop(***REMOVED***;
      str = names.join(', '***REMOVED*** + ', and ' + last;
  ***REMOVED*** else {
      str = should.inspect(names[0]***REMOVED***;
  ***REMOVED***

    // message
    str = 'have ' + (len > 1 ? 'properties ' : 'a property '***REMOVED*** + str;

    this.assert(
      ok
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to ' + str }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not ' + str }***REMOVED***;

    return this;
***REMOVED***,

  /**
   * Assert own property _name_ exists.
   *
   * @param {String} name
   * @param {String} description
   * @api public
   */

  ownProperty: function(name, description***REMOVED***{
    this.assert(
      hasOwnProperty.call(this.obj, name***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to have own property ' + should.inspect(name***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not have own property ' + should.inspect(name***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    this.obj = this.obj[name];
    return this;
***REMOVED***,

  /**
   * Assert that string starts with `str`.
   * @param {String} str
   * @param {String} description
   * @api public
   */

  startWith: function(str, description***REMOVED*** {
    this.assert(0 === this.obj.indexOf(str***REMOVED***
    , function(***REMOVED*** { return 'expected ' + this.inspect + ' to start with ' + should.inspect(str***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
    , function(***REMOVED*** { return 'expected ' + this.inspect + ' to not start with ' + should.inspect(str***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
    , void 0
    , void 0
    , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert that string ends with `str`.
   * @param {String} str
   * @param {String} description
   * @api public
   */

  endWith: function(str, description***REMOVED*** {
    this.assert(-1 !== this.obj.indexOf(str, this.obj.length - str.length***REMOVED***
    , function(***REMOVED*** { return 'expected ' + this.inspect + ' to end with ' + should.inspect(str***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
    , function(***REMOVED*** { return 'expected ' + this.inspect + ' to not end with ' + should.inspect(str***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
    , void 0
    , void 0
    , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert that `obj` is present via `.indexOf(***REMOVED***` or that `obj` contains some sub-object.
   *
   * @param {*} obj
   * @param {String} description
   * @api public
   */

  include: function(obj, description***REMOVED***{
    if (!Array.isArray(this.obj***REMOVED*** && !util.isString(this.obj***REMOVED******REMOVED***{
      var cmp = {};
      for (var key in obj***REMOVED*** cmp[key] = this.obj[key];
      this.assert(
          eql(cmp, obj***REMOVED***
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to include an object equal to ' + should.inspect(obj***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not include an object equal to ' + should.inspect(obj***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , void 0
        , void 0
        , description***REMOVED***;
  ***REMOVED*** else {
      this.assert(
          ~this.obj.indexOf(obj***REMOVED***
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to include ' + should.inspect(obj***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not include ' + should.inspect(obj***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
        , void 0
        , void 0
        , description***REMOVED***;
  ***REMOVED***
    return this;
***REMOVED***,

  /**
   * Assert that an object equal to `obj` is present.
   *
   * @param {Array} obj
   * @param {String} description
   * @api public
   */

  includeEql: function(obj, description***REMOVED***{
    this.assert(
      this.obj.some(function(item***REMOVED*** { return eql(obj, item***REMOVED***; }***REMOVED***
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to include an object equal to ' + should.inspect(obj***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not include an object equal to ' + should.inspect(obj***REMOVED*** + (description ? " | " + description : ""***REMOVED*** }
      , void 0
      , void 0
      , description***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert exact keys or inclusion of keys by using
   * the `.include` modifier.
   *
   * @param {Array|String ...} keys
   * @api public
   */

  keys: function(keys***REMOVED***{
    var str
      , ok = true;

    keys = keys instanceof Array
      ? keys
      : Array.prototype.slice.call(arguments***REMOVED***;

    if (!keys.length***REMOVED*** throw new Error('keys required'***REMOVED***;

    var actual = Object.keys(this.obj***REMOVED***
      , len = keys.length;

    // make sure they're all present
    ok = keys.every(function(key***REMOVED***{
      return ~actual.indexOf(key***REMOVED***;
  ***REMOVED******REMOVED***;

    // matching length
    ok = ok && keys.length == actual.length;

    // key string
    if (len > 1***REMOVED*** {
      keys = keys.map(function(key***REMOVED***{
        return should.inspect(key***REMOVED***;
    ***REMOVED******REMOVED***;
      var last = keys.pop(***REMOVED***;
      str = keys.join(', '***REMOVED*** + ', and ' + last;
  ***REMOVED*** else {
      str = should.inspect(keys[0]***REMOVED***;
  ***REMOVED***

    // message
    str = 'have ' + (len > 1 ? 'keys ' : 'key '***REMOVED*** + str;

    this.assert(
        ok
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to ' + str }
      , function(***REMOVED***{ return 'expected ' + this.inspect + ' to not ' + str }***REMOVED***;

    return this;
***REMOVED***,

  /**
   * Assert that header `field` has the given `val`.
   *
   * @param {String} field
   * @param {String} val
   * @return {Assertion} for chaining
   * @api public
   */

  header: function(field, val***REMOVED***{
    this.obj.should
      .have.property('headers'***REMOVED***.and
      .have.property(field.toLowerCase(***REMOVED***, val***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert `.statusCode` of `code`.
   *
   * @param {Number} code
   * @return {Assertion} for chaining
   * @api public
   */

  status:  function(code***REMOVED***{
    this.obj.should.have.property('statusCode'***REMOVED***;
    var status = this.obj.statusCode;

    this.assert(
        code == status
      , function(***REMOVED***{ return 'expected response code of ' + code + ' ' + should.inspect(statusCodes[code]***REMOVED***
        + ', but got ' + status + ' ' + should.inspect(statusCodes[status]***REMOVED*** }
      , function(***REMOVED***{ return 'expected to not respond with ' + code + ' ' + should.inspect(statusCodes[code]***REMOVED*** }***REMOVED***;

    return this;
***REMOVED***,

  /**
   * Assert that this response has content-type: application/json.
   *
   * @return {Assertion} for chaining
   * @api public
   */

  get json(***REMOVED*** {
    this.obj.should.have.property('headers'***REMOVED***;
    this.obj.headers.should.have.property('content-type'***REMOVED***;
    this.obj.headers['content-type'].should.include('application/json'***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert that this response has content-type: text/html.
   *
   * @return {Assertion} for chaining
   * @api public
   */

  get html(***REMOVED*** {
    this.obj.should.have.property('headers'***REMOVED***;
    this.obj.headers.should.have.property('content-type'***REMOVED***;
    this.obj.headers['content-type'].should.include('text/html'***REMOVED***;
    return this;
***REMOVED***,

  /**
   * Assert that this function will or will not
   * throw an exception.
   *
   * @return {Assertion} for chaining
   * @api public
   */

  throw: function(message***REMOVED***{
    var fn = this.obj
      , err = {}
      , errorInfo = ''
      , ok = true;

    try {
      fn(***REMOVED***;
      ok = false;
  ***REMOVED*** catch (e***REMOVED*** {
      err = e;
  ***REMOVED***

    if (ok***REMOVED*** {
      if ('string' == typeof message***REMOVED*** {
        ok = message == err.message;
    ***REMOVED*** else if (message instanceof RegExp***REMOVED*** {
        ok = message.test(err.message***REMOVED***;
    ***REMOVED*** else if ('function' == typeof message***REMOVED*** {
        ok = err instanceof message;
    ***REMOVED***

      if (message && !ok***REMOVED*** {
        if ('string' == typeof message***REMOVED*** {
          errorInfo = " with a message matching '" + message + "', but got '" + err.message + "'";
      ***REMOVED*** else if (message instanceof RegExp***REMOVED*** {
          errorInfo = " with a message matching " + message + ", but got '" + err.message + "'";
      ***REMOVED*** else if ('function' == typeof message***REMOVED*** {
          errorInfo = " of type " + message.name + ", but got " + err.constructor.name;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

    this.assert(
        ok
      , function(***REMOVED***{ return 'expected an exception to be thrown' + errorInfo }
      , function(***REMOVED***{ return 'expected no exception to be thrown, got "' + err.message + '"' }***REMOVED***;

    return this;
***REMOVED***
};

/**
 * Aliases.
 */

(function alias(name, as***REMOVED***{
  Assertion.prototype[as] = Assertion.prototype[name];
  return alias;
}***REMOVED***
('instanceof', 'instanceOf'***REMOVED***
('throw', 'throwError'***REMOVED***
('length', 'lengthOf'***REMOVED***
('keys', 'key'***REMOVED***
('ownProperty', 'haveOwnProperty'***REMOVED***
('above', 'greaterThan'***REMOVED***
('below', 'lessThan'***REMOVED***
('include', 'contain'***REMOVED***
('equal', 'exactly'***REMOVED***;
