// transformers/lib/transformers.js
var dirname = require('path'***REMOVED***.dirname;
var Transformer = require('./shared'***REMOVED***;

/**
 * minifiers must be first in order to be incorporated inside instances of respective output formats
 */
var uglifyJS = require('uglify-js'***REMOVED***;
exports.uglify = exports.uglifyJS = exports['uglify-js'] = new Transformer({
  name: 'uglify-js',
  engines: ['.'],
  outputFormat: 'js',
  isMinifier: true,
  sync: function (str, options***REMOVED*** {
    options.fromString = true;
    return this.cache(options***REMOVED*** || this.cache(options, uglifyJS.minify(str, options***REMOVED***.code***REMOVED***;
***REMOVED***
}***REMOVED***;
var uglifyCSS = require('css'***REMOVED***;
exports.uglifyCSS = exports['uglify-css'] = new Transformer({
  name: 'uglify-css',
  engines: ['.'],
  outputFormat: 'css',
  isMinifier: true,
  sync: function (str, options***REMOVED*** {
    options.compress = options.compress != false && options.beautify != true;
    return this.cache(options***REMOVED*** || this.cache(options, uglifyCSS.stringify(uglifyCSS.parse(str***REMOVED***, options***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports.uglifyJSON = exports['uglify-json'] = new Transformer({
  name: 'uglify-json',
  engines: ['.'],
  outputFormat: 'json',
  isMinifier: true,
  sync: function (str, options***REMOVED*** {
    return JSON.stringify(JSON.parse(str***REMOVED***, null, options.beautify***REMOVED***;
***REMOVED***
}***REMOVED***;


/**
 * Syncronous Templating Languages
 */

function sync(str, options***REMOVED*** {
  var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
  return tmpl(options***REMOVED***;
}

exports.swig = new Transformer({
  name: 'swig',
  engines: ['swig'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports.atpl = new Transformer({
  name: 'atpl',
  engines: ['atpl'],
  outputFormat: 'xml',
  sync: function sync(str, options***REMOVED*** {
    var tmpl = this.cache(options***REMOVED***;
    if (!tmpl***REMOVED*** {
      var cInfo = {cache: options.cache, filename: options.filename};
      if (options.filename***REMOVED*** {
        delete options.filename; //atpl can't handle absolute windows file paths properly
    ***REMOVED***
      tmpl = this.cache(cInfo, this.engine.compile(str, options***REMOVED******REMOVED***;
  ***REMOVED***
    return tmpl(options***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.dot = new Transformer({
  name: 'dot',
  engines: ['dot'],
  outputFormat: 'xml',
  sync: function sync(str, options***REMOVED*** {
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.template(str***REMOVED******REMOVED***;
    return tmpl(options***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.liquor = new Transformer({
  name: 'liquor',
  engines: ['liquor'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports.ejs = new Transformer({
  name: 'ejs',
  engines: ['ejs'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports.eco = new Transformer({
  name: 'eco',
  engines: ['eco'],
  outputFormat: 'xml',
  sync: sync//N.B. eco's internal this.cache isn't quite right but this bypasses it
}***REMOVED***;

exports.jqtpl = new Transformer({
  name: 'jqtpl',
  engines: ['jqtpl'],
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    var engine = this.engine;
    var key = (options.cache && options.filename***REMOVED*** ? options.filename : '@';
    engine.compile(str, key***REMOVED***;
    var res = this.engine.render(key, options***REMOVED***;
    if (!(options.cache && options.filename***REMOVED******REMOVED*** {
      delete engine.cache[key];
  ***REMOVED***
    this.cache(options, true***REMOVED***; // caching handled internally
    return res;
***REMOVED***
}***REMOVED***;

exports.haml = new Transformer({
  name: 'haml',
  engines: ['hamljs'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports['haml-coffee'] = new Transformer({
  name: 'haml-coffee',
  engines: ['haml-coffee'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports.whiskers = new Transformer({
  name: 'whiskers',
  engines: ['whiskers'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports.hogan = new Transformer({
  name: 'hogan',
  engines: ['hogan.js'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
    return tmpl.render(options, options.partials***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.handlebars = new Transformer({
  name: 'handlebars',
  engines: ['handlebars'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    for (var partial in options.partials***REMOVED*** {
      this.engine.registerPartial(partial, options.partials[partial]***REMOVED***;
  ***REMOVED***
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
    return tmpl(options***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.underscore = new Transformer({
  name: 'underscore',
  engines: ['underscore'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.template(str***REMOVED******REMOVED***;
    return tmpl(options***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.walrus = new Transformer({
  name: 'walrus',
  engines: ['walrus'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.parse(str***REMOVED******REMOVED***;
    return tmpl.compile(options***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.mustache = new Transformer({
  name: 'mustache',
  engines: ['mustache'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str***REMOVED******REMOVED***;
    return tmpl(options, options.partials***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.templayed = new Transformer({
  name: 'templayed',
  engines: ['templayed'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine(str***REMOVED******REMOVED***;
    return tmpl(options***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.plates = new Transformer({
  name: 'plates',
  engines: ['plates'],
  outputFormat: 'xml',
  sync: function(str, options***REMOVED***{
    str = this.cache(options***REMOVED*** || this.cache(options, str***REMOVED***;
    return this.engine.bind(str, options, options.map***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.mote = new Transformer({
  name: 'mote',
  engines: ['mote'],
  outputFormat: 'xml',
  sync: sync
}***REMOVED***;

exports.toffee = new Transformer({
  name: 'toffee',
  engines: ['toffee'],
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    var View = this.engine.view;
    var v = this.cache(options***REMOVED*** || this.cache(options, new View(str, options***REMOVED******REMOVED***;
    var res = v.run(options, require('vm'***REMOVED***.createContext({}***REMOVED******REMOVED***;
    if (res[0]***REMOVED*** throw res[0];
    else return res[1];
***REMOVED***
}***REMOVED***;

exports.coffeekup = exports.coffeecup = new Transformer({
  name: 'coffeecup',
  engines: ['coffeecup', 'coffeekup'],
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    var compiled = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
    return compiled(options***REMOVED***;
***REMOVED***
}***REMOVED***;

/**
 * Asyncronous Templating Languages
 */

exports.just = new Transformer({
  name: 'just',
  engines: ['just'],
  outputFormat: 'xml',
  sudoSync: true,
  async: function (str, options, cb***REMOVED*** {
    var JUST = this.engine;
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, new JUST({ root: { page: str }}***REMOVED******REMOVED***;
    tmpl.render('page', options, cb***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.ect = new Transformer({
  name: 'ect',
  engines: ['ect'],
  outputFormat: 'xml',
  sudoSync: true, // Always runs syncronously
  async: function (str, options, cb***REMOVED*** {
    var ECT = this.engine;
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, new ECT({ root: { page: str }}***REMOVED******REMOVED***;
    tmpl.render('page', options, cb***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.jade = new Transformer({
  name: 'jade',
  engines: ['jade', 'then-jade'],
  outputFormat: 'xml',
  sudoSync: 'The jade file FILENAME could not be rendered syncronously.  N.B. then-jade does not support syncronous rendering.',
  async: function (str, options, cb***REMOVED*** {
    this.cache(options, true***REMOVED***;//jade handles this.cache internally
    this.engine.render(str, options, cb***REMOVED***;
***REMOVED***
}***REMOVED***

exports.dust = new Transformer({
  name: 'dust',
  engines: ['dust', 'dustjs-linkedin'],
  outputFormat: 'xml',
  sudoSync: false,
  async: function (str, options, cb***REMOVED*** {
    var ext = 'dust'
      , views = '.';

    if (options***REMOVED*** {
      if (options.ext***REMOVED*** ext = options.ext;
      if (options.views***REMOVED*** views = options.views;
      if (options.settings && options.settings.views***REMOVED*** views = options.settings.views;
  ***REMOVED***

    this.engine.onLoad = function(path, callback***REMOVED***{
      if ('' == extname(path***REMOVED******REMOVED*** path += '.' + ext;
      if ('/' !== path[0]***REMOVED*** path = views + '/' + path;
      read(path, options, callback***REMOVED***;
  ***REMOVED***;

    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compileFn(str***REMOVED******REMOVED***;
    if (options && !options.cache***REMOVED*** this.engine.cache = {};//invalidate dust's internal cache
    tmpl(options, cb***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.jazz = new Transformer({
  name: 'jazz',
  engines: ['jazz'],
  outputFormat: 'xml',
  sudoSync: true, // except when an async function is passed to locals
  async: function (str, options, cb***REMOVED*** {
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
    tmpl.eval(options, function(str***REMOVED***{
      cb(null, str***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports.qejs = new Transformer({
  name: 'qejs',
  engines: ['qejs'],
  outputFormat: 'xml',
  sudoSync: false,
  async: function (str, options, cb***REMOVED*** {
    var tmpl = this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
    tmpl(options***REMOVED***.done(function (result***REMOVED*** {
        cb(null, result***REMOVED***;
  ***REMOVED***, function (err***REMOVED*** {
        cb(err***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

/**
 * Stylsheet Languages
 */

exports.less = new Transformer({
  name: 'less',
  engines: ['less'],
  outputFormat: 'css',
  sudoSync: 'The less file FILENAME could not be rendered syncronously.  This is usually because the file contains `@import url` statements.',
  async: function (str, options, cb***REMOVED*** {
    var self = this;
    if (self.cache(options***REMOVED******REMOVED*** return cb(null, self.cache(options***REMOVED******REMOVED***;
    if (options.filename***REMOVED*** {
      options.paths = options.paths || [dirname(options.filename***REMOVED***];
  ***REMOVED***
    //If this.cache is enabled, compress by default
    if (options.compress !== true && options.compress !== false***REMOVED*** {
      options.compress = options.cache || false;
  ***REMOVED***
    if (options.sudoSync***REMOVED*** {
      options.syncImport = true;
  ***REMOVED***
    var parser = new(this.engine.Parser***REMOVED***(options***REMOVED***;
    parser.parse(str, function (err, tree***REMOVED*** {
      try {
        if (err***REMOVED*** throw err;
        var res = tree.toCSS(options***REMOVED***;
        self.cache(options, res***REMOVED***;
        cb(null, res***REMOVED***;
    ***REMOVED*** catch (ex***REMOVED*** {
        if (ex.constructor.name === 'LessError' && typeof ex === 'object'***REMOVED*** {
          ex.filename = ex.filename || '"Unkown Source"';
          var err = new Error(self.engine.formatError(ex, options***REMOVED***.replace(/^[^:]+:/, ''***REMOVED***, ex.filename, ex.line***REMOVED***;
          err.name = ex.type;
          ex = err;
      ***REMOVED***
        return cb(ex***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports.styl = exports.stylus = new Transformer({
  name: 'stylus',
  engines: ['stylus'],
  outputFormat: 'css',
  sudoSync: true,// always runs syncronously
  async: function (str, options, cb***REMOVED*** {
    var self = this;
    if (self.cache(options***REMOVED******REMOVED*** return cb(null, self.cache(options***REMOVED******REMOVED***;
    if (options.filename***REMOVED*** {
      options.paths = options.paths || [dirname(options.filename***REMOVED***];
  ***REMOVED***
    //If this.cache is enabled, compress by default
    if (options.compress !== true && options.compress !== false***REMOVED*** {
      options.compress = options.cache || false;
  ***REMOVED***
    this.engine.render(str, options, function (err, res***REMOVED*** {
      if (err***REMOVED*** return cb(err***REMOVED***;
      self.cache(options, res***REMOVED***;
      cb(null, res***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***

exports.sass = new Transformer({
  name: 'sass',
  engines: ['sass'],
  outputFormat: 'css',
  sync: function (str, options***REMOVED*** {
    try {
      return this.cache(options***REMOVED*** || this.cache(options, this.engine.render(str***REMOVED******REMOVED***;
  ***REMOVED*** catch (ex***REMOVED*** {
      if (options.filename***REMOVED*** ex.message += ' in ' + options.filename;
      throw ex;
  ***REMOVED***
***REMOVED***
}***REMOVED***;

/**
 * Miscelaneous
 */

exports.md = exports.markdown = new Transformer({
  name: 'markdown',
  engines: ['marked', 'supermarked', 'markdown-js', 'markdown'],
  outputFormat: 'html',
  sync: function (str, options***REMOVED*** {
    var arg = options;
    if (this.engineName === 'markdown'***REMOVED*** arg = options.dialect; //even if undefined
    return this.cache(options***REMOVED*** || this.cache(options, this.engine.parse(str, arg***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;


exports.coffee = exports['coffee-script'] = exports.coffeescript = exports.coffeeScript = new Transformer({
  name: 'coffee-script',
  engines: ['coffee-script'],
  outputFormat: 'js',
  sync: function (str, options***REMOVED*** {
    return this.cache(options***REMOVED*** || this.cache(options, this.engine.compile(str, options***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports.cson = new Transformer({
  name: 'cson',
  engines: ['cson'],
  outputFormat: 'json',
  sync: function (str, options***REMOVED*** {
    //todo: remove once https://github.com/rstacruz/js2coffee/pull/174 accepted & released
    if (global.Narcissus***REMOVED*** delete global.Narcissus; //prevent global leak
    return this.cache(options***REMOVED*** || this.cache(options, JSON.stringify(this.engine.parseSync(str***REMOVED******REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports.cdata = new Transformer({
  name: 'cdata',
  engines: ['.'],// `.` means "no dependency"
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    var escaped = str.replace(/\]\]>/g, "]]]]><![CDATA[>"***REMOVED***;
    return this.cache(options***REMOVED*** || this.cache(options, '<![CDATA[' + escaped + ']]>'***REMOVED***;
***REMOVED***
}***REMOVED***;

exports["cdata-js"] = new Transformer({
  name: 'cdata-js',
  engines: ['.'],// `.` means "no dependency"
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    var escaped = str.replace(/\]\]>/g, "]]]]><![CDATA[>"***REMOVED***;
    return this.cache(options***REMOVED*** || this.cache(options, '//<![CDATA[\n' + escaped + '\n//]]>'***REMOVED***;
***REMOVED***
}***REMOVED***;

exports["cdata-css"] = new Transformer({
  name: 'cdata-css',
  engines: ['.'],// `.` means "no dependency"
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    var escaped = str.replace(/\]\]>/g, "]]]]><![CDATA[>"***REMOVED***;
    return this.cache(options***REMOVED*** || this.cache(options, '/*<![CDATA[*/\n' + escaped + '\n/*]]>*/'***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.verbatim = new Transformer({
  name: 'verbatim',
  engines: ['.'],// `.` means "no dependency"
  outputFormat: 'xml',
  sync: function (str, options***REMOVED*** {
    return this.cache(options***REMOVED*** || this.cache(options, str***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.component = exports['component-js'] = new Transformer({
  name: 'component-js',
  engines: ['component-builder'],
  outputFormat: 'js',
  async: function (str, options, cb***REMOVED*** {
    if (this.cache(options***REMOVED******REMOVED*** return this.cache(options***REMOVED***;
    var self = this;
    var builder = new this.engine(dirname(options.filename***REMOVED******REMOVED***;
    if (options.development***REMOVED*** {
      builder.development(***REMOVED***;
  ***REMOVED***
    if (options.sourceURLs === true || (options.sourceURLs !== false && options.development***REMOVED******REMOVED*** {
      builder.addSourceURLs(***REMOVED***;
  ***REMOVED***
    var ***REMOVED***;
    builder.paths = (options.paths || ['components']***REMOVED***.map(function (p***REMOVED*** {
      if (path.resolve(p***REMOVED*** === p***REMOVED*** {
        return p;
    ***REMOVED*** else {
        return path.join(dirname(options.filename***REMOVED***, p***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
    builder.build(function (err, obj***REMOVED*** {
      if (err***REMOVED*** return cb(err***REMOVED***;
      else return cb(null, self.cache(options, obj.require + obj.js***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports['component-css'] = new Transformer({
  name: 'component-css',
  engines: ['component-builder'],
  outputFormat: 'css',
  async: function (str, options, cb***REMOVED*** {
    if (this.cache(options***REMOVED******REMOVED*** return this.cache(options***REMOVED***;
    var self = this;
    var builder = new this.engine(dirname(options.filename***REMOVED******REMOVED***;
    if (options.development***REMOVED*** {
      builder.development(***REMOVED***;
  ***REMOVED***
    if (options.sourceURLs === true || (options.sourceURLs !== false && options.development***REMOVED******REMOVED*** {
      builder.addSourceURLs(***REMOVED***;
  ***REMOVED***
    var ***REMOVED***;
    builder.paths = (options.paths || ['components']***REMOVED***.map(function (p***REMOVED*** {
      if (path.resolve(p***REMOVED*** === p***REMOVED*** {
        return p;
    ***REMOVED*** else {
        return path.join(dirname(options.filename***REMOVED***, p***REMOVED***;
    ***REMOVED***
  ***REMOVED******REMOVED***;
    builder.build(function (err, obj***REMOVED*** {
      if (err***REMOVED*** return cb(err***REMOVED***;
      else return cb(null, self.cache(options, obj.css***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports['html2jade'] = new Transformer({
  name: 'html2jade',
  engines: ['html2jade'],
  outputFormat: 'jade',
  async: function (str, options, cb***REMOVED*** {
    return this.cache(options***REMOVED*** || this.cache(options, this.engine.convertHtml(str, options, cb***REMOVED******REMOVED***;
***REMOVED***
}***REMOVED***;

exports['highlight'] = new Transformer({
  name: 'highlight',
  engines: ['highlight.js'],
  outputFormat: 'xml',
  sync: function (str, options, cb***REMOVED*** {
    if (this.cache(options***REMOVED******REMOVED*** return this.cache(options***REMOVED***;
    if (options.lang***REMOVED*** {
      try {
        return this.cache(options, this.engine.highlight(options.lang, str***REMOVED***.value***REMOVED***;
    ***REMOVED*** catch (ex***REMOVED*** {}
  ***REMOVED***
    if (options.auto || !options.lang***REMOVED*** {
      try {
        return this.cache(options, this.engine.highlightAuto(str***REMOVED***.value***REMOVED***;
    ***REMOVED*** catch (ex***REMOVED*** {}
  ***REMOVED***
    return this.cache(options, str***REMOVED***;
***REMOVED***
}***REMOVED***;


/**
 * Marker transformers (they don't actually apply a transformation, but let you declare the 'outputFormat'***REMOVED***
 */

exports.css = new Transformer({
  name: 'css',
  engines: ['.'],// `.` means "no dependency"
  outputFormat: 'css',
  sync: function (str, options***REMOVED*** {
    return this.cache(options***REMOVED*** || this.cache(options, str***REMOVED***;
***REMOVED***
}***REMOVED***;

exports.js = new Transformer({
  name: 'js',
  engines: ['.'],// `.` means "no dependency"
  outputFormat: 'js',
  sync: function (str, options***REMOVED*** {
    return this.cache(options***REMOVED*** || this.cache(options, str***REMOVED***;
***REMOVED***
}***REMOVED***;
