// mongodb/lib/mongodb/commands/db_command.js
var QueryCommand = require('./query_command'***REMOVED***.QueryCommand,
  InsertCommand = require('./insert_command'***REMOVED***.InsertCommand,
  inherits = require('util'***REMOVED***.inherits,
  utils = require('../utils'***REMOVED***,
  crypto = require('crypto'***REMOVED***;

/**
  Db Command
**/
var DbCommand = exports.DbCommand = function(dbInstance, collectionName, queryOptions, numberToSkip, numberToReturn, query, returnFieldSelector, options***REMOVED*** {
  QueryCommand.call(this***REMOVED***;
  this.collectionName = collectionName;
  this.queryOptions = queryOptions;
  this.numberToSkip = numberToSkip;
  this.numberToReturn = numberToReturn;
  this.query = query;
  this.returnFieldSelector = returnFieldSelector;
  this.db = dbInstance;

  // Set the slave ok bit
  if(this.db && this.db.slaveOk***REMOVED*** {
    this.queryOptions |= QueryCommand.OPTS_SLAVE;
***REMOVED***

  // Make sure we don't get a null exception
  options = options == null ? {} : options;

  // Allow for overriding the BSON checkKeys function
  this.checkKeys = typeof options['checkKeys'] == 'boolean' ? options["checkKeys"] : true;

  // Let us defined on a command basis if we want functions to be serialized or not
  if(options['serializeFunctions'] != null && options['serializeFunctions']***REMOVED*** {
    this.serializeFunctions = true;
***REMOVED***
};

inherits(DbCommand, QueryCommand***REMOVED***;

// Constants
DbCommand.SYSTEM_NAMESPACE_COLLECTION = "system.namespaces";
DbCommand.SYSTEM_INDEX_COLLECTION = "system.indexes";
DbCommand.SYSTEM_PROFILE_COLLECTION = "system.profile";
DbCommand.SYSTEM_USER_COLLECTION = "system.users";
DbCommand.SYSTEM_COMMAND_COLLECTION = "$cmd";
DbCommand.SYSTEM_JS_COLLECTION = "system.js";

// New commands
DbCommand.NcreateIsMasterCommand = function(db, databaseName***REMOVED*** {
  return new DbCommand(db, databaseName + "." + DbCommand.SYSTEM_COMMAND_COLLECTION, QueryCommand.OPTS_NO_CURSOR_TIMEOUT, 0, -1, {'ismaster':1}, null***REMOVED***;
};

// Provide constructors for different db commands
DbCommand.createIsMasterCommand = function(db***REMOVED*** {
  return new DbCommand(db, db.databaseName + "." + DbCommand.SYSTEM_COMMAND_COLLECTION, QueryCommand.OPTS_NO_CURSOR_TIMEOUT, 0, -1, {'ismaster':1}, null***REMOVED***;
};

DbCommand.createGetLastErrorCommand = function(options, db***REMOVED*** {
  if (typeof db === 'undefined'***REMOVED*** {
    db =  options;
    options = {};
***REMOVED***
  // Final command
  var command = {'getlasterror':1};
  // If we have an options Object let's merge in the fields (fsync/wtimeout/w***REMOVED***
  if('object' === typeof options***REMOVED*** {
    for(var name in options***REMOVED*** {
      command[name] = options[name]
  ***REMOVED***
***REMOVED***

  // Special case for w == 1, remove the w
  if(1 == command.w***REMOVED*** {
    delete command.w;
***REMOVED***

  // Execute command
  return new DbCommand(db, db.databaseName + "." + DbCommand.SYSTEM_COMMAND_COLLECTION, QueryCommand.OPTS_NO_CURSOR_TIMEOUT, 0, -1, command, null***REMOVED***;
};

DbCommand.createGetLastStatusCommand = DbCommand.createGetLastErrorCommand;

DbCommand.createDbCommand = function(db, command_hash, options, auth_db***REMOVED*** {
  var db_name = (auth_db ? auth_db : db.databaseName***REMOVED*** + "." + DbCommand.SYSTEM_COMMAND_COLLECTION;
  options = options == null ? {checkKeys: false} : options;
  return new DbCommand(db, db_name, QueryCommand.OPTS_NO_CURSOR_TIMEOUT, 0, -1, command_hash, null, options***REMOVED***;
};

DbCommand.createAdminDbCommand = function(db, command_hash***REMOVED*** {
  return new DbCommand(db, "admin." + DbCommand.SYSTEM_COMMAND_COLLECTION, QueryCommand.OPTS_NO_CURSOR_TIMEOUT, 0, -1, command_hash, null***REMOVED***;
};

DbCommand.createAdminDbCommandSlaveOk = function(db, command_hash***REMOVED*** {
  return new DbCommand(db, "admin." + DbCommand.SYSTEM_COMMAND_COLLECTION, QueryCommand.OPTS_NO_CURSOR_TIMEOUT | QueryCommand.OPTS_SLAVE, 0, -1, command_hash, null***REMOVED***;
};

DbCommand.createDbSlaveOkCommand = function(db, command_hash, options***REMOVED*** {
  options = options == null ? {checkKeys: false} : options;
  var dbName = options.dbName ? options.dbName : db.databaseName;
  var flags = options.slaveOk ? QueryCommand.OPTS_NO_CURSOR_TIMEOUT | QueryCommand.OPTS_SLAVE : QueryCommand.OPTS_NO_CURSOR_TIMEOUT;
  return new DbCommand(db, dbName + "." + DbCommand.SYSTEM_COMMAND_COLLECTION, flags, 0, -1, command_hash, null, options***REMOVED***;
};
