// mongodb/lib/mongodb/commands/get_more_command.js
var BaseCommand = require('./base_command'***REMOVED***.BaseCommand,
  inherits = require('util'***REMOVED***.inherits,
  binaryutils = require('../utils'***REMOVED***;

/**
  Get More Document Command
**/
var GetMoreCommand = exports.GetMoreCommand = function(db, collectionName, numberToReturn, cursorId***REMOVED*** {
  BaseCommand.call(this***REMOVED***;

  this.collectionName = collectionName;
  this.numberToReturn = numberToReturn;
  this.cursorId = cursorId;
  this.db = db;
};

inherits(GetMoreCommand, BaseCommand***REMOVED***;

GetMoreCommand.OP_GET_MORE = 2005;

GetMoreCommand.prototype.toBinary = function(***REMOVED*** {
  // Validate that we are not passing 0x00 in the colletion name
  if(!!~this.collectionName.indexOf("\x00"***REMOVED******REMOVED*** {
    throw new Error("namespace cannot contain a null character"***REMOVED***;
***REMOVED***

  // Calculate total length of the document
  var totalLengthOfCommand = 4 + Buffer.byteLength(this.collectionName***REMOVED*** + 1 + 4 + 8 + (4 * 4***REMOVED***;
  // Let's build the single pass buffer command
  var _index = 0;
  var _command = new Buffer(totalLengthOfCommand***REMOVED***;
  // Write the header information to the buffer
  _command[_index++] = totalLengthOfCommand & 0xff;
  _command[_index++] = (totalLengthOfCommand >> 8***REMOVED*** & 0xff;
  _command[_index++] = (totalLengthOfCommand >> 16***REMOVED*** & 0xff;
  _command[_index++] = (totalLengthOfCommand >> 24***REMOVED*** & 0xff;     

  // Write the request ID
  _command[_index++] = this.requestId & 0xff;
  _command[_index++] = (this.requestId >> 8***REMOVED*** & 0xff;
  _command[_index++] = (this.requestId >> 16***REMOVED*** & 0xff;
  _command[_index++] = (this.requestId >> 24***REMOVED*** & 0xff;     

  // Write zero
  _command[_index++] = 0;
  _command[_index++] = 0;
  _command[_index++] = 0;
  _command[_index++] = 0;

  // Write the op_code for the command
  _command[_index++] = GetMoreCommand.OP_GET_MORE & 0xff;
  _command[_index++] = (GetMoreCommand.OP_GET_MORE >> 8***REMOVED*** & 0xff;
  _command[_index++] = (GetMoreCommand.OP_GET_MORE >> 16***REMOVED*** & 0xff;
  _command[_index++] = (GetMoreCommand.OP_GET_MORE >> 24***REMOVED*** & 0xff;     

  // Write zero
  _command[_index++] = 0;
  _command[_index++] = 0;
  _command[_index++] = 0;
  _command[_index++] = 0;

  // Write the collection name to the command
  _index = _index + _command.write(this.collectionName, _index, 'utf8'***REMOVED*** + 1;
  _command[_index - 1] = 0;    

  // Number of documents to return
  _command[_index++] = this.numberToReturn & 0xff;
  _command[_index++] = (this.numberToReturn >> 8***REMOVED*** & 0xff;
  _command[_index++] = (this.numberToReturn >> 16***REMOVED*** & 0xff;
  _command[_index++] = (this.numberToReturn >> 24***REMOVED*** & 0xff;     
  
  // Encode the cursor id
  var low_bits = this.cursorId.getLowBits(***REMOVED***;
  // Encode low bits
  _command[_index++] = low_bits & 0xff;
  _command[_index++] = (low_bits >> 8***REMOVED*** & 0xff;
  _command[_index++] = (low_bits >> 16***REMOVED*** & 0xff;
  _command[_index++] = (low_bits >> 24***REMOVED*** & 0xff;     
  
  var high_bits = this.cursorId.getHighBits(***REMOVED***;
  // Encode high bits
  _command[_index++] = high_bits & 0xff;
  _command[_index++] = (high_bits >> 8***REMOVED*** & 0xff;
  _command[_index++] = (high_bits >> 16***REMOVED*** & 0xff;
  _command[_index++] = (high_bits >> 24***REMOVED*** & 0xff;     
  // Return command
  return _command;
};
