// morgan/index.js
/*!
 * Connect - logger
 * Copyright(c***REMOVED*** 2010 Sencha Inc.
 * Copyright(c***REMOVED*** 2011 TJ Holowaychuk
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var bytes = require('bytes'***REMOVED***;

/*!
 * Log buffer.
 */

var buf = [];

/*!
 * Default log buffer duration.
 */

var defaultBufferDuration = 1000;

/**
 * Logger:
 *
 * Log requests with the given `options` or a `format` string.
 *
 * Options:
 *
 *   - `format`  Format string, see below for tokens
 *   - `stream`  Output stream, defaults to _stdout_
 *   - `buffer`  Buffer duration, defaults to 1000ms when _true_
 *   - `immediate`  Write log line on request instead of response (for response times***REMOVED***
 *   - `skip`    Function to determine if logging is skipped, called as
 *               `skip(req, res***REMOVED***`, defaults to always false.
 *
 * Tokens:
 *
 *   - `:req[header]` ex: `:req[Accept]`
 *   - `:res[header]` ex: `:res[Content-Length]`
 *   - `:http-version`
 *   - `:response-time`
 *   - `:remote-addr`
 *   - `:date`
 *   - `:method`
 *   - `:url`
 *   - `:referrer`
 *   - `:user-agent`
 *   - `:status`
 *
 * Formats:
 *
 *   Pre-defined formats that ship with connect:
 *
 *    - `default` ':remote-addr - - [:date] ":method :url HTTP/:http-version" :status :res[content-length] ":referrer" ":user-agent"'
 *    - `short` ':remote-addr - :method :url HTTP/:http-version :status :res[content-length] - :response-time ms'
 *    - `tiny`  ':method :url :status :res[content-length] - :response-time ms'
 *    - `dev` concise output colored by response status for development use
 *
 * Examples:
 *
 *      connect.logger(***REMOVED*** // default
 *      connect.logger('short'***REMOVED***
 *      connect.logger('tiny'***REMOVED***
 *      connect.logger({ immediate: true, format: 'dev' }***REMOVED***
 *      connect.logger(':method :url - :referrer'***REMOVED***
 *      connect.logger(':req[content-type] -> :res[content-type]'***REMOVED***
 *      connect.logger(function(tokens, req, res***REMOVED***{ return 'some format string' }***REMOVED***
 *      connect.logger({ format: 'dev', skip: function(req, res***REMOVED***{ return res.statusCode === 304; }}***REMOVED***
 *
 * Defining Tokens:
 *
 *   To define a token, simply invoke `connect.logger.token(***REMOVED***` with the
 *   name and a callback function. The value returned is then available
 *   as ":type" in this case.
 *
 *      connect.logger.token('type', function(req, res***REMOVED***{ return req.headers['content-type']; }***REMOVED***
 *
 * Defining Formats:
 *
 *   All default formats are defined this way, however it's public API as well:
 *
 *       connect.logger.format('name', 'string or function'***REMOVED***
 *
 * @param {String|Function|Object} format or options
 * @return {Function}
 * @api public
 */

exports = module.exports = function logger(options***REMOVED*** {
  if ('object' == typeof options***REMOVED*** {
    options = options || {};
***REMOVED*** else if (options***REMOVED*** {
    options = { format: options };
***REMOVED*** else {
    options = {};
***REMOVED***

  // output on request instead of response
  var immediate = options.immediate;

  // check if log entry should be skipped
  var skip = options.skip || function (***REMOVED*** { return false; };

  // format name
  var fmt = exports[options.format] || options.format || exports.default;

  // compile format
  if ('function' != typeof fmt***REMOVED*** fmt = compile(fmt***REMOVED***;

  // options
  var stream = options.stream || process.stdout
    , buffer = options.buffer;

  // buffering support
  if (buffer***REMOVED*** {
    var realStream = stream
      , interval = 'number' == typeof buffer
        ? buffer
        : defaultBufferDuration;

    // flush interval
    setInterval(function(***REMOVED***{
      if (buf.length***REMOVED*** {
        realStream.write(buf.join(''***REMOVED******REMOVED***;
        buf.length = 0;
    ***REMOVED***
  ***REMOVED***, interval***REMOVED***;

    // swap the stream
    stream = {
      write: function(str***REMOVED***{
        buf.push(str***REMOVED***;
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

  return function logger(req, res, next***REMOVED*** {
    var sock = req.socket;
    req._startTime = new Date;
    req._remoteAddress = sock.socket ? sock.socket.remoteAddress : sock.remoteAddress;

    function logRequest(***REMOVED***{
      res.removeListener('finish', logRequest***REMOVED***;
      res.removeListener('close', logRequest***REMOVED***;
      if (skip(req, res***REMOVED******REMOVED*** return;
      var line = fmt(exports, req, res***REMOVED***;
      if (null == line***REMOVED*** return;
      stream.write(line + '\n'***REMOVED***;
  ***REMOVED***;

    // immediate
    if (immediate***REMOVED*** {
      logRequest(***REMOVED***;
    // proxy end to output logging
  ***REMOVED*** else {
      res.on('finish', logRequest***REMOVED***;
      res.on('close', logRequest***REMOVED***;
  ***REMOVED***


    next(***REMOVED***;
***REMOVED***;
};

/**
 * Compile `fmt` into a function.
 *
 * @param {String} fmt
 * @return {Function}
 * @api private
 */

function compile(fmt***REMOVED*** {
  fmt = fmt.replace(/"/g, '\\"'***REMOVED***;
  var js = '  return "' + fmt.replace(/:([-\w]{2,}***REMOVED***(?:\[([^\]]+***REMOVED***\]***REMOVED***?/g, function(_, name, arg***REMOVED***{
    return '"\n    + (tokens["' + name + '"](req, res, "' + arg + '"***REMOVED*** || "-"***REMOVED*** + "';
***REMOVED******REMOVED*** + '";'
  return new Function('tokens, req, res', js***REMOVED***;
};

/**
 * Define a token function with the given `name`,
 * and callback `fn(req, res***REMOVED***`.
 *
 * @param {String} name
 * @param {Function} fn
 * @return {Object} exports for chaining
 * @api public
 */

exports.token = function(name, fn***REMOVED*** {
  exports[name] = fn;
  return this;
};

/**
 * Define a `fmt` with the given `name`.
 *
 * @param {String} name
 * @param {String|Function} fmt
 * @return {Object} exports for chaining
 * @api public
 */

exports.format = function(name, str***REMOVED***{
  exports[name] = str;
  return this;
};

/**
 * Default format.
 */

exports.format('default', ':remote-addr - - [:date] ":method :url HTTP/:http-version" :status :res[content-length] ":referrer" ":user-agent"'***REMOVED***;

/**
 * Short format.
 */

exports.format('short', ':remote-addr - :method :url HTTP/:http-version :status :res[content-length] - :response-time ms'***REMOVED***;

/**
 * Tiny format.
 */

exports.format('tiny', ':method :url :status :res[content-length] - :response-time ms'***REMOVED***;

/**
 * dev (colored***REMOVED***
 */

exports.format('dev', function(tokens, req, res***REMOVED***{
  var status = res.statusCode
    , len = parseInt(res.getHeader('Content-Length'***REMOVED***, 10***REMOVED***
    , color = 32;

  if (status >= 500***REMOVED*** color = 31
  else if (status >= 400***REMOVED*** color = 33
  else if (status >= 300***REMOVED*** color = 36;

  len = isNaN(len***REMOVED***
    ? ''
    : len = ' - ' + bytes(len***REMOVED***;

  return '\x1b[90m' + req.method
    + ' ' + (req.originalUrl || req.url***REMOVED*** + ' '
    + '\x1b[' + color + 'm' + res.statusCode
    + ' \x1b[90m'
    + (new Date - req._startTime***REMOVED***
    + 'ms' + len
    + '\x1b[0m';
}***REMOVED***;

/**
 * request url
 */

exports.token('url', function(req***REMOVED***{
  return req.originalUrl || req.url;
}***REMOVED***;

/**
 * request method
 */

exports.token('method', function(req***REMOVED***{
  return req.method;
}***REMOVED***;

/**
 * response time in milliseconds
 */

exports.token('response-time', function(req***REMOVED***{
  return String(Date.now(***REMOVED*** - req._startTime***REMOVED***;
}***REMOVED***;

/**
 * UTC date
 */

exports.token('date', function(***REMOVED***{
  return new Date(***REMOVED***.toUTCString(***REMOVED***;
}***REMOVED***;

/**
 * response status code
 */

exports.token('status', function(req, res***REMOVED***{
  return res.headersSent ? res.statusCode : null;
}***REMOVED***;

/**
 * normalized referrer
 */

exports.token('referrer', function(req***REMOVED***{
  return req.headers['referer'] || req.headers['referrer'];
}***REMOVED***;

/**
 * remote address
 */

exports.token('remote-addr', function(req***REMOVED***{
  if (req.ip***REMOVED*** return req.ip;
  if (req._remoteAddress***REMOVED*** return req._remoteAddress;
  var sock = req.socket;
  if (sock.socket***REMOVED*** return sock.socket.remoteAddress;
  return sock.remoteAddress;
}***REMOVED***;

/**
 * HTTP version
 */

exports.token('http-version', function(req***REMOVED***{
  return req.httpVersionMajor + '.' + req.httpVersionMinor;
}***REMOVED***;

/**
 * UA string
 */

exports.token('user-agent', function(req***REMOVED***{
  return req.headers['user-agent'];
}***REMOVED***;

/**
 * request header
 */

exports.token('req', function(req, res, field***REMOVED***{
  return req.headers[field.toLowerCase(***REMOVED***];
}***REMOVED***;

/**
 * response header
 */

exports.token('res', function(req, res, field***REMOVED***{
  return (res._headers || {}***REMOVED***[field.toLowerCase(***REMOVED***];
}***REMOVED***;
