// stylus/lib/nodes/function.js

/*!
 * Stylus - Function
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***;

/**
 * Initialize a new `Function` with `name`, `params`, and `body`.
 *
 * @param {String} name
 * @param {Params|Function} params
 * @param {Block} body
 * @api public
 */

var Function = module.exports = function Function(name, params, body***REMOVED***{
  Node.call(this***REMOVED***;
  this.name = name;
  this.params = params;
  this.block = body;
  if ('function' == typeof params***REMOVED*** this.fn = params;
};

/**
 * Check function arity.
 *
 * @return {Boolean}
 * @api public
 */

Function.prototype.__defineGetter__('arity', function(***REMOVED***{
  return this.params.length;
}***REMOVED***;

/**
 * Inherit from `Node.prototype`.
 */

Function.prototype.__proto__ = Node.prototype;

/**
 * Return hash.
 *
 * @return {String}
 * @api public
 */

Function.prototype.__defineGetter__('hash', function(***REMOVED***{
  return 'function ' + this.name;
}***REMOVED***;

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

Function.prototype.clone = function(***REMOVED***{
  if (this.fn***REMOVED*** {
    var clone = new Function(
        this.name
      , this.fn***REMOVED***;
***REMOVED*** else {
    var clone = new Function(
        this.name
      , this.params.clone(***REMOVED***
      , this.block.clone(***REMOVED******REMOVED***;
***REMOVED***
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  return clone;
};

/**
 * Return <name>(param1, param2, ...***REMOVED***.
 *
 * @return {String}
 * @api public
 */

Function.prototype.toString = function(***REMOVED***{
  if (this.fn***REMOVED*** {
    return this.name
      + '('
      + this.fn.toString(***REMOVED***
        .match(/^function *\w*\((.*?***REMOVED***\***REMOVED***/***REMOVED***
        .slice(1***REMOVED***
        .join(', '***REMOVED***
      + '***REMOVED***';
***REMOVED*** else {
    return this.name
      + '('
      + this.params.nodes.join(', '***REMOVED***
      + '***REMOVED***';
***REMOVED***
};
