// stylus/lib/nodes/hsla.js

/*!
 * Stylus - HSLA
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***
  , nodes = require('./'***REMOVED***;

/**
 * Initialize a new `HSLA` with the given h,s,l,a component values.
 *
 * @param {Number} h
 * @param {Number} s
 * @param {Number} l
 * @param {Number} a
 * @api public
 */

var HSLA = exports = module.exports = function HSLA(h,s,l,a***REMOVED***{
  Node.call(this***REMOVED***;
  this.h = clampDegrees(h***REMOVED***;
  this.s = clampPercentage(s***REMOVED***;
  this.l = clampPercentage(l***REMOVED***;
  this.a = clampAlpha(a***REMOVED***;
  this.hsla = this;
};

/**
 * Inherit from `Node.prototype`.
 */

HSLA.prototype.__proto__ = Node.prototype;

/**
 * Return hsla(n,n,n,n***REMOVED***.
 *
 * @return {String}
 * @api public
 */

HSLA.prototype.toString = function(***REMOVED***{
  return 'hsla('
    + this.h + ','
    + this.s.toFixed(0***REMOVED*** + ','
    + this.l.toFixed(0***REMOVED*** + ','
    + this.a + '***REMOVED***';
};

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

HSLA.prototype.clone = function(***REMOVED***{
  var clone = new HSLA(
      this.h
    , this.s
    , this.l
    , this.a***REMOVED***;
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  return clone;
};

/**
 * Return rgba `RGBA` representation.
 *
 * @return {RGBA}
 * @api public
 */

HSLA.prototype.__defineGetter__('rgba', function(***REMOVED***{
  return nodes.RGBA.fromHSLA(this***REMOVED***;
}***REMOVED***;

/**
 * Return hash.
 *
 * @return {String}
 * @api public
 */

HSLA.prototype.__defineGetter__('hash', function(***REMOVED***{
  return this.rgba.toString(***REMOVED***;
}***REMOVED***;

/**
 * Add h,s,l to the current component values.
 *
 * @param {Number} h
 * @param {Number} s
 * @param {Number} l
 * @return {HSLA} new node
 * @api public
 */

HSLA.prototype.add = function(h,s,l***REMOVED***{
  return new HSLA(
      this.h + h
    , this.s + s
    , this.l + l
    , this.a***REMOVED***;
};

/**
 * Subtract h,s,l from the current component values.
 *
 * @param {Number} h
 * @param {Number} s
 * @param {Number} l
 * @return {HSLA} new node
 * @api public
 */

HSLA.prototype.sub = function(h,s,l***REMOVED***{
  return this.add(-h, -s, -l***REMOVED***;
};

/**
 * Operate on `right` with the given `op`.
 *
 * @param {String} op
 * @param {Node} right
 * @return {Node}
 * @api public
 */

HSLA.prototype.operate = function(op, right***REMOVED***{
  switch (op***REMOVED*** {
    case '==':
    case '!=':
    case '<=':
    case '>=':
    case '<':
    case '>':
    case 'is a':
    case '||':
    case '&&':
      return this.rgba.operate(op, right***REMOVED***;
    default:
      return this.rgba.operate(op, right***REMOVED***.hsla;
***REMOVED***
};

/**
 * Return `HSLA` representation of the given `color`.
 *
 * @param {RGBA} color
 * @return {HSLA}
 * @api public
 */

exports.fromRGBA = function(rgba***REMOVED***{
  var r = rgba.r / 255
    , g = rgba.g / 255
    , b = rgba.b / 255
    , a = rgba.a;

  var min = Math.min(r,g,b***REMOVED***
    , max = Math.max(r,g,b***REMOVED***
    , l = (max + min***REMOVED*** / 2
    , d = max - min
    , h, s;

  switch (max***REMOVED*** {
    case min: h = 0; break;
    case r: h = 60 * (g-b***REMOVED*** / d; break;
    case g: h = 60 * (b-r***REMOVED*** / d + 120; break;
    case b: h = 60 * (r-g***REMOVED*** / d + 240; break;
***REMOVED***

  if (max == min***REMOVED*** {
    s = 0;
***REMOVED*** else if (l < .5***REMOVED*** {
    s = d / (2 * l***REMOVED***;
***REMOVED*** else {
    s = d / (2 - 2 * l***REMOVED***;
***REMOVED***

  h %= 360;
  s *= 100;
  l *= 100;

  return new HSLA(h,s,l,a***REMOVED***;
};

/**
 * Adjust lightness by `percent`.
 *
 * @param {Number} percent
 * @return {HSLA} for chaining
 * @api public
 */

HSLA.prototype.adjustLightness = function(percent***REMOVED***{
  this.l = clampPercentage(this.l + this.l * (percent / 100***REMOVED******REMOVED***;
  return this;
};

/**
 * Adjust hue by `deg`.
 *
 * @param {Number} deg
 * @return {HSLA} for chaining
 * @api public
 */

HSLA.prototype.adjustHue = function(deg***REMOVED***{
  this.h = clampDegrees(this.h + deg***REMOVED***;
  return this;
};

/**
 * Clamp degree `n` >= 0 and <= 360.
 *
 * @param {Number} n
 * @return {Number}
 * @api private
 */

function clampDegrees(n***REMOVED*** {
  n = n % 360;
  return n >= 0 ? n : 360 + n;
}

/**
 * Clamp percentage `n` >= 0 and <= 100.
 *
 * @param {Number} n
 * @return {Number}
 * @api private
 */

function clampPercentage(n***REMOVED*** {
  return Math.max(0, Math.min(n, 100***REMOVED******REMOVED***;
}

/**
 * Clamp alpha `n` >= 0 and <= 1.
 *
 * @param {Number} n
 * @return {Number}
 * @api private
 */

function clampAlpha(n***REMOVED*** {
  return Math.max(0, Math.min(n, 1***REMOVED******REMOVED***;
}
