// stylus/lib/nodes/ident.js

/*!
 * Stylus - Ident
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***
  , nodes = require('./'***REMOVED***;

/**
 * Initialize a new `Ident` by `name` with the given `val` node.
 *
 * @param {String} name
 * @param {Node} val
 * @api public
 */

var Ident = module.exports = function Ident(name, val, mixin***REMOVED***{
  Node.call(this***REMOVED***;
  this.name = name;
  this.string = name;
  this.val = val || nodes.null;
  this.mixin = !!mixin;
};

/**
 * Check if the variable has a value.
 *
 * @return {Boolean}
 * @api public
 */

Ident.prototype.__defineGetter__('isEmpty', function(***REMOVED***{
  return undefined == this.val;
}***REMOVED***;

/**
 * Return hash.
 *
 * @return {String}
 * @api public
 */

Ident.prototype.__defineGetter__('hash', function(***REMOVED***{
  return this.name;
}***REMOVED***;

/**
 * Inherit from `Node.prototype`.
 */

Ident.prototype.__proto__ = Node.prototype;

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

Ident.prototype.clone = function(***REMOVED***{
  var clone = new Ident(this.name, this.val.clone(***REMOVED***, this.mixin***REMOVED***;
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  clone.property = this.property;
  return clone;
};

/**
 * Return <name>.
 *
 * @return {String}
 * @api public
 */

Ident.prototype.toString = function(***REMOVED***{
  return this.name;
};

/**
 * Coerce `other` to an ident.
 *
 * @param {Node} other
 * @return {String}
 * @api public
 */

Ident.prototype.coerce = function(other***REMOVED***{
  switch (other.nodeName***REMOVED*** {
    case 'ident':
    case 'string':
    case 'literal':
      return new Ident(other.string***REMOVED***;
    default:
      return Node.prototype.coerce.call(this, other***REMOVED***;
***REMOVED***
};

/**
 * Operate on `right` with the given `op`.
 *
 * @param {String} op
 * @param {Node} right
 * @return {Node}
 * @api public
 */

Ident.prototype.operate = function(op, right***REMOVED***{
  var val = right.first;
  switch (op***REMOVED*** {
    case '-':
      if ('unit' == val.nodeName***REMOVED*** {
        var expr = new nodes.Expression;
        val = val.clone(***REMOVED***;
        val.val = -val.val;
        expr.push(this***REMOVED***;
        expr.push(val***REMOVED***;
        return expr;
    ***REMOVED***
    case '+':
      return new nodes.Ident(this.string + this.coerce(val***REMOVED***.string***REMOVED***;
***REMOVED***
  return Node.prototype.operate.call(this, op, right***REMOVED***;
};
