// stylus/lib/nodes/if.js

/*!
 * Stylus - If
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***;

/**
 * Initialize a new `If` with the given `cond`.
 *
 * @param {Expression} cond
 * @param {Boolean|Block} negate, block
 * @api public
 */

var If = module.exports = function If(cond, negate***REMOVED***{
  Node.call(this***REMOVED***;
  this.cond = cond;
  this.elses = [];
  if (negate && negate.nodeName***REMOVED*** {
    this.block = negate;
***REMOVED*** else {
    this.negate = negate;
***REMOVED***
};

/**
 * Inherit from `Node.prototype`.
 */

If.prototype.__proto__ = Node.prototype;

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

If.prototype.clone = function(***REMOVED***{
  var cond = this.cond.clone(***REMOVED***
    , block = this.block.clone(***REMOVED***;
  var clone = new If(cond, block***REMOVED***;
  clone.elses = this.elses.map(function(node***REMOVED***{ return node.clone(***REMOVED***; }***REMOVED***;
  clone.negate = this.negate;
  clone.postfix = this.postfix;
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  return clone;
};
