// stylus/lib/nodes/property.js

/*!
 * Stylus - Property
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***;

/**
 * Initialize a new `Property` with the given `segs` and optional `expr`.
 *
 * @param {Array} segs
 * @param {Expression} expr
 * @api public
 */

var Property = module.exports = function Property(segs, expr***REMOVED***{
  Node.call(this***REMOVED***;
  this.segments = segs;
  this.expr = expr;
};

/**
 * Inherit from `Node.prototype`.
 */

Property.prototype.__proto__ = Node.prototype;

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

Property.prototype.clone = function(***REMOVED***{
  var clone = new Property(this.segments***REMOVED***;
  clone.name = this.name;
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  clone.segments = this.segments.map(function(node***REMOVED***{ return node.clone(***REMOVED***; }***REMOVED***;
  if (this.expr***REMOVED*** clone.expr = this.expr.clone(***REMOVED***;
  return clone;
};

/**
 * Return string representation of this node.
 *
 * @return {String}
 * @api public
 */

Property.prototype.toString = function(***REMOVED***{
  return 'property(' + this.segments.join(''***REMOVED*** + ', ' + this.expr + '***REMOVED***';
};

/**
 * Operate on the property expression.
 *
 * @param {String} op
 * @param {Node} right
 * @return {Node}
 * @api public
 */

Property.prototype.operate = function(op, right, val***REMOVED***{
  return this.expr.operate(op, right, val***REMOVED***;
};
