// stylus/lib/nodes/query.js

/*!
 * Stylus - Query
 * Copyright(c***REMOVED*** 2014 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***;

/**
 * Initialize a new `Query`.
 *
 * @param {QueryExpr} query
 * @api public
 */

var Query = module.exports = function Query(query***REMOVED***{
  Node.call(this***REMOVED***;
  this.nodes = query ? [query] : [];
  this.predicate = '';
};

/**
 * Inherit from `Node.prototype`.
 */

Query.prototype.__proto__ = Node.prototype;

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

Query.prototype.clone = function(***REMOVED***{
  var clone = new Query;
  clone.predicate = this.predicate;
  for (var i = 0, len = this.nodes.length; i < len; ++i***REMOVED*** {
    clone.push(this.nodes[i].clone(***REMOVED******REMOVED***;
***REMOVED***
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  return clone;
};

/**
 * Push the given `expr`.
 *
 * @param {QueryExpr} expr
 * @api public
 */

Query.prototype.push = function(expr***REMOVED***{
  this.nodes.push(expr***REMOVED***;
};

/**
 * Return "<a> and <b> and <c>"
 *
 * @return {String}
 * @api public
 */

Query.prototype.toString = function(***REMOVED***{
  var pred = this.predicate ? this.predicate + ' ' : '';
  return pred + this.nodes.map(function(expr***REMOVED***{
    return expr.toString(***REMOVED***;
***REMOVED******REMOVED***.join(' and '***REMOVED***;
};
