// stylus/lib/nodes/string.js
/*!
 * Stylus - String
 * Copyright(c***REMOVED*** 2010 LearnBoost <dev@learnboost.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var Node = require('./node'***REMOVED***
  , sprintf = require('../functions'***REMOVED***.s
  , utils = require('../utils'***REMOVED***
  , nodes = require('./'***REMOVED***;

/**
 * Initialize a new `String` with the given `val`.
 *
 * @param {String} val
 * @param {String} quote
 * @api public
 */

var String = module.exports = function String(val, quote***REMOVED***{
  Node.call(this***REMOVED***;
  this.val = val;
  this.string = val;
  this.prefixed = false;
  if (typeof quote !== 'string'***REMOVED*** {
    this.quote = "'";
***REMOVED*** else {
    this.quote = quote;
***REMOVED***
};

/**
 * Inherit from `Node.prototype`.
 */

String.prototype.__proto__ = Node.prototype;

/**
 * Return quoted string.
 *
 * @return {String}
 * @api public
 */

String.prototype.toString = function(***REMOVED***{
  return this.quote + this.val + this.quote;
};

/**
 * Return a clone of this node.
 * 
 * @return {Node}
 * @api public
 */

String.prototype.clone = function(***REMOVED***{
  var clone = new String(this.val, this.quote***REMOVED***;
  clone.lineno = this.lineno;
  clone.filename = this.filename;
  return clone;
};

/**
 * Return Boolean based on the length of this string.
 *
 * @return {Boolean}
 * @api public
 */

String.prototype.toBoolean = function(***REMOVED***{
  return nodes.Boolean(this.val.length***REMOVED***;
};

/**
 * Coerce `other` to a string.
 *
 * @param {Node} other
 * @return {String}
 * @api public
 */

String.prototype.coerce = function(other***REMOVED***{
  switch (other.nodeName***REMOVED*** {
    case 'string':
      return other;
    case 'expression':
      return new String(other.nodes.map(function(node***REMOVED***{
        return this.coerce(node***REMOVED***.val;
    ***REMOVED***, this***REMOVED***.join(' '***REMOVED******REMOVED***;
    default:
      return new String(other.toString(***REMOVED******REMOVED***;
***REMOVED***
};

/**
 * Operate on `right` with the given `op`.
 *
 * @param {String} op
 * @param {Node} right
 * @return {Node}
 * @api public
 */

String.prototype.operate = function(op, right***REMOVED***{
  switch (op***REMOVED*** {
    case '%':
      var expr = new nodes.Expression;
      expr.push(this***REMOVED***;

      // constructargs
      var args = 'expression' == right.nodeName
        ? utils.unwrap(right***REMOVED***.nodes
        : [right];

      // apply
      return sprintf.apply(null, [expr].concat(args***REMOVED******REMOVED***;
    case '+':
      var expr = new nodes.Expression;
      expr.push(new String(this.val + this.coerce(right***REMOVED***.val***REMOVED******REMOVED***;
      return expr;
    default:
      return Node.prototype.operate.call(this, op, right***REMOVED***;
***REMOVED***
};
