// bson/lib/bson/long.js
// Licensed under the Apache License, Version 2.0 (the "License"***REMOVED***;
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright 2009 Google Inc. All Rights Reserved

/**
 * Defines a Long class for representing a 64-bit two's-complement
 * integer value, which faithfully simulates the behavior of a Java "Long". This
 * implementation is derived from LongLib in GWT.
 *
 * Constructs a 64-bit two's-complement integer, given its low and high 32-bit
 * values as *signed* integers.  See the from* functions below for more
 * convenient ways of constructing Longs.
 *
 * The internal representation of a Long is the two given signed, 32-bit values.
 * We use 32-bit pieces because these are the size of integers on which
 * Javascript performs bit-operations.  For operations like addition and
 * multiplication, we split each number into 16-bit pieces, which can easily be
 * multiplied within Javascript's floating-point representation without overflow
 * or change in sign.
 *
 * In the algorithms below, we frequently reduce the negative case to the
 * positive case by negating the input(s***REMOVED*** and then post-processing the result.
 * Note that we must ALWAYS check specially whether those values are MIN_VALUE
 * (-2^63***REMOVED*** because -MIN_VALUE == MIN_VALUE (since 2^63 cannot be represented as
 * a positive number, it overflows back into a negative***REMOVED***.  Not handling this
 * case would often result in infinite recursion.
 *
 * @class Represents the BSON Long type.
 * @param {Number} low  the low (signed***REMOVED*** 32 bits of the Long.
 * @param {Number} high the high (signed***REMOVED*** 32 bits of the Long.
 */
function Long(low, high***REMOVED*** {
  if(!(this instanceof Long***REMOVED******REMOVED*** return new Long(low, high***REMOVED***;
  
  this._bsontype = 'Long';
  /**
   * @type {number}
   * @api private
   */
  this.low_ = low | 0;  // force into 32 signed bits.

  /**
   * @type {number}
   * @api private
   */
  this.high_ = high | 0;  // force into 32 signed bits.
};

/**
 * Return the int value.
 *
 * @return {Number} the value, assuming it is a 32-bit integer.
 * @api public
 */
Long.prototype.toInt = function(***REMOVED*** {
  return this.low_;
};

/**
 * Return the Number value.
 *
 * @return {Number} the closest floating-point representation to this value.
 * @api public
 */
Long.prototype.toNumber = function(***REMOVED*** {
  return this.high_ * Long.TWO_PWR_32_DBL_ +
         this.getLowBitsUnsigned(***REMOVED***;
};

/**
 * Return the JSON value.
 *
 * @return {String} the JSON representation.
 * @api public
 */
Long.prototype.toJSON = function(***REMOVED*** {
  return this.toString(***REMOVED***;
}

/**
 * Return the String value.
 *
 * @param {Number} [opt_radix] the radix in which the text should be written.
 * @return {String} the textual representation of this value.
 * @api public
 */
Long.prototype.toString = function(opt_radix***REMOVED*** {
  var radix = opt_radix || 10;
  if (radix < 2 || 36 < radix***REMOVED*** {
    throw Error('radix out of range: ' + radix***REMOVED***;
***REMOVED***

  if (this.isZero(***REMOVED******REMOVED*** {
    return '0';
***REMOVED***

  if (this.isNegative(***REMOVED******REMOVED*** {
    if (this.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
      // We need to change the Long value before it can be negated, so we remove
      // the bottom-most digit in this base and then recurse to do the rest.
      var radixLong = Long.fromNumber(radix***REMOVED***;
      var div = this.div(radixLong***REMOVED***;
      var rem = div.multiply(radixLong***REMOVED***.subtract(this***REMOVED***;
      return div.toString(radix***REMOVED*** + rem.toInt(***REMOVED***.toString(radix***REMOVED***;
  ***REMOVED*** else {
      return '-' + this.negate(***REMOVED***.toString(radix***REMOVED***;
  ***REMOVED***
***REMOVED***

  // Do several (6***REMOVED*** digits each time through the loop, so as to
  // minimize the calls to the very expensive emulated div.
  var radixToPower = Long.fromNumber(Math.pow(radix, 6***REMOVED******REMOVED***;

  var rem = this;
  var result = '';
  while (true***REMOVED*** {
    var remDiv = rem.div(radixToPower***REMOVED***;
    var intval = rem.subtract(remDiv.multiply(radixToPower***REMOVED******REMOVED***.toInt(***REMOVED***;
    var digits = intval.toString(radix***REMOVED***;

    rem = remDiv;
    if (rem.isZero(***REMOVED******REMOVED*** {
      return digits + result;
  ***REMOVED*** else {
      while (digits.length < 6***REMOVED*** {
        digits = '0' + digits;
    ***REMOVED***
      result = '' + digits + result;
  ***REMOVED***
***REMOVED***
};

/**
 * Return the high 32-bits value.
 *
 * @return {Number} the high 32-bits as a signed value.
 * @api public
 */
Long.prototype.getHighBits = function(***REMOVED*** {
  return this.high_;
};

/**
 * Return the low 32-bits value.
 *
 * @return {Number} the low 32-bits as a signed value.
 * @api public
 */
Long.prototype.getLowBits = function(***REMOVED*** {
  return this.low_;
};

/**
 * Return the low unsigned 32-bits value.
 *
 * @return {Number} the low 32-bits as an unsigned value.
 * @api public
 */
Long.prototype.getLowBitsUnsigned = function(***REMOVED*** {
  return (this.low_ >= 0***REMOVED*** ?
      this.low_ : Long.TWO_PWR_32_DBL_ + this.low_;
};

/**
 * Returns the number of bits needed to represent the absolute value of this Long.
 *
 * @return {Number} Returns the number of bits needed to represent the absolute value of this Long.
 * @api public
 */
Long.prototype.getNumBitsAbs = function(***REMOVED*** {
  if (this.isNegative(***REMOVED******REMOVED*** {
    if (this.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
      return 64;
  ***REMOVED*** else {
      return this.negate(***REMOVED***.getNumBitsAbs(***REMOVED***;
  ***REMOVED***
***REMOVED*** else {
    var val = this.high_ != 0 ? this.high_ : this.low_;
    for (var bit = 31; bit > 0; bit--***REMOVED*** {
      if ((val & (1 << bit***REMOVED******REMOVED*** != 0***REMOVED*** {
        break;
    ***REMOVED***
  ***REMOVED***
    return this.high_ != 0 ? bit + 33 : bit + 1;
***REMOVED***
};

/**
 * Return whether this value is zero.
 *
 * @return {Boolean} whether this value is zero.
 * @api public
 */
Long.prototype.isZero = function(***REMOVED*** {
  return this.high_ == 0 && this.low_ == 0;
};

/**
 * Return whether this value is negative.
 *
 * @return {Boolean} whether this value is negative.
 * @api public
 */
Long.prototype.isNegative = function(***REMOVED*** {
  return this.high_ < 0;
};

/**
 * Return whether this value is odd.
 *
 * @return {Boolean} whether this value is odd.
 * @api public
 */
Long.prototype.isOdd = function(***REMOVED*** {
  return (this.low_ & 1***REMOVED*** == 1;
};

/**
 * Return whether this Long equals the other
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} whether this Long equals the other
 * @api public
 */
Long.prototype.equals = function(other***REMOVED*** {
  return (this.high_ == other.high_***REMOVED*** && (this.low_ == other.low_***REMOVED***;
};

/**
 * Return whether this Long does not equal the other.
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} whether this Long does not equal the other.
 * @api public
 */
Long.prototype.notEquals = function(other***REMOVED*** {
  return (this.high_ != other.high_***REMOVED*** || (this.low_ != other.low_***REMOVED***;
};

/**
 * Return whether this Long is less than the other.
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} whether this Long is less than the other.
 * @api public
 */
Long.prototype.lessThan = function(other***REMOVED*** {
  return this.compare(other***REMOVED*** < 0;
};

/**
 * Return whether this Long is less than or equal to the other.
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} whether this Long is less than or equal to the other.
 * @api public
 */
Long.prototype.lessThanOrEqual = function(other***REMOVED*** {
  return this.compare(other***REMOVED*** <= 0;
};

/**
 * Return whether this Long is greater than the other.
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} whether this Long is greater than the other.
 * @api public
 */
Long.prototype.greaterThan = function(other***REMOVED*** {
  return this.compare(other***REMOVED*** > 0;
};

/**
 * Return whether this Long is greater than or equal to the other.
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} whether this Long is greater than or equal to the other.
 * @api public
 */
Long.prototype.greaterThanOrEqual = function(other***REMOVED*** {
  return this.compare(other***REMOVED*** >= 0;
};

/**
 * Compares this Long with the given one.
 *
 * @param {Long} other Long to compare against.
 * @return {Boolean} 0 if they are the same, 1 if the this is greater, and -1 if the given one is greater.
 * @api public
 */
Long.prototype.compare = function(other***REMOVED*** {
  if (this.equals(other***REMOVED******REMOVED*** {
    return 0;
***REMOVED***

  var thisNeg = this.isNegative(***REMOVED***;
  var otherNeg = other.isNegative(***REMOVED***;
  if (thisNeg && !otherNeg***REMOVED*** {
    return -1;
***REMOVED***
  if (!thisNeg && otherNeg***REMOVED*** {
    return 1;
***REMOVED***

  // at this point, the signs are the same, so subtraction will not overflow
  if (this.subtract(other***REMOVED***.isNegative(***REMOVED******REMOVED*** {
    return -1;
***REMOVED*** else {
    return 1;
***REMOVED***
};

/**
 * The negation of this value.
 *
 * @return {Long} the negation of this value.
 * @api public
 */
Long.prototype.negate = function(***REMOVED*** {
  if (this.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
    return Long.MIN_VALUE;
***REMOVED*** else {
    return this.not(***REMOVED***.add(Long.ONE***REMOVED***;
***REMOVED***
};

/**
 * Returns the sum of this and the given Long.
 *
 * @param {Long} other Long to add to this one.
 * @return {Long} the sum of this and the given Long.
 * @api public
 */
Long.prototype.add = function(other***REMOVED*** {
  // Divide each number into 4 chunks of 16 bits, and then sum the chunks.

  var a48 = this.high_ >>> 16;
  var a32 = this.high_ & 0xFFFF;
  var a16 = this.low_ >>> 16;
  var a00 = this.low_ & 0xFFFF;

  var b48 = other.high_ >>> 16;
  var b32 = other.high_ & 0xFFFF;
  var b16 = other.low_ >>> 16;
  var b00 = other.low_ & 0xFFFF;

  var c48 = 0, c32 = 0, c16 = 0, c00 = 0;
  c00 += a00 + b00;
  c16 += c00 >>> 16;
  c00 &= 0xFFFF;
  c16 += a16 + b16;
  c32 += c16 >>> 16;
  c16 &= 0xFFFF;
  c32 += a32 + b32;
  c48 += c32 >>> 16;
  c32 &= 0xFFFF;
  c48 += a48 + b48;
  c48 &= 0xFFFF;
  return Long.fromBits((c16 << 16***REMOVED*** | c00, (c48 << 16***REMOVED*** | c32***REMOVED***;
};

/**
 * Returns the difference of this and the given Long.
 *
 * @param {Long} other Long to subtract from this.
 * @return {Long} the difference of this and the given Long.
 * @api public
 */
Long.prototype.subtract = function(other***REMOVED*** {
  return this.add(other.negate(***REMOVED******REMOVED***;
};

/**
 * Returns the product of this and the given Long.
 *
 * @param {Long} other Long to multiply with this.
 * @return {Long} the product of this and the other.
 * @api public
 */
Long.prototype.multiply = function(other***REMOVED*** {
  if (this.isZero(***REMOVED******REMOVED*** {
    return Long.ZERO;
***REMOVED*** else if (other.isZero(***REMOVED******REMOVED*** {
    return Long.ZERO;
***REMOVED***

  if (this.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
    return other.isOdd(***REMOVED*** ? Long.MIN_VALUE : Long.ZERO;
***REMOVED*** else if (other.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
    return this.isOdd(***REMOVED*** ? Long.MIN_VALUE : Long.ZERO;
***REMOVED***

  if (this.isNegative(***REMOVED******REMOVED*** {
    if (other.isNegative(***REMOVED******REMOVED*** {
      return this.negate(***REMOVED***.multiply(other.negate(***REMOVED******REMOVED***;
  ***REMOVED*** else {
      return this.negate(***REMOVED***.multiply(other***REMOVED***.negate(***REMOVED***;
  ***REMOVED***
***REMOVED*** else if (other.isNegative(***REMOVED******REMOVED*** {
    return this.multiply(other.negate(***REMOVED******REMOVED***.negate(***REMOVED***;
***REMOVED***

  // If both Longs are small, use float multiplication
  if (this.lessThan(Long.TWO_PWR_24_***REMOVED*** &&
      other.lessThan(Long.TWO_PWR_24_***REMOVED******REMOVED*** {
    return Long.fromNumber(this.toNumber(***REMOVED*** * other.toNumber(***REMOVED******REMOVED***;
***REMOVED***

  // Divide each Long into 4 chunks of 16 bits, and then add up 4x4 products.
  // We can skip products that would overflow.

  var a48 = this.high_ >>> 16;
  var a32 = this.high_ & 0xFFFF;
  var a16 = this.low_ >>> 16;
  var a00 = this.low_ & 0xFFFF;

  var b48 = other.high_ >>> 16;
  var b32 = other.high_ & 0xFFFF;
  var b16 = other.low_ >>> 16;
  var b00 = other.low_ & 0xFFFF;

  var c48 = 0, c32 = 0, c16 = 0, c00 = 0;
  c00 += a00 * b00;
  c16 += c00 >>> 16;
  c00 &= 0xFFFF;
  c16 += a16 * b00;
  c32 += c16 >>> 16;
  c16 &= 0xFFFF;
  c16 += a00 * b16;
  c32 += c16 >>> 16;
  c16 &= 0xFFFF;
  c32 += a32 * b00;
  c48 += c32 >>> 16;
  c32 &= 0xFFFF;
  c32 += a16 * b16;
  c48 += c32 >>> 16;
  c32 &= 0xFFFF;
  c32 += a00 * b32;
  c48 += c32 >>> 16;
  c32 &= 0xFFFF;
  c48 += a48 * b00 + a32 * b16 + a16 * b32 + a00 * b48;
  c48 &= 0xFFFF;
  return Long.fromBits((c16 << 16***REMOVED*** | c00, (c48 << 16***REMOVED*** | c32***REMOVED***;
};

/**
 * Returns this Long divided by the given one.
 *
 * @param {Long} other Long by which to divide.
 * @return {Long} this Long divided by the given one.
 * @api public
 */
Long.prototype.div = function(other***REMOVED*** {
  if (other.isZero(***REMOVED******REMOVED*** {
    throw Error('division by zero'***REMOVED***;
***REMOVED*** else if (this.isZero(***REMOVED******REMOVED*** {
    return Long.ZERO;
***REMOVED***

  if (this.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
    if (other.equals(Long.ONE***REMOVED*** ||
        other.equals(Long.NEG_ONE***REMOVED******REMOVED*** {
      return Long.MIN_VALUE;  // recall that -MIN_VALUE == MIN_VALUE
  ***REMOVED*** else if (other.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
      return Long.ONE;
  ***REMOVED*** else {
      // At this point, we have |other| >= 2, so |this/other| < |MIN_VALUE|.
      var halfThis = this.shiftRight(1***REMOVED***;
      var approx = halfThis.div(other***REMOVED***.shiftLeft(1***REMOVED***;
      if (approx.equals(Long.ZERO***REMOVED******REMOVED*** {
        return other.isNegative(***REMOVED*** ? Long.ONE : Long.NEG_ONE;
    ***REMOVED*** else {
        var rem = this.subtract(other.multiply(approx***REMOVED******REMOVED***;
        var result = approx.add(rem.div(other***REMOVED******REMOVED***;
        return result;
    ***REMOVED***
  ***REMOVED***
***REMOVED*** else if (other.equals(Long.MIN_VALUE***REMOVED******REMOVED*** {
    return Long.ZERO;
***REMOVED***

  if (this.isNegative(***REMOVED******REMOVED*** {
    if (other.isNegative(***REMOVED******REMOVED*** {
      return this.negate(***REMOVED***.div(other.negate(***REMOVED******REMOVED***;
  ***REMOVED*** else {
      return this.negate(***REMOVED***.div(other***REMOVED***.negate(***REMOVED***;
  ***REMOVED***
***REMOVED*** else if (other.isNegative(***REMOVED******REMOVED*** {
    return this.div(other.negate(***REMOVED******REMOVED***.negate(***REMOVED***;
***REMOVED***

  // Repeat the following until the remainder is less than other:  find a
  // floating-point that approximates remainder / other *from below*, add this
  // into the result, and subtract it from the remainder.  It is critical that
  // the approximate value is less than or equal to the real value so that the
  // remainder never becomes negative.
  var res = Long.ZERO;
  var rem = this;
  while (rem.greaterThanOrEqual(other***REMOVED******REMOVED*** {
    // Approximate the result of division. This may be a little greater or
    // smaller than the actual value.
    var approx = Math.max(1, Math.floor(rem.toNumber(***REMOVED*** / other.toNumber(***REMOVED******REMOVED******REMOVED***;

    // We will tweak the approximate result by changing it in the 48-th digit or
    // the smallest non-fractional digit, whichever is larger.
    var log2 = Math.ceil(Math.log(approx***REMOVED*** / Math.LN2***REMOVED***;
    var delta = (log2 <= 48***REMOVED*** ? 1 : Math.pow(2, log2 - 48***REMOVED***;

    // Decrease the approximation until it is smaller than the remainder.  Note
    // that if it is too large, the product overflows and is negative.
    var approxRes = Long.fromNumber(approx***REMOVED***;
    var approxRem = approxRes.multiply(other***REMOVED***;
    while (approxRem.isNegative(***REMOVED*** || approxRem.greaterThan(rem***REMOVED******REMOVED*** {
      approx -= delta;
      approxRes = Long.fromNumber(approx***REMOVED***;
      approxRem = approxRes.multiply(other***REMOVED***;
  ***REMOVED***

    // We know the answer can't be zero... and actually, zero would cause
    // infinite recursion since we would make no progress.
    if (approxRes.isZero(***REMOVED******REMOVED*** {
      approxRes = Long.ONE;
  ***REMOVED***

    res = res.add(approxRes***REMOVED***;
    rem = rem.subtract(approxRem***REMOVED***;
***REMOVED***
  return res;
};

/**
 * Returns this Long modulo the given one.
 *
 * @param {Long} other Long by which to mod.
 * @return {Long} this Long modulo the given one.
 * @api public
 */
Long.prototype.modulo = function(other***REMOVED*** {
  return this.subtract(this.div(other***REMOVED***.multiply(other***REMOVED******REMOVED***;
};

/**
 * The bitwise-NOT of this value.
 *
 * @return {Long} the bitwise-NOT of this value.
 * @api public
 */
Long.prototype.not = function(***REMOVED*** {
  return Long.fromBits(~this.low_, ~this.high_***REMOVED***;
};

/**
 * Returns the bitwise-AND of this Long and the given one.
 *
 * @param {Long} other the Long with which to AND.
 * @return {Long} the bitwise-AND of this and the other.
 * @api public
 */
Long.prototype.and = function(other***REMOVED*** {
  return Long.fromBits(this.low_ & other.low_, this.high_ & other.high_***REMOVED***;
};

/**
 * Returns the bitwise-OR of this Long and the given one.
 *
 * @param {Long} other the Long with which to OR.
 * @return {Long} the bitwise-OR of this and the other.
 * @api public
 */
Long.prototype.or = function(other***REMOVED*** {
  return Long.fromBits(this.low_ | other.low_, this.high_ | other.high_***REMOVED***;
};

/**
 * Returns the bitwise-XOR of this Long and the given one.
 *
 * @param {Long} other the Long with which to XOR.
 * @return {Long} the bitwise-XOR of this and the other.
 * @api public
 */
Long.prototype.xor = function(other***REMOVED*** {
  return Long.fromBits(this.low_ ^ other.low_, this.high_ ^ other.high_***REMOVED***;
};

/**
 * Returns this Long with bits shifted to the left by the given amount.
 *
 * @param {Number} numBits the number of bits by which to shift.
 * @return {Long} this shifted to the left by the given amount.
 * @api public
 */
Long.prototype.shiftLeft = function(numBits***REMOVED*** {
  numBits &= 63;
  if (numBits == 0***REMOVED*** {
    return this;
***REMOVED*** else {
    var low = this.low_;
    if (numBits < 32***REMOVED*** {
      var high = this.high_;
      return Long.fromBits(
                 low << numBits,
                 (high << numBits***REMOVED*** | (low >>> (32 - numBits***REMOVED******REMOVED******REMOVED***;
  ***REMOVED*** else {
      return Long.fromBits(0, low << (numBits - 32***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED***
};

/**
 * Returns this Long with bits shifted to the right by the given amount.
 *
 * @param {Number} numBits the number of bits by which to shift.
 * @return {Long} this shifted to the right by the given amount.
 * @api public
 */
Long.prototype.shiftRight = function(numBits***REMOVED*** {
  numBits &= 63;
  if (numBits == 0***REMOVED*** {
    return this;
***REMOVED*** else {
    var high = this.high_;
    if (numBits < 32***REMOVED*** {
      var low = this.low_;
      return Long.fromBits(
                 (low >>> numBits***REMOVED*** | (high << (32 - numBits***REMOVED******REMOVED***,
                 high >> numBits***REMOVED***;
  ***REMOVED*** else {
      return Long.fromBits(
                 high >> (numBits - 32***REMOVED***,
                 high >= 0 ? 0 : -1***REMOVED***;
  ***REMOVED***
***REMOVED***
};

/**
 * Returns this Long with bits shifted to the right by the given amount, with the new top bits matching the current sign bit.
 *
 * @param {Number} numBits the number of bits by which to shift.
 * @return {Long} this shifted to the right by the given amount, with zeros placed into the new leading bits.
 * @api public
 */
Long.prototype.shiftRightUnsigned = function(numBits***REMOVED*** {
  numBits &= 63;
  if (numBits == 0***REMOVED*** {
    return this;
***REMOVED*** else {
    var high = this.high_;
    if (numBits < 32***REMOVED*** {
      var low = this.low_;
      return Long.fromBits(
                 (low >>> numBits***REMOVED*** | (high << (32 - numBits***REMOVED******REMOVED***,
                 high >>> numBits***REMOVED***;
  ***REMOVED*** else if (numBits == 32***REMOVED*** {
      return Long.fromBits(high, 0***REMOVED***;
  ***REMOVED*** else {
      return Long.fromBits(high >>> (numBits - 32***REMOVED***, 0***REMOVED***;
  ***REMOVED***
***REMOVED***
};

/**
 * Returns a Long representing the given (32-bit***REMOVED*** integer value.
 *
 * @param {Number} value the 32-bit integer in question.
 * @return {Long} the corresponding Long value.
 * @api public
 */
Long.fromInt = function(value***REMOVED*** {
  if (-128 <= value && value < 128***REMOVED*** {
    var cachedObj = Long.INT_CACHE_[value];
    if (cachedObj***REMOVED*** {
      return cachedObj;
  ***REMOVED***
***REMOVED***

  var obj = new Long(value | 0, value < 0 ? -1 : 0***REMOVED***;
  if (-128 <= value && value < 128***REMOVED*** {
    Long.INT_CACHE_[value] = obj;
***REMOVED***
  return obj;
};

/**
 * Returns a Long representing the given value, provided that it is a finite number. Otherwise, zero is returned.
 *
 * @param {Number} value the number in question.
 * @return {Long} the corresponding Long value.
 * @api public
 */
Long.fromNumber = function(value***REMOVED*** {
  if (isNaN(value***REMOVED*** || !isFinite(value***REMOVED******REMOVED*** {
    return Long.ZERO;
***REMOVED*** else if (value <= -Long.TWO_PWR_63_DBL_***REMOVED*** {
    return Long.MIN_VALUE;
***REMOVED*** else if (value + 1 >= Long.TWO_PWR_63_DBL_***REMOVED*** {
    return Long.MAX_VALUE;
***REMOVED*** else if (value < 0***REMOVED*** {
    return Long.fromNumber(-value***REMOVED***.negate(***REMOVED***;
***REMOVED*** else {
    return new Long(
               (value % Long.TWO_PWR_32_DBL_***REMOVED*** | 0,
               (value / Long.TWO_PWR_32_DBL_***REMOVED*** | 0***REMOVED***;
***REMOVED***
};

/**
 * Returns a Long representing the 64-bit integer that comes by concatenating the given high and low bits. Each is assumed to use 32 bits.
 *
 * @param {Number} lowBits the low 32-bits.
 * @param {Number} highBits the high 32-bits.
 * @return {Long} the corresponding Long value.
 * @api public
 */
Long.fromBits = function(lowBits, highBits***REMOVED*** {
  return new Long(lowBits, highBits***REMOVED***;
};

/**
 * Returns a Long representation of the given string, written using the given radix.
 *
 * @param {String} str the textual representation of the Long.
 * @param {Number} opt_radix the radix in which the text is written.
 * @return {Long} the corresponding Long value.
 * @api public
 */
Long.fromString = function(str, opt_radix***REMOVED*** {
  if (str.length == 0***REMOVED*** {
    throw Error('number format error: empty string'***REMOVED***;
***REMOVED***

  var radix = opt_radix || 10;
  if (radix < 2 || 36 < radix***REMOVED*** {
    throw Error('radix out of range: ' + radix***REMOVED***;
***REMOVED***

  if (str.charAt(0***REMOVED*** == '-'***REMOVED*** {
    return Long.fromString(str.substring(1***REMOVED***, radix***REMOVED***.negate(***REMOVED***;
***REMOVED*** else if (str.indexOf('-'***REMOVED*** >= 0***REMOVED*** {
    throw Error('number format error: interior "-" character: ' + str***REMOVED***;
***REMOVED***

  // Do several (8***REMOVED*** digits each time through the loop, so as to
  // minimize the calls to the very expensive emulated div.
  var radixToPower = Long.fromNumber(Math.pow(radix, 8***REMOVED******REMOVED***;

  var result = Long.ZERO;
  for (var i = 0; i < str.length; i += 8***REMOVED*** {
    var size = Math.min(8, str.length - i***REMOVED***;
    var value = parseInt(str.substring(i, i + size***REMOVED***, radix***REMOVED***;
    if (size < 8***REMOVED*** {
      var power = Long.fromNumber(Math.pow(radix, size***REMOVED******REMOVED***;
      result = result.multiply(power***REMOVED***.add(Long.fromNumber(value***REMOVED******REMOVED***;
  ***REMOVED*** else {
      result = result.multiply(radixToPower***REMOVED***;
      result = result.add(Long.fromNumber(value***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED***
  return result;
};

// NOTE: Common constant values ZERO, ONE, NEG_ONE, etc. are defined below the
// from* methods on which they depend.


/**
 * A cache of the Long representations of small integer values.
 * @type {Object}
 * @api private
 */
Long.INT_CACHE_ = {};

// NOTE: the compiler should inline these constant values below and then remove
// these variables, so there should be no runtime penalty for these.

/**
 * Number used repeated below in calculations.  This must appear before the
 * first call to any from* function below.
 * @type {number}
 * @api private
 */
Long.TWO_PWR_16_DBL_ = 1 << 16;

/**
 * @type {number}
 * @api private
 */
Long.TWO_PWR_24_DBL_ = 1 << 24;

/**
 * @type {number}
 * @api private
 */
Long.TWO_PWR_32_DBL_ = Long.TWO_PWR_16_DBL_ * Long.TWO_PWR_16_DBL_;

/**
 * @type {number}
 * @api private
 */
Long.TWO_PWR_31_DBL_ = Long.TWO_PWR_32_DBL_ / 2;

/**
 * @type {number}
 * @api private
 */
Long.TWO_PWR_48_DBL_ = Long.TWO_PWR_32_DBL_ * Long.TWO_PWR_16_DBL_;

/**
 * @type {number}
 * @api private
 */
Long.TWO_PWR_64_DBL_ = Long.TWO_PWR_32_DBL_ * Long.TWO_PWR_32_DBL_;

/**
 * @type {number}
 * @api private
 */
Long.TWO_PWR_63_DBL_ = Long.TWO_PWR_64_DBL_ / 2;

/** @type {Long} */
Long.ZERO = Long.fromInt(0***REMOVED***;

/** @type {Long} */
Long.ONE = Long.fromInt(1***REMOVED***;

/** @type {Long} */
Long.NEG_ONE = Long.fromInt(-1***REMOVED***;

/** @type {Long} */
Long.MAX_VALUE =
    Long.fromBits(0xFFFFFFFF | 0, 0x7FFFFFFF | 0***REMOVED***;

/** @type {Long} */
Long.MIN_VALUE = Long.fromBits(0, 0x80000000 | 0***REMOVED***;

/**
 * @type {Long}
 * @api private
 */
Long.TWO_PWR_24_ = Long.fromInt(1 << 24***REMOVED***;

/**
 * Expose.
 */
exports.Long = Long;
