// tough-cookie/generate-pubsuffix.js
/*
 * Copyright GoInstant, Inc. and other contributors. All rights reserved.
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"***REMOVED***, to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */
'use strict';
var fs = require('fs'***REMOVED***;
var assert = require('assert'***REMOVED***;
var punycode = require('punycode'***REMOVED***;

fs.readFile('./public-suffix.txt', 'utf8', function(err,string***REMOVED*** {
  if (err***REMOVED*** {
    throw err;
***REMOVED***
  var lines = string.split("\n"***REMOVED***;
  process.nextTick(function(***REMOVED*** {
    processList(lines***REMOVED***;
***REMOVED******REMOVED***;
}***REMOVED***;

var index = {};

var COMMENT = new RegExp('//.+'***REMOVED***;
function processList(lines***REMOVED*** {
  while (lines.length***REMOVED*** {
    var line = lines.shift(***REMOVED***;
    line = line.replace(COMMENT,''***REMOVED***.trim(***REMOVED***;
    if (!line***REMOVED*** {
      continue;
  ***REMOVED***
    addToIndex(index,line***REMOVED***;
***REMOVED***

  pubSufTest(***REMOVED***;

  var w = fs.createWriteStream('./lib/pubsuffix.js',{
    flags: 'w',
    encoding: 'utf8',
    mode: parseInt('644',8***REMOVED***
***REMOVED******REMOVED***;
  w.on('end', process.exit***REMOVED***;
  w.write("/****************************************************\n"***REMOVED***;
  w.write(" * AUTOMATICALLY GENERATED by generate-pubsuffix.js *\n"***REMOVED***;
  w.write(" *                  DO NOT EDIT!                    *\n"***REMOVED***;
  w.write(" ****************************************************/\n\n"***REMOVED***;

  w.write("module.exports.getPublicSuffix = "***REMOVED***;
  w.write(getPublicSuffix.toString(***REMOVED******REMOVED***;
  w.write(";\n\n"***REMOVED***;

  w.write("// The following generated structure is used under the MPL version 1.1\n"***REMOVED***;
  w.write("// See public-suffix.txt for more information\n\n"***REMOVED***;
  w.write("var index = module.exports.index = Object.freeze(\n"***REMOVED***;
  w.write(JSON.stringify(index***REMOVED******REMOVED***;
  w.write("***REMOVED***;\n\n"***REMOVED***;
  w.write("// END of automatically generated file\n"***REMOVED***;

  w.end(***REMOVED***;
}

function addToIndex(index,line***REMOVED*** {
  var prefix = '';
  if (line.replace(/^(!|\*\.***REMOVED***/***REMOVED******REMOVED*** {
    prefix = RegExp.$1;
    line = line.slice(prefix.length***REMOVED***;
***REMOVED***
  line = prefix + punycode.toASCII(line***REMOVED***;

  if (line.substr(0,1***REMOVED*** == '!'***REMOVED*** {
    index[line.substr(1***REMOVED***] = false;
***REMOVED*** else {
    index[line] = true;
***REMOVED***
}

// include the licence in the function since it gets written to pubsuffix.js
function getPublicSuffix(domain***REMOVED*** {
  /*
   * Copyright GoInstant, Inc. and other contributors. All rights reserved.
   * Permission is hereby granted, free of charge, to any person obtaining a copy
   * of this software and associated documentation files (the "Software"***REMOVED***, to
   * deal in the Software without restriction, including without limitation the
   * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
   * sell copies of the Software, and to permit persons to whom the Software is
   * furnished to do so, subject to the following conditions:
   *
   * The above copyright notice and this permission notice shall be included in
   * all copies or substantial portions of the Software.
   *
   * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
   * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
   * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
   * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
   * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
   * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
   * IN THE SOFTWARE.
   */
  if (!domain***REMOVED*** {
    return null;
***REMOVED***
  if (domain.match(/^\./***REMOVED******REMOVED*** {
    return null;
***REMOVED***

  domain = domain.toLowerCase(***REMOVED***;
  var parts = domain.split('.'***REMOVED***.reverse(***REMOVED***;

  var suffix = '';
  var suffixLen = 0;
  for (var i=0; i<parts.length; i++***REMOVED*** {
    var part = parts[i];
    var starstr = '*'+suffix;
    var partstr = part+suffix;

    if (index[starstr]***REMOVED*** { // star rule matches
      suffixLen = i+1;
      if (index[partstr] === false***REMOVED*** { // exception rule matches (NB: false, not undefined***REMOVED***
        suffixLen--;
    ***REMOVED***
  ***REMOVED*** else if (index[partstr]***REMOVED*** { // exact match, not exception
      suffixLen = i+1;
  ***REMOVED***

    suffix = '.'+part+suffix;
***REMOVED***

  if (index['*'+suffix]***REMOVED*** { // *.domain exists (e.g. *.kyoto.jp for domain='kyoto.jp'***REMOVED***;
    return null;
***REMOVED***

  if (suffixLen && parts.length > suffixLen***REMOVED*** {
    return parts.slice(0,suffixLen+1***REMOVED***.reverse(***REMOVED***.join('.'***REMOVED***;
***REMOVED***

  return null;
}

function checkPublicSuffix(give,get***REMOVED*** {
  var got = getPublicSuffix(give***REMOVED***;
  assert.equal(got, get, give+' should be '+(get==null?'NULL':get***REMOVED***+' but got '+got***REMOVED***;
}

// pubSufTest(***REMOVED*** was converted to JavaScript from http://publicsuffix.org/list/test.txt
function pubSufTest(***REMOVED*** {
  // For this function-scope and this function-scope ONLY:
  // Any copyright is dedicated to the Public Domain.
  // http://creativecommons.org/publicdomain/zero/1.0/

  // NULL input.
  checkPublicSuffix(null, null***REMOVED***;
  // Mixed case.
  checkPublicSuffix('COM', null***REMOVED***;
  checkPublicSuffix('example.COM', 'example.com'***REMOVED***;
  checkPublicSuffix('WwW.example.COM', 'example.com'***REMOVED***;
  // Leading dot.
  checkPublicSuffix('.com', null***REMOVED***;
  checkPublicSuffix('.example', null***REMOVED***;
  checkPublicSuffix('.example.com', null***REMOVED***;
  checkPublicSuffix('.example.example', null***REMOVED***;
  // Unlisted TLD.
  checkPublicSuffix('example', null***REMOVED***;
  checkPublicSuffix('example.example', null***REMOVED***;
  checkPublicSuffix('b.example.example', null***REMOVED***;
  checkPublicSuffix('a.b.example.example', null***REMOVED***;
  // Listed, but non-Internet, TLD.
  checkPublicSuffix('local', null***REMOVED***;
  checkPublicSuffix('example.local', null***REMOVED***;
  checkPublicSuffix('b.example.local', null***REMOVED***;
  checkPublicSuffix('a.b.example.local', null***REMOVED***;
  // TLD with only 1 rule.
  checkPublicSuffix('biz', null***REMOVED***;
  checkPublicSuffix('domain.biz', 'domain.biz'***REMOVED***;
  checkPublicSuffix('b.domain.biz', 'domain.biz'***REMOVED***;
  checkPublicSuffix('a.b.domain.biz', 'domain.biz'***REMOVED***;
  // TLD with some 2-level rules.
  checkPublicSuffix('com', null***REMOVED***;
  checkPublicSuffix('example.com', 'example.com'***REMOVED***;
  checkPublicSuffix('b.example.com', 'example.com'***REMOVED***;
  checkPublicSuffix('a.b.example.com', 'example.com'***REMOVED***;
  checkPublicSuffix('uk.com', null***REMOVED***;
  checkPublicSuffix('example.uk.com', 'example.uk.com'***REMOVED***;
  checkPublicSuffix('b.example.uk.com', 'example.uk.com'***REMOVED***;
  checkPublicSuffix('a.b.example.uk.com', 'example.uk.com'***REMOVED***;
  checkPublicSuffix('test.ac', 'test.ac'***REMOVED***;
  // TLD with only 1 (wildcard***REMOVED*** rule.
  checkPublicSuffix('cy', null***REMOVED***;
  checkPublicSuffix('c.cy', null***REMOVED***;
  checkPublicSuffix('b.c.cy', 'b.c.cy'***REMOVED***;
  checkPublicSuffix('a.b.c.cy', 'b.c.cy'***REMOVED***;
  // More complex TLD.
  checkPublicSuffix('jp', null***REMOVED***;
  checkPublicSuffix('test.jp', 'test.jp'***REMOVED***;
  checkPublicSuffix('www.test.jp', 'test.jp'***REMOVED***;
  checkPublicSuffix('ac.jp', null***REMOVED***;
  checkPublicSuffix('test.ac.jp', 'test.ac.jp'***REMOVED***;
  checkPublicSuffix('www.test.ac.jp', 'test.ac.jp'***REMOVED***;
  checkPublicSuffix('kyoto.jp', null***REMOVED***;
  checkPublicSuffix('c.kyoto.jp', null***REMOVED***;
  checkPublicSuffix('b.c.kyoto.jp', 'b.c.kyoto.jp'***REMOVED***;
  checkPublicSuffix('a.b.c.kyoto.jp', 'b.c.kyoto.jp'***REMOVED***;
  checkPublicSuffix('pref.kyoto.jp', 'pref.kyoto.jp'***REMOVED***;  // Exception rule.
  checkPublicSuffix('www.pref.kyoto.jp', 'pref.kyoto.jp'***REMOVED***;  // Exception rule.
  checkPublicSuffix('city.kyoto.jp', 'city.kyoto.jp'***REMOVED***;  // Exception rule.
  checkPublicSuffix('www.city.kyoto.jp', 'city.kyoto.jp'***REMOVED***;  // Exception rule.
  // TLD with a wildcard rule and exceptions.
  checkPublicSuffix('om', null***REMOVED***;
  checkPublicSuffix('test.om', null***REMOVED***;
  checkPublicSuffix('b.test.om', 'b.test.om'***REMOVED***;
  checkPublicSuffix('a.b.test.om', 'b.test.om'***REMOVED***;
  checkPublicSuffix('songfest.om', 'songfest.om'***REMOVED***;
  checkPublicSuffix('www.songfest.om', 'songfest.om'***REMOVED***;
  // US K12.
  checkPublicSuffix('us', null***REMOVED***;
  checkPublicSuffix('test.us', 'test.us'***REMOVED***;
  checkPublicSuffix('www.test.us', 'test.us'***REMOVED***;
  checkPublicSuffix('ak.us', null***REMOVED***;
  checkPublicSuffix('test.ak.us', 'test.ak.us'***REMOVED***;
  checkPublicSuffix('www.test.ak.us', 'test.ak.us'***REMOVED***;
  checkPublicSuffix('k12.ak.us', null***REMOVED***;
  checkPublicSuffix('test.k12.ak.us', 'test.k12.ak.us'***REMOVED***;
  checkPublicSuffix('www.test.k12.ak.us', 'test.k12.ak.us'***REMOVED***;


}
