// jade/lib/nodes/attrs.js
'use strict';

var Node = require('./node'***REMOVED***;

/**
 * Initialize a `Attrs` node.
 *
 * @api public
 */

var Attrs = module.exports = function Attrs(***REMOVED*** {
  this.attributeNames = [];
  this.attrs = [];
  this.attributeBlocks = [];
};

// Inherit from `Node`.
Attrs.prototype = Object.create(Node.prototype***REMOVED***;
Attrs.prototype.constructor = Attrs;

Attrs.prototype.type = 'Attrs';

/**
 * Set attribute `name` to `val`, keep in mind these become
 * part of a raw js object literal, so to quote a value you must
 * '"quote me"', otherwise or example 'user.name' is literal JavaScript.
 *
 * @param {String} name
 * @param {String} val
 * @param {Boolean} escaped
 * @return {Tag} for chaining
 * @api public
 */

Attrs.prototype.setAttribute = function(name, val, escaped***REMOVED***{
  if (name !== 'class' && this.attributeNames.indexOf(name***REMOVED*** !== -1***REMOVED*** {
    throw new Error('Duplicate attribute "' + name + '" is not allowed.'***REMOVED***;
***REMOVED***
  this.attributeNames.push(name***REMOVED***;
  this.attrs.push({ name: name, val: val, escaped: escaped }***REMOVED***;
  return this;
};

/**
 * Remove attribute `name` when present.
 *
 * @param {String} name
 * @api public
 */

Attrs.prototype.removeAttribute = function(name***REMOVED***{
  var err = new Error('attrs.removeAttribute is deprecated and will be removed in v2.0.0'***REMOVED***;
  console.warn(err.stack***REMOVED***;

  for (var i = 0, len = this.attrs.length; i < len; ++i***REMOVED*** {
    if (this.attrs[i] && this.attrs[i].name == name***REMOVED*** {
      delete this.attrs[i];
  ***REMOVED***
***REMOVED***
};

/**
 * Get attribute value by `name`.
 *
 * @param {String} name
 * @return {String}
 * @api public
 */

Attrs.prototype.getAttribute = function(name***REMOVED***{
  var err = new Error('attrs.getAttribute is deprecated and will be removed in v2.0.0'***REMOVED***;
  console.warn(err.stack***REMOVED***;

  for (var i = 0, len = this.attrs.length; i < len; ++i***REMOVED*** {
    if (this.attrs[i] && this.attrs[i].name == name***REMOVED*** {
      return this.attrs[i].val;
  ***REMOVED***
***REMOVED***
};

Attrs.prototype.addAttributes = function (src***REMOVED*** {
  this.attributeBlocks.push(src***REMOVED***;
};
