// jade/lib/nodes/block.js
'use strict';

var Node = require('./node'***REMOVED***;

/**
 * Initialize a new `Block` with an optional `node`.
 *
 * @param {Node} node
 * @api public
 */

var Block = module.exports = function Block(node***REMOVED***{
  this.nodes = [];
  if (node***REMOVED*** this.push(node***REMOVED***;
};

// Inherit from `Node`.
Block.prototype = Object.create(Node.prototype***REMOVED***;
Block.prototype.constructor = Block;

Block.prototype.type = 'Block';

/**
 * Block flag.
 */

Block.prototype.isBlock = true;

/**
 * Replace the nodes in `other` with the nodes
 * in `this` block.
 *
 * @param {Block} other
 * @api private
 */

Block.prototype.replace = function(other***REMOVED***{
  var err = new Error('block.replace is deprecated and will be removed in v2.0.0'***REMOVED***;
  console.warn(err.stack***REMOVED***;

  other.nodes = this.nodes;
};

/**
 * Push the given `node`.
 *
 * @param {Node} node
 * @return {Number}
 * @api public
 */

Block.prototype.push = function(node***REMOVED***{
  return this.nodes.push(node***REMOVED***;
};

/**
 * Check if this block is empty.
 *
 * @return {Boolean}
 * @api public
 */

Block.prototype.isEmpty = function(***REMOVED***{
  return 0 == this.nodes.length;
};

/**
 * Unshift the given `node`.
 *
 * @param {Node} node
 * @return {Number}
 * @api public
 */

Block.prototype.unshift = function(node***REMOVED***{
  return this.nodes.unshift(node***REMOVED***;
};

/**
 * Return the "last" block, or the first `yield` node.
 *
 * @return {Block}
 * @api private
 */

Block.prototype.includeBlock = function(***REMOVED***{
  var ret = this
    , node;

  for (var i = 0, len = this.nodes.length; i < len; ++i***REMOVED*** {
    node = this.nodes[i];
    if (node.yield***REMOVED*** return node;
    else if (node.textOnly***REMOVED*** continue;
    else if (node.includeBlock***REMOVED*** ret = node.includeBlock(***REMOVED***;
    else if (node.block && !node.block.isEmpty(***REMOVED******REMOVED*** ret = node.block.includeBlock(***REMOVED***;
    if (ret.yield***REMOVED*** return ret;
***REMOVED***

  return ret;
};

/**
 * Return a clone of this block.
 *
 * @return {Block}
 * @api private
 */

Block.prototype.clone = function(***REMOVED***{
  var err = new Error('block.clone is deprecated and will be removed in v2.0.0'***REMOVED***;
  console.warn(err.stack***REMOVED***;

  var clone = new Block;
  for (var i = 0, len = this.nodes.length; i < len; ++i***REMOVED*** {
    clone.push(this.nodes[i].clone(***REMOVED******REMOVED***;
***REMOVED***
  return clone;
};
