// mongodb/lib/mongodb/gridfs/gridstore.js
/**
 * @fileOverview GridFS is a tool for MongoDB to store files to the database.
 * Because of the restrictions of the object size the database can hold, a
 * facility to split a file into several chunks is needed. The {@link GridStore}
 * class offers a simplified api to interact with files while managing the
 * chunks of split files behind the scenes. More information about GridFS can be
 * found <a href="http://www.mongodb.org/display/DOCS/GridFS">here</a>.
 */
var Chunk = require('./chunk'***REMOVED***.Chunk,
  DbCommand = require('../commands/db_command'***REMOVED***.DbCommand,
  ObjectID = require('bson'***REMOVED***.ObjectID,
  Buffer = require('buffer'***REMOVED***.Buffer,
  fs = require('fs'***REMOVED***,
  timers = require('timers'***REMOVED***,
  util = require('util'***REMOVED***,
  inherits = util.inherits,
  ReadStream = require('./readstream'***REMOVED***.ReadStream,
  Stream = require('stream'***REMOVED***;

// Set processor, setImmediate if 0.10 otherwise nextTick
var processor = require('../utils'***REMOVED***.processor(***REMOVED***;

var REFERENCE_BY_FILENAME = 0,
  REFERENCE_BY_ID = 1;

/**
 * A class representation of a file stored in GridFS.
 *
 * Modes
 *  - **"r"** - read only. This is the default mode.
 *  - **"w"** - write in truncate mode. Existing data will be overwriten.
 *  - **w+"** - write in edit mode (append is not guaranteed for concurrent operations***REMOVED***
 *
 * Options
 *  - **root** {String}, root collection to use. Defaults to **{GridStore.DEFAULT_ROOT_COLLECTION}**.
 *  - **content_type** {String}, mime type of the file. Defaults to **{GridStore.DEFAULT_CONTENT_TYPE}**.
 *  - **chunk_size** {Number}, size for the chunk. Defaults to **{Chunk.DEFAULT_CHUNK_SIZE}**.
 *  - **metadata** {Object}, arbitrary data the user wants to store.
 *  - **readPreference** {String}, the prefered read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *  - **w**, {Number/String, > -1 || 'majority' || tag name} the write concern for the operation where < 1 is no acknowlegement of write and w >= 1, w = 'majority' or tag acknowledges the write
 *  - **wtimeout**, {Number, 0} set the timeout for waiting for write concern to finish (combines with w option***REMOVED***
 *  - **fsync**, (Boolean, default:false***REMOVED*** write waits for fsync before returning, from MongoDB 2.6 on, fsync cannot be combined with journal
 *  - **j**, (Boolean, default:false***REMOVED*** write waits for journal sync before returning
 *
 * @class Represents the GridStore.
 * @param {Db} db A database instance to interact with.
 * @param {Any} [id] optional unique id for this file
 * @param {String} [filename] optional filename for this file, no unique constrain on the field
 * @param {String} mode set the mode for this file.
 * @param {Object} options optional properties to specify.
 * @return {GridStore}
 */
var GridStore = function GridStore(db, id, filename, mode, options***REMOVED*** {
  if(!(this instanceof GridStore***REMOVED******REMOVED*** return new GridStore(db, id, filename, mode, options***REMOVED***;

  var self = this;
  this.db = db;

  // Call stream constructor
  if(typeof Stream == 'function'***REMOVED*** {
    Stream.call(this***REMOVED***;
***REMOVED***

  // Handle options
  if(typeof options === 'undefined'***REMOVED*** options = {};
  // Handle mode
  if(typeof mode === 'undefined'***REMOVED*** {
    mode = filename;
    filename = undefined;
***REMOVED*** else if(typeof mode == 'object'***REMOVED*** {
    options = mode;
    mode = filename;
    filename = undefined;
***REMOVED***

  if(id instanceof ObjectID***REMOVED*** {
    this.referenceBy = REFERENCE_BY_ID;
    this.fileId = id;
    this.filename = filename;
***REMOVED*** else if(typeof filename == 'undefined'***REMOVED*** {
    this.referenceBy = REFERENCE_BY_FILENAME;
    this.filename = id;
    if (mode.indexOf('w'***REMOVED*** != null***REMOVED*** {
      this.fileId = new ObjectID(***REMOVED***;
  ***REMOVED***
***REMOVED*** else {
    this.referenceBy = REFERENCE_BY_ID;
    this.fileId = id;
    this.filename = filename;
***REMOVED***

  // Set up the rest
  this.mode = mode == null ? "r" : mode;
  this.options = options || {};
  
  // Set the root if overridden
  this.root = this.options['root'] == null ? exports.GridStore.DEFAULT_ROOT_COLLECTION : this.options['root'];
  this.position = 0;
  this.readPreference = this.options.readPreference || 'primary';
  this.writeConcern = _getWriteConcern(db, this.options***REMOVED***;
  // Set default chunk size
  this.internalChunkSize = this.options['chunkSize'] == null ? Chunk.DEFAULT_CHUNK_SIZE : this.options['chunkSize'];
}

/**
 *  Code for the streaming capabilities of the gridstore object
 *  Most code from Aaron heckmanns project https://github.com/aheckmann/gridfs-stream
 *  Modified to work on the gridstore object itself
 *  @ignore
 */
if(typeof Stream == 'function'***REMOVED*** {
  GridStore.prototype = { __proto__: Stream.prototype }
} else {
  // Node 0.4.X compatibility code
  GridStore.prototype = { __proto__: Stream.Stream.prototype }
}

// Move pipe to _pipe
GridStore.prototype._pipe = GridStore.prototype.pipe;

/**
 * Opens the file from the database and initialize this object. Also creates a
 * new one if file does not exist.
 *
 * @param {Function} callback this will be called after executing this method. The first parameter will contain an **{Error}** object and the second parameter will be null if an error occured. Otherwise, the first parameter will be null and the second will contain the reference to this object.
 * @return {null}
 * @api public
 */
GridStore.prototype.open = function(callback***REMOVED*** {
  if( this.mode != "w" && this.mode != "w+" && this.mode != "r"***REMOVED***{
    callback(new Error("Illegal mode " + this.mode***REMOVED***, null***REMOVED***;
    return;
***REMOVED***

  var self = this;

  // Get the write concern
  var writeConcern = _getWriteConcern(this.db, this.options***REMOVED***;

  // If we are writing we need to ensure we have the right indexes for md5's
  if((self.mode == "w" || self.mode == "w+"***REMOVED******REMOVED*** {
    // Get files collection
    var collection = self.collection(***REMOVED***;
    // Put index on filename
    collection.ensureIndex([['filename', 1]], writeConcern, function(err, index***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;

      // Get chunk collection
      var chunkCollection = self.chunkCollection(***REMOVED***;
      // Ensure index on chunk collection
      chunkCollection.ensureIndex([['files_id', 1], ['n', 1]], writeConcern, function(err, index***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        _open(self, writeConcern, callback***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED*** else {
    // Open the gridstore
    _open(self, writeConcern, callback***REMOVED***;
***REMOVED***
};

/**
 * Hidding the _open function
 * @ignore
 * @api private
 */
var _open = function(self, options, callback***REMOVED*** {
  var collection = self.collection(***REMOVED***;
  // Create the query
  var query = self.referenceBy == REFERENCE_BY_ID ? {_id:self.fileId} : {filename:self.filename};
  query = null == self.fileId && this.filename == null ? null : query;
  options.readPreference = self.readPreference;

  // Fetch the chunks
  if(query != null***REMOVED*** {
    collection.find(query, options, function(err, cursor***REMOVED*** {
      if(err***REMOVED*** return error(err***REMOVED***;

      // Fetch the file
      cursor.nextObject(function(err, doc***REMOVED*** {
        if(err***REMOVED*** return error(err***REMOVED***;

        // Check if the collection for the files exists otherwise prepare the new one
        if(doc != null***REMOVED*** {
          self.fileId = doc._id;
          self.filename = doc.filename;
          self.contentType = doc.contentType;
          self.internalChunkSize = doc.chunkSize;
          self.uploadDate = doc.uploadDate;
          self.aliases = doc.aliases;
          self.length = doc.length;
          self.metadata = doc.metadata;
          self.internalMd5 = doc.md5;
      ***REMOVED*** else if (self.mode != 'r'***REMOVED*** {
          self.fileId = self.fileId == null ? new ObjectID(***REMOVED*** : self.fileId;
          self.contentType = exports.GridStore.DEFAULT_CONTENT_TYPE;
          self.internalChunkSize = self.internalChunkSize == null ? Chunk.DEFAULT_CHUNK_SIZE : self.internalChunkSize;
          self.length = 0;
      ***REMOVED*** else {
          self.length = 0;
          var txtId = self.fileId instanceof ObjectID ? self.fileId.toHexString(***REMOVED*** : self.fileId;
          return error(new Error((self.referenceBy == REFERENCE_BY_ID ? txtId : self.filename***REMOVED*** + " does not exist", self***REMOVED******REMOVED***;
      ***REMOVED***

        // Process the mode of the object
        if(self.mode == "r"***REMOVED*** {
          nthChunk(self, 0, options, function(err, chunk***REMOVED*** {
            if(err***REMOVED*** return error(err***REMOVED***;
            self.currentChunk = chunk;
            self.position = 0;
            callback(null, self***REMOVED***;
        ***REMOVED******REMOVED***;
      ***REMOVED*** else if(self.mode == "w"***REMOVED*** {
          // Delete any existing chunks
          deleteChunks(self, options, function(err, result***REMOVED*** {
            if(err***REMOVED*** return error(err***REMOVED***;
            self.currentChunk = new Chunk(self, {'n':0}, self.writeConcern***REMOVED***;
            self.contentType = self.options['content_type'] == null ? self.contentType : self.options['content_type'];
            self.internalChunkSize = self.options['chunk_size'] == null ? self.internalChunkSize : self.options['chunk_size'];
            self.metadata = self.options['metadata'] == null ? self.metadata : self.options['metadata'];
            self.aliases = self.options['aliases'] == null ? self.aliases : self.options['aliases'];
            self.position = 0;
            callback(null, self***REMOVED***;
        ***REMOVED******REMOVED***;
      ***REMOVED*** else if(self.mode == "w+"***REMOVED*** {
          nthChunk(self, lastChunkNumber(self***REMOVED***, options, function(err, chunk***REMOVED*** {
            if(err***REMOVED*** return error(err***REMOVED***;
            // Set the current chunk
            self.currentChunk = chunk == null ? new Chunk(self, {'n':0}, self.writeConcern***REMOVED*** : chunk;
            self.currentChunk.position = self.currentChunk.data.length(***REMOVED***;
            self.metadata = self.options['metadata'] == null ? self.metadata : self.options['metadata'];
            self.aliases = self.options['aliases'] == null ? self.aliases : self.options['aliases'];
            self.position = self.length;
            callback(null, self***REMOVED***;
        ***REMOVED******REMOVED***;
      ***REMOVED***
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED*** else {
    // Write only mode
    self.fileId = null == self.fileId ? new ObjectID(***REMOVED*** : self.fileId;
    self.contentType = exports.GridStore.DEFAULT_CONTENT_TYPE;
    self.internalChunkSize = self.internalChunkSize == null ? Chunk.DEFAULT_CHUNK_SIZE : self.internalChunkSize;
    self.length = 0;

    var collection2 = self.chunkCollection(***REMOVED***;
    // No file exists set up write mode
    if(self.mode == "w"***REMOVED*** {
      // Delete any existing chunks
      deleteChunks(self, options, function(err, result***REMOVED*** {
        if(err***REMOVED*** return error(err***REMOVED***;
        self.currentChunk = new Chunk(self, {'n':0}, self.writeConcern***REMOVED***;
        self.contentType = self.options['content_type'] == null ? self.contentType : self.options['content_type'];
        self.internalChunkSize = self.options['chunk_size'] == null ? self.internalChunkSize : self.options['chunk_size'];
        self.metadata = self.options['metadata'] == null ? self.metadata : self.options['metadata'];
        self.aliases = self.options['aliases'] == null ? self.aliases : self.options['aliases'];
        self.position = 0;
        callback(null, self***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED*** else if(self.mode == "w+"***REMOVED*** {
      nthChunk(self, lastChunkNumber(self***REMOVED***, options, function(err, chunk***REMOVED*** {
        if(err***REMOVED*** return error(err***REMOVED***;
        // Set the current chunk
        self.currentChunk = chunk == null ? new Chunk(self, {'n':0}, self.writeConcern***REMOVED*** : chunk;
        self.currentChunk.position = self.currentChunk.data.length(***REMOVED***;
        self.metadata = self.options['metadata'] == null ? self.metadata : self.options['metadata'];
        self.aliases = self.options['aliases'] == null ? self.aliases : self.options['aliases'];
        self.position = self.length;
        callback(null, self***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED***

  // only pass error to callback once
  function error (err***REMOVED*** {
    if(error.err***REMOVED*** return;
    callback(error.err = err***REMOVED***;
***REMOVED***
};

/**
 * Stores a file from the file system to the GridFS database.
 *
 * @param {String|Buffer|FileHandle} file the file to store.
 * @param {Function} callback this will be called after this method is executed. The first parameter will be null and the the second will contain the reference to this object.
 * @return {null}
 * @api public
 */
GridStore.prototype.writeFile = function (file, callback***REMOVED*** {
  var self = this;
  if (typeof file === 'string'***REMOVED*** {
    fs.open(file, 'r', function (err, fd***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;
      self.writeFile(fd, callback***REMOVED***;
  ***REMOVED******REMOVED***;
    return;
***REMOVED***

  self.open(function (err, self***REMOVED*** {
    if(err***REMOVED*** return callback(err, self***REMOVED***;

    fs.fstat(file, function (err, stats***REMOVED*** {
      if(err***REMOVED*** return callback(err, self***REMOVED***;

      var offset = 0;
      var index = 0;
      var numberOfChunksLeft = Math.min(stats.size / self.chunkSize***REMOVED***;

      // Write a chunk
      var writeChunk = function(***REMOVED*** {
        fs.read(file, self.chunkSize, offset, 'binary', function(err, data, bytesRead***REMOVED*** {
          if(err***REMOVED*** return callback(err, self***REMOVED***;

          offset = offset + bytesRead;

          // Create a new chunk for the data
          var chunk = new Chunk(self, {n:index++}, self.writeConcern***REMOVED***;
          chunk.write(data, function(err, chunk***REMOVED*** {
            if(err***REMOVED*** return callback(err, self***REMOVED***;

            chunk.save({}, function(err, result***REMOVED*** {
              if(err***REMOVED*** return callback(err, self***REMOVED***;

              self.position = self.position + data.length;

              // Point to current chunk
              self.currentChunk = chunk;

              if(offset >= stats.size***REMOVED*** {
                fs.close(file***REMOVED***;
                self.close(function(err, result***REMOVED*** {
                  if(err***REMOVED*** return callback(err, self***REMOVED***;
                  return callback(null, self***REMOVED***;
              ***REMOVED******REMOVED***;
            ***REMOVED*** else {
                return processor(writeChunk***REMOVED***;
            ***REMOVED***
          ***REMOVED******REMOVED***;
        ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
    ***REMOVED***

      // Process the first write
      processor(writeChunk***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Writes some data. This method will work properly only if initialized with mode
 * "w" or "w+".
 *
 * @param string {string} The data to write.
 * @param close {boolean=false} opt_argument Closes this file after writing if
 *     true.
 * @param callback {function(*, GridStore***REMOVED***} This will be called after executing
 *     this method. The first parameter will contain null and the second one
 *     will contain a reference to this object.
 *
 * @ignore
 * @api private
 */
var writeBuffer = function(self, buffer, close, callback***REMOVED*** {
  if(typeof close === "function"***REMOVED*** { callback = close; close = null; }
  var finalClose = typeof close == 'boolean' ? close : false;

  if(self.mode[0] != "w"***REMOVED*** {
    callback(new Error((self.referenceBy == REFERENCE_BY_ID ? self.toHexString(***REMOVED*** : self.filename***REMOVED*** + " not opened for writing"***REMOVED***, null***REMOVED***;
***REMOVED*** else {
    if(self.currentChunk.position + buffer.length >= self.chunkSize***REMOVED*** {
      // Write out the current Chunk and then keep writing until we have less data left than a chunkSize left
      // to a new chunk (recursively***REMOVED***
      var previousChunkNumber = self.currentChunk.chunkNumber;
      var leftOverDataSize = self.chunkSize - self.currentChunk.position;
      var firstChunkData = buffer.slice(0, leftOverDataSize***REMOVED***;
      var leftOverData = buffer.slice(leftOverDataSize***REMOVED***;
      // A list of chunks to write out
      var chunksToWrite = [self.currentChunk.write(firstChunkData***REMOVED***];
      // If we have more data left than the chunk size let's keep writing new chunks
      while(leftOverData.length >= self.chunkSize***REMOVED*** {
        // Create a new chunk and write to it
        var newChunk = new Chunk(self, {'n': (previousChunkNumber + 1***REMOVED***}, self.writeConcern***REMOVED***;
        var firstChunkData = leftOverData.slice(0, self.chunkSize***REMOVED***;
        leftOverData = leftOverData.slice(self.chunkSize***REMOVED***;
        // Update chunk number
        previousChunkNumber = previousChunkNumber + 1;
        // Write data
        newChunk.write(firstChunkData***REMOVED***;
        // Push chunk to save list
        chunksToWrite.push(newChunk***REMOVED***;
    ***REMOVED***

      // Set current chunk with remaining data
      self.currentChunk = new Chunk(self, {'n': (previousChunkNumber + 1***REMOVED***}, self.writeConcern***REMOVED***;
      // If we have left over data write it
      if(leftOverData.length > 0***REMOVED*** self.currentChunk.write(leftOverData***REMOVED***;

      // Update the position for the gridstore
      self.position = self.position + buffer.length;
      // Total number of chunks to write
      var numberOfChunksToWrite = chunksToWrite.length;

      for(var i = 0; i < chunksToWrite.length; i++***REMOVED*** {
        chunksToWrite[i].save({}, function(err, result***REMOVED*** {
          if(err***REMOVED*** return callback(err***REMOVED***;

          numberOfChunksToWrite = numberOfChunksToWrite - 1;

          if(numberOfChunksToWrite <= 0***REMOVED*** {
            // We care closing the file before returning
            if(finalClose***REMOVED*** {
              return self.close(function(err, result***REMOVED*** {
                callback(err, self***REMOVED***;
            ***REMOVED******REMOVED***;
          ***REMOVED***
            
            // Return normally
            return callback(null, self***REMOVED***;
        ***REMOVED***
      ***REMOVED******REMOVED***;
    ***REMOVED***
  ***REMOVED*** else {
      // Update the position for the gridstore
      self.position = self.position + buffer.length;
      // We have less data than the chunk size just write it and callback
      self.currentChunk.write(buffer***REMOVED***;
      // We care closing the file before returning
      if(finalClose***REMOVED*** {
        return self.close(function(err, result***REMOVED*** {
          callback(err, self***REMOVED***;
      ***REMOVED******REMOVED***;
    ***REMOVED***
      // Return normally
      return callback(null, self***REMOVED***;
  ***REMOVED***
***REMOVED***
};

/**
 * Creates a mongoDB object representation of this object.
 *
 * @param callback {function(object***REMOVED***} This will be called after executing this
 *     method. The object will be passed to the first parameter and will have
 *     the structure:
 *
 *        <pre><code>
 *        {
 *          '_id' : , // {number} id for this file
 *          'filename' : , // {string} name for this file
 *          'contentType' : , // {string} mime type for this file
 *          'length' : , // {number} size of this file?
 *          'chunksize' : , // {number} chunk size used by this file
 *          'uploadDate' : , // {Date}
 *          'aliases' : , // {array of string}
 *          'metadata' : , // {string}
 *      ***REMOVED***
 *        </code></pre>
 *
 * @ignore
 * @api private
 */
var buildMongoObject = function(self, callback***REMOVED*** {
  // Calcuate the length
  var mongoObject = {
    '_id': self.fileId,
    'filename': self.filename,
    'contentType': self.contentType,
    'length': self.position ? self.position : 0,
    'chunkSize': self.chunkSize,
    'uploadDate': self.uploadDate,
    'aliases': self.aliases,
    'metadata': self.metadata
***REMOVED***;

  var md5Command = {filemd5:self.fileId, root:self.root};
  self.db.command(md5Command, function(err, results***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;

    mongoObject.md5 = results.md5;
    callback(null, mongoObject***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Saves this file to the database. This will overwrite the old entry if it
 * already exists. This will work properly only if mode was initialized to
 * "w" or "w+".
 *
 * @param {Function} callback this will be called after executing this method. Passes an **{Error}** object to the first parameter and null to the second if an error occured. Otherwise, passes null to the first and a reference to this object to the second.
 * @return {null}
 * @api public
 */
GridStore.prototype.close = function(callback***REMOVED*** {
  var self = this;

  if(self.mode[0] == "w"***REMOVED*** {
    // Set up options
    var options = self.writeConcern;

    if(self.currentChunk != null && self.currentChunk.position > 0***REMOVED*** {
      self.currentChunk.save({}, function(err, chunk***REMOVED*** {
        if(err && typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***;

        self.collection(function(err, files***REMOVED*** {
          if(err && typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***;

          // Build the mongo object
          if(self.uploadDate != null***REMOVED*** {
            files.remove({'_id':self.fileId}, self.writeConcern, function(err, collection***REMOVED*** {
              if(err && typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***;

              buildMongoObject(self, function(err, mongoObject***REMOVED*** {
                if(err***REMOVED*** {
                  if(typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***; else throw err;
              ***REMOVED***

                files.save(mongoObject, options, function(err***REMOVED*** {
                  if(typeof callback == 'function'***REMOVED***
                    callback(err, mongoObject***REMOVED***;
              ***REMOVED******REMOVED***;
            ***REMOVED******REMOVED***;
          ***REMOVED******REMOVED***;
        ***REMOVED*** else {
            self.uploadDate = new Date(***REMOVED***;
            buildMongoObject(self, function(err, mongoObject***REMOVED*** {
              if(err***REMOVED*** {
                if(typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***; else throw err;
            ***REMOVED***

              files.save(mongoObject, options, function(err***REMOVED*** {
                if(typeof callback == 'function'***REMOVED***
                  callback(err, mongoObject***REMOVED***;
            ***REMOVED******REMOVED***;
          ***REMOVED******REMOVED***;
        ***REMOVED***
      ***REMOVED******REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED*** else {
      self.collection(function(err, files***REMOVED*** {
        if(err && typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***;

        self.uploadDate = new Date(***REMOVED***;
        buildMongoObject(self, function(err, mongoObject***REMOVED*** {
          if(err***REMOVED*** {
            if(typeof callback == 'function'***REMOVED*** return callback(err***REMOVED***; else throw err;
        ***REMOVED***

          files.save(mongoObject, options, function(err***REMOVED*** {
            if(typeof callback == 'function'***REMOVED***
              callback(err, mongoObject***REMOVED***;
        ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED*** else if(self.mode[0] == "r"***REMOVED*** {
    if(typeof callback == 'function'***REMOVED***
      callback(null, null***REMOVED***;
***REMOVED*** else {
    if(typeof callback == 'function'***REMOVED***
      callback(new Error("Illegal mode " + self.mode***REMOVED***, null***REMOVED***;
***REMOVED***
};

/**
 * Gets the nth chunk of this file.
 *
 * @param chunkNumber {number} The nth chunk to retrieve.
 * @param callback {function(*, Chunk|object***REMOVED***} This will be called after
 *     executing this method. null will be passed to the first parameter while
 *     a new {@link Chunk} instance will be passed to the second parameter if
 *     the chunk was found or an empty object {} if not.
 *
 * @ignore
 * @api private
 */
var nthChunk = function(self, chunkNumber, options, callback***REMOVED*** {
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  options = options || self.writeConcern;
  options.readPreference = self.readPreference;
  // Get the nth chunk
  self.chunkCollection(***REMOVED***.find({'files_id':self.fileId, 'n':chunkNumber}, options, function(err, cursor***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;

    cursor.nextObject(function(err, chunk***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;

      var finalChunk = chunk == null ? {} : chunk;
      callback(null, new Chunk(self, finalChunk, self.writeConcern***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 *
 * @ignore
 * @api private
 */
GridStore.prototype._nthChunk = function(chunkNumber, callback***REMOVED*** {
  nthChunk(this, chunkNumber, callback***REMOVED***;
}

/**
 * @return {Number} The last chunk number of this file.
 *
 * @ignore
 * @api private
 */
var lastChunkNumber = function(self***REMOVED*** {
  return Math.floor((self.length ? self.length - 1 : 0***REMOVED***/self.chunkSize***REMOVED***;
};

/**
 * Retrieve this file's chunks collection.
 *
 * @param {Function} callback this will be called after executing this method. An exception object will be passed to the first parameter when an error occured or null otherwise. A new **{Collection}** object will be passed to the second parameter if no error occured.
 * @return {null}
 * @api public
 */
GridStore.prototype.chunkCollection = function(callback***REMOVED*** {
  if(typeof callback == 'function'***REMOVED***
    return this.db.collection((this.root + ".chunks"***REMOVED***, callback***REMOVED***;
  return this.db.collection((this.root + ".chunks"***REMOVED******REMOVED***;
};

/**
 * Deletes all the chunks of this file in the database.
 *
 * @param callback {function(*, boolean***REMOVED***} This will be called after this method
 *     executes. Passes null to the first and true to the second argument.
 *
 * @ignore
 * @api private
 */
var deleteChunks = function(self, options, callback***REMOVED*** {
  if(typeof options == 'function'***REMOVED*** {
    callback = options;
    options = {};
***REMOVED***

  options = options || self.writeConcern;

  if(self.fileId != null***REMOVED*** {
    self.chunkCollection(***REMOVED***.remove({'files_id':self.fileId}, options, function(err, result***REMOVED*** {
      if(err***REMOVED*** return callback(err, false***REMOVED***;
      callback(null, true***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED*** else {
    callback(null, true***REMOVED***;
***REMOVED***
};

/**
 * Deletes all the chunks of this file in the database.
 *
 * @param {Function} callback this will be called after this method executes. Passes null to the first and true to the second argument.
 * @return {null}
 * @api public
 */
GridStore.prototype.unlink = function(callback***REMOVED*** {
  var self = this;
  deleteChunks(this, function(err***REMOVED*** {
    if(err!==null***REMOVED*** {
      err.message = "at deleteChunks: " + err.message;
      return callback(err***REMOVED***;
  ***REMOVED***

    self.collection(function(err, collection***REMOVED*** {
      if(err!==null***REMOVED*** {
        err.message = "at collection: " + err.message;
        return callback(err***REMOVED***;
    ***REMOVED***

      collection.remove({'_id':self.fileId}, self.writeConcern, function(err***REMOVED*** {
        callback(err, self***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Retrieves the file collection associated with this object.
 *
 * @param {Function} callback this will be called after executing this method. An exception object will be passed to the first parameter when an error occured or null otherwise. A new **{Collection}** object will be passed to the second parameter if no error occured.
 * @return {null}
 * @api public
 */
GridStore.prototype.collection = function(callback***REMOVED*** {
  if(typeof callback == 'function'***REMOVED***
    this.db.collection(this.root + ".files", callback***REMOVED***;
  return this.db.collection(this.root + ".files"***REMOVED***;
};

/**
 * Reads the data of this file.
 *
 * @param {String} [separator] the character to be recognized as the newline separator.
 * @param {Function} callback This will be called after this method is executed. The first parameter will be null and the second parameter will contain an array of strings representing the entire data, each element representing a line including the separator character.
 * @return {null}
 * @api public
 */
GridStore.prototype.readlines = function(separator, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 0***REMOVED***;
  callback = args.pop(***REMOVED***;
  separator = args.length ? args.shift(***REMOVED*** : "\n";

  this.read(function(err, data***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;

    var items = data.toString(***REMOVED***.split(separator***REMOVED***;
    items = items.length > 0 ? items.splice(0, items.length - 1***REMOVED*** : [];
    for(var i = 0; i < items.length; i++***REMOVED*** {
      items[i] = items[i] + separator;
  ***REMOVED***

    callback(null, items***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Deletes all the chunks of this file in the database if mode was set to "w" or
 * "w+" and resets the read/write head to the initial position.
 *
 * @param {Function} callback this will be called after executing this method. The first parameter will contain null and the second one will contain a reference to this object.
 * @return {null}
 * @api public
 */
GridStore.prototype.rewind = function(callback***REMOVED*** {
  var self = this;

  if(this.currentChunk.chunkNumber != 0***REMOVED*** {
    if(this.mode[0] == "w"***REMOVED*** {
      deleteChunks(self, function(err, gridStore***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        self.currentChunk = new Chunk(self, {'n': 0}, self.writeConcern***REMOVED***;
        self.position = 0;
        callback(null, self***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED*** else {
      self.currentChunk(0, function(err, chunk***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        self.currentChunk = chunk;
        self.currentChunk.rewind(***REMOVED***;
        self.position = 0;
        callback(null, self***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED*** else {
    self.currentChunk.rewind(***REMOVED***;
    self.position = 0;
    callback(null, self***REMOVED***;
***REMOVED***
};

/**
 * Retrieves the contents of this file and advances the read/write head. Works with Buffers only.
 *
 * There are 3 signatures for this method:
 *
 * (callback***REMOVED***
 * (length, callback***REMOVED***
 * (length, buffer, callback***REMOVED***
 *
 * @param {Number} [length] the number of characters to read. Reads all the characters from the read/write head to the EOF if not specified.
 * @param {String|Buffer} [buffer] a string to hold temporary data. This is used for storing the string data read so far when recursively calling this method.
 * @param {Function} callback this will be called after this method is executed. null will be passed to the first parameter and a string containing the contents of the buffer concatenated with the contents read from this file will be passed to the second.
 * @return {null}
 * @api public
 */
GridStore.prototype.read = function(length, buffer, callback***REMOVED*** {
  var self = this;

  var args = Array.prototype.slice.call(arguments, 0***REMOVED***;
  callback = args.pop(***REMOVED***;
  length = args.length ? args.shift(***REMOVED*** : null;
  buffer = args.length ? args.shift(***REMOVED*** : null;

  // The data is a c-terminated string and thus the length - 1
  var finalLength = length == null ? self.length - self.position : length;
  var finalBuffer = buffer == null ? new Buffer(finalLength***REMOVED*** : buffer;
  // Add a index to buffer to keep track of writing position or apply current index
  finalBuffer._index = buffer != null && buffer._index != null ? buffer._index : 0;

  if((self.currentChunk.length(***REMOVED*** - self.currentChunk.position + finalBuffer._index***REMOVED*** >= finalLength***REMOVED*** {
    var slice = self.currentChunk.readSlice(finalLength - finalBuffer._index***REMOVED***;
    // Copy content to final buffer
    slice.copy(finalBuffer, finalBuffer._index***REMOVED***;
    // Update internal position
    self.position = self.position + finalBuffer.length;
    // Check if we don't have a file at all
    if(finalLength == 0 && finalBuffer.length == 0***REMOVED*** return callback(new Error("File does not exist"***REMOVED***, null***REMOVED***;
    // Else return data
    callback(null, finalBuffer***REMOVED***;
***REMOVED*** else {
    var slice = self.currentChunk.readSlice(self.currentChunk.length(***REMOVED*** - self.currentChunk.position***REMOVED***;
    // Copy content to final buffer
    slice.copy(finalBuffer, finalBuffer._index***REMOVED***;
    // Update index position
    finalBuffer._index += slice.length;

    // Load next chunk and read more
    nthChunk(self, self.currentChunk.chunkNumber + 1, function(err, chunk***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;

      if(chunk.length(***REMOVED*** > 0***REMOVED*** {
        self.currentChunk = chunk;
        self.read(length, finalBuffer, callback***REMOVED***;
    ***REMOVED*** else {
        if (finalBuffer._index > 0***REMOVED*** {
          callback(null, finalBuffer***REMOVED***
      ***REMOVED*** else {
          callback(new Error("no chunks found for file, possibly corrupt"***REMOVED***, null***REMOVED***;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED***
}

/**
 * Retrieves the position of the read/write head of this file.
 *
 * @param {Function} callback This gets called after this method terminates. null is passed to the first parameter and the position is passed to the second.
 * @return {null}
 * @api public
 */
GridStore.prototype.tell = function(callback***REMOVED*** {
  callback(null, this.position***REMOVED***;
};

/**
 * Moves the read/write head to a new location.
 *
 * There are 3 signatures for this method
 *
 * Seek Location Modes
 *  - **GridStore.IO_SEEK_SET**, **(default***REMOVED***** set the position from the start of the file.
 *  - **GridStore.IO_SEEK_CUR**, set the position from the current position in the file.
 *  - **GridStore.IO_SEEK_END**, set the position from the end of the file.
 *
 * @param {Number} [position] the position to seek to
 * @param {Number} [seekLocation] seek mode. Use one of the Seek Location modes.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain null and the second one will contain a reference to this object.
 * @return {null}
 * @api public
 */
GridStore.prototype.seek = function(position, seekLocation, callback***REMOVED*** {
  var self = this;

  var args = Array.prototype.slice.call(arguments, 1***REMOVED***;
  callback = args.pop(***REMOVED***;
  seekLocation = args.length ? args.shift(***REMOVED*** : null;

  var seekLocationFinal = seekLocation == null ? exports.GridStore.IO_SEEK_SET : seekLocation;
  var finalPosition = position;
  var targetPosition = 0;

  // Calculate the position
  if(seekLocationFinal == exports.GridStore.IO_SEEK_CUR***REMOVED*** {
    targetPosition = self.position + finalPosition;
***REMOVED*** else if(seekLocationFinal == exports.GridStore.IO_SEEK_END***REMOVED*** {
    targetPosition = self.length + finalPosition;
***REMOVED*** else {
    targetPosition = finalPosition;
***REMOVED***

  // Get the chunk
  var newChunkNumber = Math.floor(targetPosition/self.chunkSize***REMOVED***;
  if(newChunkNumber != self.currentChunk.chunkNumber***REMOVED*** {
    var seekChunk = function(***REMOVED*** {
      nthChunk(self, newChunkNumber, function(err, chunk***REMOVED*** {
        self.currentChunk = chunk;
        self.position = targetPosition;
        self.currentChunk.position = (self.position % self.chunkSize***REMOVED***;
        callback(err, self***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED***;

    if(self.mode[0] == 'w'***REMOVED*** {
      self.currentChunk.save({}, function(err***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        seekChunk(***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED*** else {
      seekChunk(***REMOVED***;
  ***REMOVED***
***REMOVED*** else {
    self.position = targetPosition;
    self.currentChunk.position = (self.position % self.chunkSize***REMOVED***;
    callback(null, self***REMOVED***;
***REMOVED***
};

/**
 * Verify if the file is at EOF.
 *
 * @return {Boolean} true if the read/write head is at the end of this file.
 * @api public
 */
GridStore.prototype.eof = function(***REMOVED*** {
  return this.position == this.length ? true : false;
};

/**
 * Retrieves a single character from this file.
 *
 * @param {Function} callback this gets called after this method is executed. Passes null to the first parameter and the character read to the second or null to the second if the read/write head is at the end of the file.
 * @return {null}
 * @api public
 */
GridStore.prototype.getc = function(callback***REMOVED*** {
  var self = this;

  if(self.eof(***REMOVED******REMOVED*** {
    callback(null, null***REMOVED***;
***REMOVED*** else if(self.currentChunk.eof(***REMOVED******REMOVED*** {
    nthChunk(self, self.currentChunk.chunkNumber + 1, function(err, chunk***REMOVED*** {
      self.currentChunk = chunk;
      self.position = self.position + 1;
      callback(err, self.currentChunk.getc(***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED*** else {
    self.position = self.position + 1;
    callback(null, self.currentChunk.getc(***REMOVED******REMOVED***;
***REMOVED***
};

/**
 * Writes a string to the file with a newline character appended at the end if
 * the given string does not have one.
 *
 * @param {String} string the string to write.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain null and the second one will contain a reference to this object.
 * @return {null}
 * @api public
 */
GridStore.prototype.puts = function(string, callback***REMOVED*** {
  var finalString = string.match(/\n$/***REMOVED*** == null ? string + "\n" : string;
  this.write(finalString, callback***REMOVED***;
};

/**
 * Returns read stream based on this GridStore file
 *
 * Events
 *  - **data** {function(item***REMOVED*** {}} the data event triggers when a document is ready.
 *  - **end** {function(***REMOVED*** {}} the end event triggers when there is no more documents available.
 *  - **close** {function(***REMOVED*** {}} the close event triggers when the stream is closed.
 *  - **error** {function(err***REMOVED*** {}} the error event triggers if an error happens.
 *
 * @param {Boolean} autoclose if true current GridStore will be closed when EOF and 'close' event will be fired
 * @return {null}
 * @api public
 */
GridStore.prototype.stream = function(autoclose***REMOVED*** {
  return new ReadStream(autoclose, this***REMOVED***;
};

/**
* The collection to be used for holding the files and chunks collection.
*
* @classconstant DEFAULT_ROOT_COLLECTION
**/
GridStore.DEFAULT_ROOT_COLLECTION = 'fs';

/**
* Default file mime type
*
* @classconstant DEFAULT_CONTENT_TYPE
**/
GridStore.DEFAULT_CONTENT_TYPE = 'binary/octet-stream';

/**
* Seek mode where the given length is absolute.
*
* @classconstant IO_SEEK_SET
**/
GridStore.IO_SEEK_SET = 0;

/**
* Seek mode where the given length is an offset to the current read/write head.
*
* @classconstant IO_SEEK_CUR
**/
GridStore.IO_SEEK_CUR = 1;

/**
* Seek mode where the given length is an offset to the end of the file.
*
* @classconstant IO_SEEK_END
**/
GridStore.IO_SEEK_END = 2;

/**
 * Checks if a file exists in the database.
 *
 * Options
 *  - **readPreference** {String}, the prefered read preference (ReadPreference.PRIMARY, ReadPreference.PRIMARY_PREFERRED, ReadPreference.SECONDARY, ReadPreference.SECONDARY_PREFERRED, ReadPreference.NEAREST***REMOVED***.
 *
 * @param {Db} db the database to query.
 * @param {String} name the name of the file to look for.
 * @param {String} [rootCollection] the root collection that holds the files and chunks collection. Defaults to **{GridStore.DEFAULT_ROOT_COLLECTION}**.
 * @param {Function} callback this will be called after this method executes. Passes null to the first and passes true to the second if the file exists and false otherwise.
 * @return {null}
 * @api public
 */
GridStore.exist = function(db, fileIdObject, rootCollection, options, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  rootCollection = args.length ? args.shift(***REMOVED*** : null;
  options = args.length ? args.shift(***REMOVED*** : {};

  // Establish read preference
  var readPreference = options.readPreference || 'primary';
  // Fetch collection
  var rootCollectionFinal = rootCollection != null ? rootCollection : GridStore.DEFAULT_ROOT_COLLECTION;
  db.collection(rootCollectionFinal + ".files", function(err, collection***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;

    // Build query
    var query = (typeof fileIdObject == 'string' || Object.prototype.toString.call(fileIdObject***REMOVED*** == '[object RegExp]' ***REMOVED***
      ? {'filename':fileIdObject}
      : {'_id':fileIdObject};    // Attempt to locate file

    collection.find(query, {readPreference:readPreference}, function(err, cursor***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;

      cursor.nextObject(function(err, item***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        callback(null, item == null ? false : true***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Gets the list of files stored in the GridFS.
 *
 * @param {Db} db the database to query.
 * @param {String} [rootCollection] the root collection that holds the files and chunks collection. Defaults to **{GridStore.DEFAULT_ROOT_COLLECTION}**.
 * @param {Function} callback this will be called after this method executes. Passes null to the first and passes an array of strings containing the names of the files.
 * @return {null}
 * @api public
 */
GridStore.list = function(db, rootCollection, options, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 1***REMOVED***;
  callback = args.pop(***REMOVED***;
  rootCollection = args.length ? args.shift(***REMOVED*** : null;
  options = args.length ? args.shift(***REMOVED*** : {};

  // Ensure we have correct values
  if(rootCollection != null && typeof rootCollection == 'object'***REMOVED*** {
    options = rootCollection;
    rootCollection = null;
***REMOVED***

  // Establish read preference
  var readPreference = options.readPreference || 'primary';
  // Check if we are returning by id not filename
  var byId = options['id'] != null ? options['id'] : false;
  // Fetch item
  var rootCollectionFinal = rootCollection != null ? rootCollection : GridStore.DEFAULT_ROOT_COLLECTION;
  var items = [];
  db.collection((rootCollectionFinal + ".files"***REMOVED***, function(err, collection***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;

    collection.find({}, {readPreference:readPreference}, function(err, cursor***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;

      cursor.each(function(err, item***REMOVED*** {
        if(item != null***REMOVED*** {
          items.push(byId ? item._id : item.filename***REMOVED***;
      ***REMOVED*** else {
          callback(err, items***REMOVED***;
      ***REMOVED***
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Reads the contents of a file.
 *
 * This method has the following signatures
 *
 * (db, name, callback***REMOVED***
 * (db, name, length, callback***REMOVED***
 * (db, name, length, offset, callback***REMOVED***
 * (db, name, length, offset, options, callback***REMOVED***
 *
 * @param {Db} db the database to query.
 * @param {String} name the name of the file.
 * @param {Number} [length] the size of data to read.
 * @param {Number} [offset] the offset from the head of the file of which to start reading from.
 * @param {Object} [options] the options for the file.
 * @param {Function} callback this will be called after this method executes. A string with an error message will be passed to the first parameter when the length and offset combination exceeds the length of the file while an Error object will be passed if other forms of error occured, otherwise, a string is passed. The second parameter will contain the data read if successful or null if an error occured.
 * @return {null}
 * @api public
 */
GridStore.read = function(db, name, length, offset, options, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  length = args.length ? args.shift(***REMOVED*** : null;
  offset = args.length ? args.shift(***REMOVED*** : null;
  options = args.length ? args.shift(***REMOVED*** : null;

  new GridStore(db, name, "r", options***REMOVED***.open(function(err, gridStore***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;
    // Make sure we are not reading out of bounds
    if(offset && offset >= gridStore.length***REMOVED*** return callback("offset larger than size of file", null***REMOVED***;
    if(length && length > gridStore.length***REMOVED*** return callback("length is larger than the size of the file", null***REMOVED***;
    if(offset && length && (offset + length***REMOVED*** > gridStore.length***REMOVED*** return callback("offset and length is larger than the size of the file", null***REMOVED***;

    if(offset != null***REMOVED*** {
      gridStore.seek(offset, function(err, gridStore***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        gridStore.read(length, callback***REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED*** else {
      gridStore.read(length, callback***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;
};

/**
 * Reads the data of this file.
 *
 * @param {Db} db the database to query.
 * @param {String} name the name of the file.
 * @param {String} [separator] the character to be recognized as the newline separator.
 * @param {Object} [options] file options.
 * @param {Function} callback this will be called after this method is executed. The first parameter will be null and the second parameter will contain an array of strings representing the entire data, each element representing a line including the separator character.
 * @return {null}
 * @api public
 */
GridStore.readlines = function(db, name, separator, options, callback***REMOVED*** {
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  separator = args.length ? args.shift(***REMOVED*** : null;
  options = args.length ? args.shift(***REMOVED*** : null;

  var finalSeperator = separator == null ? "\n" : separator;
  new GridStore(db, name, "r", options***REMOVED***.open(function(err, gridStore***REMOVED*** {
    if(err***REMOVED*** return callback(err***REMOVED***;
    gridStore.readlines(finalSeperator, callback***REMOVED***;
***REMOVED******REMOVED***;
};

/**
 * Deletes the chunks and metadata information of a file from GridFS.
 *
 * @param {Db} db the database to interact with.
 * @param {String|Array} names the name/names of the files to delete.
 * @param {Object} [options] the options for the files.
 * @callback {Function} this will be called after this method is executed. The first parameter will contain an Error object if an error occured or null otherwise. The second parameter will contain a reference to this object.
 * @return {null}
 * @api public
 */
GridStore.unlink = function(db, names, options, callback***REMOVED*** {
  var self = this;
  var args = Array.prototype.slice.call(arguments, 2***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** : {};

  // Get the write concern
  var writeConcern = _getWriteConcern(db, options***REMOVED***;

  // List of names
  if(names.constructor == Array***REMOVED*** {
    var tc = 0;
    for(var i = 0; i < names.length; i++***REMOVED*** {
      ++tc;
      GridStore.unlink(db, names[i], options, function(result***REMOVED*** {
        if(--tc == 0***REMOVED*** {
            callback(null, self***REMOVED***;
      ***REMOVED***
    ***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED*** else {
    new GridStore(db, names, "w", options***REMOVED***.open(function(err, gridStore***REMOVED*** {
      if(err***REMOVED*** return callback(err***REMOVED***;
      deleteChunks(gridStore, function(err, result***REMOVED*** {
        if(err***REMOVED*** return callback(err***REMOVED***;
        gridStore.collection(function(err, collection***REMOVED*** {
          if(err***REMOVED*** return callback(err***REMOVED***;
          collection.remove({'_id':gridStore.fileId}, writeConcern, function(err, result***REMOVED*** {
            callback(err, self***REMOVED***;
        ***REMOVED******REMOVED***;
      ***REMOVED******REMOVED***;
    ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED***
};

/**
 * Returns the current chunksize of the file.
 *
 * @field chunkSize
 * @type {Number}
 * @getter
 * @setter
 * @property return number of bytes in the current chunkSize.
 */
Object.defineProperty(GridStore.prototype, "chunkSize", { enumerable: true
 , get: function (***REMOVED*** {
     return this.internalChunkSize;
 ***REMOVED***
 , set: function(value***REMOVED*** {
     if(!(this.mode[0] == "w" && this.position == 0 && this.uploadDate == null***REMOVED******REMOVED*** {
       this.internalChunkSize = this.internalChunkSize;
   ***REMOVED*** else {
       this.internalChunkSize = value;
   ***REMOVED***
 ***REMOVED***
}***REMOVED***;

/**
 * The md5 checksum for this file.
 *
 * @field md5
 * @type {Number}
 * @getter
 * @setter
 * @property return this files md5 checksum.
 */
Object.defineProperty(GridStore.prototype, "md5", { enumerable: true
 , get: function (***REMOVED*** {
     return this.internalMd5;
 ***REMOVED***
}***REMOVED***;

/**
 *  GridStore Streaming methods
 *  Handles the correct return of the writeable stream status
 *  @ignore
 */
Object.defineProperty(GridStore.prototype, "writable", { enumerable: true
 , get: function (***REMOVED*** {
    if(this._writeable == null***REMOVED*** {
      this._writeable = this.mode != null && this.mode.indexOf("w"***REMOVED*** != -1;
  ***REMOVED***
    // Return the _writeable
    return this._writeable;
***REMOVED***
 , set: function(value***REMOVED*** {
    this._writeable = value;
***REMOVED***
}***REMOVED***;

/**
 *  Handles the correct return of the readable stream status
 *  @ignore
 */
Object.defineProperty(GridStore.prototype, "readable", { enumerable: true
 , get: function (***REMOVED*** {
    if(this._readable == null***REMOVED*** {
      this._readable = this.mode != null && this.mode.indexOf("r"***REMOVED*** != -1;
  ***REMOVED***
    return this._readable;
***REMOVED***
 , set: function(value***REMOVED*** {
    this._readable = value;
***REMOVED***
}***REMOVED***;

GridStore.prototype.paused;

/**
 *  Handles the correct setting of encoding for the stream
 *  @ignore
 */
GridStore.prototype.setEncoding = fs.ReadStream.prototype.setEncoding;

/**
 *  Handles the end events
 *  @ignore
 */
GridStore.prototype.end = function end(data***REMOVED*** {
  var self = this;
  // allow queued data to write before closing
  if(!this.writable***REMOVED*** return;
  this.writable = false;

  if(data***REMOVED*** {
    this._q.push(data***REMOVED***;
***REMOVED***

  this.on('drain', function (***REMOVED*** {
    self.close(function (err***REMOVED*** {
      if (err***REMOVED*** return _error(self, err***REMOVED***;
      self.emit('close'***REMOVED***;
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***;

  _flush(self***REMOVED***;
}

/**
 *  Handles the normal writes to gridstore
 *  @ignore
 */
var _writeNormal = function(self, data, close, callback***REMOVED*** {
  // If we have a buffer write it using the writeBuffer method
  if(Buffer.isBuffer(data***REMOVED******REMOVED*** {
    return writeBuffer(self, data, close, callback***REMOVED***;
***REMOVED*** else {
    return writeBuffer(self, new Buffer(data, 'binary'***REMOVED***, close, callback***REMOVED***;
***REMOVED***
}

/**
 * Writes some data. This method will work properly only if initialized with mode "w" or "w+".
 *
 * @param {String|Buffer} data the data to write.
 * @param {Boolean} [close] closes this file after writing if set to true.
 * @param {Function} callback this will be called after executing this method. The first parameter will contain null and the second one will contain a reference to this object.
 * @return {null}
 * @api public
 */
GridStore.prototype.write = function write(data, close, callback***REMOVED*** {
  // If it's a normal write delegate the call
  if(typeof close == 'function' || typeof callback == 'function'***REMOVED*** {
    return _writeNormal(this, data, close, callback***REMOVED***;
***REMOVED***

  // Otherwise it's a stream write
  var self = this;
  if (!this.writable***REMOVED*** {
    throw new Error('GridWriteStream is not writable'***REMOVED***;
***REMOVED***

  // queue data until we open.
  if(!this._opened***REMOVED*** {
    // Set up a queue to save data until gridstore object is ready
    this._q = [];
    _openStream(self***REMOVED***;
    this._q.push(data***REMOVED***;
    return false;
***REMOVED***

  // Push data to queue
  this._q.push(data***REMOVED***;
  _flush(this***REMOVED***;
  // Return write successful
  return true;
}

/**
 *  Handles the destroy part of a stream
 *  @ignore
 */
GridStore.prototype.destroy = function destroy(***REMOVED*** {
  // close and do not emit any more events. queued data is not sent.
  if(!this.writable***REMOVED*** return;
  this.readable = false;
  if(this.writable***REMOVED*** {
    this.writable = false;
    this._q.length = 0;
    this.emit('close'***REMOVED***;
***REMOVED***
}

/**
 *  Handles the destroySoon part of a stream
 *  @ignore
 */
GridStore.prototype.destroySoon = function destroySoon(***REMOVED*** {
  // as soon as write queue is drained, destroy.
  // may call destroy immediately if no data is queued.
  if(!this._q.length***REMOVED*** {
    return this.destroy(***REMOVED***;
***REMOVED***
  this._destroying = true;
}

/**
 *  Handles the pipe part of the stream
 *  @ignore
 */
GridStore.prototype.pipe = function(destination, options***REMOVED*** {
  var self = this;
  // Open the gridstore
  this.open(function(err, result***REMOVED*** {
    if(err***REMOVED*** _errorRead(self, err***REMOVED***;
    if(!self.readable***REMOVED*** return;
    // Set up the pipe
    self._pipe(destination, options***REMOVED***;
    // Emit the stream is open
    self.emit('open'***REMOVED***;
    // Read from the stream
    _read(self***REMOVED***;
***REMOVED******REMOVED***
}

/**
 *  Internal module methods
 *  @ignore
 */
var _read = function _read(self***REMOVED*** {
  if (!self.readable || self.paused || self.reading***REMOVED*** {
    return;
***REMOVED***

  self.reading = true;
  var stream = self._stream = self.stream(***REMOVED***;
  stream.paused = self.paused;

  stream.on('data', function (data***REMOVED*** {
    if (self._decoder***REMOVED*** {
      var str = self._decoder.write(data***REMOVED***;
      if (str.length***REMOVED*** self.emit('data', str***REMOVED***;
  ***REMOVED*** else {
      self.emit('data', data***REMOVED***;
  ***REMOVED***
***REMOVED******REMOVED***;

  stream.on('end', function (data***REMOVED*** {
    self.emit('end', data***REMOVED***;
***REMOVED******REMOVED***;

  stream.on('error', function (data***REMOVED*** {
    _errorRead(self, data***REMOVED***;
***REMOVED******REMOVED***;

  stream.on('close', function (data***REMOVED*** {
    self.emit('close', data***REMOVED***;
***REMOVED******REMOVED***;

  self.pause = function (***REMOVED*** {
    // native doesn't always pause.
    // bypass its pause(***REMOVED*** method to hack it
    self.paused = stream.paused = true;
***REMOVED***

  self.resume = function (***REMOVED*** {
    if(!self.paused***REMOVED*** return;

    self.paused = false;
    stream.resume(***REMOVED***;
    self.readable = stream.readable;
***REMOVED***

  self.destroy = function (***REMOVED*** {
    self.readable = false;
    stream.destroy(***REMOVED***;
***REMOVED***
}

/**
 * pause
 * @ignore
 */
GridStore.prototype.pause = function pause (***REMOVED*** {
  // Overridden when the GridStore opens.
  this.paused = true;
}

/**
 * resume
 * @ignore
 */
GridStore.prototype.resume = function resume (***REMOVED*** {
  // Overridden when the GridStore opens.
  this.paused = false;
}

/**
 *  Internal module methods
 *  @ignore
 */
var _flush = function _flush(self, _force***REMOVED*** {
  if (!self._opened***REMOVED*** return;
  if (!_force && self._flushing***REMOVED*** return;
  self._flushing = true;

  // write the entire q to gridfs
  if (!self._q.length***REMOVED*** {
    self._flushing = false;
    self.emit('drain'***REMOVED***;

    if(self._destroying***REMOVED*** {
      self.destroy(***REMOVED***;
  ***REMOVED***
    return;
***REMOVED***

  self.write(self._q.shift(***REMOVED***, function (err, store***REMOVED*** {
    if (err***REMOVED*** return _error(self, err***REMOVED***;
    self.emit('progress', store.position***REMOVED***;
    _flush(self, true***REMOVED***;
***REMOVED******REMOVED***;
}

var _openStream = function _openStream (self***REMOVED*** {
  if(self._opening == true***REMOVED*** return;
  self._opening = true;

  // Open the store
  self.open(function (err, gridstore***REMOVED*** {
    if (err***REMOVED*** return _error(self, err***REMOVED***;
    self._opened = true;
    self.emit('open'***REMOVED***;
    _flush(self***REMOVED***;
***REMOVED******REMOVED***;
}

var _error = function _error(self, err***REMOVED*** {
  self.destroy(***REMOVED***;
  self.emit('error', err***REMOVED***;
}

var _errorRead = function _errorRead (self, err***REMOVED*** {
  self.readable = false;
  self.emit('error', err***REMOVED***;
}

/**
 * @ignore
 */
var _hasWriteConcern = function(errorOptions***REMOVED*** {
  return errorOptions == true
    || errorOptions.w > 0
    || errorOptions.w == 'majority'
    || errorOptions.j == true
    || errorOptions.journal == true
    || errorOptions.fsync == true
}

/**
 * @ignore
 */
var _setWriteConcernHash = function(options***REMOVED*** {
  var finalOptions = {};
  if(options.w != null***REMOVED*** finalOptions.w = options.w;  
  if(options.journal == true***REMOVED*** finalOptions.j = options.journal;
  if(options.j == true***REMOVED*** finalOptions.j = options.j;
  if(options.fsync == true***REMOVED*** finalOptions.fsync = options.fsync;
  if(options.wtimeout != null***REMOVED*** finalOptions.wtimeout = options.wtimeout;  
  return finalOptions;
}

/**
 * @ignore
 */
var _getWriteConcern = function(self, options***REMOVED*** {
  // Final options
  var finalOptions = {w:1};
  options = options || {};

  // Local options verification
  if(options.w != null || typeof options.j == 'boolean' || typeof options.journal == 'boolean' || typeof options.fsync == 'boolean'***REMOVED*** {
    finalOptions = _setWriteConcernHash(options***REMOVED***;
***REMOVED*** else if(options.safe != null && typeof options.safe == 'object'***REMOVED*** {
    finalOptions = _setWriteConcernHash(options.safe***REMOVED***;
***REMOVED*** else if(typeof options.safe == "boolean"***REMOVED*** {
    finalOptions = {w: (options.safe ? 1 : 0***REMOVED***};
***REMOVED*** else if(self.options.w != null || typeof self.options.j == 'boolean' || typeof self.options.journal == 'boolean' || typeof self.options.fsync == 'boolean'***REMOVED*** {
    finalOptions = _setWriteConcernHash(self.options***REMOVED***;
***REMOVED*** else if(self.safe.w != null || typeof self.safe.j == 'boolean' || typeof self.safe.journal == 'boolean' || typeof self.safe.fsync == 'boolean'***REMOVED*** {
    finalOptions = _setWriteConcernHash(self.safe***REMOVED***;
***REMOVED*** else if(typeof self.safe == "boolean"***REMOVED*** {
    finalOptions = {w: (self.safe ? 1 : 0***REMOVED***};
***REMOVED***

  // Ensure we don't have an invalid combination of write concerns
  if(finalOptions.w < 1 
    && (finalOptions.journal == true || finalOptions.j == true || finalOptions.fsync == true***REMOVED******REMOVED*** throw new Error("No acknowledgement using w < 1 cannot be combined with journal:true or fsync:true"***REMOVED***;

  // Return the options
  return finalOptions;
}

/**
 * @ignore
 * @api private
 */
exports.GridStore = GridStore;
