// mongodb/lib/mongodb/gridfs/grid.js
var GridStore = require('./gridstore'***REMOVED***.GridStore,
  ObjectID = require('bson'***REMOVED***.ObjectID;

/**
 * A class representation of a simple Grid interface.
 *
 * @class Represents the Grid.
 * @param {Db} db A database instance to interact with.
 * @param {String} [fsName] optional different root collection for GridFS.
 * @return {Grid}
 */
function Grid(db, fsName***REMOVED*** {

  if(!(this instanceof Grid***REMOVED******REMOVED*** return new Grid(db, fsName***REMOVED***;

  this.db = db;
  this.fsName = fsName == null ? GridStore.DEFAULT_ROOT_COLLECTION : fsName;
}

/**
 * Puts binary data to the grid
 *
 * Options
 *  - **_id** {Any}, unique id for this file
 *  - **root** {String}, root collection to use. Defaults to **{GridStore.DEFAULT_ROOT_COLLECTION}**.
 *  - **content_type** {String}, mime type of the file. Defaults to **{GridStore.DEFAULT_CONTENT_TYPE}**.
 *  - **chunk_size** {Number}, size for the chunk. Defaults to **{Chunk.DEFAULT_CHUNK_SIZE}**.
 *  - **metadata** {Object}, arbitrary data the user wants to store.
 *
 * @param {Buffer} data buffer with Binary Data.
 * @param {Object} [options] the options for the files.
 * @param {Function} callback this will be called after this method is executed. The first parameter will contain an Error object if an error occured or null otherwise. The second parameter will contain a reference to this object.
 * @return {null}
 * @api public
 */
Grid.prototype.put = function(data, options, callback***REMOVED*** {
  var self = this;
  var args = Array.prototype.slice.call(arguments, 1***REMOVED***;
  callback = args.pop(***REMOVED***;
  options = args.length ? args.shift(***REMOVED*** : {};
  // If root is not defined add our default one
  options['root'] = options['root'] == null ? this.fsName : options['root'];

  // Return if we don't have a buffer object as data
  if(!(Buffer.isBuffer(data***REMOVED******REMOVED******REMOVED*** return callback(new Error("Data object must be a buffer object"***REMOVED***, null***REMOVED***;
  // Get filename if we are using it
  var filename = options['filename'] || null;
  // Get id if we are using it
  var id = options['_id'] || null;
  // Create gridstore
  var gridStore = new GridStore(this.db, id, filename, "w", options***REMOVED***;
  gridStore.open(function(err, gridStore***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;

    gridStore.write(data, function(err, result***REMOVED*** {
      if(err***REMOVED*** return callback(err, null***REMOVED***;

      gridStore.close(function(err, result***REMOVED*** {
        if(err***REMOVED*** return callback(err, null***REMOVED***;
        callback(null, result***REMOVED***;
    ***REMOVED******REMOVED***
  ***REMOVED******REMOVED***
***REMOVED******REMOVED***
}

/**
 * Get binary data to the grid
 *
 * @param {Any} id for file.
 * @param {Function} callback this will be called after this method is executed. The first parameter will contain an Error object if an error occured or null otherwise. The second parameter will contain a reference to this object.
 * @return {null}
 * @api public
 */
Grid.prototype.get = function(id, callback***REMOVED*** {
  // Create gridstore
  var gridStore = new GridStore(this.db, id, null, "r", {root:this.fsName}***REMOVED***;
  gridStore.open(function(err, gridStore***REMOVED*** {
    if(err***REMOVED*** return callback(err, null***REMOVED***;

    // Return the data
    gridStore.read(function(err, data***REMOVED*** {
      return callback(err, data***REMOVED***
  ***REMOVED******REMOVED***;
***REMOVED******REMOVED***
}

/**
 * Delete file from grid
 *
 * @param {Any} id for file.
 * @param {Function} callback this will be called after this method is executed. The first parameter will contain an Error object if an error occured or null otherwise. The second parameter will contain a reference to this object.
 * @return {null}
 * @api public
 */
Grid.prototype.delete = function(id, callback***REMOVED*** {
  // Create gridstore
  GridStore.unlink(this.db, id, {root:this.fsName}, function(err, result***REMOVED*** {
    if(err***REMOVED*** return callback(err, false***REMOVED***;
    return callback(null, true***REMOVED***;
***REMOVED******REMOVED***;
}

exports.Grid = Grid;
