// express/lib/view.js
/**
 * Module dependencies.
 */

var ***REMOVED***
  , fs = require('fs'***REMOVED***
  , utils = require('./utils'***REMOVED***
  , dirname = path.dirname
  , basename = path.basename
  , extname = path.extname
  , exists = fs.existsSync || path.existsSync
  , join = path.join;

/**
 * Expose `View`.
 */

module.exports = View;

/**
 * Initialize a new `View` with the given `name`.
 *
 * Options:
 *
 *   - `defaultEngine` the default template engine name
 *   - `engines` template engine require(***REMOVED*** cache
 *   - `root` root path for view lookup
 *
 * @param {String} name
 * @param {Object} options
 * @api private
 */

function View(name, options***REMOVED*** {
  options = options || {};
  this.name = name;
  this.root = options.root;
  var engines = options.engines;
  this.defaultEngine = options.defaultEngine;
  var ext = this.ext = extname(name***REMOVED***;
  if (!ext && !this.defaultEngine***REMOVED*** throw new Error('No default engine was specified and no extension was provided.'***REMOVED***;
  if (!ext***REMOVED*** name += (ext = this.ext = ('.' != this.defaultEngine[0] ? '.' : ''***REMOVED*** + this.defaultEngine***REMOVED***;
  this.engine = engines[ext] || (engines[ext] = require(ext.slice(1***REMOVED******REMOVED***.__express***REMOVED***;
  this.path = this.lookup(name***REMOVED***;
}

/**
 * Lookup view by the given `path`
 *
 * @param {String} path
 * @return {String}
 * @api private
 */

View.prototype.lookup = function(path***REMOVED***{
  var ext = this.ext;

  // <path>.<engine>
  if (!utils.isAbsolute(path***REMOVED******REMOVED*** path = join(this.root, path***REMOVED***;
  if (exists(path***REMOVED******REMOVED*** return path;

  // <path>/index.<engine>
  path = join(dirname(path***REMOVED***, basename(path, ext***REMOVED***, 'index' + ext***REMOVED***;
  if (exists(path***REMOVED******REMOVED*** return path;
};

/**
 * Render with the given `options` and callback `fn(err, str***REMOVED***`.
 *
 * @param {Object} options
 * @param {Function} fn
 * @api private
 */

View.prototype.render = function(options, fn***REMOVED***{
  this.engine(this.path, options, fn***REMOVED***;
};
