// uglify-js/lib/parse.js
/***********************************************************************

  A JavaScript tokenizer / parser / beautifier / compressor.
  https://github.com/mishoo/UglifyJS2

  -------------------------------- (C***REMOVED*** ---------------------------------

                           Author: Mihai Bazon
                         <mihai.bazon@gmail.com>
                       http://mihai.bazon.net/blog

  Distributed under the BSD license:

    Copyright 2012 (c***REMOVED*** Mihai Bazon <mihai.bazon@gmail.com>
    Parser based on parse-js (http://marijn.haverbeke.nl/parse-js/***REMOVED***.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions
    are met:

        * Redistributions of source code must retain the above
          copyright notice, this list of conditions and the following
          disclaimer.

        * Redistributions in binary form must reproduce the above
          copyright notice, this list of conditions and the following
          disclaimer in the documentation and/or other materials
          provided with the distribution.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDER “AS IS” AND ANY
    EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
    PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER BE
    LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
    OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
    PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION***REMOVED*** HOWEVER CAUSED AND ON ANY
    THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
    TORT (INCLUDING NEGLIGENCE OR OTHERWISE***REMOVED*** ARISING IN ANY WAY OUT OF
    THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
    SUCH DAMAGE.

 ***********************************************************************/

"use strict";

var KEYWORDS = 'break case catch const continue debugger default delete do else finally for function if in instanceof new return switch throw try typeof var void while with';
var KEYWORDS_ATOM = 'false null true';
var RESERVED_WORDS = 'abstract boolean byte char class double enum export extends final float goto implements import int interface long native package private protected public short static super synchronized this throws transient volatile yield'
    + " " + KEYWORDS_ATOM + " " + KEYWORDS;
var KEYWORDS_BEFORE_EXPRESSION = 'return new delete throw else case';

KEYWORDS = makePredicate(KEYWORDS***REMOVED***;
RESERVED_WORDS = makePredicate(RESERVED_WORDS***REMOVED***;
KEYWORDS_BEFORE_EXPRESSION = makePredicate(KEYWORDS_BEFORE_EXPRESSION***REMOVED***;
KEYWORDS_ATOM = makePredicate(KEYWORDS_ATOM***REMOVED***;

var OPERATOR_CHARS = makePredicate(characters("+-*&%=<>!?|~^"***REMOVED******REMOVED***;

var RE_HEX_NUMBER = /^0x[0-9a-f]+$/i;
var RE_OCT_NUMBER = /^0[0-7]+$/;
var RE_DEC_NUMBER = /^\d*\.?\d*(?:e[+-]?\d*(?:\d\.?|\.?\d***REMOVED***\d****REMOVED***?$/i;

var OPERATORS = makePredicate([
    "in",
    "instanceof",
    "typeof",
    "new",
    "void",
    "delete",
    "++",
    "--",
    "+",
    "-",
    "!",
    "~",
    "&",
    "|",
    "^",
    "*",
    "/",
    "%",
    ">>",
    "<<",
    ">>>",
    "<",
    ">",
    "<=",
    ">=",
    "==",
    "===",
    "!=",
    "!==",
    "?",
    "=",
    "+=",
    "-=",
    "/=",
    "*=",
    "%=",
    ">>=",
    "<<=",
    ">>>=",
    "|=",
    "^=",
    "&=",
    "&&",
    "||"
]***REMOVED***;

var WHITESPACE_CHARS = makePredicate(characters(" \u00a0\n\r\t\f\u000b\u200b\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000"***REMOVED******REMOVED***;

var PUNC_BEFORE_EXPRESSION = makePredicate(characters("[{(,.;:"***REMOVED******REMOVED***;

var PUNC_CHARS = makePredicate(characters("[]{}(***REMOVED***,;:"***REMOVED******REMOVED***;

var REGEXP_MODIFIERS = makePredicate(characters("gmsiy"***REMOVED******REMOVED***;

/* -----[ Tokenizer ]----- */

// regexps adapted from http://xregexp.com/plugins/#unicode
var UNICODE = {
    letter: new RegExp("[\\u0041-\\u005A\\u0061-\\u007A\\u00AA\\u00B5\\u00BA\\u00C0-\\u00D6\\u00D8-\\u00F6\\u00F8-\\u02C1\\u02C6-\\u02D1\\u02E0-\\u02E4\\u02EC\\u02EE\\u0370-\\u0374\\u0376\\u0377\\u037A-\\u037D\\u0386\\u0388-\\u038A\\u038C\\u038E-\\u03A1\\u03A3-\\u03F5\\u03F7-\\u0481\\u048A-\\u0523\\u0531-\\u0556\\u0559\\u0561-\\u0587\\u05D0-\\u05EA\\u05F0-\\u05F2\\u0621-\\u064A\\u066E\\u066F\\u0671-\\u06D3\\u06D5\\u06E5\\u06E6\\u06EE\\u06EF\\u06FA-\\u06FC\\u06FF\\u0710\\u0712-\\u072F\\u074D-\\u07A5\\u07B1\\u07CA-\\u07EA\\u07F4\\u07F5\\u07FA\\u0904-\\u0939\\u093D\\u0950\\u0958-\\u0961\\u0971\\u0972\\u097B-\\u097F\\u0985-\\u098C\\u098F\\u0990\\u0993-\\u09A8\\u09AA-\\u09B0\\u09B2\\u09B6-\\u09B9\\u09BD\\u09CE\\u09DC\\u09DD\\u09DF-\\u09E1\\u09F0\\u09F1\\u0A05-\\u0A0A\\u0A0F\\u0A10\\u0A13-\\u0A28\\u0A2A-\\u0A30\\u0A32\\u0A33\\u0A35\\u0A36\\u0A38\\u0A39\\u0A59-\\u0A5C\\u0A5E\\u0A72-\\u0A74\\u0A85-\\u0A8D\\u0A8F-\\u0A91\\u0A93-\\u0AA8\\u0AAA-\\u0AB0\\u0AB2\\u0AB3\\u0AB5-\\u0AB9\\u0ABD\\u0AD0\\u0AE0\\u0AE1\\u0B05-\\u0B0C\\u0B0F\\u0B10\\u0B13-\\u0B28\\u0B2A-\\u0B30\\u0B32\\u0B33\\u0B35-\\u0B39\\u0B3D\\u0B5C\\u0B5D\\u0B5F-\\u0B61\\u0B71\\u0B83\\u0B85-\\u0B8A\\u0B8E-\\u0B90\\u0B92-\\u0B95\\u0B99\\u0B9A\\u0B9C\\u0B9E\\u0B9F\\u0BA3\\u0BA4\\u0BA8-\\u0BAA\\u0BAE-\\u0BB9\\u0BD0\\u0C05-\\u0C0C\\u0C0E-\\u0C10\\u0C12-\\u0C28\\u0C2A-\\u0C33\\u0C35-\\u0C39\\u0C3D\\u0C58\\u0C59\\u0C60\\u0C61\\u0C85-\\u0C8C\\u0C8E-\\u0C90\\u0C92-\\u0CA8\\u0CAA-\\u0CB3\\u0CB5-\\u0CB9\\u0CBD\\u0CDE\\u0CE0\\u0CE1\\u0D05-\\u0D0C\\u0D0E-\\u0D10\\u0D12-\\u0D28\\u0D2A-\\u0D39\\u0D3D\\u0D60\\u0D61\\u0D7A-\\u0D7F\\u0D85-\\u0D96\\u0D9A-\\u0DB1\\u0DB3-\\u0DBB\\u0DBD\\u0DC0-\\u0DC6\\u0E01-\\u0E30\\u0E32\\u0E33\\u0E40-\\u0E46\\u0E81\\u0E82\\u0E84\\u0E87\\u0E88\\u0E8A\\u0E8D\\u0E94-\\u0E97\\u0E99-\\u0E9F\\u0EA1-\\u0EA3\\u0EA5\\u0EA7\\u0EAA\\u0EAB\\u0EAD-\\u0EB0\\u0EB2\\u0EB3\\u0EBD\\u0EC0-\\u0EC4\\u0EC6\\u0EDC\\u0EDD\\u0F00\\u0F40-\\u0F47\\u0F49-\\u0F6C\\u0F88-\\u0F8B\\u1000-\\u102A\\u103F\\u1050-\\u1055\\u105A-\\u105D\\u1061\\u1065\\u1066\\u106E-\\u1070\\u1075-\\u1081\\u108E\\u10A0-\\u10C5\\u10D0-\\u10FA\\u10FC\\u1100-\\u1159\\u115F-\\u11A2\\u11A8-\\u11F9\\u1200-\\u1248\\u124A-\\u124D\\u1250-\\u1256\\u1258\\u125A-\\u125D\\u1260-\\u1288\\u128A-\\u128D\\u1290-\\u12B0\\u12B2-\\u12B5\\u12B8-\\u12BE\\u12C0\\u12C2-\\u12C5\\u12C8-\\u12D6\\u12D8-\\u1310\\u1312-\\u1315\\u1318-\\u135A\\u1380-\\u138F\\u13A0-\\u13F4\\u1401-\\u166C\\u166F-\\u1676\\u1681-\\u169A\\u16A0-\\u16EA\\u1700-\\u170C\\u170E-\\u1711\\u1720-\\u1731\\u1740-\\u1751\\u1760-\\u176C\\u176E-\\u1770\\u1780-\\u17B3\\u17D7\\u17DC\\u1820-\\u1877\\u1880-\\u18A8\\u18AA\\u1900-\\u191C\\u1950-\\u196D\\u1970-\\u1974\\u1980-\\u19A9\\u19C1-\\u19C7\\u1A00-\\u1A16\\u1B05-\\u1B33\\u1B45-\\u1B4B\\u1B83-\\u1BA0\\u1BAE\\u1BAF\\u1C00-\\u1C23\\u1C4D-\\u1C4F\\u1C5A-\\u1C7D\\u1D00-\\u1DBF\\u1E00-\\u1F15\\u1F18-\\u1F1D\\u1F20-\\u1F45\\u1F48-\\u1F4D\\u1F50-\\u1F57\\u1F59\\u1F5B\\u1F5D\\u1F5F-\\u1F7D\\u1F80-\\u1FB4\\u1FB6-\\u1FBC\\u1FBE\\u1FC2-\\u1FC4\\u1FC6-\\u1FCC\\u1FD0-\\u1FD3\\u1FD6-\\u1FDB\\u1FE0-\\u1FEC\\u1FF2-\\u1FF4\\u1FF6-\\u1FFC\\u2071\\u207F\\u2090-\\u2094\\u2102\\u2107\\u210A-\\u2113\\u2115\\u2119-\\u211D\\u2124\\u2126\\u2128\\u212A-\\u212D\\u212F-\\u2139\\u213C-\\u213F\\u2145-\\u2149\\u214E\\u2183\\u2184\\u2C00-\\u2C2E\\u2C30-\\u2C5E\\u2C60-\\u2C6F\\u2C71-\\u2C7D\\u2C80-\\u2CE4\\u2D00-\\u2D25\\u2D30-\\u2D65\\u2D6F\\u2D80-\\u2D96\\u2DA0-\\u2DA6\\u2DA8-\\u2DAE\\u2DB0-\\u2DB6\\u2DB8-\\u2DBE\\u2DC0-\\u2DC6\\u2DC8-\\u2DCE\\u2DD0-\\u2DD6\\u2DD8-\\u2DDE\\u2E2F\\u3005\\u3006\\u3031-\\u3035\\u303B\\u303C\\u3041-\\u3096\\u309D-\\u309F\\u30A1-\\u30FA\\u30FC-\\u30FF\\u3105-\\u312D\\u3131-\\u318E\\u31A0-\\u31B7\\u31F0-\\u31FF\\u3400\\u4DB5\\u4E00\\u9FC3\\uA000-\\uA48C\\uA500-\\uA60C\\uA610-\\uA61F\\uA62A\\uA62B\\uA640-\\uA65F\\uA662-\\uA66E\\uA67F-\\uA697\\uA717-\\uA71F\\uA722-\\uA788\\uA78B\\uA78C\\uA7FB-\\uA801\\uA803-\\uA805\\uA807-\\uA80A\\uA80C-\\uA822\\uA840-\\uA873\\uA882-\\uA8B3\\uA90A-\\uA925\\uA930-\\uA946\\uAA00-\\uAA28\\uAA40-\\uAA42\\uAA44-\\uAA4B\\uAC00\\uD7A3\\uF900-\\uFA2D\\uFA30-\\uFA6A\\uFA70-\\uFAD9\\uFB00-\\uFB06\\uFB13-\\uFB17\\uFB1D\\uFB1F-\\uFB28\\uFB2A-\\uFB36\\uFB38-\\uFB3C\\uFB3E\\uFB40\\uFB41\\uFB43\\uFB44\\uFB46-\\uFBB1\\uFBD3-\\uFD3D\\uFD50-\\uFD8F\\uFD92-\\uFDC7\\uFDF0-\\uFDFB\\uFE70-\\uFE74\\uFE76-\\uFEFC\\uFF21-\\uFF3A\\uFF41-\\uFF5A\\uFF66-\\uFFBE\\uFFC2-\\uFFC7\\uFFCA-\\uFFCF\\uFFD2-\\uFFD7\\uFFDA-\\uFFDC]"***REMOVED***,
    non_spacing_mark: new RegExp("[\\u0300-\\u036F\\u0483-\\u0487\\u0591-\\u05BD\\u05BF\\u05C1\\u05C2\\u05C4\\u05C5\\u05C7\\u0610-\\u061A\\u064B-\\u065E\\u0670\\u06D6-\\u06DC\\u06DF-\\u06E4\\u06E7\\u06E8\\u06EA-\\u06ED\\u0711\\u0730-\\u074A\\u07A6-\\u07B0\\u07EB-\\u07F3\\u0816-\\u0819\\u081B-\\u0823\\u0825-\\u0827\\u0829-\\u082D\\u0900-\\u0902\\u093C\\u0941-\\u0948\\u094D\\u0951-\\u0955\\u0962\\u0963\\u0981\\u09BC\\u09C1-\\u09C4\\u09CD\\u09E2\\u09E3\\u0A01\\u0A02\\u0A3C\\u0A41\\u0A42\\u0A47\\u0A48\\u0A4B-\\u0A4D\\u0A51\\u0A70\\u0A71\\u0A75\\u0A81\\u0A82\\u0ABC\\u0AC1-\\u0AC5\\u0AC7\\u0AC8\\u0ACD\\u0AE2\\u0AE3\\u0B01\\u0B3C\\u0B3F\\u0B41-\\u0B44\\u0B4D\\u0B56\\u0B62\\u0B63\\u0B82\\u0BC0\\u0BCD\\u0C3E-\\u0C40\\u0C46-\\u0C48\\u0C4A-\\u0C4D\\u0C55\\u0C56\\u0C62\\u0C63\\u0CBC\\u0CBF\\u0CC6\\u0CCC\\u0CCD\\u0CE2\\u0CE3\\u0D41-\\u0D44\\u0D4D\\u0D62\\u0D63\\u0DCA\\u0DD2-\\u0DD4\\u0DD6\\u0E31\\u0E34-\\u0E3A\\u0E47-\\u0E4E\\u0EB1\\u0EB4-\\u0EB9\\u0EBB\\u0EBC\\u0EC8-\\u0ECD\\u0F18\\u0F19\\u0F35\\u0F37\\u0F39\\u0F71-\\u0F7E\\u0F80-\\u0F84\\u0F86\\u0F87\\u0F90-\\u0F97\\u0F99-\\u0FBC\\u0FC6\\u102D-\\u1030\\u1032-\\u1037\\u1039\\u103A\\u103D\\u103E\\u1058\\u1059\\u105E-\\u1060\\u1071-\\u1074\\u1082\\u1085\\u1086\\u108D\\u109D\\u135F\\u1712-\\u1714\\u1732-\\u1734\\u1752\\u1753\\u1772\\u1773\\u17B7-\\u17BD\\u17C6\\u17C9-\\u17D3\\u17DD\\u180B-\\u180D\\u18A9\\u1920-\\u1922\\u1927\\u1928\\u1932\\u1939-\\u193B\\u1A17\\u1A18\\u1A56\\u1A58-\\u1A5E\\u1A60\\u1A62\\u1A65-\\u1A6C\\u1A73-\\u1A7C\\u1A7F\\u1B00-\\u1B03\\u1B34\\u1B36-\\u1B3A\\u1B3C\\u1B42\\u1B6B-\\u1B73\\u1B80\\u1B81\\u1BA2-\\u1BA5\\u1BA8\\u1BA9\\u1C2C-\\u1C33\\u1C36\\u1C37\\u1CD0-\\u1CD2\\u1CD4-\\u1CE0\\u1CE2-\\u1CE8\\u1CED\\u1DC0-\\u1DE6\\u1DFD-\\u1DFF\\u20D0-\\u20DC\\u20E1\\u20E5-\\u20F0\\u2CEF-\\u2CF1\\u2DE0-\\u2DFF\\u302A-\\u302F\\u3099\\u309A\\uA66F\\uA67C\\uA67D\\uA6F0\\uA6F1\\uA802\\uA806\\uA80B\\uA825\\uA826\\uA8C4\\uA8E0-\\uA8F1\\uA926-\\uA92D\\uA947-\\uA951\\uA980-\\uA982\\uA9B3\\uA9B6-\\uA9B9\\uA9BC\\uAA29-\\uAA2E\\uAA31\\uAA32\\uAA35\\uAA36\\uAA43\\uAA4C\\uAAB0\\uAAB2-\\uAAB4\\uAAB7\\uAAB8\\uAABE\\uAABF\\uAAC1\\uABE5\\uABE8\\uABED\\uFB1E\\uFE00-\\uFE0F\\uFE20-\\uFE26]"***REMOVED***,
    space_combining_mark: new RegExp("[\\u0903\\u093E-\\u0940\\u0949-\\u094C\\u094E\\u0982\\u0983\\u09BE-\\u09C0\\u09C7\\u09C8\\u09CB\\u09CC\\u09D7\\u0A03\\u0A3E-\\u0A40\\u0A83\\u0ABE-\\u0AC0\\u0AC9\\u0ACB\\u0ACC\\u0B02\\u0B03\\u0B3E\\u0B40\\u0B47\\u0B48\\u0B4B\\u0B4C\\u0B57\\u0BBE\\u0BBF\\u0BC1\\u0BC2\\u0BC6-\\u0BC8\\u0BCA-\\u0BCC\\u0BD7\\u0C01-\\u0C03\\u0C41-\\u0C44\\u0C82\\u0C83\\u0CBE\\u0CC0-\\u0CC4\\u0CC7\\u0CC8\\u0CCA\\u0CCB\\u0CD5\\u0CD6\\u0D02\\u0D03\\u0D3E-\\u0D40\\u0D46-\\u0D48\\u0D4A-\\u0D4C\\u0D57\\u0D82\\u0D83\\u0DCF-\\u0DD1\\u0DD8-\\u0DDF\\u0DF2\\u0DF3\\u0F3E\\u0F3F\\u0F7F\\u102B\\u102C\\u1031\\u1038\\u103B\\u103C\\u1056\\u1057\\u1062-\\u1064\\u1067-\\u106D\\u1083\\u1084\\u1087-\\u108C\\u108F\\u109A-\\u109C\\u17B6\\u17BE-\\u17C5\\u17C7\\u17C8\\u1923-\\u1926\\u1929-\\u192B\\u1930\\u1931\\u1933-\\u1938\\u19B0-\\u19C0\\u19C8\\u19C9\\u1A19-\\u1A1B\\u1A55\\u1A57\\u1A61\\u1A63\\u1A64\\u1A6D-\\u1A72\\u1B04\\u1B35\\u1B3B\\u1B3D-\\u1B41\\u1B43\\u1B44\\u1B82\\u1BA1\\u1BA6\\u1BA7\\u1BAA\\u1C24-\\u1C2B\\u1C34\\u1C35\\u1CE1\\u1CF2\\uA823\\uA824\\uA827\\uA880\\uA881\\uA8B4-\\uA8C3\\uA952\\uA953\\uA983\\uA9B4\\uA9B5\\uA9BA\\uA9BB\\uA9BD-\\uA9C0\\uAA2F\\uAA30\\uAA33\\uAA34\\uAA4D\\uAA7B\\uABE3\\uABE4\\uABE6\\uABE7\\uABE9\\uABEA\\uABEC]"***REMOVED***,
    connector_punctuation: new RegExp("[\\u005F\\u203F\\u2040\\u2054\\uFE33\\uFE34\\uFE4D-\\uFE4F\\uFF3F]"***REMOVED***
};

function is_letter(code***REMOVED*** {
    return (code >= 97 && code <= 122***REMOVED***
        || (code >= 65 && code <= 90***REMOVED***
        || (code >= 0xaa && UNICODE.letter.test(String.fromCharCode(code***REMOVED******REMOVED******REMOVED***;
};

function is_digit(code***REMOVED*** {
    return code >= 48 && code <= 57; //XXX: find out if "UnicodeDigit" means something else than 0..9
};

function is_alphanumeric_char(code***REMOVED*** {
    return is_digit(code***REMOVED*** || is_letter(code***REMOVED***;
};

function is_unicode_combining_mark(ch***REMOVED*** {
    return UNICODE.non_spacing_mark.test(ch***REMOVED*** || UNICODE.space_combining_mark.test(ch***REMOVED***;
};

function is_unicode_connector_punctuation(ch***REMOVED*** {
    return UNICODE.connector_punctuation.test(ch***REMOVED***;
};

function is_identifier(name***REMOVED*** {
    return !RESERVED_WORDS(name***REMOVED*** && /^[a-z_$][a-z0-9_$]*$/i.test(name***REMOVED***;
};

function is_identifier_start(code***REMOVED*** {
    return code == 36 || code == 95 || is_letter(code***REMOVED***;
};

function is_identifier_char(ch***REMOVED*** {
    var code = ch.charCodeAt(0***REMOVED***;
    return is_identifier_start(code***REMOVED***
        || is_digit(code***REMOVED***
        || code == 8204 // \u200c: zero-width non-joiner <ZWNJ>
        || code == 8205 // \u200d: zero-width joiner <ZWJ> (in my ECMA-262 PDF, this is also 200c***REMOVED***
        || is_unicode_combining_mark(ch***REMOVED***
        || is_unicode_connector_punctuation(ch***REMOVED***
    ;
};

function is_identifier_string(str***REMOVED***{
    var i = str.length;
    if (i == 0***REMOVED*** return false;
    if (!is_identifier_start(str.charCodeAt(0***REMOVED******REMOVED******REMOVED*** return false;
    while (--i >= 0***REMOVED*** {
        if (!is_identifier_char(str.charAt(i***REMOVED******REMOVED******REMOVED***
            return false;
  ***REMOVED***
    return true;
};

function parse_js_number(num***REMOVED*** {
    if (RE_HEX_NUMBER.test(num***REMOVED******REMOVED*** {
        return parseInt(num.substr(2***REMOVED***, 16***REMOVED***;
  ***REMOVED*** else if (RE_OCT_NUMBER.test(num***REMOVED******REMOVED*** {
        return parseInt(num.substr(1***REMOVED***, 8***REMOVED***;
  ***REMOVED*** else if (RE_DEC_NUMBER.test(num***REMOVED******REMOVED*** {
        return parseFloat(num***REMOVED***;
  ***REMOVED***
};

function JS_Parse_Error(message, line, col, pos***REMOVED*** {
    this.message = message;
    this.line = line;
    this.col = col;
    this.pos = pos;
    this.stack = new Error(***REMOVED***.stack;
};

JS_Parse_Error.prototype.toString = function(***REMOVED*** {
    return this.message + " (line: " + this.line + ", col: " + this.col + ", pos: " + this.pos + "***REMOVED***" + "\n\n" + this.stack;
};

function js_error(message, filename, line, col, pos***REMOVED*** {
    throw new JS_Parse_Error(message, line, col, pos***REMOVED***;
};

function is_token(token, type, val***REMOVED*** {
    return token.type == type && (val == null || token.value == val***REMOVED***;
};

var EX_EOF = {};

function tokenizer($TEXT, filename, html5_comments***REMOVED*** {

    var S = {
        text            : $TEXT.replace(/\r\n?|[\n\u2028\u2029]/g, "\n"***REMOVED***.replace(/\uFEFF/g, ''***REMOVED***,
        filename        : filename,
        pos             : 0,
        tokpos          : 0,
        line            : 1,
        tokline         : 0,
        col             : 0,
        tokcol          : 0,
        newline_before  : false,
        regex_allowed   : false,
        comments_before : []
  ***REMOVED***;

    function peek(***REMOVED*** { return S.text.charAt(S.pos***REMOVED***; };

    function next(signal_eof, in_string***REMOVED*** {
        var ch = S.text.charAt(S.pos++***REMOVED***;
        if (signal_eof && !ch***REMOVED***
            throw EX_EOF;
        if (ch == "\n"***REMOVED*** {
            S.newline_before = S.newline_before || !in_string;
            ++S.line;
            S.col = 0;
      ***REMOVED*** else {
            ++S.col;
      ***REMOVED***
        return ch;
  ***REMOVED***;

    function forward(i***REMOVED*** {
        while (i-- > 0***REMOVED*** next(***REMOVED***;
  ***REMOVED***;

    function looking_at(str***REMOVED*** {
        return S.text.substr(S.pos, str.length***REMOVED*** == str;
  ***REMOVED***;

    function find(what, signal_eof***REMOVED*** {
        var pos = S.text.indexOf(what, S.pos***REMOVED***;
        if (signal_eof && pos == -1***REMOVED*** throw EX_EOF;
        return pos;
  ***REMOVED***;

    function start_token(***REMOVED*** {
        S.tokline = S.line;
        S.tokcol = S.col;
        S.tokpos = S.pos;
  ***REMOVED***;

    var prev_was_dot = false;
    function token(type, value, is_comment***REMOVED*** {
        S.regex_allowed = ((type == "operator" && !UNARY_POSTFIX(value***REMOVED******REMOVED*** ||
                           (type == "keyword" && KEYWORDS_BEFORE_EXPRESSION(value***REMOVED******REMOVED*** ||
                           (type == "punc" && PUNC_BEFORE_EXPRESSION(value***REMOVED******REMOVED******REMOVED***;
        prev_was_dot = (type == "punc" && value == "."***REMOVED***;
        var ret = {
            type   : type,
            value  : value,
            line   : S.tokline,
            col    : S.tokcol,
            pos    : S.tokpos,
            endpos : S.pos,
            nlb    : S.newline_before,
            file   : filename
      ***REMOVED***;
        if (!is_comment***REMOVED*** {
            ret.comments_before = S.comments_before;
            S.comments_before = [];
            // make note of any newlines in the comments that came before
            for (var i = 0, len = ret.comments_before.length; i < len; i++***REMOVED*** {
                ret.nlb = ret.nlb || ret.comments_before[i].nlb;
          ***REMOVED***
      ***REMOVED***
        S.newline_before = false;
        return new AST_Token(ret***REMOVED***;
  ***REMOVED***;

    function skip_whitespace(***REMOVED*** {
        while (WHITESPACE_CHARS(peek(***REMOVED******REMOVED******REMOVED***
            next(***REMOVED***;
  ***REMOVED***;

    function read_while(pred***REMOVED*** {
        var ret = "", ch, i = 0;
        while ((ch = peek(***REMOVED******REMOVED*** && pred(ch, i++***REMOVED******REMOVED***
            ret += next(***REMOVED***;
        return ret;
  ***REMOVED***;

    function parse_error(err***REMOVED*** {
        js_error(err, filename, S.tokline, S.tokcol, S.tokpos***REMOVED***;
  ***REMOVED***;

    function read_num(prefix***REMOVED*** {
        var has_e = false, after_e = false, has_x = false, has_dot = prefix == ".";
        var num = read_while(function(ch, i***REMOVED***{
            var code = ch.charCodeAt(0***REMOVED***;
            switch (code***REMOVED*** {
              case 120: case 88: // xX
                return has_x ? false : (has_x = true***REMOVED***;
              case 101: case 69: // eE
                return has_x ? true : has_e ? false : (has_e = after_e = true***REMOVED***;
              case 45: // -
                return after_e || (i == 0 && !prefix***REMOVED***;
              case 43: // +
                return after_e;
              case (after_e = false, 46***REMOVED***: // .
                return (!has_dot && !has_x && !has_e***REMOVED*** ? (has_dot = true***REMOVED*** : false;
          ***REMOVED***
            return is_alphanumeric_char(code***REMOVED***;
      ***REMOVED******REMOVED***;
        if (prefix***REMOVED*** num = prefix + num;
        var valid = parse_js_number(num***REMOVED***;
        if (!isNaN(valid***REMOVED******REMOVED*** {
            return token("num", valid***REMOVED***;
      ***REMOVED*** else {
            parse_error("Invalid syntax: " + num***REMOVED***;
      ***REMOVED***
  ***REMOVED***;

    function read_escaped_char(in_string***REMOVED*** {
        var ch = next(true, in_string***REMOVED***;
        switch (ch.charCodeAt(0***REMOVED******REMOVED*** {
          case 110 : return "\n";
          case 114 : return "\r";
          case 116 : return "\t";
          case 98  : return "\b";
          case 118 : return "\u000b"; // \v
          case 102 : return "\f";
          case 48  : return "\0";
          case 120 : return String.fromCharCode(hex_bytes(2***REMOVED******REMOVED***; // \x
          case 117 : return String.fromCharCode(hex_bytes(4***REMOVED******REMOVED***; // \u
          case 10  : return ""; // newline
          default  : return ch;
      ***REMOVED***
  ***REMOVED***;

    function hex_bytes(n***REMOVED*** {
        var num = 0;
        for (; n > 0; --n***REMOVED*** {
            var digit = parseInt(next(true***REMOVED***, 16***REMOVED***;
            if (isNaN(digit***REMOVED******REMOVED***
                parse_error("Invalid hex-character pattern in string"***REMOVED***;
            num = (num << 4***REMOVED*** | digit;
      ***REMOVED***
        return num;
  ***REMOVED***;

    var read_string = with_eof_error("Unterminated string constant", function(***REMOVED***{
        var quote = next(***REMOVED***, ret = "";
        for (;;***REMOVED*** {
            var ch = next(true***REMOVED***;
            if (ch == "\\"***REMOVED*** {
                // read OctalEscapeSequence (XXX: deprecated if "strict mode"***REMOVED***
                // https://github.com/mishoo/UglifyJS/issues/178
                var octal_len = 0, first = null;
                ch = read_while(function(ch***REMOVED***{
                    if (ch >= "0" && ch <= "7"***REMOVED*** {
                        if (!first***REMOVED*** {
                            first = ch;
                            return ++octal_len;
                      ***REMOVED***
                        else if (first <= "3" && octal_len <= 2***REMOVED*** return ++octal_len;
                        else if (first >= "4" && octal_len <= 1***REMOVED*** return ++octal_len;
                  ***REMOVED***
                    return false;
              ***REMOVED******REMOVED***;
                if (octal_len > 0***REMOVED*** ch = String.fromCharCode(parseInt(ch, 8***REMOVED******REMOVED***;
                else ch = read_escaped_char(true***REMOVED***;
          ***REMOVED***
            else if (ch == quote***REMOVED*** break;
            ret += ch;
      ***REMOVED***
        return token("string", ret***REMOVED***;
  ***REMOVED******REMOVED***;

    function skip_line_comment(type***REMOVED*** {
        var regex_allowed = S.regex_allowed;
        var i = find("\n"***REMOVED***, ret;
        if (i == -1***REMOVED*** {
            ret = S.text.substr(S.pos***REMOVED***;
            S.pos = S.text.length;
      ***REMOVED*** else {
            ret = S.text.substring(S.pos, i***REMOVED***;
            S.pos = i;
      ***REMOVED***
        S.comments_before.push(token(type, ret, true***REMOVED******REMOVED***;
        S.regex_allowed = regex_allowed;
        return next_token(***REMOVED***;
  ***REMOVED***;

    var skip_multiline_comment = with_eof_error("Unterminated multiline comment", function(***REMOVED***{
        var regex_allowed = S.regex_allowed;
        var i = find("*/", true***REMOVED***;
        var text = S.text.substring(S.pos, i***REMOVED***;
        var a = text.split("\n"***REMOVED***, n = a.length;
        // update stream position
        S.pos = i + 2;
        S.line += n - 1;
        if (n > 1***REMOVED*** S.col = a[n - 1].length;
        else S.col += a[n - 1].length;
        S.col += 2;
        var nlb = S.newline_before = S.newline_before || text.indexOf("\n"***REMOVED*** >= 0;
        S.comments_before.push(token("comment2", text, true***REMOVED******REMOVED***;
        S.regex_allowed = regex_allowed;
        S.newline_before = nlb;
        return next_token(***REMOVED***;
  ***REMOVED******REMOVED***;

    function read_name(***REMOVED*** {
        var backslash = false, name = "", ch, escaped = false, hex;
        while ((ch = peek(***REMOVED******REMOVED*** != null***REMOVED*** {
            if (!backslash***REMOVED*** {
                if (ch == "\\"***REMOVED*** escaped = backslash = true, next(***REMOVED***;
                else if (is_identifier_char(ch***REMOVED******REMOVED*** name += next(***REMOVED***;
                else break;
          ***REMOVED***
            else {
                if (ch != "u"***REMOVED*** parse_error("Expecting UnicodeEscapeSequence -- uXXXX"***REMOVED***;
                ch = read_escaped_char(***REMOVED***;
                if (!is_identifier_char(ch***REMOVED******REMOVED*** parse_error("Unicode char: " + ch.charCodeAt(0***REMOVED*** + " is not valid in identifier"***REMOVED***;
                name += ch;
                backslash = false;
          ***REMOVED***
      ***REMOVED***
        if (KEYWORDS(name***REMOVED*** && escaped***REMOVED*** {
            hex = name.charCodeAt(0***REMOVED***.toString(16***REMOVED***.toUpperCase(***REMOVED***;
            name = "\\u" + "0000".substr(hex.length***REMOVED*** + hex + name.slice(1***REMOVED***;
      ***REMOVED***
        return name;
  ***REMOVED***;

    var read_regexp = with_eof_error("Unterminated regular expression", function(regexp***REMOVED***{
        var prev_backslash = false, ch, in_class = false;
        while ((ch = next(true***REMOVED******REMOVED******REMOVED*** if (prev_backslash***REMOVED*** {
            regexp += "\\" + ch;
            prev_backslash = false;
      ***REMOVED*** else if (ch == "["***REMOVED*** {
            in_class = true;
            regexp += ch;
      ***REMOVED*** else if (ch == "]" && in_class***REMOVED*** {
            in_class = false;
            regexp += ch;
      ***REMOVED*** else if (ch == "/" && !in_class***REMOVED*** {
            break;
      ***REMOVED*** else if (ch == "\\"***REMOVED*** {
            prev_backslash = true;
      ***REMOVED*** else {
            regexp += ch;
      ***REMOVED***
        var mods = read_name(***REMOVED***;
        return token("regexp", new RegExp(regexp, mods***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;

    function read_operator(prefix***REMOVED*** {
        function grow(op***REMOVED*** {
            if (!peek(***REMOVED******REMOVED*** return op;
            var bigger = op + peek(***REMOVED***;
            if (OPERATORS(bigger***REMOVED******REMOVED*** {
                next(***REMOVED***;
                return grow(bigger***REMOVED***;
          ***REMOVED*** else {
                return op;
          ***REMOVED***
      ***REMOVED***;
        return token("operator", grow(prefix || next(***REMOVED******REMOVED******REMOVED***;
  ***REMOVED***;

    function handle_slash(***REMOVED*** {
        next(***REMOVED***;
        switch (peek(***REMOVED******REMOVED*** {
          case "/":
            next(***REMOVED***;
            return skip_line_comment("comment1"***REMOVED***;
          case "*":
            next(***REMOVED***;
            return skip_multiline_comment(***REMOVED***;
      ***REMOVED***
        return S.regex_allowed ? read_regexp(""***REMOVED*** : read_operator("/"***REMOVED***;
  ***REMOVED***;

    function handle_dot(***REMOVED*** {
        next(***REMOVED***;
        return is_digit(peek(***REMOVED***.charCodeAt(0***REMOVED******REMOVED***
            ? read_num("."***REMOVED***
            : token("punc", "."***REMOVED***;
  ***REMOVED***;

    function read_word(***REMOVED*** {
        var word = read_name(***REMOVED***;
        if (prev_was_dot***REMOVED*** return token("name", word***REMOVED***;
        return KEYWORDS_ATOM(word***REMOVED*** ? token("atom", word***REMOVED***
            : !KEYWORDS(word***REMOVED*** ? token("name", word***REMOVED***
            : OPERATORS(word***REMOVED*** ? token("operator", word***REMOVED***
            : token("keyword", word***REMOVED***;
  ***REMOVED***;

    function with_eof_error(eof_error, cont***REMOVED*** {
        return function(x***REMOVED*** {
            try {
                return cont(x***REMOVED***;
          ***REMOVED*** catch(ex***REMOVED*** {
                if (ex === EX_EOF***REMOVED*** parse_error(eof_error***REMOVED***;
                else throw ex;
          ***REMOVED***
      ***REMOVED***;
  ***REMOVED***;

    function next_token(force_regexp***REMOVED*** {
        if (force_regexp != null***REMOVED***
            return read_regexp(force_regexp***REMOVED***;
        skip_whitespace(***REMOVED***;
        start_token(***REMOVED***;
        if (html5_comments***REMOVED*** {
            if (looking_at("<!--"***REMOVED******REMOVED*** {
                forward(4***REMOVED***;
                return skip_line_comment("comment3"***REMOVED***;
          ***REMOVED***
            if (looking_at("-->"***REMOVED*** && S.newline_before***REMOVED*** {
                forward(3***REMOVED***;
                return skip_line_comment("comment4"***REMOVED***;
          ***REMOVED***
      ***REMOVED***
        var ch = peek(***REMOVED***;
        if (!ch***REMOVED*** return token("eof"***REMOVED***;
        var code = ch.charCodeAt(0***REMOVED***;
        switch (code***REMOVED*** {
          case 34: case 39: return read_string(***REMOVED***;
          case 46: return handle_dot(***REMOVED***;
          case 47: return handle_slash(***REMOVED***;
      ***REMOVED***
        if (is_digit(code***REMOVED******REMOVED*** return read_num(***REMOVED***;
        if (PUNC_CHARS(ch***REMOVED******REMOVED*** return token("punc", next(***REMOVED******REMOVED***;
        if (OPERATOR_CHARS(ch***REMOVED******REMOVED*** return read_operator(***REMOVED***;
        if (code == 92 || is_identifier_start(code***REMOVED******REMOVED*** return read_word(***REMOVED***;
        parse_error("Unexpected character '" + ch + "'"***REMOVED***;
  ***REMOVED***;

    next_token.context = function(nc***REMOVED*** {
        if (nc***REMOVED*** S = nc;
        return S;
  ***REMOVED***;

    return next_token;

};

/* -----[ Parser (constants***REMOVED*** ]----- */

var UNARY_PREFIX = makePredicate([
    "typeof",
    "void",
    "delete",
    "--",
    "++",
    "!",
    "~",
    "-",
    "+"
]***REMOVED***;

var UNARY_POSTFIX = makePredicate([ "--", "++" ]***REMOVED***;

var ASSIGNMENT = makePredicate([ "=", "+=", "-=", "/=", "*=", "%=", ">>=", "<<=", ">>>=", "|=", "^=", "&=" ]***REMOVED***;

var PRECEDENCE = (function(a, ret***REMOVED***{
    for (var i = 0; i < a.length; ++i***REMOVED*** {
        var b = a[i];
        for (var j = 0; j < b.length; ++j***REMOVED*** {
            ret[b[j]] = i + 1;
      ***REMOVED***
  ***REMOVED***
    return ret;
}***REMOVED***(
    [
        ["||"],
        ["&&"],
        ["|"],
        ["^"],
        ["&"],
        ["==", "===", "!=", "!=="],
        ["<", ">", "<=", ">=", "in", "instanceof"],
        [">>", "<<", ">>>"],
        ["+", "-"],
        ["*", "/", "%"]
    ],
    {}
***REMOVED***;

var STATEMENTS_WITH_LABELS = array_to_hash([ "for", "do", "while", "switch" ]***REMOVED***;

var ATOMIC_START_TOKEN = array_to_hash([ "atom", "num", "string", "regexp", "name" ]***REMOVED***;

/* -----[ Parser ]----- */

function parse($TEXT, options***REMOVED*** {

    options = defaults(options, {
        strict         : false,
        filename       : null,
        toplevel       : null,
        expression     : false,
        html5_comments : true,
  ***REMOVED******REMOVED***;

    var S = {
        input         : (typeof $TEXT == "string"
                         ? tokenizer($TEXT, options.filename,
                                     options.html5_comments***REMOVED***
                         : $TEXT***REMOVED***,
        token         : null,
        prev          : null,
        peeked        : null,
        in_function   : 0,
        in_directives : true,
        in_loop       : 0,
        labels        : []
  ***REMOVED***;

    S.token = next(***REMOVED***;

    function is(type, value***REMOVED*** {
        return is_token(S.token, type, value***REMOVED***;
  ***REMOVED***;

    function peek(***REMOVED*** { return S.peeked || (S.peeked = S.input(***REMOVED******REMOVED***; };

    function next(***REMOVED*** {
        S.prev = S.token;
        if (S.peeked***REMOVED*** {
            S.token = S.peeked;
            S.peeked = null;
      ***REMOVED*** else {
            S.token = S.input(***REMOVED***;
      ***REMOVED***
        S.in_directives = S.in_directives && (
            S.token.type == "string" || is("punc", ";"***REMOVED***
        ***REMOVED***;
        return S.token;
  ***REMOVED***;

    function prev(***REMOVED*** {
        return S.prev;
  ***REMOVED***;

    function croak(msg, line, col, pos***REMOVED*** {
        var ctx = S.input.context(***REMOVED***;
        js_error(msg,
                 ctx.filename,
                 line != null ? line : ctx.tokline,
                 col != null ? col : ctx.tokcol,
                 pos != null ? pos : ctx.tokpos***REMOVED***;
  ***REMOVED***;

    function token_error(token, msg***REMOVED*** {
        croak(msg, token.line, token.col***REMOVED***;
  ***REMOVED***;

    function unexpected(token***REMOVED*** {
        if (token == null***REMOVED***
            token = S.token;
        token_error(token, "Unexpected token: " + token.type + " (" + token.value + "***REMOVED***"***REMOVED***;
  ***REMOVED***;

    function expect_token(type, val***REMOVED*** {
        if (is(type, val***REMOVED******REMOVED*** {
            return next(***REMOVED***;
      ***REMOVED***
        token_error(S.token, "Unexpected token " + S.token.type + " «" + S.token.value + "»" + ", expected " + type + " «" + val + "»"***REMOVED***;
  ***REMOVED***;

    function expect(punc***REMOVED*** { return expect_token("punc", punc***REMOVED***; };

    function can_insert_semicolon(***REMOVED*** {
        return !options.strict && (
            S.token.nlb || is("eof"***REMOVED*** || is("punc", "}"***REMOVED***
        ***REMOVED***;
  ***REMOVED***;

    function semicolon(***REMOVED*** {
        if (is("punc", ";"***REMOVED******REMOVED*** next(***REMOVED***;
        else if (!can_insert_semicolon(***REMOVED******REMOVED*** unexpected(***REMOVED***;
  ***REMOVED***;

    function parenthesised(***REMOVED*** {
        expect("("***REMOVED***;
        var exp = expression(true***REMOVED***;
        expect("***REMOVED***"***REMOVED***;
        return exp;
  ***REMOVED***;

    function embed_tokens(parser***REMOVED*** {
        return function(***REMOVED*** {
            var start = S.token;
            var expr = parser(***REMOVED***;
            var end = prev(***REMOVED***;
            expr.start = start;
            expr.end = end;
            return expr;
      ***REMOVED***;
  ***REMOVED***;

    function handle_regexp(***REMOVED*** {
        if (is("operator", "/"***REMOVED*** || is("operator", "/="***REMOVED******REMOVED*** {
            S.peeked = null;
            S.token = S.input(S.token.value.substr(1***REMOVED******REMOVED***; // force regexp
      ***REMOVED***
  ***REMOVED***;

    var statement = embed_tokens(function(***REMOVED*** {
        var tmp;
        handle_regexp(***REMOVED***;
        switch (S.token.type***REMOVED*** {
          case "string":
            var dir = S.in_directives, stat = simple_statement(***REMOVED***;
            // XXXv2: decide how to fix directives
            if (dir && stat.body instanceof AST_String && !is("punc", ","***REMOVED******REMOVED***
                return new AST_Directive({ value: stat.body.value }***REMOVED***;
            return stat;
          case "num":
          case "regexp":
          case "operator":
          case "atom":
            return simple_statement(***REMOVED***;

          case "name":
            return is_token(peek(***REMOVED***, "punc", ":"***REMOVED***
                ? labeled_statement(***REMOVED***
                : simple_statement(***REMOVED***;

          case "punc":
            switch (S.token.value***REMOVED*** {
              case "{":
                return new AST_BlockStatement({
                    start : S.token,
                    body  : block_(***REMOVED***,
                    end   : prev(***REMOVED***
              ***REMOVED******REMOVED***;
              case "[":
              case "(":
                return simple_statement(***REMOVED***;
              case ";":
                next(***REMOVED***;
                return new AST_EmptyStatement(***REMOVED***;
              default:
                unexpected(***REMOVED***;
          ***REMOVED***

          case "keyword":
            switch (tmp = S.token.value, next(***REMOVED***, tmp***REMOVED*** {
              case "break":
                return break_cont(AST_Break***REMOVED***;

              case "continue":
                return break_cont(AST_Continue***REMOVED***;

              case "debugger":
                semicolon(***REMOVED***;
                return new AST_Debugger(***REMOVED***;

              case "do":
                return new AST_Do({
                    body      : in_loop(statement***REMOVED***,
                    condition : (expect_token("keyword", "while"***REMOVED***, tmp = parenthesised(***REMOVED***, semicolon(***REMOVED***, tmp***REMOVED***
              ***REMOVED******REMOVED***;

              case "while":
                return new AST_While({
                    condition : parenthesised(***REMOVED***,
                    body      : in_loop(statement***REMOVED***
              ***REMOVED******REMOVED***;

              case "for":
                return for_(***REMOVED***;

              case "function":
                return function_(AST_Defun***REMOVED***;

              case "if":
                return if_(***REMOVED***;

              case "return":
                if (S.in_function == 0***REMOVED***
                    croak("'return' outside of function"***REMOVED***;
                return new AST_Return({
                    value: ( is("punc", ";"***REMOVED***
                             ? (next(***REMOVED***, null***REMOVED***
                             : can_insert_semicolon(***REMOVED***
                             ? null
                             : (tmp = expression(true***REMOVED***, semicolon(***REMOVED***, tmp***REMOVED*** ***REMOVED***
              ***REMOVED******REMOVED***;

              case "switch":
                return new AST_Switch({
                    expression : parenthesised(***REMOVED***,
                    body       : in_loop(switch_body_***REMOVED***
              ***REMOVED******REMOVED***;

              case "throw":
                if (S.token.nlb***REMOVED***
                    croak("Illegal newline after 'throw'"***REMOVED***;
                return new AST_Throw({
                    value: (tmp = expression(true***REMOVED***, semicolon(***REMOVED***, tmp***REMOVED***
              ***REMOVED******REMOVED***;

              case "try":
                return try_(***REMOVED***;

              case "var":
                return tmp = var_(***REMOVED***, semicolon(***REMOVED***, tmp;

              case "const":
                return tmp = const_(***REMOVED***, semicolon(***REMOVED***, tmp;

              case "with":
                return new AST_With({
                    expression : parenthesised(***REMOVED***,
                    body       : statement(***REMOVED***
              ***REMOVED******REMOVED***;

              default:
                unexpected(***REMOVED***;
          ***REMOVED***
      ***REMOVED***
  ***REMOVED******REMOVED***;

    function labeled_statement(***REMOVED*** {
        var label = as_symbol(AST_Label***REMOVED***;
        if (find_if(function(l***REMOVED***{ return l.name == label.name }, S.labels***REMOVED******REMOVED*** {
            // ECMA-262, 12.12: An ECMAScript program is considered
            // syntactically incorrect if it contains a
            // LabelledStatement that is enclosed by a
            // LabelledStatement with the same Identifier as label.
            croak("Label " + label.name + " defined twice"***REMOVED***;
      ***REMOVED***
        expect(":"***REMOVED***;
        S.labels.push(label***REMOVED***;
        var stat = statement(***REMOVED***;
        S.labels.pop(***REMOVED***;
        if (!(stat instanceof AST_IterationStatement***REMOVED******REMOVED*** {
            // check for `continue` that refers to this label.
            // those should be reported as syntax errors.
            // https://github.com/mishoo/UglifyJS2/issues/287
            label.references.forEach(function(ref***REMOVED***{
                if (ref instanceof AST_Continue***REMOVED*** {
                    ref = ref.label.start;
                    croak("Continue label `" + label.name + "` refers to non-IterationStatement.",
                          ref.line, ref.col, ref.pos***REMOVED***;
              ***REMOVED***
          ***REMOVED******REMOVED***;
      ***REMOVED***
        return new AST_LabeledStatement({ body: stat, label: label }***REMOVED***;
  ***REMOVED***;

    function simple_statement(tmp***REMOVED*** {
        return new AST_SimpleStatement({ body: (tmp = expression(true***REMOVED***, semicolon(***REMOVED***, tmp***REMOVED*** }***REMOVED***;
  ***REMOVED***;

    function break_cont(type***REMOVED*** {
        var label = null, ldef;
        if (!can_insert_semicolon(***REMOVED******REMOVED*** {
            label = as_symbol(AST_LabelRef, true***REMOVED***;
      ***REMOVED***
        if (label != null***REMOVED*** {
            ldef = find_if(function(l***REMOVED***{ return l.name == label.name }, S.labels***REMOVED***;
            if (!ldef***REMOVED***
                croak("Undefined label " + label.name***REMOVED***;
            label.thedef = ldef;
      ***REMOVED***
        else if (S.in_loop == 0***REMOVED***
            croak(type.TYPE + " not inside a loop or switch"***REMOVED***;
        semicolon(***REMOVED***;
        var stat = new type({ label: label }***REMOVED***;
        if (ldef***REMOVED*** ldef.references.push(stat***REMOVED***;
        return stat;
  ***REMOVED***;

    function for_(***REMOVED*** {
        expect("("***REMOVED***;
        var init = null;
        if (!is("punc", ";"***REMOVED******REMOVED*** {
            init = is("keyword", "var"***REMOVED***
                ? (next(***REMOVED***, var_(true***REMOVED******REMOVED***
                : expression(true, true***REMOVED***;
            if (is("operator", "in"***REMOVED******REMOVED*** {
                if (init instanceof AST_Var && init.definitions.length > 1***REMOVED***
                    croak("Only one variable declaration allowed in for..in loop"***REMOVED***;
                next(***REMOVED***;
                return for_in(init***REMOVED***;
          ***REMOVED***
      ***REMOVED***
        return regular_for(init***REMOVED***;
  ***REMOVED***;

    function regular_for(init***REMOVED*** {
        expect(";"***REMOVED***;
        var test = is("punc", ";"***REMOVED*** ? null : expression(true***REMOVED***;
        expect(";"***REMOVED***;
        var step = is("punc", "***REMOVED***"***REMOVED*** ? null : expression(true***REMOVED***;
        expect("***REMOVED***"***REMOVED***;
        return new AST_For({
            init      : init,
            condition : test,
            step      : step,
            body      : in_loop(statement***REMOVED***
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    function for_in(init***REMOVED*** {
        var lhs = init instanceof AST_Var ? init.definitions[0].name : null;
        var obj = expression(true***REMOVED***;
        expect("***REMOVED***"***REMOVED***;
        return new AST_ForIn({
            init   : init,
            name   : lhs,
            object : obj,
            body   : in_loop(statement***REMOVED***
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    var function_ = function(ctor***REMOVED*** {
        var in_statement = ctor === AST_Defun;
        var name = is("name"***REMOVED*** ? as_symbol(in_statement ? AST_SymbolDefun : AST_SymbolLambda***REMOVED*** : null;
        if (in_statement && !name***REMOVED***
            unexpected(***REMOVED***;
        expect("("***REMOVED***;
        return new ctor({
            name: name,
            argnames: (function(first, a***REMOVED***{
                while (!is("punc", "***REMOVED***"***REMOVED******REMOVED*** {
                    if (first***REMOVED*** first = false; else expect(","***REMOVED***;
                    a.push(as_symbol(AST_SymbolFunarg***REMOVED******REMOVED***;
              ***REMOVED***
                next(***REMOVED***;
                return a;
          ***REMOVED******REMOVED***(true, []***REMOVED***,
            body: (function(loop, labels***REMOVED***{
                ++S.in_function;
                S.in_directives = true;
                S.in_loop = 0;
                S.labels = [];
                var a = block_(***REMOVED***;
                --S.in_function;
                S.in_loop = loop;
                S.labels = labels;
                return a;
          ***REMOVED******REMOVED***(S.in_loop, S.labels***REMOVED***
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    function if_(***REMOVED*** {
        var cond = parenthesised(***REMOVED***, body = statement(***REMOVED***, belse = null;
        if (is("keyword", "else"***REMOVED******REMOVED*** {
            next(***REMOVED***;
            belse = statement(***REMOVED***;
      ***REMOVED***
        return new AST_If({
            condition   : cond,
            body        : body,
            alternative : belse
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    function block_(***REMOVED*** {
        expect("{"***REMOVED***;
        var a = [];
        while (!is("punc", "}"***REMOVED******REMOVED*** {
            if (is("eof"***REMOVED******REMOVED*** unexpected(***REMOVED***;
            a.push(statement(***REMOVED******REMOVED***;
      ***REMOVED***
        next(***REMOVED***;
        return a;
  ***REMOVED***;

    function switch_body_(***REMOVED*** {
        expect("{"***REMOVED***;
        var a = [], cur = null, branch = null, tmp;
        while (!is("punc", "}"***REMOVED******REMOVED*** {
            if (is("eof"***REMOVED******REMOVED*** unexpected(***REMOVED***;
            if (is("keyword", "case"***REMOVED******REMOVED*** {
                if (branch***REMOVED*** branch.end = prev(***REMOVED***;
                cur = [];
                branch = new AST_Case({
                    start      : (tmp = S.token, next(***REMOVED***, tmp***REMOVED***,
                    expression : expression(true***REMOVED***,
                    body       : cur
              ***REMOVED******REMOVED***;
                a.push(branch***REMOVED***;
                expect(":"***REMOVED***;
          ***REMOVED***
            else if (is("keyword", "default"***REMOVED******REMOVED*** {
                if (branch***REMOVED*** branch.end = prev(***REMOVED***;
                cur = [];
                branch = new AST_Default({
                    start : (tmp = S.token, next(***REMOVED***, expect(":"***REMOVED***, tmp***REMOVED***,
                    body  : cur
              ***REMOVED******REMOVED***;
                a.push(branch***REMOVED***;
          ***REMOVED***
            else {
                if (!cur***REMOVED*** unexpected(***REMOVED***;
                cur.push(statement(***REMOVED******REMOVED***;
          ***REMOVED***
      ***REMOVED***
        if (branch***REMOVED*** branch.end = prev(***REMOVED***;
        next(***REMOVED***;
        return a;
  ***REMOVED***;

    function try_(***REMOVED*** {
        var body = block_(***REMOVED***, bcatch = null, bfinally = null;
        if (is("keyword", "catch"***REMOVED******REMOVED*** {
            var start = S.token;
            next(***REMOVED***;
            expect("("***REMOVED***;
            var name = as_symbol(AST_SymbolCatch***REMOVED***;
            expect("***REMOVED***"***REMOVED***;
            bcatch = new AST_Catch({
                start   : start,
                argname : name,
                body    : block_(***REMOVED***,
                end     : prev(***REMOVED***
          ***REMOVED******REMOVED***;
      ***REMOVED***
        if (is("keyword", "finally"***REMOVED******REMOVED*** {
            var start = S.token;
            next(***REMOVED***;
            bfinally = new AST_Finally({
                start : start,
                body  : block_(***REMOVED***,
                end   : prev(***REMOVED***
          ***REMOVED******REMOVED***;
      ***REMOVED***
        if (!bcatch && !bfinally***REMOVED***
            croak("Missing catch/finally blocks"***REMOVED***;
        return new AST_Try({
            body     : body,
            bcatch   : bcatch,
            bfinally : bfinally
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    function vardefs(no_in, in_const***REMOVED*** {
        var a = [];
        for (;;***REMOVED*** {
            a.push(new AST_VarDef({
                start : S.token,
                name  : as_symbol(in_const ? AST_SymbolConst : AST_SymbolVar***REMOVED***,
                value : is("operator", "="***REMOVED*** ? (next(***REMOVED***, expression(false, no_in***REMOVED******REMOVED*** : null,
                end   : prev(***REMOVED***
          ***REMOVED******REMOVED******REMOVED***;
            if (!is("punc", ","***REMOVED******REMOVED***
                break;
            next(***REMOVED***;
      ***REMOVED***
        return a;
  ***REMOVED***;

    var var_ = function(no_in***REMOVED*** {
        return new AST_Var({
            start       : prev(***REMOVED***,
            definitions : vardefs(no_in, false***REMOVED***,
            end         : prev(***REMOVED***
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    var const_ = function(***REMOVED*** {
        return new AST_Const({
            start       : prev(***REMOVED***,
            definitions : vardefs(false, true***REMOVED***,
            end         : prev(***REMOVED***
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    var new_ = function(***REMOVED*** {
        var start = S.token;
        expect_token("operator", "new"***REMOVED***;
        var newexp = expr_atom(false***REMOVED***, args;
        if (is("punc", "("***REMOVED******REMOVED*** {
            next(***REMOVED***;
            args = expr_list("***REMOVED***"***REMOVED***;
      ***REMOVED*** else {
            args = [];
      ***REMOVED***
        return subscripts(new AST_New({
            start      : start,
            expression : newexp,
            args       : args,
            end        : prev(***REMOVED***
      ***REMOVED******REMOVED***, true***REMOVED***;
  ***REMOVED***;

    function as_atom_node(***REMOVED*** {
        var tok = S.token, ret;
        switch (tok.type***REMOVED*** {
          case "name":
          case "keyword":
            ret = _make_symbol(AST_SymbolRef***REMOVED***;
            break;
          case "num":
            ret = new AST_Number({ start: tok, end: tok, value: tok.value }***REMOVED***;
            break;
          case "string":
            ret = new AST_String({ start: tok, end: tok, value: tok.value }***REMOVED***;
            break;
          case "regexp":
            ret = new AST_RegExp({ start: tok, end: tok, value: tok.value }***REMOVED***;
            break;
          case "atom":
            switch (tok.value***REMOVED*** {
              case "false":
                ret = new AST_False({ start: tok, end: tok }***REMOVED***;
                break;
              case "true":
                ret = new AST_True({ start: tok, end: tok }***REMOVED***;
                break;
              case "null":
                ret = new AST_Null({ start: tok, end: tok }***REMOVED***;
                break;
          ***REMOVED***
            break;
      ***REMOVED***
        next(***REMOVED***;
        return ret;
  ***REMOVED***;

    var expr_atom = function(allow_calls***REMOVED*** {
        if (is("operator", "new"***REMOVED******REMOVED*** {
            return new_(***REMOVED***;
      ***REMOVED***
        var start = S.token;
        if (is("punc"***REMOVED******REMOVED*** {
            switch (start.value***REMOVED*** {
              case "(":
                next(***REMOVED***;
                var ex = expression(true***REMOVED***;
                ex.start = start;
                ex.end = S.token;
                expect("***REMOVED***"***REMOVED***;
                return subscripts(ex, allow_calls***REMOVED***;
              case "[":
                return subscripts(array_(***REMOVED***, allow_calls***REMOVED***;
              case "{":
                return subscripts(object_(***REMOVED***, allow_calls***REMOVED***;
          ***REMOVED***
            unexpected(***REMOVED***;
      ***REMOVED***
        if (is("keyword", "function"***REMOVED******REMOVED*** {
            next(***REMOVED***;
            var func = function_(AST_Function***REMOVED***;
            func.start = start;
            func.end = prev(***REMOVED***;
            return subscripts(func, allow_calls***REMOVED***;
      ***REMOVED***
        if (ATOMIC_START_TOKEN[S.token.type]***REMOVED*** {
            return subscripts(as_atom_node(***REMOVED***, allow_calls***REMOVED***;
      ***REMOVED***
        unexpected(***REMOVED***;
  ***REMOVED***;

    function expr_list(closing, allow_trailing_comma, allow_empty***REMOVED*** {
        var first = true, a = [];
        while (!is("punc", closing***REMOVED******REMOVED*** {
            if (first***REMOVED*** first = false; else expect(","***REMOVED***;
            if (allow_trailing_comma && is("punc", closing***REMOVED******REMOVED*** break;
            if (is("punc", ","***REMOVED*** && allow_empty***REMOVED*** {
                a.push(new AST_Hole({ start: S.token, end: S.token }***REMOVED******REMOVED***;
          ***REMOVED*** else {
                a.push(expression(false***REMOVED******REMOVED***;
          ***REMOVED***
      ***REMOVED***
        next(***REMOVED***;
        return a;
  ***REMOVED***;

    var array_ = embed_tokens(function(***REMOVED*** {
        expect("["***REMOVED***;
        return new AST_Array({
            elements: expr_list("]", !options.strict, true***REMOVED***
      ***REMOVED******REMOVED***;
  ***REMOVED******REMOVED***;

    var object_ = embed_tokens(function(***REMOVED*** {
        expect("{"***REMOVED***;
        var first = true, a = [];
        while (!is("punc", "}"***REMOVED******REMOVED*** {
            if (first***REMOVED*** first = false; else expect(","***REMOVED***;
            if (!options.strict && is("punc", "}"***REMOVED******REMOVED***
                // allow trailing comma
                break;
            var start = S.token;
            var type = start.type;
            var name = as_property_name(***REMOVED***;
            if (type == "name" && !is("punc", ":"***REMOVED******REMOVED*** {
                if (name == "get"***REMOVED*** {
                    a.push(new AST_ObjectGetter({
                        start : start,
                        key   : as_atom_node(***REMOVED***,
                        value : function_(AST_Accessor***REMOVED***,
                        end   : prev(***REMOVED***
                  ***REMOVED******REMOVED******REMOVED***;
                    continue;
              ***REMOVED***
                if (name == "set"***REMOVED*** {
                    a.push(new AST_ObjectSetter({
                        start : start,
                        key   : as_atom_node(***REMOVED***,
                        value : function_(AST_Accessor***REMOVED***,
                        end   : prev(***REMOVED***
                  ***REMOVED******REMOVED******REMOVED***;
                    continue;
              ***REMOVED***
          ***REMOVED***
            expect(":"***REMOVED***;
            a.push(new AST_ObjectKeyVal({
                start : start,
                key   : name,
                value : expression(false***REMOVED***,
                end   : prev(***REMOVED***
          ***REMOVED******REMOVED******REMOVED***;
      ***REMOVED***
        next(***REMOVED***;
        return new AST_Object({ properties: a }***REMOVED***;
  ***REMOVED******REMOVED***;

    function as_property_name(***REMOVED*** {
        var tmp = S.token;
        next(***REMOVED***;
        switch (tmp.type***REMOVED*** {
          case "num":
          case "string":
          case "name":
          case "operator":
          case "keyword":
          case "atom":
            return tmp.value;
          default:
            unexpected(***REMOVED***;
      ***REMOVED***
  ***REMOVED***;

    function as_name(***REMOVED*** {
        var tmp = S.token;
        next(***REMOVED***;
        switch (tmp.type***REMOVED*** {
          case "name":
          case "operator":
          case "keyword":
          case "atom":
            return tmp.value;
          default:
            unexpected(***REMOVED***;
      ***REMOVED***
  ***REMOVED***;

    function _make_symbol(type***REMOVED*** {
        var name = S.token.value;
        return new (name == "this" ? AST_This : type***REMOVED***({
            name  : String(name***REMOVED***,
            start : S.token,
            end   : S.token
      ***REMOVED******REMOVED***;
  ***REMOVED***;

    function as_symbol(type, noerror***REMOVED*** {
        if (!is("name"***REMOVED******REMOVED*** {
            if (!noerror***REMOVED*** croak("Name expected"***REMOVED***;
            return null;
      ***REMOVED***
        var sym = _make_symbol(type***REMOVED***;
        next(***REMOVED***;
        return sym;
  ***REMOVED***;

    var subscripts = function(expr, allow_calls***REMOVED*** {
        var start = expr.start;
        if (is("punc", "."***REMOVED******REMOVED*** {
            next(***REMOVED***;
            return subscripts(new AST_Dot({
                start      : start,
                expression : expr,
                property   : as_name(***REMOVED***,
                end        : prev(***REMOVED***
          ***REMOVED******REMOVED***, allow_calls***REMOVED***;
      ***REMOVED***
        if (is("punc", "["***REMOVED******REMOVED*** {
            next(***REMOVED***;
            var prop = expression(true***REMOVED***;
            expect("]"***REMOVED***;
            return subscripts(new AST_Sub({
                start      : start,
                expression : expr,
                property   : prop,
                end        : prev(***REMOVED***
          ***REMOVED******REMOVED***, allow_calls***REMOVED***;
      ***REMOVED***
        if (allow_calls && is("punc", "("***REMOVED******REMOVED*** {
            next(***REMOVED***;
            return subscripts(new AST_Call({
                start      : start,
                expression : expr,
                args       : expr_list("***REMOVED***"***REMOVED***,
                end        : prev(***REMOVED***
          ***REMOVED******REMOVED***, true***REMOVED***;
      ***REMOVED***
        return expr;
  ***REMOVED***;

    var maybe_unary = function(allow_calls***REMOVED*** {
        var start = S.token;
        if (is("operator"***REMOVED*** && UNARY_PREFIX(start.value***REMOVED******REMOVED*** {
            next(***REMOVED***;
            handle_regexp(***REMOVED***;
            var ex = make_unary(AST_UnaryPrefix, start.value, maybe_unary(allow_calls***REMOVED******REMOVED***;
            ex.start = start;
            ex.end = prev(***REMOVED***;
            return ex;
      ***REMOVED***
        var val = expr_atom(allow_calls***REMOVED***;
        while (is("operator"***REMOVED*** && UNARY_POSTFIX(S.token.value***REMOVED*** && !S.token.nlb***REMOVED*** {
            val = make_unary(AST_UnaryPostfix, S.token.value, val***REMOVED***;
            val.start = start;
            val.end = S.token;
            next(***REMOVED***;
      ***REMOVED***
        return val;
  ***REMOVED***;

    function make_unary(ctor, op, expr***REMOVED*** {
        if ((op == "++" || op == "--"***REMOVED*** && !is_assignable(expr***REMOVED******REMOVED***
            croak("Invalid use of " + op + " operator"***REMOVED***;
        return new ctor({ operator: op, expression: expr }***REMOVED***;
  ***REMOVED***;

    var expr_op = function(left, min_prec, no_in***REMOVED*** {
        var op = is("operator"***REMOVED*** ? S.token.value : null;
        if (op == "in" && no_in***REMOVED*** op = null;
        var prec = op != null ? PRECEDENCE[op] : null;
        if (prec != null && prec > min_prec***REMOVED*** {
            next(***REMOVED***;
            var right = expr_op(maybe_unary(true***REMOVED***, prec, no_in***REMOVED***;
            return expr_op(new AST_Binary({
                start    : left.start,
                left     : left,
                operator : op,
                right    : right,
                end      : right.end
          ***REMOVED******REMOVED***, min_prec, no_in***REMOVED***;
      ***REMOVED***
        return left;
  ***REMOVED***;

    function expr_ops(no_in***REMOVED*** {
        return expr_op(maybe_unary(true***REMOVED***, 0, no_in***REMOVED***;
  ***REMOVED***;

    var maybe_conditional = function(no_in***REMOVED*** {
        var start = S.token;
        var expr = expr_ops(no_in***REMOVED***;
        if (is("operator", "?"***REMOVED******REMOVED*** {
            next(***REMOVED***;
            var yes = expression(false***REMOVED***;
            expect(":"***REMOVED***;
            return new AST_Conditional({
                start       : start,
                condition   : expr,
                consequent  : yes,
                alternative : expression(false, no_in***REMOVED***,
                end         : prev(***REMOVED***
          ***REMOVED******REMOVED***;
      ***REMOVED***
        return expr;
  ***REMOVED***;

    function is_assignable(expr***REMOVED*** {
        if (!options.strict***REMOVED*** return true;
        if (expr instanceof AST_This***REMOVED*** return false;
        return (expr instanceof AST_PropAccess || expr instanceof AST_Symbol***REMOVED***;
  ***REMOVED***;

    var maybe_assign = function(no_in***REMOVED*** {
        var start = S.token;
        var left = maybe_conditional(no_in***REMOVED***, val = S.token.value;
        if (is("operator"***REMOVED*** && ASSIGNMENT(val***REMOVED******REMOVED*** {
            if (is_assignable(left***REMOVED******REMOVED*** {
                next(***REMOVED***;
                return new AST_Assign({
                    start    : start,
                    left     : left,
                    operator : val,
                    right    : maybe_assign(no_in***REMOVED***,
                    end      : prev(***REMOVED***
              ***REMOVED******REMOVED***;
          ***REMOVED***
            croak("Invalid assignment"***REMOVED***;
      ***REMOVED***
        return left;
  ***REMOVED***;

    var expression = function(commas, no_in***REMOVED*** {
        var start = S.token;
        var expr = maybe_assign(no_in***REMOVED***;
        if (commas && is("punc", ","***REMOVED******REMOVED*** {
            next(***REMOVED***;
            return new AST_Seq({
                start  : start,
                car    : expr,
                cdr    : expression(true, no_in***REMOVED***,
                end    : peek(***REMOVED***
          ***REMOVED******REMOVED***;
      ***REMOVED***
        return expr;
  ***REMOVED***;

    function in_loop(cont***REMOVED*** {
        ++S.in_loop;
        var ret = cont(***REMOVED***;
        --S.in_loop;
        return ret;
  ***REMOVED***;

    if (options.expression***REMOVED*** {
        return expression(true***REMOVED***;
  ***REMOVED***

    return (function(***REMOVED***{
        var start = S.token;
        var body = [];
        while (!is("eof"***REMOVED******REMOVED***
            body.push(statement(***REMOVED******REMOVED***;
        var end = prev(***REMOVED***;
        var toplevel = options.toplevel;
        if (toplevel***REMOVED*** {
            toplevel.body = toplevel.body.concat(body***REMOVED***;
            toplevel.end = end;
      ***REMOVED*** else {
            toplevel = new AST_Toplevel({ start: start, body: body, end: end }***REMOVED***;
      ***REMOVED***
        return toplevel;
  ***REMOVED******REMOVED***(***REMOVED***;

};
