// tough-cookie/lib/store.js
'use strict';
/*jshint unused:false */

function Store(***REMOVED*** {
}
exports.Store = Store;

// Stores may be synchronous, but are still required to use a
// Continuation-Passing Style API.  The CookieJar itself will expose a "*Sync"
// API that converts from synchronous-callbacks to imperative style.
Store.prototype.synchronous = false;

Store.prototype.findCookie = function(domain, path, key, cb***REMOVED*** {
  throw new Error('findCookie is not implemented'***REMOVED***;
};

Store.prototype.findCookies = function(domain, path, cb***REMOVED*** {
  throw new Error('findCookies is not implemented'***REMOVED***;
};

Store.prototype.putCookie = function(cookie, cb***REMOVED*** {
  throw new Error('putCookie is not implemented'***REMOVED***;
};

Store.prototype.updateCookie = function(oldCookie, newCookie, cb***REMOVED*** {
  // recommended default implementation:
  // return this.putCookie(newCookie, cb***REMOVED***;
  throw new Error('updateCookie is not implemented'***REMOVED***;
};

Store.prototype.removeCookie = function(domain, path, key, cb***REMOVED*** {
  throw new Error('removeCookie is not implemented'***REMOVED***;
};

Store.prototype.removeCookies = function removeCookies(domain, path, cb***REMOVED*** {
  throw new Error('removeCookies is not implemented'***REMOVED***;
};
