// xmlbuilder/lib/XMLDTDEntity.js
// Generated by CoffeeScript 1.6.3
(function(***REMOVED*** {
  var XMLDTDEntity, _;

  _ = require('lodash-node'***REMOVED***;

  module.exports = XMLDTDEntity = (function(***REMOVED*** {
    function XMLDTDEntity(parent, pe, name, value***REMOVED*** {
      this.stringify = parent.stringify;
      if (name == null***REMOVED*** {
        throw new Error("Missing entity name"***REMOVED***;
    ***REMOVED***
      if (value == null***REMOVED*** {
        throw new Error("Missing entity value"***REMOVED***;
    ***REMOVED***
      this.pe = !!pe;
      this.name = this.stringify.eleName(name***REMOVED***;
      if (!_.isObject(value***REMOVED******REMOVED*** {
        this.value = this.stringify.dtdEntityValue(value***REMOVED***;
    ***REMOVED*** else {
        if (!value.pubID && !value.sysID***REMOVED*** {
          throw new Error("Public and/or system identifiers are required for an external entity"***REMOVED***;
      ***REMOVED***
        if (value.pubID && !value.sysID***REMOVED*** {
          throw new Error("System identifier is required for a public external entity"***REMOVED***;
      ***REMOVED***
        if (value.pubID != null***REMOVED*** {
          this.pubID = this.stringify.dtdPubID(value.pubID***REMOVED***;
      ***REMOVED***
        if (value.sysID != null***REMOVED*** {
          this.sysID = this.stringify.dtdSysID(value.sysID***REMOVED***;
      ***REMOVED***
        if (value.nData != null***REMOVED*** {
          this.nData = this.stringify.dtdNData(value.nData***REMOVED***;
      ***REMOVED***
        if (this.pe && this.nData***REMOVED*** {
          throw new Error("Notation declaration is not allowed in a parameter entity"***REMOVED***;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

    XMLDTDEntity.prototype.clone = function(***REMOVED*** {
      return _.create(XMLDTDEntity.prototype, this***REMOVED***;
  ***REMOVED***;

    XMLDTDEntity.prototype.toString = function(options, level***REMOVED*** {
      var indent, newline, pretty, r, space;
      pretty = (options != null ? options.pretty : void 0***REMOVED*** || false;
      indent = (options != null ? options.indent : void 0***REMOVED*** || '  ';
      newline = (options != null ? options.newline : void 0***REMOVED*** || '\n';
      level || (level = 0***REMOVED***;
      space = new Array(level + 1***REMOVED***.join(indent***REMOVED***;
      r = '';
      if (pretty***REMOVED*** {
        r += space;
    ***REMOVED***
      r += '<!ENTITY';
      if (this.pe***REMOVED*** {
        r += ' %';
    ***REMOVED***
      r += ' ' + this.name;
      if (this.value***REMOVED*** {
        r += ' "' + this.value + '"';
    ***REMOVED*** else {
        if (this.pubID && this.sysID***REMOVED*** {
          r += ' PUBLIC "' + this.pubID + '" "' + this.sysID + '"';
      ***REMOVED*** else if (this.sysID***REMOVED*** {
          r += ' SYSTEM "' + this.sysID + '"';
      ***REMOVED***
        if (this.nData***REMOVED*** {
          r += ' NDATA ' + this.nData;
      ***REMOVED***
    ***REMOVED***
      r += '>';
      if (pretty***REMOVED*** {
        r += newline;
    ***REMOVED***
      return r;
  ***REMOVED***;

    return XMLDTDEntity;

***REMOVED******REMOVED***(***REMOVED***;

}***REMOVED***.call(this***REMOVED***;
