// xmlbuilder/lib/XMLDTDAttList.js
// Generated by CoffeeScript 1.6.3
(function(***REMOVED*** {
  var XMLDTDAttList, _;

  _ = require('lodash-node'***REMOVED***;

  module.exports = XMLDTDAttList = (function(***REMOVED*** {
    function XMLDTDAttList(parent, elementName, attributeName, attributeType, defaultValueType, defaultValue***REMOVED*** {
      this.stringify = parent.stringify;
      if (elementName == null***REMOVED*** {
        throw new Error("Missing DTD element name"***REMOVED***;
    ***REMOVED***
      if (attributeName == null***REMOVED*** {
        throw new Error("Missing DTD attribute name"***REMOVED***;
    ***REMOVED***
      if (!attributeType***REMOVED*** {
        throw new Error("Missing DTD attribute type"***REMOVED***;
    ***REMOVED***
      if (!defaultValueType***REMOVED*** {
        throw new Error("Missing DTD attribute default"***REMOVED***;
    ***REMOVED***
      if (defaultValueType.indexOf('#'***REMOVED*** !== 0***REMOVED*** {
        defaultValueType = '#' + defaultValueType;
    ***REMOVED***
      if (!defaultValueType.match(/^(#REQUIRED|#IMPLIED|#FIXED|#DEFAULT***REMOVED***$/***REMOVED******REMOVED*** {
        throw new Error("Invalid default value type; expected: #REQUIRED, #IMPLIED, #FIXED or #DEFAULT"***REMOVED***;
    ***REMOVED***
      if (defaultValue && !defaultValueType.match(/^(#FIXED|#DEFAULT***REMOVED***$/***REMOVED******REMOVED*** {
        throw new Error("Default value only applies to #FIXED or #DEFAULT"***REMOVED***;
    ***REMOVED***
      this.elementName = this.stringify.eleName(elementName***REMOVED***;
      this.attributeName = this.stringify.attName(attributeName***REMOVED***;
      this.attributeType = this.stringify.dtdAttType(attributeType***REMOVED***;
      this.defaultValue = this.stringify.dtdAttDefault(defaultValue***REMOVED***;
      this.defaultValueType = defaultValueType;
  ***REMOVED***

    XMLDTDAttList.prototype.clone = function(***REMOVED*** {
      return _.create(XMLDTDAttList.prototype, this***REMOVED***;
  ***REMOVED***;

    XMLDTDAttList.prototype.toString = function(options, level***REMOVED*** {
      var indent, newline, pretty, r, space;
      pretty = (options != null ? options.pretty : void 0***REMOVED*** || false;
      indent = (options != null ? options.indent : void 0***REMOVED*** || '  ';
      newline = (options != null ? options.newline : void 0***REMOVED*** || '\n';
      level || (level = 0***REMOVED***;
      space = new Array(level + 1***REMOVED***.join(indent***REMOVED***;
      r = '';
      if (pretty***REMOVED*** {
        r += space;
    ***REMOVED***
      r += '<!ATTLIST ' + this.elementName + ' ' + this.attributeName + ' ' + this.attributeType;
      if (this.defaultValueType !== '#DEFAULT'***REMOVED*** {
        r += ' ' + this.defaultValueType;
    ***REMOVED***
      if (this.defaultValue***REMOVED*** {
        r += ' "' + this.defaultValue + '"';
    ***REMOVED***
      r += '>';
      if (pretty***REMOVED*** {
        r += newline;
    ***REMOVED***
      return r;
  ***REMOVED***;

    return XMLDTDAttList;

***REMOVED******REMOVED***(***REMOVED***;

}***REMOVED***.call(this***REMOVED***;
