// xmlbuilder/lib/XMLDeclaration.js
// Generated by CoffeeScript 1.6.3
(function(***REMOVED*** {
  var XMLDeclaration, XMLNode, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent***REMOVED*** { for (var key in parent***REMOVED*** { if (__hasProp.call(parent, key***REMOVED******REMOVED*** child[key] = parent[key]; } function ctor(***REMOVED*** { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(***REMOVED***; child.__super__ = parent.prototype; return child; };

  _ = require('lodash-node'***REMOVED***;

  XMLNode = require('./XMLNode'***REMOVED***;

  module.exports = XMLDeclaration = (function(_super***REMOVED*** {
    __extends(XMLDeclaration, _super***REMOVED***;

    function XMLDeclaration(parent, version, encoding, standalone***REMOVED*** {
      var _ref;
      XMLDeclaration.__super__.constructor.call(this, parent***REMOVED***;
      if (_.isObject(version***REMOVED******REMOVED*** {
        _ref = version, version = _ref.version, encoding = _ref.encoding, standalone = _ref.standalone;
    ***REMOVED***
      if (!version***REMOVED*** {
        version = '1.0';
    ***REMOVED***
      if (version != null***REMOVED*** {
        this.version = this.stringify.xmlVersion(version***REMOVED***;
    ***REMOVED***
      if (encoding != null***REMOVED*** {
        this.encoding = this.stringify.xmlEncoding(encoding***REMOVED***;
    ***REMOVED***
      if (standalone != null***REMOVED*** {
        this.standalone = this.stringify.xmlStandalone(standalone***REMOVED***;
    ***REMOVED***
  ***REMOVED***

    XMLDeclaration.prototype.clone = function(***REMOVED*** {
      return _.create(XMLDeclaration.prototype, this***REMOVED***;
  ***REMOVED***;

    XMLDeclaration.prototype.toString = function(options, level***REMOVED*** {
      var indent, newline, pretty, r, space;
      pretty = (options != null ? options.pretty : void 0***REMOVED*** || false;
      indent = (options != null ? options.indent : void 0***REMOVED*** || '  ';
      newline = (options != null ? options.newline : void 0***REMOVED*** || '\n';
      level || (level = 0***REMOVED***;
      space = new Array(level + 1***REMOVED***.join(indent***REMOVED***;
      r = '';
      if (pretty***REMOVED*** {
        r += space;
    ***REMOVED***
      r += '<?xml';
      if (this.version != null***REMOVED*** {
        r += ' version="' + this.version + '"';
    ***REMOVED***
      if (this.encoding != null***REMOVED*** {
        r += ' encoding="' + this.encoding + '"';
    ***REMOVED***
      if (this.standalone != null***REMOVED*** {
        r += ' standalone="' + this.standalone + '"';
    ***REMOVED***
      r += '?>';
      if (pretty***REMOVED*** {
        r += newline;
    ***REMOVED***
      return r;
  ***REMOVED***;

    return XMLDeclaration;

***REMOVED******REMOVED***(XMLNode***REMOVED***;

}***REMOVED***.call(this***REMOVED***;
