// xmlbuilder/lib/XMLDTDNotation.js
// Generated by CoffeeScript 1.6.3
(function(***REMOVED*** {
  var XMLDTDNotation, _;

  _ = require('lodash-node'***REMOVED***;

  module.exports = XMLDTDNotation = (function(***REMOVED*** {
    function XMLDTDNotation(parent, name, value***REMOVED*** {
      this.stringify = parent.stringify;
      if (name == null***REMOVED*** {
        throw new Error("Missing notation name"***REMOVED***;
    ***REMOVED***
      if (!value.pubID && !value.sysID***REMOVED*** {
        throw new Error("Public or system identifiers are required for an external entity"***REMOVED***;
    ***REMOVED***
      this.name = this.stringify.eleName(name***REMOVED***;
      if (value.pubID != null***REMOVED*** {
        this.pubID = this.stringify.dtdPubID(value.pubID***REMOVED***;
    ***REMOVED***
      if (value.sysID != null***REMOVED*** {
        this.sysID = this.stringify.dtdSysID(value.sysID***REMOVED***;
    ***REMOVED***
  ***REMOVED***

    XMLDTDNotation.prototype.clone = function(***REMOVED*** {
      return _.create(XMLDTDNotation.prototype, this***REMOVED***;
  ***REMOVED***;

    XMLDTDNotation.prototype.toString = function(options, level***REMOVED*** {
      var indent, newline, pretty, r, space;
      pretty = (options != null ? options.pretty : void 0***REMOVED*** || false;
      indent = (options != null ? options.indent : void 0***REMOVED*** || '  ';
      newline = (options != null ? options.newline : void 0***REMOVED*** || '\n';
      level || (level = 0***REMOVED***;
      space = new Array(level + 1***REMOVED***.join(indent***REMOVED***;
      r = '';
      if (pretty***REMOVED*** {
        r += space;
    ***REMOVED***
      r += '<!NOTATION ' + this.name;
      if (this.pubID && this.sysID***REMOVED*** {
        r += ' PUBLIC "' + this.pubID + '" "' + this.sysID + '"';
    ***REMOVED*** else if (this.pubID***REMOVED*** {
        r += ' PUBLIC "' + this.pubID + '"';
    ***REMOVED*** else if (this.sysID***REMOVED*** {
        r += ' SYSTEM "' + this.sysID + '"';
    ***REMOVED***
      r += '>';
      if (pretty***REMOVED*** {
        r += newline;
    ***REMOVED***
      return r;
  ***REMOVED***;

    return XMLDTDNotation;

***REMOVED******REMOVED***(***REMOVED***;

}***REMOVED***.call(this***REMOVED***;
