// xmlbuilder/lib/XMLElement.js
// Generated by CoffeeScript 1.6.3
(function(***REMOVED*** {
  var XMLAttribute, XMLElement, XMLNode, XMLProcessingInstruction, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent***REMOVED*** { for (var key in parent***REMOVED*** { if (__hasProp.call(parent, key***REMOVED******REMOVED*** child[key] = parent[key]; } function ctor(***REMOVED*** { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(***REMOVED***; child.__super__ = parent.prototype; return child; };

  _ = require('lodash-node'***REMOVED***;

  XMLNode = require('./XMLNode'***REMOVED***;

  XMLAttribute = require('./XMLAttribute'***REMOVED***;

  XMLProcessingInstruction = require('./XMLProcessingInstruction'***REMOVED***;

  module.exports = XMLElement = (function(_super***REMOVED*** {
    __extends(XMLElement, _super***REMOVED***;

    function XMLElement(parent, name, attributes***REMOVED*** {
      XMLElement.__super__.constructor.call(this, parent***REMOVED***;
      if (name == null***REMOVED*** {
        throw new Error("Missing element name"***REMOVED***;
    ***REMOVED***
      this.name = this.stringify.eleName(name***REMOVED***;
      this.children = [];
      this.instructions = [];
      this.attributes = {};
      if (attributes != null***REMOVED*** {
        this.attribute(attributes***REMOVED***;
    ***REMOVED***
  ***REMOVED***

    XMLElement.prototype.clone = function(***REMOVED*** {
      var att, attName, clonedSelf, pi, _i, _len, _ref, _ref1;
      clonedSelf = _.create(XMLElement.prototype, this***REMOVED***;
      clonedSelf.attributes = {};
      _ref = this.attributes;
      for (attName in _ref***REMOVED*** {
        if (!__hasProp.call(_ref, attName***REMOVED******REMOVED*** continue;
        att = _ref[attName];
        clonedSelf.attributes[attName] = att.clone(***REMOVED***;
    ***REMOVED***
      clonedSelf.instructions = [];
      _ref1 = this.instructions;
      for (_i = 0, _len = _ref1.length; _i < _len; _i++***REMOVED*** {
        pi = _ref1[_i];
        clonedSelf.instructions.push(pi.clone(***REMOVED******REMOVED***;
    ***REMOVED***
      clonedSelf.children = [];
      this.children.forEach(function(child***REMOVED*** {
        var clonedChild;
        clonedChild = child.clone(***REMOVED***;
        clonedChild.parent = clonedSelf;
        return clonedSelf.children.push(clonedChild***REMOVED***;
    ***REMOVED******REMOVED***;
      return clonedSelf;
  ***REMOVED***;

    XMLElement.prototype.attribute = function(name, value***REMOVED*** {
      var attName, attValue;
      if (_.isObject(name***REMOVED******REMOVED*** {
        for (attName in name***REMOVED*** {
          if (!__hasProp.call(name, attName***REMOVED******REMOVED*** continue;
          attValue = name[attName];
          this.attribute(attName, attValue***REMOVED***;
      ***REMOVED***
    ***REMOVED*** else {
        if (_.isFunction(value***REMOVED******REMOVED*** {
          value = value.apply(***REMOVED***;
      ***REMOVED***
        if (!this.options.skipNullAttributes || (value != null***REMOVED******REMOVED*** {
          this.attributes[name] = new XMLAttribute(this, name, value***REMOVED***;
      ***REMOVED***
    ***REMOVED***
      return this;
  ***REMOVED***;

    XMLElement.prototype.removeAttribute = function(name***REMOVED*** {
      var attName, _i, _len;
      if (name == null***REMOVED*** {
        throw new Error("Missing attribute name"***REMOVED***;
    ***REMOVED***
      if (_.isArray(name***REMOVED******REMOVED*** {
        for (_i = 0, _len = name.length; _i < _len; _i++***REMOVED*** {
          attName = name[_i];
          delete this.attributes[attName];
      ***REMOVED***
    ***REMOVED*** else {
        delete this.attributes[name];
    ***REMOVED***
      return this;
  ***REMOVED***;

    XMLElement.prototype.instruction = function(target, value***REMOVED*** {
      var insTarget, insValue, instruction, _i, _len;
      if (_.isArray(target***REMOVED******REMOVED*** {
        for (_i = 0, _len = target.length; _i < _len; _i++***REMOVED*** {
          insTarget = target[_i];
          this.instruction(insTarget***REMOVED***;
      ***REMOVED***
    ***REMOVED*** else if (_.isObject(target***REMOVED******REMOVED*** {
        for (insTarget in target***REMOVED*** {
          if (!__hasProp.call(target, insTarget***REMOVED******REMOVED*** continue;
          insValue = target[insTarget];
          this.instruction(insTarget, insValue***REMOVED***;
      ***REMOVED***
    ***REMOVED*** else {
        if (_.isFunction(value***REMOVED******REMOVED*** {
          value = value.apply(***REMOVED***;
      ***REMOVED***
        instruction = new XMLProcessingInstruction(this, target, value***REMOVED***;
        this.instructions.push(instruction***REMOVED***;
    ***REMOVED***
      return this;
  ***REMOVED***;

    XMLElement.prototype.toString = function(options, level***REMOVED*** {
      var att, child, indent, instruction, name, newline, pretty, r, space, _i, _j, _len, _len1, _ref, _ref1, _ref2;
      pretty = (options != null ? options.pretty : void 0***REMOVED*** || false;
      indent = (options != null ? options.indent : void 0***REMOVED*** || '  ';
      newline = (options != null ? options.newline : void 0***REMOVED*** || '\n';
      level || (level = 0***REMOVED***;
      space = new Array(level + 1***REMOVED***.join(indent***REMOVED***;
      r = '';
      _ref = this.instructions;
      for (_i = 0, _len = _ref.length; _i < _len; _i++***REMOVED*** {
        instruction = _ref[_i];
        r += instruction.toString(options, level + 1***REMOVED***;
    ***REMOVED***
      if (pretty***REMOVED*** {
        r += space;
    ***REMOVED***
      r += '<' + this.name;
      _ref1 = this.attributes;
      for (name in _ref1***REMOVED*** {
        if (!__hasProp.call(_ref1, name***REMOVED******REMOVED*** continue;
        att = _ref1[name];
        r += att.toString(options***REMOVED***;
    ***REMOVED***
      if (this.children.length === 0***REMOVED*** {
        r += '/>';
        if (pretty***REMOVED*** {
          r += newline;
      ***REMOVED***
    ***REMOVED*** else if (pretty && this.children.length === 1 && (this.children[0].value != null***REMOVED******REMOVED*** {
        r += '>';
        r += this.children[0].value;
        r += '</' + this.name + '>';
        r += newline;
    ***REMOVED*** else {
        r += '>';
        if (pretty***REMOVED*** {
          r += newline;
      ***REMOVED***
        _ref2 = this.children;
        for (_j = 0, _len1 = _ref2.length; _j < _len1; _j++***REMOVED*** {
          child = _ref2[_j];
          r += child.toString(options, level + 1***REMOVED***;
      ***REMOVED***
        if (pretty***REMOVED*** {
          r += space;
      ***REMOVED***
        r += '</' + this.name + '>';
        if (pretty***REMOVED*** {
          r += newline;
      ***REMOVED***
    ***REMOVED***
      return r;
  ***REMOVED***;

    XMLElement.prototype.att = function(name, value***REMOVED*** {
      return this.attribute(name, value***REMOVED***;
  ***REMOVED***;

    XMLElement.prototype.ins = function(target, value***REMOVED*** {
      return this.instruction(target, value***REMOVED***;
  ***REMOVED***;

    XMLElement.prototype.a = function(name, value***REMOVED*** {
      return this.attribute(name, value***REMOVED***;
  ***REMOVED***;

    XMLElement.prototype.i = function(target, value***REMOVED*** {
      return this.instruction(target, value***REMOVED***;
  ***REMOVED***;

    return XMLElement;

***REMOVED******REMOVED***(XMLNode***REMOVED***;

}***REMOVED***.call(this***REMOVED***;
